var documenterSearchIndex = {"docs":
[{"location":"steady_state_functionality/homotopy_continuation/#homotopy_continuation","page":"Finding Steady States through Homotopy Continuation","title":"Finding Steady States through Homotopy Continuation","text":"<details><summary><strong>Environment setup and package installation</strong></summary>\n\nThe following code sets up an environment for running the code on this page.\n\nusing Pkg\nPkg.activate(; temp = true) # Creates a temporary environment, which is deleted when the Julia session ends.\nPkg.add(\"Catalyst\")\nPkg.add(\"HomotopyContinuation\")\n\n</details>\n\n<details><summary><strong>Quick-start example</strong></summary>\n\nThe following code provides a brief example of how to find all of a reaction network model's steady states using HomotopyContinuation.jl package.\n\nusing Catalyst\nimport HomotopyContinuation\nwilhelm_2009_model = @reaction_network begin\n    k1, Y --> 2X\n    k2, 2X --> X + Y\n    k3, X + Y --> Y\n    k4, X --> 0\nend\n\n# Designate the parameter sets for which we want to find steady states. Finds them using `hc_steady_states`.\nps = [:k1 => 8.0, :k2 => 2.0, :k3 => 1.0, :k4 => 1.5]\nhc_steady_states(wilhelm_2009_model, ps)\n\n</details>\n\n\n  \n\nThe steady states of a dynamical system dx over dt = f(x) can be found by solving 0 = f(x). This is typically a hard problem, and generally, there is no method guaranteed to find all steady states for a system that has multiple ones. However, many chemical reaction networks generate polynomial systems (for example those which are purely mass action or have only have Hill functions with integer Hill exponents). The roots of these can reliably be found through a homotopy continuation algorithm[1][2]. This is implemented in Julia through the HomotopyContinuation.jl package[3].\n\nCatalyst contains a special homotopy continuation extension that is loaded whenever HomotopyContinuation.jl is. This exports a single function, hc_steady_states, that can be used to find the steady states of any ReactionSystem structure.\n\nFor this tutorial, we will use the Wilhelm model (which demonstrates bistability in a small chemical reaction network). We declare the model and the parameter set for which we want to find the steady states:\n\nusing Catalyst\nimport HomotopyContinuation\n\nwilhelm_2009_model = @reaction_network begin\n    k1, Y --> 2X\n    k2, 2X --> X + Y\n    k3, X + Y --> Y\n    k4, X --> 0\nend\nps = [:k1 => 8.0, :k2 => 2.0, :k3 => 1.0, :k4 => 1.5]\nnothing # hide\n\nHere, we only run import HomotopyContinuation as we do not require any of its functions, and just need it to be present in the current scope for the extension to be activated.\n\nNow we can find the steady states using:\n\nhc_steady_states(wilhelm_2009_model, ps)\n\nThe order of the species in the output vectors are the same as in species(wilhelm_2009_model).\n\nIt should be noted that the steady state multivariate polynomials produced by reaction systems may have both imaginary and negative roots, which are filtered away by hc_steady_states. If you want the negative roots, you can use the hc_steady_states(wilhelm_2009_model, ps; filter_negative=false) argument.","category":"section"},{"location":"steady_state_functionality/homotopy_continuation/#homotopy_continuation_conservation_laws","page":"Finding Steady States through Homotopy Continuation","title":"Systems with conservation laws","text":"Some systems are under-determined, and have an infinite number of possible steady states. These are typically systems containing a conservation law, e.g.\n\nusing Catalyst # hide\ntwo_state_model = @reaction_network begin\n    (k1,k2), X1 <--> X2\nend\n\nCatalyst allows the conservation laws of such systems to be computed using the conservationlaws function. By using these to reduce the dimensionality of the system, as well as specifying the initial amount of each species, homotopy continuation can again be used to find steady states. Here we do this by designating such an initial condition (which is used to compute the system's conserved quantities, in this case X1 + X2):\n\nimport HomotopyContinuation # hide\nps = [:k1 => 2.0, :k2 => 1.0]\nu0 = [:X1 => 1.0, :X2 => 1.0]\nhc_steady_states(two_state_model, ps; u0)","category":"section"},{"location":"steady_state_functionality/homotopy_continuation/#Final-notes","page":"Finding Steady States through Homotopy Continuation","title":"Final notes","text":"hc_steady_states support any systems where all rates are systems of rational polynomials (such as Hill functions with integer Hill coefficients).\nWhen providing initial conditions to compute conservation laws, values are only required for those species that are part of conserved quantities. If this set of species is unknown, it is recommended to provide initial conditions for all species. \nAdditional arguments provided to hc_steady_states are automatically passed to HomotopyContinuation's solve command. Use e.g. show_progress = false to disable the progress bar.\n\n","category":"section"},{"location":"steady_state_functionality/homotopy_continuation/#homotopy_continuation_citation","page":"Finding Steady States through Homotopy Continuation","title":"Citation","text":"If you use this functionality in your research, please cite the following paper to support the authors of the HomotopyContinuation package:\n\n@inproceedings{HomotopyContinuation.jl,\n  title={{H}omotopy{C}ontinuation.jl: {A} {P}ackage for {H}omotopy {C}ontinuation in {J}ulia},\n  author={Breiding, Paul and Timme, Sascha},\n  booktitle={International Congress on Mathematical Software},\n  pages={458--465},\n  year={2018},\n  organization={Springer}\n}\n\n","category":"section"},{"location":"steady_state_functionality/homotopy_continuation/#References","page":"Finding Steady States through Homotopy Continuation","title":"References","text":"[1]: Andrew J Sommese, Charles W Wampler The Numerical Solution of Systems of Polynomials Arising in Engineering and Science, World Scientific (2005).\n\n[2]: Daniel J. Bates, Paul Breiding, Tianran Chen, Jonathan D. Hauenstein, Anton Leykin, Frank Sottile, Numerical Nonlinear Algebra, arXiv (2023).\n\n[3]: Paul Breiding, Sascha Timme, HomotopyContinuation.jl: A Package for Homotopy Continuation in Julia, International Congress on Mathematical Software (2018).","category":"section"},{"location":"model_creation/reactionsystem_content_accessing/#model_accessing","page":"Accessing Model Properties","title":"Accessing Model Properties","text":"<details><summary><strong>Environment setup and package installation</strong></summary>\n\nThe following code sets up an environment for running the code on this page.\n\nusing Pkg\nPkg.activate(; temp = true) # Creates a temporary environment, which is deleted when the Julia session ends.\nPkg.add(\"Catalyst\")\nPkg.add(\"OrdinaryDiffEqDefault\")\nPkg.add(\"Plots\")\nPkg.add(\"SymbolicIndexingInterface\")\n\n</details>\n\n\n   Catalyst is based around the creation, analysis, and simulation of chemical reaction network models. Catalyst stores these models in ReactionSystem structures. This page describes some basic functions for accessing the content of these structures. This includes retrieving lists of species, parameters, or reactions that a model consists of. An extensive list of relevant functions for working with ReactionSystem models can be found in Catalyst's API.\n\nwarning: Warning\nGenerally, a field of a Julia structure can be accessed through struct.fieldname. E.g. a simulation's time vector can be retrieved using simulation.t. While Catalyst ReactionSystems are structures, one should never access their fields using this approach, but rather using the accessor functions described below and in the API (direct accessing of fields can yield unexpected behaviours). E.g. to retrieve the species of a ReactionsSystem called rs, use Catalyst.get_species(rs), not rs.species. The reason is that, as shown below, Catalyst (and more generally any ModelingToolkitBase system) reserves this type of accessing for accessing symbolic variables stored in the system. I.e. rs.X refers to the X symbolic variable, not a field in rs named \"X\".","category":"section"},{"location":"model_creation/reactionsystem_content_accessing/#model_accessing_symbolic_variables","page":"Accessing Model Properties","title":"Direct accessing of symbolic model parameter and species","text":"Previously we have described how the parameters and species that Catalyst models contain are represented using so-called symbolic variables (and how these enable the forming of symbolic expressions). We have described how, during programmatic modelling, the user has direct access to these and how this can be taken advantage of. We have also described how, during DSL-based modelling, the need for symbolic representation can be circumvented by using @unpack or by creating an observable. However, sometimes, it is easier to directly access a symbolic variable through the model itself, something which we will describe here.\n\nLet us consider the following two-state model\n\nusing Catalyst\nrs = @reaction_network begin\n    (k1,k2), X1 <--> X2\nend\n\nIf we wish to access one of the symbolic variables stored in it (here X1, X2, k1, and k2), we simply write \n\nrs.X1\n\nto access e.g. X1. This symbolic variable can be used just like those declared using @parameters and @species:\n\nusing OrdinaryDiffEqDefault\nu0 = [rs.X1 => 1.0, rs.X2 => 2.0]\nps = [rs.k1 => 2.0, rs.k2 => 4.0]\noprob = ODEProblem(rs, u0, (0.0, 10.0), ps)\nsol = solve(oprob)\nnothing # hide\n\nWe can also use them to form symbolic expressions:\n\nXtot = rs.X1 + rs.X2\n\nwhich can be used when we e.g. plot our simulation:\n\nusing Plots\nplot(sol; idxs = [rs.X1, rs.X2, Xtot])\n\nNext we create our two-state model programmatically:\n\nt = default_t()\n@species X1(t) X2(t)\n@parameters k1 k2\nrxs = [\n    Reaction(k1, [X1], [X2]),\n    Reaction(k2, [X2], [X1])\n]\n@named rs_prog = ReactionSystem(rxs, t)\nrs_prog = complete(rs_prog)\nnothing # hide\n\nHere, we can confirm that the symbolic variables we access through the model are identical to those we used to create it:\n\nisequal(rs.k1, k1)\n\nwarning: Warning\nWhen accessing model symbolic variables through the model (using e.g. rs.X1), it is important to first ensure that the model has been marked complete.","category":"section"},{"location":"model_creation/reactionsystem_content_accessing/#model_accessing_basics","page":"Accessing Model Properties","title":"Accessing basic model properties","text":"","category":"section"},{"location":"model_creation/reactionsystem_content_accessing/#model_accessing_basics_parameters_n_species","page":"Accessing Model Properties","title":"Accessing model parameter and species","text":"Previously we showed how to access individual parameters or species of a ReactionSystem model. Next, the parameters and species functions allow us to retrieve all parameters and species as vectors:\n\nusing Catalyst # hide\nsir = @reaction_network begin\n    α, S + I --> 2I\n    β, I --> R\nend\nparameters(sir)\n\nspecies(sir)\n\nThese vectors contain the exact same symbolic variables that we would access through the system:\n\nissetequal([sir.S, sir.I, sir.R], species(sir))\n\nIf we wish to count the number of parameters or species in a system, we can do this directly through the numparams and numspecies functions:\n\nnumparams(sir)\n\nnumspecies(sir)","category":"section"},{"location":"model_creation/reactionsystem_content_accessing/#model_accessing_basics_reactions","page":"Accessing Model Properties","title":"Accessing model reactions","text":"A vector containing all a model's reactions can be retrieved using the reactions function:\n\nreactions(sir)\n\nWe can count the number of reactions in a model using the numreactions function:\n\nnumreactions(sir)\n\nFinally, a vector with all the reactions' rates can be retrieved using reactionrates:\n\nreactionrates(sir)","category":"section"},{"location":"model_creation/reactionsystem_content_accessing/#model_accessing_basics_reactions-2","page":"Accessing Model Properties","title":"Accessing content of models coupled to equations","text":"Previously, we have shown how to couple equations to a chemical reaction network model, creating models containing non-species unknowns (variables). Here we create a birth-death model where some nutrient supply (modelled through the variable N) is depleted in the presence of X:\n\nusing Catalyst # hide\ncoupled_crn = @reaction_network begin\n    @equations D(N) ~ -N * X\n    (p/(1+N),d), 0 <--> X\nend\n\nHere, the unknowns function returns all unknowns (i.e. species and variables):\n\nunknowns(coupled_crn)\n\nMeanwhile, species returns the species only, while nonspecies returns the variables only:\n\nspecies(coupled_crn)\n\nnonspecies(coupled_crn)\n\nSimilarly, the equations function returns a vector with all reactions and equations of the model (ordered so that reactions occur first and equations thereafter):\n\nequations(coupled_crn)\n\nMeanwhile, reactions returns the reactions only, while nonreactions returns any algebraic or differential equations:\n\nreactions(coupled_crn)\n\nnonreactions(coupled_crn)","category":"section"},{"location":"model_creation/reactionsystem_content_accessing/#model_accessing_basics_others","page":"Accessing Model Properties","title":"Accessing other model properties","text":"There exist several other functions for accessing model properties. \n\nThe observed, continuous_events, discrete_events functions can be used to access a model's observables, continuous events, and discrete events, respectively.\n\nThe ModelingToolkitBase.get_iv function can be used to retrieve a model's independent variable:\n\nModelingToolkitBase.get_iv(sir)","category":"section"},{"location":"model_creation/reactionsystem_content_accessing/#model_accessing_basics_getname","page":"Accessing Model Properties","title":"Extracting names from symbolic variables","text":"When working with symbolic variables, it can be useful to extract their names as Symbols (e.g. for labelling or building dictionaries). The SymbolicIndexingInterface.getname function does this:\n\nusing SymbolicIndexingInterface: getname\ngetname(sir.S)\n\nThis can be broadcast to extract names from a collection of symbolic variables:\n\ngetname.(species(sir))","category":"section"},{"location":"model_creation/reactionsystem_content_accessing/#model_accessing_hierarchical","page":"Accessing Model Properties","title":"Accessing properties of hierarchical models","text":"Previously, we have described how compositional modelling can be used to create hierarchical models. There are some special considerations when accessing the content of hierarchical models, which we will describe here.\n\nFirst, we will create a simple hierarchical model. It describes a protein (X) which is created in its inactive form (Xᵢ) in the nucleus, from which it is transported to the cytoplasm, where it is activated.\n\nusing Catalyst # hide\n# Declare submodels.\nnucleus_sys = @network_component nucleus begin\n    (p,d), 0 <--> Xᵢ\nend\ncytoplasm_sys = @network_component cytoplasm begin\n    kₐ, Xᵢ --> Xₐ\n    d, (Xᵢ, Xₐ) --> 0\nend\n\n# Assembly hierarchical model.\ntransport = @reaction kₜ, $(nucleus_sys.Xᵢ) --> $(cytoplasm_sys.Xᵢ)\n@named rs = ReactionSystem([transport], default_t(); systems = [nucleus_sys, cytoplasm_sys])\nrs = complete(rs)\n\nThis model consists of a top-level system, which contains the transportation reaction only, and two subsystems. We can retrieve all the subsystems of the top-level system through Catalyst.get_systems:\n\nCatalyst.get_systems(rs)\nnothing # hide\n\nnote: Note\nIf either of the subsystems had had further subsystems, these would not be retrieved by Catalyst.get_systems (which only returns the direct subsystems of the input system).","category":"section"},{"location":"model_creation/reactionsystem_content_accessing/#model_accessing_hierarchical_symbolic_variables","page":"Accessing Model Properties","title":"Accessing parameter and species of hierarchical models","text":"Our hierarchical model consists of a top-level system (rs) with two subsystems (nucleus_sys and cytoplasm_sys). Note that we have given our subsystems names (nucleus and cytoplasm, respectively). Above, we retrieved the subsystems by calling Catalyst.get_systems on our top-level system. We can also retrieve a subsystem directly by calling:\n\nrs.nucleus\n\nrs.cytoplasm\n\nnote: Note\nWhen accessing subsystems, we use the subsystems' names, not the name of their variables (i.e. we call rs.nucleus, not rs.nucleus_sys).\n\nNext, if we wish to access a species declared as a part of one of the subsystems, we do so through it. E.g. here we access Xₐ (which is part of the cytoplasm subsystem):\n\nrs.cytoplasm.Xₐ\n\nNote that species contained in a subsystem have the subsystem's name prepended to their name when we access it.\n\nBoth subsystems contain a species Xᵢ. However, while they have the same name, these are different species when accessed through their respective models:\n\nisequal(rs.nucleus.Xᵢ, rs.cytoplasm.Xᵢ)\n\nThe same holds for the model parameters, i.e. while each subsystem contains a parameter d, these are considered different parameters:\n\nisequal(rs.nucleus.d, rs.cytoplasm.d)\n\nThe parameter kₜ is actually contained within the top-level model, and is accessed directly through it:\n\nrs.kₜ\n\nPractically, when we simulate our hierarchical model, we use all of this to designate initial conditions and parameters. I.e. below we designate values for the two Xᵢ species and d parameters separately:\n\nusing OrdinaryDiffEqDefault, Plots\nu0 = [rs.nucleus.Xᵢ => 0.0, rs.cytoplasm.Xᵢ => 0.0, rs.cytoplasm.Xₐ => 0.0]\nps = [rs.nucleus.p => 1.0, rs.nucleus.d => 0.2, rs.cytoplasm.kₐ => 5.0, rs.cytoplasm.d => 0.2, rs.kₜ => 0.1]\noprob = ODEProblem(rs, u0, (0.0, 10.0), ps)\nsol = solve(oprob)\nplot(sol)\n\nnote: Note\nWhen we access a symbolic variable through a subsystem (e.g. rs.nucleus.Xᵢ) that subsystem's name is prepended to the symbolic variable's name (we call this namespacing). This is also the case if we access it through the original model, i.e. nucleus_sys.Xᵢ. Namespacing is only performed when we access variables of incomplete systems. I.e. isequal(nucleus_sys.d, cytoplasm_sys.d) returns false (as the systems are incomplete and namespacing is performed). However, isequal(complete(nucleus_sys).d, complete(cytoplasm_sys).d) returns true  (as the systems are complete and namespacing is not performed). This is the reason that the system top-level system's name is never prepended when we do e.g. rs.kₜ (because here, rs is complete).","category":"section"},{"location":"model_creation/reactionsystem_content_accessing/#model_accessing_hierarchical_symbolic_variables-2","page":"Accessing Model Properties","title":"Accessing the content of hierarchical models","text":"In the last section, we noted that our hierarchical model contained several instances of the Xᵢ species. The species function, which retrieves all of a model's species shows that our model has three species (two types of Xᵢ, and one type of Xₐ)\n\nspecies(rs)\n\nSimilarly, parameters retrieves five different parameters. Here, we note that kₜ (which has no model name prepended) belongs to the top-level system (and not a subsystem):\n\nparameters(rs)\n\nIf we wish to retrieve the species (or parameters) that are specifically contained in the top-level system (and not only indirectly through its subsystems), we can use the Catalyst.get_species (or ModelingToolkitBase.getps) functions:\n\nCatalyst.get_species(rs)\n\nModelingToolkitBase.get_ps(rs)\n\nHere, our top-level model contains a single parameter (kₜ), and two the two versions of the Xᵢ species. These are all the symbolic variables that occur in the transportation reaction (@kₜ, $(nucleus_sys.Xᵢ) --> $(cytoplasm_sys.Xᵢ)), which is the only reaction of the top-level system. We can apply these functions to the systems as well. However, when we do so, the systems' names are not prepended:\n\nCatalyst.get_ps(rs.nucleus)\n\nGenerally, functions starting with get_ retrieve only the components stored in the input system (and do not consider its subsystems), while the corresponding function without get_ also retrieves the components stored in subsystems. I.e. compare the Catalyst.get_rxs and reactions functions:\n\nreactions(rs)\n\nCatalyst.get_rxs(rs)\n\nOther examples of similar pairs of functions are get_unknowns and unknowns, and get_observed and observed.\n\nnote: Note\nDue to the large number of available accessor functions, most get_ functions are not directly exported by Catalyst. Hence, these must be used as Catalyst.get_rxs, rather than get_rxs directly. Again, a full list of accessor functions and instructions on how to use them can be found in Catalyst's API.","category":"section"},{"location":"model_creation/functional_parameters/#time_dependent_parameters","page":"Inputs and time-dependent (or functional) parameters","title":"Inputs and time-dependent (or functional) parameters","text":"<details><summary><strong>Environment setup and package installation</strong></summary>\n\nThe following code sets up an environment for running the code on this page.\n\nusing Pkg\nPkg.activate(; temp = true) # Creates a temporary environment, which is deleted when the Julia session ends.\nPkg.add(\"Catalyst\")\nPkg.add(\"DataInterpolations\")\nPkg.add(\"OrdinaryDiffEqDefault\")\nPkg.add(\"Plots\")\n\n</details>\n\n<details><summary><strong>Quick-start example</strong></summary>\n\nSo called functional parameters have many uses. A primary one (shown in this example) is to interpolate some data points and uses as a functional value.\n\nusing Catalyst, DataInterpolations, OrdinaryDiffEqDefault, Plots\n\n# Here we create an interpolated function from some values and plots it.\ntend = 10.0\nts = collect(0.0:0.05:tend)\nspline = LinearInterpolation((2 .+ ts) ./ (1 .+ ts), ts)\n@parameters (pIn::typeof(spline))(..)\nplot(spline)\n\n# Next, we use this as a time dependent function describing the production rate of `X`.\nbd_model = @reaction_network begin\n    $pIn(t), 0 --> X\n    d, X --> 0\nend\n\n# Finally we simualte the model normally, but using `spline` as the `pIn` parameter's value.\nu0 = [:X => 0.5]\nps = [:d => 2.0, :pIn => spline]\noprob = ODEProblem(bd_model, u0, tend, ps)\nsol = solve(oprob)\nplot(sol)\n\n</details>\n\n\n   Catalyst supports the usage of \"functional parameters\". In practice, these are parameters that are given by (typically) time-dependent functions (they can also depend on e.g. species values, as discussed here). They are a way to inject custom functions into models. Functional parameters can be used when rates depend on real data, or to represent complicated functions (which use e.g. for loops or random number generation). Here, the function's values are declared as a data interpolation (which interpolates discrete samples to a continuous function). This is then used as the functional parameter's value in the simulation. This tutorial first shows how to create time-dependent functional parameters, and then gives an example where the functional parameter depends on a species value.\n\nAn alternative approach for representing complicated functions is by using @register_symbolic.","category":"section"},{"location":"model_creation/functional_parameters/#functional_parameters_basic_example","page":"Inputs and time-dependent (or functional) parameters","title":"Basic example","text":"Let us first consider an easy, quick-start example (the next section will discuss what is going on in more detail). We will consider a simple birth-death model, but where the birth rate is determined by an input parameter (for which the value depends on time). First, we define the input parameter programmatically, and its values across all time points using the DataInterpolations.jl package. In this example we will use the input function pIn(t) = (2 + t)(1 + t). Finally, we plot the input function, demonstrating how while it is defined at discrete points, DataInterpolations.jl generalises this to a continuous function.\n\nusing Catalyst, DataInterpolations, Plots\nt = default_t()\ntend = 10.0\nts = collect(0.0:0.05:tend)\nspline = LinearInterpolation((2 .+ ts) ./ (1 .+ ts), ts)\n@parameters (pIn::typeof(spline))(..)\nplot(spline)\n\nNext, we create our model, interpolating the input parameter into the @reaction_network declaration.\n\nbd_model = @reaction_network begin\n    $pIn(t), 0 --> X\n    d, X --> 0\nend\n\nFinally, we can simulate our model as normal (but where we set the value of the pIn parameter to our interpolated data).\n\nusing OrdinaryDiffEqDefault\nu0 = [:X => 0.5]\nps = [:d => 2.0, :pIn => spline]\noprob = ODEProblem(bd_model, u0, tend, ps)\nsol = solve(oprob)\nplot(sol)\n\nnote: Note\nFor this simple example, (2 + t)(1 + t) could have been used directly as a reaction rate (or written as a normal function), technically making the functional parameter approach unnecessary. However, here we used this function as a simple example of how discrete data can be made continuous using DataInterpolations, and then have its values inserted using a (functional) parameter.","category":"section"},{"location":"model_creation/functional_parameters/#functional_parameters_circ_rhythm","page":"Inputs and time-dependent (or functional) parameters","title":"Inserting a customised, time-dependent, input","text":"Let us now go through everything again, but providing some more details. Let us first consider the input parameter. We have previously described how a time-dependent rate can model a circadian rhythm. For real applications, due to e.g. clouds, sunlight is not a perfect sine wave. Here, a common solution is to take real sunlight data from some location and use in the model. Here, we will create synthetic (noisy) data as our light input:\n\nusing Plots\ntend = 120.0\nts = collect(0.0:1.0:tend)\nlight = sin.(ts/6) .+ 1\nlight = [max(0.0, l - rand()) for l in light]\nplot(ts, light; seriestype = :scatter, label = \"Experienced light\")\n\nNow this input is only actually defined at the sample points, making it incompatible with a continuous ODE simulation. To enable this, we will use the DataInterpolations package to create an interpolated version of this data, which forms the actual input:\n\nusing DataInterpolations\ninterpolated_light = LinearInterpolation(light, ts)\nplot(interpolated_light)\n\nWe are now ready to declare our model. We will consider a protein with an active and an inactive form (Pₐ and Pᵢ) where the activation is driven by the presence of sunlight. In this example we we create our model using the programmatic approach. Do note the special syntax we use to declare our input parameter, where we both designate it as a generic function and its type as the type of our interpolated input. Also note that, within the model, we mark the input parameter (light_in) as a function of t.\n\nusing Catalyst\nt = default_t()\nin_type = typeof(interpolated_light)\n@parameters kA kD (light_in::in_type)(..)\n@species Pₐ(t) Pᵢ(t)\nrxs = [\n    Reaction(kA*light_in(t), [Pᵢ], [Pₐ]),\n    Reaction(kD, [Pₐ], [Pᵢ])\n]\n@named rs = ReactionSystem(rxs, t)\nrs = complete(rs)\n\nNow we can simulate our model. Here, we use the interpolated data as the input parameter's value.\n\nusing OrdinaryDiffEqDefault\nu0 = [Pᵢ => 1.0, Pₐ => 0.0]\nps = [kA => 1.5, kD => 1.0, light_in => interpolated_light]\noprob = ODEProblem(rs, u0, tend, ps)\nsol = solve(oprob)\nplot(sol)","category":"section"},{"location":"model_creation/functional_parameters/#functional_parameters_circ_rhythm_dsl","page":"Inputs and time-dependent (or functional) parameters","title":"Interpolating the input into the DSL","text":"It is possible to use time-dependent inputs when creating models through the DSL as well. However, it can still be convenient to declare the input parameter programmatically as above. Next, we can interpolate it into our DSL-declaration (ensuring to also make it a function of t):\n\nrs_dsl = @reaction_network rs begin\n    (kA*$light_in(t), kD), Pᵢ <--> Pₐ\nend\n\nWe can confirm that this model is identical to our programmatic one (and should we wish to, we can simulate it using identical syntax).\n\nrs == rs_dsl","category":"section"},{"location":"model_creation/functional_parameters/#functional_parameters_sir","page":"Inputs and time-dependent (or functional) parameters","title":"Non-time functional parameters","text":"Previously we have demonstrated functional parameters that are functions of time. However, functional parameters can be functions of any variable (however, currently, more than one argument is not supported). Here we will demonstrate this using a SIR model, but instead of having the infection rate scale linearly with the number of infected individuals, we instead assume we have measured data of the infection rate (as dependent on the number of infected individuals) and wish to use this instead. Normally we use the following infection reaction in the SIR model:\n\n@reaction k1, S + I --> 2I\n\nFor ODE models, this would give the same equations as\n\n@reaction k1*I, S --> I\n\nDue to performance reasons (especially for jump simulations) the former approach is strongly encouraged. Here, however, we will assume that we have measured real data of how the number of infected individuals affects the infection rate, and wish to use this in our model, i.e. something like this:\n\n@reaction k1*i_rate(I), S --> I\n\nThis is a case where we can use a functional parameter, whose value depends on the value of I.\n\nWe start by declaring the functional parameter that describes how the infection rate depends on the number of infected individuals. We also plot the measured infection rate, and compare it to the theoretical rate usually used in the SIR model.\n\nusing DataInterpolations, Plots\nI_grid = collect(0.0:5.0:100.0)\nI_measured = 300.0 *(0.8*rand(length(I_grid)) .+ 0.6) .* I_grid ./ (300 .+ I_grid)\nI_rate = LinearInterpolation(I_measured, I_grid)\nplot(I_rate; label = \"Measured infection rate\")\nplot!(I_grid, I_grid; label = \"Normal SIR infection rate\")\n\nNext, we create our model (using the DSL approach).\n\nusing Catalyst\n@parameters (inf_rate::typeof(I_rate))(..)\nsir = @reaction_network rs begin\n    k1*$inf_rate(I), S --> I\n    k2, I --> R\nend\nnothing # hide\n\nFinally, we can simulate our model.\n\nusing OrdinaryDiffEqDefault\nu0 = [:S => 99.0, :I => 1.0, :R => 0.0]\nps = [:k1 => 0.002, :k2 => 0.01, :inf_rate => I_rate]\noprob = ODEProblem(sir, u0, 250.0, ps)\nsol = solve(oprob)\nplot(sol)\n\nnote: Note\nWhen declaring a functional parameter of time, it is easy to set its grid values (i.e. ensure they range from the first to the final time point). For Functional parameters that depend on species concentrations it is trickier, and one must make sure that any potential input-species values that can occur during the simulation are represented in the interpolation. ","category":"section"},{"location":"model_creation/functional_parameters/#functional_parameters_interpolation_algs","page":"Inputs and time-dependent (or functional) parameters","title":"Using different data interpolation approaches","text":"Up until now we have used linear interpolation of our data. However, DataInterpolations.jl supports other interpolation methods. To demonstrate these we here generate a data set, and then show the linear, cubic, and constant interpolations:\n\nusing DataInterpolations, Plots\nxs = collect(0.0:1.0:10.0)\nys = xs ./ (5*rand(length(xs)) .+ xs)\nspline_linear = LinearInterpolation(ys, xs)\nspline_cubuc = CubicSpline(ys, xs)\nspline_const = ConstantInterpolation(ys, xs)\nplot(spline_linear)\nplot!(spline_cubuc)\nplot!(spline_const)\n\nFinally, DataInterpolations.jl also allows various extrapolation methods.","category":"section"},{"location":"inverse_problems/global_sensitivity_analysis/#global_sensitivity_analysis","page":"Global Sensitivity Analysis","title":"Global Sensitivity Analysis","text":"<details><summary><strong>Environment setup and package installation</strong></summary>\n\nThe following code sets up an environment for running the code on this page.\n\nusing Pkg\nPkg.activate(; temp = true) # Creates a temporary environment, which is deleted when the Julia session ends.\nPkg.add(\"Catalyst\")\nPkg.add(\"GlobalSensitivity\")\nPkg.add(\"Plots\")\nPkg.add(\"OrdinaryDiffEqDefault\")\n\n</details>\n\n<details><summary><strong>Quick-start example</strong></summary>\n\nThe following code provides a brief example of how to perform global sensitivity analysis using the GlobalSensitivity.jl package.\n\nusing Catalyst, GlobalSensitivity, OrdinaryDiffEqDefault, SymbolicIndexingInterface\n\n# Designates a model, parameter set, and set of initial conditions.\n# For this infectious disease model we will determine the peak number of cases's sensitivity to the parameters.\nsir = @reaction_network begin\n    β, S + I --> 2I\n    γ, I --> R\nend\nu0 = [:S => 999.0, :I => 1.0, :R => 0.0]\n\n# Declare required structures (a base ODEProblem and a \"p_setter\" which updates the parameter sets in a performant manner).\noprob_base = ODEProblem(sir, u0, (0.0, 10000.0), [:β => 1.0, :γ => 1.0]) # Parameter valued do not matter here.\np_setter = setp_oop(oprob_base, [:β, :γ])\n\n# Creates a function that simulates the model an returns the peak number of cases.\nfunction peak_cases(p)\n    # Updates the ODEProblem with teh proposed parameter set.\n    p = p_setter(oprob_base, p)\n    oprob = remake(oprob_base; p)\n    sol = solve(oprob; maxiters = 100000, verbose = false)\n    return maximum(sol[:I])\nend\n\n# Performs a Morris sensitivity analysis on the designated domain.\nglobal_sens = gsa(peak_cases, Morris(), [(0.001, 10.0), (0.001, 10.0)])\n\n</details>\n\n\n  \n   Global sensitivity analysis (GSA) is used to study the sensitivity of a function's outputs with respect to its input[1]. Within the context of chemical reaction network modelling it is primarily used for two purposes:\n\nWhen fitting a model's parameters to data, it can be applied to the cost function of the optimisation problem. Here, GSA helps determine which parameters do, and do not, affect the model's fit to the data. This can be used to identify parameters that are less relevant to the observed data.\nWhen measuring some system behaviour or property, it can help determine which parameters influence that property. E.g. for a model of a biofuel-producing circuit in a synthetic organism, GSA could determine which system parameters have the largest impact on the total rate of biofuel production.\n\nGSA can be carried out using the GlobalSensitivity.jl package. This tutorial contains a brief introduction of how to use it for GSA on Catalyst models, with GlobalSensitivity providing a more complete documentation.","category":"section"},{"location":"inverse_problems/global_sensitivity_analysis/#global_sensitivity_analysis_global_vs_local_sensitivity","page":"Global Sensitivity Analysis","title":"Global vs local sensitivity","text":"A related concept to global sensitivity is local sensitivity. This, rather than measuring a function's sensitivity (with regards to its inputs) across its entire (or large part of its) domain, measures it at a specific point. This is equivalent to computing the function's gradients at a specific point in phase space, which is an important routine for most gradient-based optimisation methods (typically carried out through automatic differentiation). For most Catalyst-related functionalities, local sensitivities are computed using the SciMLSensitivity.jl package. While certain GSA methods can utilise local sensitivities, this is not necessarily the case.\n\nWhile local sensitivities are primarily used as a subroutine of other methodologies (such as optimisation schemes), it also has direct uses. E.g., in the context of fitting parameters to data, local sensitivity analysis can be used to, at the parameter set of the optimal fit, determine the cost function's sensitivity to the system parameters.","category":"section"},{"location":"inverse_problems/global_sensitivity_analysis/#global_sensitivity_analysis_basic_example","page":"Global Sensitivity Analysis","title":"Basic example","text":"We will consider a simple SEIR model of an infectious disease. This is an expansion of the classic SIR model with an additional exposed state, E, denoting individuals who are latently infected but currently unable to transmit their infection to others.\n\nusing Catalyst\nseir_model = @reaction_network begin\n    10^β, S + I --> E + I\n    10^a, E --> I\n    10^γ, I --> R\nend\n\nWe will study the peak number of infected cases's (max(I(t))) sensitivity to the system's three parameters. We create a function which simulates the system from a given initial condition and measures this property:\n\nusing OrdinaryDiffEqDefault\n\nu0 = [:S => 999.0, :I => 1.0, :E => 0.0, :R => 0.0]\np_dummy = [:β => 0.0, :a => 0.0, :γ => 0.0]\noprob_base = ODEProblem(seir_model, u0, (0.0, 10000.0), p_dummy)\n\nfunction peak_cases(p)\n    ps = [:β => p[1], :a => p[2], :γ => p[3]]\n    oprob = remake(oprob_base; p = ps)\n    sol = solve(oprob; maxiters = 100000, verbose = false)\n    SciMLBase.successful_retcode(sol) || return Inf\n    return maximum(sol[:I])\nend\nnothing # hide\n\nNow, GSA can be applied to our peak_cases function using GlobalSensitivity's gsa function. It takes 3 mandatory inputs:\n\nThe function for which we wish to carry out GSA.\nA method with which we wish to carry out GSA.\nA domain on which we carry out GSA. This is defined by a vector, which contains one two-valued Tuple for each parameter. These Tuples contain a lower and an upper bound for their respective parameter's value.\n\nE.g., here we carry out GSA using Morris's method:\n\nusing GlobalSensitivity\nglobal_sens = gsa(peak_cases, Morris(), [(-3.0,-1.0), (-2.0,0.0), (-2.0,0.0)])\nnothing # hide\n\non the domain 10^β  (-30-10), 10^a  (-2000), 10^γ  (-2000) (which corresponds to β  (000101), a  (00110), γ  (00110)). The output of gsa varies depending on which GSA approach is used. GlobalSensitivity implements a range of methods for GSA. Below, we will describe the most common ones, as well as how to apply them and interpret their outputs.\n\nnote: Note\nWe should make a couple of notes about the example above:Here, we write our parameters on the forms 10^β, 10^a, and 10^γ, which transforms them into log-space. As previously described, this is advantageous in the context of inverse problems such as this one.\nFor GSA, where a function is evaluated a large number of times, it is ideal to write it as performant as possible. Hence, we initially create a base ODEProblem, and then apply the remake function to it in each evaluation of peak_cases to generate a problem which is solved for that specific parameter set.\nAgain, as previously described in other inverse problem tutorials, when exploring a function over large parameter spaces, we will likely simulate our model for unsuitable parameter sets. To reduce time spent on these, and to avoid excessive warning messages, we provide the maxiters = 100000 and verbose = false arguments to solve.\nAs we have encountered in a few other cases, the gsa function is not able to take parameter inputs of the map form usually used for Catalyst. Hence, as a first step in peak_cases we convert the parameter vector to this form. Next, we remember that the order of the parameters when we e.g. evaluate the GSA output, or set the parameter bounds, corresponds to the order used in ps = [:β => p[1], :a => p[2], :γ => p[3]].","category":"section"},{"location":"inverse_problems/global_sensitivity_analysis/#global_sensitivity_analysis_sobol","page":"Global Sensitivity Analysis","title":"Sobol's method-based global sensitivity analysis","text":"The most common method for GSA is Sobol's method. This can be carried out using:\n\nglobal_sens = gsa(peak_cases, Sobol(), [(-3.0,-1.0), (-2.0,0.0), (-2.0,0.0)]; samples = 500)\nnothing # hide\n\nNote: when Sobol() is used as the method, the samples argument must also be used.\n\nSobol's method computes so-called Sobol indices, each measuring some combination of input's effect on the output. Here, when Sobol() is used, the first order, second order, and total order Sobol indices are computed. These can be accessed through the following fields:\n\nglobal_sens.S1: A vector where the i'th element is the output's sensitivity to variations in the i'th input.\nglobal_sens.S2: A matrix where element i-j contains the output's sensitivity to simultaneous variations in the i'th and j'th inputs.\nglobal_sens.ST: A vector where the i'th element is the output's sensitivity to any simultaneous variation of any combination of inputs that contain the i'th input. While only the first and second-order (and the total) Sobol indices are computed, the total order index compounds the information contained in Sobol indices across all orders.\n\nWe can plot the first-order Sobol indices to analyse their content:\n\nusing Plots\nbar([\"β\", \"a\", \"γ\"], global_sens.S1; group = [\"β\", \"a\", \"γ\"], fillrange = 1e-3)\n\nHere, we see that β has a relatively low effect on the peak in infected cases, as compared to a and γ. Plotting the total order indices suggests the same:\n\nbar([\"β\", \"a\", \"γ\"], global_sens.ST; group = [\"β\", \"a\", \"γ\"], fillrange = 1e-3)\n\nGlobalSensitivity implements several versions of Sobol's method, and also provides several options. These are described here. Specifically, it is often recommended to, due to its quick computation time, use the related extended Fourier amplitude sensitivity test (EFAST) version. We can run this using:\n\nglobal_sens = gsa(peak_cases, eFAST(), [(-3.0,-1.0), (-2.0,0.0), (-2.0,0.0)]; samples = 500)\nnothing # hide\n\nIt should be noted that when EFAST is used, only the first and total-order Sobol indices are computed (and not the second-order ones).","category":"section"},{"location":"inverse_problems/global_sensitivity_analysis/#global_sensitivity_analysis_morris","page":"Global Sensitivity Analysis","title":"Morris's method-based global sensitivity analysis","text":"An alternative to using Sobol's method is to use Morris's method. The syntax is similar to previously (however, the samples argument is no longer required):\n\nglobal_sens = gsa(peak_cases, Morris(), [(-3.0,-1.0), (-2.0,0.0), (-2.0,0.0)])\nnothing # hide\n\nMorris's method computes, for parameter samples across parameter space, their elementary effect on the output. Next, the output's sensitivity with respect to each parameter is assessed through various statistics on these elementary effects. In practice, the following two fields are considered:\n\nglobal_sens.means_star (called μ*): Measures each parameter's influence on the output. A large μ* indicates a parameter to which the output is sensitive. \nglobal_sens.variances: Measures the variance of each parameter's influence on the output. A large variance suggests that a parameter's influence on the output is highly dependent on other parameter values.\n\nWe can check these values for our example:\n\nmean_star_plot = bar([\"β\" \"a\" \"γ\"], global_sens.means_star; labels=[\"β\" \"a\" \"γ\"], title=\"μ*\")\nvariances_plot = bar([\"β\" \"a\" \"γ\"], global_sens.variances; labels=[\"β\" \"a\" \"γ\"], title=\"σ²\")\nplot(mean_star_plot, variances_plot)\n\nAs previously, we note that the peak number of infected cases is more sensitive to a and γ than to β.\n\nnote: Note\nThe syntax for plotting the output using Sobol's and Morris's methods is slightly different. The reason is that global_sens.means_star and global_sens.variances (for Morris's method) are 1x3 Matrices, while for Sobol's method, global_sens.S1 and global_sens.ST are length-3 vectors.\n\nGenerally, Morris's method is computationally less intensive, and has easier to interpret output, as compared to Sobol's method. However, if computational resources are available, Sobol's method is more comprehensive.","category":"section"},{"location":"inverse_problems/global_sensitivity_analysis/#global_sensitivity_analysis_other_methods","page":"Global Sensitivity Analysis","title":"Other global sensitivity analysis methods","text":"GlobalSensitivity also implements additional methods for GSA, more details on these can be found in the package's documentation.","category":"section"},{"location":"inverse_problems/global_sensitivity_analysis/#global_sensitivity_analysis_nonscalars","page":"Global Sensitivity Analysis","title":"Global sensitivity analysis for non-scalar outputs","text":"Previously, we have demonstrated GSA on functions with scalar outputs. However, it is also possible to apply it to functions with vector outputs. Let us consider our previous function, but where it provides both the peak number of exposed and infected individuals:\n\nfunction peak_cases_2(p)\n    ps = [:β => p[1], :a => p[2], :γ => p[3]]\n    oprob = remake(oprob_base; p = ps)\n    sol = solve(oprob; maxiters = 100000, verbose = false)\n    SciMLBase.successful_retcode(sol) || return Inf\n    return [maximum(sol[:E]), maximum(sol[:I])]\nend\nnothing # hide\n\nWe can apply gsa to this function as previously:\n\nglobal_sens = gsa(peak_cases_2, Morris(), [(-3.0,-1.0), (-2.0,0.0), (-2.0,0.0)])\nnothing # hide\n\nhowever, each output field is now a multi-row matrix, containing one row for each of the outputs. E.g., we have\n\nglobal_sens.means_star\n\nHere, the function's sensitivity is evaluated with respect to each output independently. Hence, GSA on peak_cases_2 is equivalent to first carrying out GSA on a function returning the peak number of exposed individuals, and then on one returning the peak number of infected individuals.\n\n","category":"section"},{"location":"inverse_problems/global_sensitivity_analysis/#global_sensitivity_analysis_citations","page":"Global Sensitivity Analysis","title":"Citations","text":"If you use this functionality in your research, in addition to Catalyst, please cite the following paper to support the authors of the GlobalSensitivity.jl package:\n\n@article{dixit2022globalsensitivity,\n  title={GlobalSensitivity. jl: Performant and Parallel Global Sensitivity Analysis with Julia},\n  author={Dixit, Vaibhav Kumar and Rackauckas, Christopher},\n  journal={Journal of Open Source Software},\n  volume={7},\n  number={76},\n  pages={4561},\n  year={2022}\n}\n\n","category":"section"},{"location":"inverse_problems/global_sensitivity_analysis/#References","page":"Global Sensitivity Analysis","title":"References","text":"[1]: Saltelli, A et al. Global Sensitivity Analysis. The Primer, Wiley (2008).","category":"section"},{"location":"introduction_to_catalyst/introduction_to_catalyst/#introduction_to_catalyst","page":"Introduction to Catalyst","title":"Introduction to Catalyst","text":"In this tutorial we provide an introduction to using Catalyst to specify chemical reaction networks, and then to solve ODE, jump, and SDE models generated from them [1]. At the end we show what mathematical rate laws and transition rate functions (i.e. intensities or propensities) are generated by Catalyst for ODE, SDE and jump process models. The Mathematical Models Catalyst can Generate documentation illustrates the abstract mathematical models Catalyst reaction models can be converted to, but please note it assumes one has already read this tutorial as a prerequisite.\n\nWe begin by installing Catalyst and any needed packages into a new environment. This step can be skipped if you have already installed them in your current, active environment:\n\nusing Pkg\n\n# name of the environment\nPkg.activate(\"catalyst_introduction\")\n\n# packages we will use in this tutorial\nPkg.add(\"Catalyst\")\nPkg.add(\"OrdinaryDiffEqTsit5\")\nPkg.add(\"Plots\")\nPkg.add(\"Latexify\")\nPkg.add(\"JumpProcesses\")\nPkg.add(\"StochasticDiffEq\")\n\nWe next load the basic packages we'll need for our first example:\n\nusing Catalyst, OrdinaryDiffEqTsit5, Plots, Latexify\n\nLet's start by using the Catalyst @reaction_network macro to specify a simple chemical reaction network: the well-known repressilator. We first construct the reaction network. The basic types of arrows and predefined rate laws one can use are discussed in detail within the tutorial, The Reaction DSL. Here, we use a mix of first order, zero order, and repressive Hill function rate laws. Note, varnothing corresponds to the empty state, and is used for zeroth order production and first order degradation reactions:\n\nrn = @reaction_network Repressilator begin\n    hillr(P₃,α,K,n), ∅ --> m₁\n    hillr(P₁,α,K,n), ∅ --> m₂\n    hillr(P₂,α,K,n), ∅ --> m₃\n    (δ,γ), m₁ <--> ∅\n    (δ,γ), m₂ <--> ∅\n    (δ,γ), m₃ <--> ∅\n    β, m₁ --> m₁ + P₁\n    β, m₂ --> m₂ + P₂\n    β, m₃ --> m₃ + P₃\n    μ, P₁ --> ∅\n    μ, P₂ --> ∅\n    μ, P₃ --> ∅\nend\nshow(stdout, MIME\"text/plain\"(), rn) # hide\n\nshowing that we've created a new network model named Repressilator with the listed chemical species and unknowns. @reaction_network returns a ReactionSystem, which we saved in the rn variable. It can be converted to a variety of other mathematical models represented as ModelingToolkit.AbstractSystems, or analyzed in various ways using the Catalyst.jl API. For example, to see the chemical species, parameters, and reactions we can use\n\nspecies(rn)\n\nparameters(rn)\n\nand\n\nreactions(rn)\n\nWe can also use Latexify to see the corresponding reactions in Latex, which shows what the hillr terms mathematically correspond to\n\nlatexify(rn)\n\nrn #hide\n\nCatalyst also has functionality for visualizing networks using the Makie plotting ecosystem. The relevant packages to load are Catalyst, GraphMakie, NetworkLayout, and a Makie backend such as CairoMakie. Doing so and then using the plot_network function allows us to  visualize the network: \n\nusing Catalyst\nimport CairoMakie, GraphMakie, NetworkLayout\ng = plot_network(rn)\n\nThe network graph shows a variety of information, representing each species as a blue node, and each reaction as an green node. Black arrows from species to reactions indicate reactants, and are labelled with their input stoichiometry. Similarly, black arrows from reactions to species indicate products, and are labelled with their output stoichiometry. In contrast, red arrows from a species to reactions indicate the species is used within the reactions' rate expressions. For the repressilator, the reactions\n\nhillr(P₃,α,K,n), ∅ --> m₁\nhillr(P₁,α,K,n), ∅ --> m₂\nhillr(P₂,α,K,n), ∅ --> m₃\n\nhave rates that depend on the proteins, and hence lead to red arrows from each Pᵢ.\n\nNote, from the REPL or scripts one can always use Makie's save function to save the graph.\n\nsave(\"repressilator_graph.png\", g)","category":"section"},{"location":"introduction_to_catalyst/introduction_to_catalyst/#introduction_to_catalyst_massaction_ode","page":"Introduction to Catalyst","title":"Mass action ODE models","text":"Let's now use our ReactionSystem to generate and solve a corresponding mass action ODE model. We first convert it to a pure ODE model\n\nodesys = ode_model(rn)\nlatexify(rn; form = :ode, math_delimiters = true) # hide\n\n(Here Latexify is used automatically to display odesys in Latex within Markdown documents or notebook environments like Pluto.jl.)\n\nBefore we can solve the ODEs, we need to specify the values of the parameters in the model, the initial condition, and the time interval to solve the model on. To do this we need to build mappings from the symbolic parameters and the species to the corresponding numerical values for parameters and initial conditions. We can build such mappings in several ways. One is to use Julia Symbols to specify the values like\n\npmap  = (:α => .5, :K => 40, :n => 2, :δ => log(2)/120,\n         :γ => 5e-3, :β => log(2)/6, :μ => log(2)/60)\nu₀map = [:m₁ => 0., :m₂ => 0., :m₃ => 0., :P₁ => 20., :P₂ => 0., :P₃ => 0.]\nnothing   # hide\n\nAlternatively, we can use ModelingToolkitBase-based symbolic species variables to specify these mappings like\n\npsymmap  = (rn.α => .5, rn.K => 40, rn.n => 2, rn.δ => log(2)/120,\n    rn.γ => 5e-3, rn.β => 20*log(2)/120, rn.μ => log(2)/60)\nu₀symmap = [rn.m₁ => 0., rn.m₂ => 0., rn.m₃ => 0., rn.P₁ => 20.,\n    rn.P₂ => 0., rn.P₃ => 0.]\nnothing   # hide\n\nKnowing one of the preceding mappings we can set up the ODEProblem we want to solve:\n\n# time interval to solve on\ntspan = (0., 10000.)\n\n# create the ODEProblem we want to solve\noprob = ODEProblem(rn, u₀map, tspan, pmap)\nnothing   # hide\n\nBy passing rn directly to the ODEProblem, Catalyst has to (internally) call ode_model(rn) again to generate the symbolic ODEs. We could instead pass odesys directly like\n\nodesys = complete(odesys)\noprob2 = ODEProblem(odesys, [u₀map; collect(pmap)], tspan)\nnothing   # hide\n\noprob and oprob2 are functionally equivalent, each representing the same underlying problem. Note that when constructing ODEProblems from models created through  ode_model, parameter values are bundled into the initial condition input.\n\nAt this point we are all set to solve the ODEs. We can now use any ODE solver from within the OrdinaryDiffEq.jl package. We'll use the recommended default explicit solver, Tsit5(), and then plot the solutions: \n\nsol = solve(oprob, Tsit5(), saveat=10.0)\nplot(sol)\n\nWe see the well-known oscillatory behavior of the repressilator! For more on the choices of ODE solvers, see the OrdinaryDiffEq.jl documentation.\n\nnote: Note\nIn the above example we used odesys = complete(odesys) to mark the System as complete, indicating to Catalyst and ModelingToolkitBase that these models are finalized. This must be done before any system is given as input to a ode_model (or similar) call or some problem type. ReactionSystem models created through the @reaction_network DSL, like rn above, are always marked as complete when generated. Hence complete does not need to be called on them. Symbolically generated ReactionSystems, ReactionSystems generated via the @network_component macro, and any ModelingToolkitBase models generated by ode_model (or similar functions) always needs to be manually marked as complete as we do for odesys above. An expanded description on completeness can be found here.\n\n","category":"section"},{"location":"introduction_to_catalyst/introduction_to_catalyst/#Stochastic-simulation-algorithms-(SSAs)-for-stochastic-chemical-kinetics","page":"Introduction to Catalyst","title":"Stochastic simulation algorithms (SSAs) for stochastic chemical kinetics","text":"Let's now look at a stochastic chemical kinetics model of the repressilator, modeling it with jump processes. Here, we will construct a JumpProcesses JumpProblem that uses Gillespie's Direct method, and then solve it to generate one realization of the jump process:\n\n# imports the JumpProcesses packages\nusing JumpProcesses\n\n# redefine the initial condition to be integer valued\nu₀map = [:m₁ => 0, :m₂ => 0, :m₃ => 0, :P₁ => 20, :P₂ => 0, :P₃ => 0]\n\n# now, we create a JumpProblem, and let a solver be chosen for us automatically\n# in this case Gillespie's Direct Method will be selected\njprob = JumpProblem(rn, u₀map, tspan, pmap)\n\n# now, let's solve and plot the jump process:\nsol = solve(jprob)\nplot(sol)\nCatalyst.PNG(plot(sol; fmt = :png, dpi = 200)) # hide\n\nWe see that oscillations remain, but become much noisier. Note, in constructing the JumpProblem we could have specified any of the SSAs that are part of JumpProcesses instead of letting JumpProcesses auto-select a solver, see the list of SSAs (i.e., constant rate jump aggregators) in the documentation. For example, to choose the SortingDirect method we would instead say\n\njprob = JumpProblem(rn, u₀map, tspan, pmap; aggregator = SortingDirect())\nsol = solve(jprob)\nplot(sol)\nCatalyst.PNG(plot(sol; fmt = :png, dpi = 200)) # hide\n\nCommon questions that arise in using the JumpProcesses SSAs (i.e. Gillespie methods) are collated in the JumpProcesses FAQ.\n\n","category":"section"},{"location":"introduction_to_catalyst/introduction_to_catalyst/#Chemical-Langevin-equation-(CLE)-stochastic-differential-equation-(SDE)-models","page":"Introduction to Catalyst","title":"Chemical Langevin equation (CLE) stochastic differential equation (SDE) models","text":"At an intermediate physical scale between macroscopic ODE models and microscopic stochastic chemical kinetics models lies the CLE, given by a system of SDEs that add to each ODE above a noise term. As the repressilator has species that get very close to zero in size, it is not a good candidate to model with the CLE (where solutions can then go negative and become unphysical). Let's create a simpler reaction network for a birth-death process that will stay non-negative:\n\nbdp = @reaction_network begin\n  c₁, X --> 2X\n  c₂, X --> 0\n  c₃, 0 --> X\nend\np = (:c₁ => 1.0, :c₂ => 2.0, :c₃ => 50.)\nu₀ = [:X => 5.]\ntspan = (0.,4.)\nnothing   # hide\n\nThe corresponding Chemical Langevin Equation SDE is then\n\ndX(t) = left( c_1 Xleft( t right) - c_2 Xleft( t right) + c_3 right) dt + sqrtc_1 X(t) dW_1(t) - sqrtc_2 X(t) dW_2(t) + sqrtc_3 dW_3(t)\n\nwhere each W_i(t) denotes an independent Brownian Motion. We can solve the CLE model by creating an SDEProblem and solving it similarly to what we did for ODEs above:\n\n# imports the StochasticDiffEq package for SDE simulations\nusing StochasticDiffEq\n\n# SDEProblem for CLE\nsprob = SDEProblem(bdp, u₀, tspan, p)\n\n# solve and plot, tstops is used to specify enough points\n# that the plot looks well-resolved\nsol = solve(sprob, LambaEM(), tstops = range(0., step = 4e-3, length = 1001))\nplot(sol)\n\nWe again have complete freedom to select any of the StochasticDiffEq.jl SDE solvers, see the documentation.\n\n","category":"section"},{"location":"introduction_to_catalyst/introduction_to_catalyst/#Specifying-a-complete-model-via-the-DSL","page":"Introduction to Catalyst","title":"Specifying a complete model via the DSL","text":"In the previous examples we specified initial conditions and parameter values via mappings that were constructed after building our ReactionSystem. Catalyst also supports specifying default values for these during ReactionSystem construction. For example, for the last SDE example we could have also built and simulated the complete model using the DSL like\n\nbdp2 = @reaction_network begin\n    @parameters c₁ = 1.0 c₂ = 2.0 c₃ = 50.0\n    @species X(t) = 5.0\n    c₁, X --> 2X\n    c₂, X --> 0\n    c₃, 0 --> X\nend\ntspan = (0., 4.)\nsprob2 = SDEProblem(bdp2, [], tspan)\n\nLet's now simulate both models, starting from the same random number generator seed, and check we get the same solutions\n\nusing Random\nRandom.seed!(1)\nsol = solve(sprob, LambaEM(), tstops = range(0., step = 4e-3, length = 1001))\np1 = plot(sol)\nRandom.seed!(1)\nsol2 = solve(sprob2, LambaEM(), tstops = range(0., step = 4e-3, length = 1001))\np2 = plot(sol2)\nplot(p1, p2, layout = (2,1))\n\nFor details on what information can be specified via the DSL see the The Reaction DSL tutorial.\n\n","category":"section"},{"location":"introduction_to_catalyst/introduction_to_catalyst/#introduction_to_catalyst_ratelaws","page":"Introduction to Catalyst","title":"Reaction rate laws used in simulations","text":"In generating mathematical models from a ReactionSystem, reaction rates are treated as microscopic rates. That is, for a general mass action reaction of the form n_1 S_1 + n_2 S_2 + dots n_M S_M to dots with stoichiometric substrate coefficients n_i_i=1^M and rate constant k, the corresponding ODE and SDE rate laws are taken to be\n\nk prod_i=1^M frac(S_i)^n_in_i\n\nwhile the jump process transition rate (i.e., the propensity or intensity function) is\n\nk prod_i=1^M fracS_i (S_i-1) dots (S_i-n_i+1)n_i\n\nFor example, the rate law of the reaction 2X + 3Y to Z with rate constant k would be\n\nk fracX^22 fracY^33 \n\ngiving the ODE model\n\nbeginalign*\nfracdXdt =  -2 k fracX^22 fracY^33 \nfracdYdt =  -3 k fracX^22 fracY^33 \nfracdZdt = k fracX^22 fracY^33\nendalign*\n\nThis implicit rescaling of rate constants can be disabled through explicit conversion of a ReactionSystem to another system via Base.convert using the combinatoric_ratelaws=false keyword argument, i.e.\n\nrn = @reaction_network ...\node_model(rn; combinatoric_ratelaws=false)\n\nFor the previous example using this keyword argument would give the rate law\n\nk X^2 Y^3\n\nand the ODE model\n\nbeginalign*\nfracdXdt =  -2 k X^2 Y^3 \nfracdYdt =  -3 k X^2 Y^3 \nfracdZdt = k X^2 Y^3\nendalign*\n\nA more detailed summary of the precise mathematical equations Catalyst can generate is available in the Mathematical Models Catalyst can Generate documentation.\n\n","category":"section"},{"location":"introduction_to_catalyst/introduction_to_catalyst/#Notes","page":"Introduction to Catalyst","title":"Notes","text":"For each of the preceding models we converted the ReactionSystem to, i.e., ODEs, jumps, or SDEs, we had two paths for conversion:\na. Convert to the corresponding ModelingToolkit system and then use it in     creating the corresponding problem.\nb. Directly create the desired problem type from the ReactionSystem.\nThe latter is more convenient, however, the former will be more efficient if one needs to repeatedly create the associated Problem.\nModelingToolkit offers many options for optimizing the generated ODEs and SDEs, including options to build functions for evaluating Jacobians and/or multithreaded versions of derivative evaluation functions. See the options for ODEProblems and SDEProblems.\n\n","category":"section"},{"location":"introduction_to_catalyst/introduction_to_catalyst/#References","page":"Introduction to Catalyst","title":"References","text":"Torkel E. Loman, Yingbo Ma, Vasily Ilin, Shashi Gowda, Niklas Korsbo, Nikhil Yewale, Chris Rackauckas, Samuel A. Isaacson, Catalyst: Fast and flexible modeling of reaction networks, PLOS Computational Biology (2023).","category":"section"},{"location":"inverse_problems/optimization_ode_param_fitting/#optimization_parameter_fitting","page":"Parameter Fitting for ODEs using Optimization.jl","title":"Parameter Fitting for ODEs using Optimization.jl","text":"<details><summary><strong>Environment setup and package installation</strong></summary>\n\nThe following code sets up an environment for running the code on this page.\n\nusing Pkg\nPkg.activate(; temp = true) # Creates a temporary environment, which is deleted when the Julia session ends.\nPkg.add(\"Catalyst\")\nPkg.add(\"OptimizationBase\")\nPkg.add(\"OptimizationNLopt\")\nPkg.add(\"OptimizationEvolutionary\")\nPkg.add(\"OptimizationOptimJL\")\nPkg.add(\"OrdinaryDiffEqDefault\")\nPkg.add(\"Plots\")\n\n</details>\n\n<details><summary><strong>Quick-start example</strong></summary>\n\nThe following code provides a brief example of how to perform parameter fitting using the Optimization.jl package.\n\nusing Catalyst, OrdinaryDiffEqDefault, OptimizationBase, OptimizationNLopt, SymbolicIndexingInterface\n\n# What we know: A model, an initial condition, and sampled datapoints.\nrn = @reaction_network begin\n    (p,d), 0 <--> X\nend\nu0 = [:X => 0.1]\nt_samples = 1:5\nX_samples = [1.6, 2.5, 3.1, 3.5, 3.7]\n\n# Declare required structures (a base ODEProblem and a \"p_setter\" which updates the parameter sets in a performant manner).\noprob_base = ODEProblem(rn, u0, (0.0, 5.0), [:p => 1.0, :d => 1.0]) # Parameter valued do not matter here.\np_setter = setp_oop(oprob_base, [:p, :d])\n\n# A loss function measuring the sum-of-square distance between a simulation and the data.\nfunction loss(p, (oprob_base, p_setter, t_samples, X_samples))\n    # Updates the ODEProblem with teh proposed parameter set.\n    p = p_setter(oprob_base, p)\n    oprob = remake(oprob_base; p)\n\n    # Simulate the model. If sucesfull, return sum-of-squares distance as loss.\n    sol = solve(oprob; saveat = t_samples, verbose = false, maxiters = 10000)\n    SciMLBase.successful_retcode(sol) || return Inf\n    return sum(abs2, sol[:X] .- X_samples)\nend\n\n# Creates an OptimizationProblem and solves it.\np_init = [1.0, 1.0] # Initial guess for optimization run.\nopt_prob = OptimizationProblem(loss, p_init, (oprob_base, p_setter, t_samples, X_samples))\nopt_sol = solve(opt_prob, NLopt.LN_NELDERMEAD())\n\n# Simulates the model for the found parameter set.\np = p_setter(oprob_base, opt_sol.u)\noprob_fitted = remake(oprob_base; p)\nsol_fitted = solve(oprob_fitted)\n\n</details>\n\n\n  \n\nFitting parameters to data involves solving an optimisation problem (that is, finding the parameter set that optimally fits your model to your data, typically by minimising an objective function)[1]. The SciML ecosystem's primary package for solving optimisation problems is Optimization.jl. It provides access to a variety of solvers via a single common interface by wrapping a large number of optimisation libraries that have been implemented in Julia.\n\nThis tutorial demonstrates how to \n\nCreate a custom objective function which minimiser corresponds to the parameter set optimally fitting the data.\nUse Optimization.jl to minimize this objective function and find the parameter set providing the optimal fit.\n\nFor simple parameter fitting problems (such as the one outlined below), PEtab.jl often provides a more straightforward parameter fitting interface. However, Optimization.jl provides additional flexibility in defining your objective function. Indeed, it can also be used in other contexts, such as finding parameter sets that maximise the magnitude of some system behaviour. More details on how to use Optimization.jl can be found in its documentation.","category":"section"},{"location":"inverse_problems/optimization_ode_param_fitting/#optimization_parameter_fitting_basics","page":"Parameter Fitting for ODEs using Optimization.jl","title":"Basic example","text":"Let us consider a Michaelis-Menten enzyme kinetics model, where an enzyme (E) converts a substrate (S) into a product (P):\n\nusing Catalyst\nrn = @reaction_network begin\n    kB, S + E --> SE\n    kD, SE --> S + E\n    kP, SE --> P + E\nend\n\nFrom some known initial condition, and a true parameter set (which we later want to recover from the data) we generate synthetic data (on which we will demonstrate the fitting process).\n\n# Define initial conditions and parameters.\nu0 = [:S => 1.0, :E => 1.0, :SE => 0.0, :P => 0.0]\nps_true = [:kB => 1.0, :kD => 0.1, :kP => 0.5]\n\n# Generate synthetic data.\nusing OrdinaryDiffEqDefault\noprob_true = ODEProblem(rn, u0, (0.0, 10.0), ps_true)\ntrue_sol = solve(oprob_true)\ndata_sol = solve(oprob_true; saveat = 1.0)\ndata_ts = data_sol.t[2:end]\ndata_vals = (0.8 .+ 0.4*rand(10)) .* data_sol[:P][2:end]\n\n# Plots the true solutions and the (synthetic) data measurements.\nusing Plots\nplot(true_sol; idxs = :P, label = \"True solution\", lw = 8)\nplot!(data_ts, data_vals; label = \"Measurements\", seriestype=:scatter, ms = 6, color = :blue)\nplt = plot(true_sol; idxs = :P, label = \"True solution\", lw = 8) # hide\nplot!(plt, data_ts, data_vals; label = \"Measurements\", seriestype=:scatter, ms = 6, color = :blue) # hide\nCatalyst.PNG(plot(plt; fmt = :png, dpi = 200)) # hide\n\nNext, we will formulate an objective function which, for a single parameter set, simulates our model and computes the sum-of-square distance between the data and the simulation (non-sum-of-square approaches can be used, but this is the most straightforward one).\n\nps_init = [:kB => 1.0, :kD => 1.0, :kP => 1.0]\noprob_base = ODEProblem(rn, u0, (0.0, 10.0), ps_init)\nfunction objective_function(p, _)\n    p = Pair.([:kB, :kD, :kP], p)\n    oprob = remake(oprob_base; p)\n    sol = solve(oprob; saveat = data_ts, save_idxs = :P, verbose = false, maxiters = 10000)\n    SciMLBase.successful_retcode(sol) || return Inf\n    return sum((sol .- data_vals) .^2)\nend\n\nWhen our optimisation algorithm searches parameter space it will likely consider many highly non-plausible parameter sets. To better handle this we:\n\nAdd maxiters = 10000 to our solve command. As most well-behaved ODEs can be solved in relatively few timesteps, this speeds up the optimisation procedure by preventing us from spending too much time trying to simulate (for the model) unsuitable parameter sets.\nAdd verbose = false to our solve command. This prevents (potentially a very large number of) warnings from being printed to our output as unsuitable parameter sets are simulated.\nAdd the line SciMLBase.successful_retcode(sol) || return Inf, which returns an infinite value for parameter sets which does not lead to successful simulations.\n\nTo improve optimisation performance, rather than creating a new ODEProblem in each iteration, we pre-declare one which we apply remake to. We also use the saveat = data_ts, save_idxs = :P arguments to only save the values of the measured species at the measured time points.\n\nWe can now create an OptimizationProblem using our objective_function and some initial guess of parameter values from which the optimiser will start:\n\nusing OptimizationBase\np_guess = [1.0, 1.0, 1.0]\noptprob = OptimizationProblem(objective_function, p_guess)\nnothing # hide\n\nnote: Note\nOptimizationProblems cannot currently accept parameter values in the form of a map (e.g. [:kB => 1.0, :kD => 1.0, :kP => 1.0]). These must be provided as individual values (using the same order as the parameters occur in in the parameters(rs) vector). This should hopefully be remedied in future Optimization releases.\n\nnote: Note\nEspecially if you check Optimization.jl's documentation, you will note that objective functions have the f(u,p) form. This is because OptimizationProblems (like e.g. ODEProblems) can take both variables (which are varied during the optimisation procedure), but also parameters that are fixed. In our case, the optimisation variables correspond to our model parameters. Hence, our model parameter values (p) are the first argument (u). This is also why we find the optimisation solution (our optimised parameter set) in opt_sol's u field. Our optimisation problem does not actually have any parameters, hence, the second argument of objective_function is unused (that is why we call it _, a name commonly indicating unused function arguments).There are several modifications to our problem where it would actually have parameters. E.g. we might want to run the optimisation where one parameter has a known fixed value. If we then would like to rerun this for alternative fixed values, this value could be encoded as an OptimizationProblem parameter.\n\nFinally, we can solve optprob to find the parameter set that best fits our data. Optimization.jl only provides a few optimisation methods natively. However, for each supported optimisation package, it provides a corresponding wrapper package to import that optimisation package for use with Optimization.jl. E.g., if we wish to use NLopt.jl's Nelder-Mead method, we must install and import the OptimizationNLopt package. A summary of all, by Optimization.jl supported, optimisation packages can be found here. Here, we import the NLopt.jl package and uses it to minimise our objective function (thus finding a parameter set that fits the data):\n\nusing OptimizationNLopt\noptsol = solve(optprob, NLopt.LN_NELDERMEAD())\nnothing # hide\n\nWe can now simulate our model for the found parameter set (stored in optsol.u), checking that it fits our data.\n\noprob_fitted = remake(oprob_base; p = Pair.([:kB, :kD, :kP], optsol.u))\nfitted_sol = solve(oprob_fitted)\nplot!(fitted_sol; idxs = :P, label = \"Fitted solution\", linestyle = :dash, lw = 6, color = :lightblue)\nplot!(plt, fitted_sol; idxs = :P, label = \"Fitted solution\", linestyle = :dash, lw = 6, color = :lightblue) # hide\nCatalyst.PNG(plot(plt; fmt = :png, dpi = 200)) # hide\n\nnote: Note\nHere, a good exercise is to check the resulting parameter set and note that, while it creates a good fit to the data, it does not actually correspond to the original parameter set. Identifiability is a concept that studies how to deal with this problem.<!–NTS: re-add ref when identifiablity works again–>\n\nSay that we instead would like to use a genetic algorithm approach, as implemented by the Evolutionary.jl package. In this case we can run:\n\nusing OptimizationEvolutionary\nsol = solve(optprob, Evolutionary.GA())\nnothing # hide\n\nto solve optprob for this combination of solve and implementation.","category":"section"},{"location":"inverse_problems/optimization_ode_param_fitting/#optimization_parameter_fitting_AD","page":"Parameter Fitting for ODEs using Optimization.jl","title":"Utilising automatic differentiation","text":"Optimisation methods can be divided into differentiation-free and differentiation-based optimisation methods. E.g. consider finding the minimum of the function f(x) = x^2, given some initial guess of x. Here, we can simply compute the differential and descend along it until we find x=0 (admittedly, for this simple problem the minimum can be computed directly). This principle forms the basis of optimisation methods such as gradient descent, which utilises information of a function's differential to minimise it. When attempting to find a global minimum, to avoid getting stuck in local minimums, these methods are often augmented by additional routines. While the differentiation of most algebraic functions is trivial, it turns out that even complicated functions (such as the one we used above) can be differentiated computationally through the use of automatic differentiation (AD).\n\nThrough packages such as ForwardDiff.jl, ReverseDiff.jl, and Zygote.jl, Julia supports AD for most code. Specifically for code including simulation of differential equations, differentiation is supported by SciMLSensitivity.jl. Generally, AD can be used without specific knowledge from the user, however, it requires an additional step in the construction of our OptimizationProblem. Here, we create a specialised OptimizationFunction from our objective function. To it, we will also provide our choice of AD method. There are several alternatives, and in our case we will use AutoForwardDiff() (a good choice for small optimisation problems). We can then create a new OptimizationProblem using our updated objective function:\n\nopt_func = OptimizationFunction(objective_function, AutoForwardDiff())\nopt_prob = OptimizationProblem(opt_func, p_guess)\nnothing # hide\n\nFinally, we can find the optimum using some differentiation-based optimisation methods. Here we will use Optim.jl's Broyden–Fletcher–Goldfarb–Shanno algorithm implementation:\n\nusing OptimizationOptimJL\nopt_sol = solve(opt_prob, OptimizationOptimJL.BFGS())","category":"section"},{"location":"inverse_problems/optimization_ode_param_fitting/#optimization_parameter_fitting_multiple_species","page":"Parameter Fitting for ODEs using Optimization.jl","title":"Optimisation problems with data for multiple species","text":"Imagine that, in our previous example, we had measurements of the concentration of both S and P:\n\ndata_vals_S = (0.8 .+ 0.4*rand(10)) .* data_sol[:S][2:end]\ndata_vals_P = (0.8 .+ 0.4*rand(10)) .* data_sol[:P][2:end]\n\nplot(true_sol; idxs = [:S, :P], label = [\"True S\" \"True P\"], lw = 8)\nplot!(data_ts, data_vals_S; label = \"Measured S\", seriestype=:scatter, ms = 6, color = :blue)\nplot!(data_ts, data_vals_P; label = \"Measured P\", seriestype=:scatter, ms = 6, color = :red)\nplt2 = plot(true_sol; idxs = [:S, :P], label = [\"True S\" \"True P\"], lw = 8)  # hide\nplot!(plt2, data_ts, data_vals_S; label = \"Measured S\", seriestype = :scatter, ms = 6, color = :blue)  # hide\nplot!(plt2, data_ts, data_vals_P; label = \"Measured P\", seriestype = :scatter, ms = 6, color = :red)  # hide\nCatalyst.PNG(plot(plt2; fmt = :png, dpi = 200)) # hide\n\nIn this case we simply modify our objective function to take this into account:\n\nfunction objective_function_S_P(p, _)\n    p = Pair.([:kB, :kD, :kP], p)\n    oprob = remake(oprob_base; p)\n    sol = solve(oprob; saveat = data_ts, save_idxs = [:S, :P], verbose = false, maxiters = 10000)\n    SciMLBase.successful_retcode(sol) || return Inf\n    return sum((sol[:S] .- data_vals_S) .^2 + (sol[:P] .- data_vals_P) .^2)\nend\n\nHere we do not normalise the contribution from each species to the objective function. However, if species are present at different concentration levels this might be necessary (or you might essentially only take the highest concentration species(s) into account).\n\nWe can now fit our model to data and plot the results:\n\noptprob_S_P = OptimizationProblem(objective_function_S_P, p_guess)\noptsol_S_P = solve(optprob_S_P, NLopt.LN_NELDERMEAD())\np = Pair.([:kB, :kD, :kP], optsol_S_P.u)\noprob_fitted_S_P = remake(oprob_base; p)\nfitted_sol_S_P = solve(oprob_fitted_S_P)\nplot!(fitted_sol_S_P; idxs = [:S, :P], label = \"Fitted solution\", linestyle = :dash, lw = 6, color = [:lightblue :pink])\nplot!(plt2, fitted_sol_S_P; idxs = [:S, :P], label = \"Fitted solution\", linestyle = :dash, lw = 6, color = [:lightblue :pink]) # hide\nCatalyst.PNG(plot(plt2; fmt = :png, dpi = 200)) # hide","category":"section"},{"location":"inverse_problems/optimization_ode_param_fitting/#optimization_parameter_fitting_constraints","page":"Parameter Fitting for ODEs using Optimization.jl","title":"Setting parameter constraints and boundaries","text":"Sometimes, it is desirable to set boundaries on parameter values. Indeed, this can speed up the optimisation process (by preventing searching through unfeasible parts of parameter space), and can also be a requirement for some optimisation methods. This can be done by passing the lb (lower bounds) and up (upper bounds) arguments to OptimizationProblem. These are vectors (of the same length as the number of parameters), with each argument corresponding to the boundary value of the parameter with the same index. If we wish to constrain each parameter to the interval (01 100) this can be done through:\n\noptprob = OptimizationProblem(objective_function, [1.0, 1.0, 1.0]; lb = [1e-1, 1e-1, 1e-1], ub = [1e1, 1e1, 1e1])\nnothing # hide\n\nIn addition to boundaries, Optimization.jl also supports setting linear and non-linear constraints on its output solution (only available for some optimisers).","category":"section"},{"location":"inverse_problems/optimization_ode_param_fitting/#optimization_parameter_fitting_known_parameters","page":"Parameter Fitting for ODEs using Optimization.jl","title":"Parameter fitting with known parameters","text":"If we from previous knowledge know that kD = 01, and only want to fit the values of kB and kP, this can be achieved by making corresponding changes to our objective function.\n\nfunction objective_function_known_kD(p, _)\n    p = Pair.([:kB, :kD, :kP], [p[1], 0.1, p[2]])\n    oprob = remake(oprob_base; p)\n    sol = solve(oprob; saveat = data_ts, save_idxs = :P, verbose = false, maxiters = 10000)\n    SciMLBase.successful_retcode(sol) || return Inf\n    return sum((sol .- data_vals) .^2)\nend\n\nWe can now create and solve the corresponding OptimizationProblem, but with only two parameters in the initial guess.\n\noptprob_known_kD = OptimizationProblem(objective_function_known_kD, [1.0, 1.0])\noptsol_known_kD = solve(optprob_known_kD, NLopt.LN_NELDERMEAD())\nnothing # hide","category":"section"},{"location":"inverse_problems/optimization_ode_param_fitting/#optimization_parameter_fitting_solver_options","page":"Parameter Fitting for ODEs using Optimization.jl","title":"Optimisation solver options","text":"Optimization.jl supports various optimisation solver options that can be supplied to the solve command. For example, to set a maximum number of seconds (after which the optimisation process is terminated), you can use the maxtime argument:\n\noptsol_fixed_kD = solve(optprob, NLopt.LN_NELDERMEAD(); maxtime = 100)\nnothing # hide\n\nIt should be noted that not all solver options are available to all optimisation solvers.","category":"section"},{"location":"inverse_problems/optimization_ode_param_fitting/#optimization_parameter_fitting_log_scale","page":"Parameter Fitting for ODEs using Optimization.jl","title":"Fitting parameters on the logarithmic scale","text":"Often it can be advantageous to fit parameters on a logarithmic scale (rather than on a linear scale). The most straightforward way to do this is to simply replace each parameter in the model definition by its logarithmic version:\n\nusing Catalyst\nrn = @reaction_network begin\n    10^kB, S + E --> SE\n    10^kD, SE --> S + E\n    10^kP, SE --> P + E\nend\n\nAnd then going forward, by keeping in mind that parameter values are logarithmic. Here, setting\n\np_true = [:kB => 0.0, :kD => -1.0, :kP => 10^(0.5)]\nnothing # hide\n\ncorresponds to the same true parameter values as used previously ([:kB => 1.0, :kD => 0.1, :kP => 0.5]). Alternatively, we can provide the log-transform in the objective function:\n\nfunction objective_function_logtransformed(p, _)\n    p = Pair.([:kB, :kD, :kP], 10.0 .^ p)\n    oprob = remake(oprob_base; p)\n    sol = solve(oprob; saveat = data_ts, save_idxs = :P, verbose = false, maxiters = 10000)\n    SciMLBase.successful_retcode(sol) || return Inf\n    return sum((sol .- data_vals) .^2)\nend\n\n","category":"section"},{"location":"inverse_problems/optimization_ode_param_fitting/#optimization_parameter_fitting_citation","page":"Parameter Fitting for ODEs using Optimization.jl","title":"Citation","text":"If you use this functionality in your research, please cite the following paper to support the authors of the Optimization.jl package:\n\n@software{vaibhav_kumar_dixit_2023_7738525,\n\tauthor = {Vaibhav Kumar Dixit and Christopher Rackauckas},\n\tmonth = mar,\n\tpublisher = {Zenodo},\n\ttitle = {Optimization.jl: A Unified Optimization Package},\n\tversion = {v3.12.1},\n\tdoi = {10.5281/zenodo.7738525},\n  \turl = {https://doi.org/10.5281/zenodo.7738525},\n\tyear = 2023\n}\n\n","category":"section"},{"location":"inverse_problems/optimization_ode_param_fitting/#References","page":"Parameter Fitting for ODEs using Optimization.jl","title":"References","text":"[1]: Alejandro F. Villaverde, Dilan Pathirana, Fabian Fröhlich, Jan Hasenauer, Julio R. Banga, A protocol for dynamic model calibration, Briefings in Bioinformatics (2023).","category":"section"},{"location":"steady_state_functionality/steady_state_stability_computation/#steady_state_stability","page":"Steady state stability computation","title":"Steady state stability computation","text":"<details><summary><strong>Environment setup and package installation</strong></summary>\n\nThe following code sets up an environment for running the code on this page.\n\nusing Pkg\nPkg.activate(; temp = true) # Creates a temporary environment, which is deleted when the Julia session ends.\nPkg.add(\"Catalyst\")\nPkg.add(\"HomotopyContinuation\")\n\n</details>\n\n<details><summary><strong>Quick-start example</strong></summary>\n\nThe following code provides a brief example of how a steady state's stability can be determine using the steady_state_stability function.\n\nusing Catalyst\nrn = @reaction_network begin \n    (p,d), 0 <--> X\nend\nps = [:p => 2.0, :d => 0.5]\nsteady_state = [:X => 4.0] # This must be a true steady state of the system.\nsteady_state_stability(steady_state, rn, ps)\n\n</details>\n\n\n  \n\nAfter system steady states have been found using HomotopyContinuation.jl, NonlinearSolve.jl, or other means, their stability can be computed using Catalyst's steady_state_stability function. Systems with conservation laws will automatically have these removed, permitting stability computation on systems with singular Jacobian.\n\nwarning: Warning\nCatalyst currently computes steady state stabilities using the naive approach of checking whether a system's largest eigenvalue real part is negative. Furthermore, Catalyst uses a tolerance tol = 10*sqrt(eps()) to determine whether a computed eigenvalue is far away enough from 0 to be reliably considered non-zero. This threshold can be changed through the tol keyword argument.","category":"section"},{"location":"steady_state_functionality/steady_state_stability_computation/#steady_state_stability_basics","page":"Steady state stability computation","title":"Basic examples","text":"Let us consider the following basic example:\n\nusing Catalyst\nrn = @reaction_network begin \n    (p,d), 0 <--> X\nend\n\nIt has a single (stable) steady state at X = pd. We can confirm stability using the steady_state_stability function, to which we provide the steady state, the reaction system, and the parameter values:\n\nps = [:p => 2.0, :d => 0.5]\nsteady_state = [:X => 4.0]\nsteady_state_stability(steady_state, rn, ps)\n\nNext, let us consider the following self-activation loop:\n\nsa_loop = @reaction_network begin \n    (hill(X,v,K,n),d), 0 <--> X\nend\n\nFor certain parameter choices, this system exhibits multi-stability. Here, we can find the steady states using homotopy continuation:\n\nimport HomotopyContinuation\nps = [:v => 2.0, :K => 0.5, :n => 3, :d => 1.0]\nsteady_states = hc_steady_states(sa_loop, ps)\n\nNext, we can apply steady_state_stability to each steady state yielding a vector of their stabilities:\n\n[steady_state_stability(sstate, sa_loop, ps) for sstate in steady_states]\n\nFinally, as described above, Catalyst uses an optional argument, tol, to determine how strict to make the stability check.  I.e. below we set the tolerance to 1e-6 (a larger value, that is stricter, than the default of 10*sqrt(eps()))\n\n[steady_state_stability(sstate, sa_loop, ps; tol = 1e-6) for sstate in steady_states]\nnothing# hide","category":"section"},{"location":"steady_state_functionality/steady_state_stability_computation/#steady_state_stability_jacobian","page":"Steady state stability computation","title":"Pre-computing the Jacobian to increase performance when computing stability for many steady states","text":"Catalyst uses the system Jacobian to compute steady state stability, and the Jacobian is computed once for each call to steady_state_stability. If you repeatedly compute stability for steady states of the same system, pre-computing the Jacobian and supplying it to the steady_state_stability function can improve performance. \n\nIn this example we use the self-activation loop from previously, pre-computes its Jacobian, and uses it to multiple steady_state_stability calls:\n\nss_jac = steady_state_jac(sa_loop)\n\nps_1 = [:v => 2.0, :K => 0.5, :n => 3, :d => 1.0]\nsteady_states_1 = hc_steady_states(sa_loop, ps)\nstabs_1 = [steady_state_stability(st, sa_loop, ps_1; ss_jac) for st in steady_states_1]\n\nps_2 = [:v => 4.0, :K => 1.5, :n => 2, :d => 1.0]\nsteady_states_2 = hc_steady_states(sa_loop, ps)\nstabs_2 = [steady_state_stability(st, sa_loop, ps_2; ss_jac) for st in steady_states_2]\nnothing # hide\n\nwarning: Warning\nFor systems with conservation laws, steady_state_jac must be supplied a u0 vector (indicating species concentrations for conservation law computation). This is required to eliminate the conserved quantities, preventing a singular Jacobian. These are supplied using the u0 optional argument.","category":"section"},{"location":"model_creation/model_file_loading_and_export/#model_file_import_export","page":"Loading Chemical Reaction Network Models from Files","title":"Loading Chemical Reaction Network Models from Files","text":"<details><summary><strong>Environment setup and package installation</strong></summary>\n\nThe following code sets up an environment for running the code on this page.\n\nusing Pkg\nPkg.activate(; temp = true) # Creates a temporary environment, which is deleted when the Julia session ends.\nPkg.add(\"Catalyst\")\nPkg.add(\"OrdinaryDiffEqDefault\")\nPkg.add(\"Plots\")\nPkg.add(\"ReactionNetworkImporters\")\nPkg.add(\"SBMLImporter\")\n\n</details>\n\n\n   Catalyst stores chemical reaction network (CRN) models in ReactionSystem structures. This tutorial describes how to load such ReactionSystems from, and save them to, files. This can be used to save models between Julia sessions, or transfer them from one session to another. Furthermore, to facilitate the computation modelling of CRNs, several standardised file formats have been created to represent CRN models (e.g. SBML). This enables CRN models to be shared between different software and programming languages. While Catalyst itself does not have the functionality for loading such files, we will here (briefly) introduce a few packages that can load different file types to Catalyst ReactionSystems.","category":"section"},{"location":"model_creation/model_file_loading_and_export/#model_file_import_export_crn_serialization","page":"Loading Chemical Reaction Network Models from Files","title":"Saving Catalyst models to, and loading them from, Julia files","text":"Catalyst provides a save_reactionsystem function, enabling the user to save a ReactionSystem to a file. Here we demonstrate this by first creating a simple cross-coupling model:\n\nusing Catalyst\ncc_system = @reaction_network begin\n    k₁, S₁ + C --> S₁C\n    k₂, S₁C + S₂ --> CP\n    k₃, CP --> C + P\nend\n\nand next saving it to a file\n\nsave_reactionsystem(\"cross_coupling.jl\", cc_system)\n\nHere, save_reactionsystem's first argument is the path to the file where we wish to save it. The second argument is the ReactionSystem we wish to save. To load the file, we use Julia's include function:\n\ncc_loaded = include(\"cross_coupling.jl\")\nrm(\"cross_coupling.jl\") # hide\ncc_loaded # hide\n\nnote: Note\nThe destination file can be in a folder. E.g. save_reactionsystem(\"my\\_folder/reaction_network.jl\", rn) saves the model to the file \"reaction_network.jl\" in the folder \"my_folder\".\n\nHere, include is used to execute the Julia code from any file. This means that save_reactionsystem actually saves the model as executable code which re-generates the exact model which was saved (this is the reason why we use the \".jl\" extension for the saved file). Indeed, we can confirm this if we check what is printed in the file:\n\nlet\n\n# Independent variable:\n@parameters t\n\n# Parameters:\nps = @parameters kB kD kP\n\n# Species:\nsps = @species S(t) E(t) SE(t) P(t)\n\n# Reactions:\nrxs = [\n    Reaction(kB, [S, E], [SE], [1, 1], [1]),\n    Reaction(kD, [SE], [S, E], [1], [1, 1]),\n    Reaction(kP, [SE], [P, E], [1], [1, 1])\n]\n\n# Declares ReactionSystem model:\nrs = ReactionSystem(rxs, t, sps, ps; name = Symbol(\"##ReactionSystem#12592\"))\ncomplete(rs)\n\nend\n\nnote: Note\nThe code that save_reactionsystem prints uses programmatic modelling to generate the written model.\n\nIn addition to transferring models between Julia sessions, the save_reactionsystem function can also be used or print a model to a text file where you can easily inspect its components.","category":"section"},{"location":"model_creation/model_file_loading_and_export/#model_file_import_export_julia_serialisation","page":"Loading Chemical Reaction Network Models from Files","title":"Loading and saving arbitrary Julia variables using Serialization.jl","text":"Julia provides a general and lightweight interface for loading and saving Julia structures to and from files that it can be good to be aware of. It is called Serialization.jl and provides two functions, serialize and deserialize. The first allows us to write a Julia structure to a file. E.g. if we wish to save a parameter set associated with our model, we can use\n\nusing Serialization\nps = [:k₁ => 1.0, :k₂ => 0.1, :k₃ => 2.0]\nserialize(\"saved_parameters.jls\", ps)\n\nHere, we use the extension \".jls\" (standing for JuLia Serialization), however, any extension code can be used. To load a structure, we can then use\n\nloaded_sol = deserialize(\"saved_parameters.jls\")\nrm(\"saved_parameters.jls\") # hide\nloaded_sol # hide","category":"section"},{"location":"model_creation/model_file_loading_and_export/#model_file_import_export_sbml_rni_net","page":"Loading Chemical Reaction Network Models from Files","title":"Loading .net files using ReactionNetworkImporters.jl","text":"A general-purpose format for storing CRN models is so-called .net files. These can be generated by e.g. BioNetGen. The ReactionNetworkImporters.jl package enables the loading of such files to Catalyst ReactionSystem. Here we load a Repressilator model stored in the \"repressilator.net\" file:\n\nusing ReactionNetworkImporters\nprn = loadrxnetwork(BNGNetwork(), \"repressilator.net\")\n\nHere, .net files not only contain information regarding the reaction network itself, but also the numeric values (initial conditions and parameter values) required for simulating it. Hence, loadrxnetwork generates a ParsedReactionNetwork structure, containing all this information. You can access the model as prn.rn, the initial conditions as prn.u0, and the parameter values as prn.p. Furthermore, these initial conditions and parameter values are also made default values of the model.\n\nA parsed reaction network's content can then be provided to various problem types for simulation. E.g. here we perform an ODE simulation of our repressilator model:\n\nusing Catalyst, OrdinaryDiffEqDefault, Plots\ntspan = (0.0, 10000.0)\noprob = ODEProblem(prn.rn, Float64[], tspan, Float64[])\nsol = solve(oprob)\nplot(sol; idxs = [:mTetR, :mLacI, :mCI])\n\n(Image: Repressilator Simulation)\n\nNote that, as all initial conditions and parameters have default values, we can provide empty vectors for these into our ODEProblem.\n\nnote: Note\nIt should be noted that .net files support a wide range of potential model features, not all of which are currently supported by ReactionNetworkImporters. Hence, there might be some .net files which loadrxnetwork will not be able to load.\n\nA more detailed description of ReactionNetworkImporter's features can be found in its documentation.","category":"section"},{"location":"model_creation/model_file_loading_and_export/#model_file_import_export_sbml","page":"Loading Chemical Reaction Network Models from Files","title":"Loading SBML files using SBMLImporter.jl and SBMLToolkit.jl","text":"The Systems Biology Markup Language (SBML) is the most widespread format for representing CRN models. Currently, there exist two different Julia packages, SBMLImporter.jl and SBMLToolkit.jl, that are able to load SBML files to Catalyst ReactionSystem structures. SBML is able to represent a very wide range of model features, with both packages supporting most features. However, there exist SBML files (typically containing obscure model features such as events with time delays) that currently cannot be loaded into Catalyst models.\n\nSBMLImporter's load_SBML function can be used to load SBML files. Here, we load a Brusselator model stored in the \"brusselator.xml\" file:\n\nusing SBMLImporter\nprn, cbs = load_SBML(\"brusselator.xml\", massaction = true)\n\nHere, while ReactionNetworkImporters generates a ParsedReactionSystem only, SBMLImporter generates a ParsedReactionSystem (here stored in prn) and a so-called CallbackSet (here stored in cbs). While prn can be used to create various problems, when we simulate them, we must also supply cbs. E.g. to simulate our brusselator we use:\n\nusing Catalyst, OrdinaryDiffEqDefault, Plots\ntspan = (0.0, 50.0)\noprob = ODEProblem(prn.rn, prn.u0, tspan, prn.p)\nsol = solve(oprob; callback = cbs)\nplot(sol)\n\n(Image: Brusselator Simulation)\n\nNote that, while ReactionNetworkImporters adds initial condition and species values as default to the imported model, SBMLImporter does not do this. These must hence be provided to the ODEProblem directly.\n\nA more detailed description of SBMLImporter's features can be found in its documentation.\n\nnote: Note\nThe massaction = true option informs the importer that the target model follows mass-action principles. When given, this enables SBMLImporter to make appropriate modifications to the model (which are important for e.g. jump simulation performance).","category":"section"},{"location":"model_creation/model_file_loading_and_export/#model_file_import_export_package_alts","page":"Loading Chemical Reaction Network Models from Files","title":"SBMLImporter and SBMLToolkit","text":"Above, we described how to use SBMLImporter to import SBML files. Alternatively, SBMLToolkit can be used instead. It has a slightly different syntax, which is described in its documentation, and does not support as wide a range of SBML features as SBMLImporter. A short comparison of the two packages can be found here. Generally, while they both perform well, we note that for jump simulations SBMLImporter is preferable (its way for internally representing reaction event enables more performant jump simulations).","category":"section"},{"location":"model_creation/model_file_loading_and_export/#model_file_import_export_matrix_representations","page":"Loading Chemical Reaction Network Models from Files","title":"Loading models from matrix representation using ReactionNetworkImporters.jl","text":"While CRN models can be represented through various file formats, they can also be represented in various matrix forms. E.g. a CRN with m species and n reactions (and with constant rates) can be represented with either\n\nAn mxn substrate matrix (with each species's substrate stoichiometry in each reaction) and an nxm product matrix (with each species's product stoichiometry in each reaction).\n\nOr\n\nAn mxn complex stoichiometric matrix (...) and a 2mxn incidence matrix (...).\n\nThe advantage of these forms is that they offer a compact and very general way to represent a large class of CRNs. ReactionNetworkImporters have the functionality for converting matrices of these forms directly into Catalyst ReactionSystem models. Instructions on how to do this are available in ReactionNetworkImporter's documentation.\n\n","category":"section"},{"location":"model_creation/model_file_loading_and_export/#petab_citations","page":"Loading Chemical Reaction Network Models from Files","title":"Citations","text":"If you use any of this functionality in your research, in addition to Catalyst, please cite the paper(s) corresponding to whichever package(s) you used:\n\n@software{2022ReactionNetworkImporters,\n  author       = {Isaacson, Samuel},\n  title        = {{ReactionNetworkImporters.jl}},\n  howpublished = {\\url{https://github.com/SciML/ReactionNetworkImporters.jl}},\n  year         = {2022}\n}\n\n@software{2024SBMLImporter,\n  author       = {Persson, Sebastian},\n  title        = {{SBMLImporter.jl}},\n  howpublished = {\\url{https://github.com/sebapersson/SBMLImporter.jl}},\n  year         = {2024}\n}\n\n@article{LangJainRackauckas+2024,\n    url = {https://doi.org/10.1515/jib-2024-0003},\n    title = {SBMLToolkit.jl: a Julia package for importing SBML into the SciML ecosystem},\n    title = {},\n    author = {Paul F. Lang and Anand Jain and Christopher Rackauckas},\n    pages = {20240003},\n    journal = {Journal of Integrative Bioinformatics},\n    doi = {doi:10.1515/jib-2024-0003},\n    year = {2024},\n    lastchecked = {2024-06-02}\n}","category":"section"},{"location":"steady_state_functionality/examples/bifurcationkit_periodic_orbits/#bifurcationkit_periodic_orbits","page":"Computing Periodic Orbits (Oscillations) Using BifurcationKit.jl","title":"Computing Periodic Orbits (Oscillations) Using BifurcationKit.jl","text":"<details><summary><strong>Environment setup and package installation</strong></summary>\n\nThe following code sets up an environment for running the code on this page.\n\nusing Pkg\nPkg.activate(; temp = true) # Creates a temporary environment, which is deleted when the Julia session ends.\nPkg.add(\"BifurcationKit\")\nPkg.add(\"Catalyst\")\nPkg.add(\"OrdinaryDiffEqDefault\")\nPkg.add(\"Plots\")\n\n</details>\n\n\n Previously, we have shown how to compute bifurcation diagrams using BifurcationKit.jl. In this example we will consider a system which exhibits an oscillation and show how to use BifurcationKit to track not just the system's (potentially unstable) steady state, but also the periodic orbit itself. More information on how to track periodic orbits can be found in the BifurcationKit documentation.","category":"section"},{"location":"steady_state_functionality/examples/bifurcationkit_periodic_orbits/#bifurcationkit_periodic_orbits_bifdia","page":"Computing Periodic Orbits (Oscillations) Using BifurcationKit.jl","title":"Computing the bifurcation diagram for the Repressilator","text":"We will first compute the bifurcation diagram, using the same approach as in the corresponding tutorial. For this example, we will use the oscillating Repressilator model.\n\nusing Catalyst\nrepressilator = @reaction_network begin\n    hillr(Z,v,K,n), ∅ --> X\n    hillr(X,v,K,n), ∅ --> Y\n    hillr(Y,v,K,n), ∅ --> Z\n    d, (X, Y, Z) --> ∅\nend\n\nNext, we create a BifurcationProblem for our model. We will compute the bifurcation diagram with respect to the parameter v, and plot the species X in the diagram.\n\nusing BifurcationKit\nbif_par = :v\nu_guess = [:X => 20.0, :Y => 15.0, :Z => 15.0]\np_start = [:v => 10.0, :K => 15.0, :n => 3, :d => 0.2]\nplot_var = :X\nbprob = BifurcationProblem(repressilator, u_guess, p_start, bif_par; plot_var)\nnothing # hide\n\nWe compute the bifurcation diagram using the bifurcationdiagram function. We will compute it across the interval v in (020).\n\nv_span = (0.0, 20.0)\nopts_br = ContinuationPar(p_min = v_span[1], p_max = v_span[2])\nbifdia = bifurcationdiagram(bprob, PALC(), 3, opts_br; bothside = true)\nnothing # hide\n\nFinally, we plot the bifurcation diagram.\n\nusing Plots\nplot(bifdia; xguide = bif_par, yguide = plot_var, xlimit = v_span, branchlabel = \"Steady state concentration\",\n    linewidthstable = 5)\n\nWe note that for small values of v the system's single steady state is stable (where the line is thicker). After a Hopf bifurcation (the red point), the state turns unstable (where the line is thinner). For chemical reaction networks (which mostly are well-behaved) a single unstable steady state typically corresponds to an oscillation. We can confirm that the system oscillates in the unstable region (while it reaches a stable steady state in the stable region) using simulations:\n\nusing OrdinaryDiffEqDefault\np_nosc = [:v => 5.0, :K => 15.0, :n => 3, :d => 0.2]\np_osc = [:v => 15.0, :K => 15.0, :n => 3, :d => 0.2]\nprob_nosc = ODEProblem(repressilator, u_guess, 80.0, p_nosc)\nprob_osc = ODEProblem(repressilator, u_guess, 80.0, p_osc)\nsol_nosc = OrdinaryDiffEqDefault.solve(prob_nosc)\nsol_osc = OrdinaryDiffEqDefault.solve(prob_osc)\nplot(plot(sol_nosc; title = \"v = 5\"), plot(sol_osc; title = \"v = 15\"), size = (1000,400), lw = 4)","category":"section"},{"location":"steady_state_functionality/examples/bifurcationkit_periodic_orbits/#bifurcationkit_periodic_orbits_pos","page":"Computing Periodic Orbits (Oscillations) Using BifurcationKit.jl","title":"Tracking the periodic orbits","text":"Next, we will use BifurcationKit.jl's continuation function (the bifurcationdiagram function, which we previously have used, works by calling continuation recursively) to track the periodic orbit which appears with the Hopf bifurcation point.\n\nFirst, we set the options for the continuation. Just like for bifurcation diagrams we must set our continuation parameters. Here we will use the same one as for the initial diagram (however, additional ones can be supplied).\n\nopts_po = ContinuationPar(opts_br)\nnothing # hide\n\nDuring the continuation we will compute the periodic orbit using the Collocation method (however, the Trapezoid and Shooting method also exist, with each having their advantages and disadvantages). For this, we create a PeriodicOrbitOCollProblem which we will supply to our continuation computation.\n\npoprob = PeriodicOrbitOCollProblem(50, 4)\nnothing # hide\n\nFinally, we will also create a record_from_solution function. This is a function which records information of the solution at each step of the continuation (which we can later plot or investigate using other means).\n\nX_idx = findfirst(isequal(repressilator.X), unknowns(complete(ss_ode_model(repressilator))))\nfunction record_from_solution(x, p; kwargs...)\n xtt = get_periodic_orbit(p.prob, x, p.p)\n min, max = extrema(xtt[1,:])\n period = getperiod(p.prob, x, p.p)\n    return (; min, max, period)\nend\nnothing # hide\n\nHere, get_periodic_orbit computes the system's periodic orbit. From it we extract the X species's minimum and maximum values (using extrema) and the period length (using getperiod). We return these quantities as a NamedTuple.\n\nWe can now compute the periodic orbit. First we must extract a branch from our bifurcation diagram (using get_branch), as continuation do not work on bifurcation diagrams directly (our bifurcation diagram consists of a single branch, which we here extract). Next, we can call continuation on it, designating that we wish to do continuation from the second point on the branch (which corresponds to the Hopf bifurcation, the first point is one of the branch's two endpoints).\n\nbranch = get_branch(bifdia, ()).γ\nbr_po = continuation(branch, 2, opts_po, poprob; record_from_solution)\nnothing # hide\n\nFinally, we can plot the periodic orbit. We use the vars argument to designate what we wish to plot. Here we first provide (:param, :min) (designating that we wish to plot the min value returned by record_from_solution, i.e. the minimum value throughout the periodic orbit) against the continuation parameter's (v) value. Next, we plot the maximum periodic orbit value using (:param, :max).\n\nplot(bifdia; xlimit = v_span, branchlabel = \"Steady state concentration\", linewidthstable = 5)\nplot!(br_po, vars = (:param, :min); color = 2, branchlabel = \"Oscillation amplitude (min/max)\")\nplot!(br_po, vars = (:param, :max); color = 2, xguide = bif_par, yguide = plot_var, branchlabel = \"\")\n\nHere we can see that, as v increases, the oscillation amplitude increases with it.\n\nPreviously, we had record_from_solution record the periodic orbit's period. This means that we can plot it as well. Here, we plot it against v using vars = (:param, :period).\n\nplot(br_po, vars = (:param, :period); xguide = bif_par, yguide = \"Period length\", xlimit = v_span, ylimit = (0.0, Inf))\n\nIn the plot we see that the period starts at around 18 time units, and slowly increase with v.\n\n","category":"section"},{"location":"steady_state_functionality/examples/bifurcationkit_periodic_orbits/#bifurcationkit_periodic_orbits_citation","page":"Computing Periodic Orbits (Oscillations) Using BifurcationKit.jl","title":"Citation","text":"If you use BifurcationKit.jl for your work, we ask that you cite the following paper!! Open source development strongly depends on this. It is referenced on HAL-Inria with bibtex entry CITATION.bib.","category":"section"},{"location":"model_simulation/simulation_plotting/#simulation_plotting","page":"Simulation Plotting","title":"Simulation Plotting","text":"<details><summary><strong>Environment setup and package installation</strong></summary>\n\nThe following code sets up an environment for running the code on this page.\n\nusing Pkg\nPkg.activate(; temp = true) # Creates a temporary environment, which is deleted when the Julia session ends.\nPkg.add(\"Catalyst\")\nPkg.add(\"OrdinaryDiffEqDefault\")\nPkg.add(\"Plots\")\n\n</details>\n\n\n   Catalyst uses the Plots.jl package for performing all plots. This section provides a brief summary of some useful plotting options, while Plots.jl's documentation provides a more throughout description of how to tune your plots.\n\nnote: Note\nMakie.jl is a popular alternative to the Plots.jl package. While it is not used within Catalyst's documentation, it is worth considering (especially for users interested in interactivity, or increased control over their plots).","category":"section"},{"location":"model_simulation/simulation_plotting/#simulation_plotting_options","page":"Simulation Plotting","title":"Common plotting options","text":"Let us consider the oscillating Brusselator model. We have previously shown how model simulation solutions can be plotted using the plot function. Here we plot an ODE simulation from the Brusselator:\n\nusing Catalyst, OrdinaryDiffEqDefault, Plots\n\nbrusselator = @reaction_network begin\n    A, ∅ --> X\n    1, 2X + Y --> 3X\n    B, X --> Y\n    1, X --> ∅\nend\nu0 = [:X => 1.0, :Y => 0.0]\ntspan = (0.0, 50.0)\nps = [:A => 1.0, :B => 4.0]\n\noprob = ODEProblem(brusselator, u0, tspan, ps)\nsol = solve(oprob)\nplot(sol)\n\nVarious plotting options can be provided as optional arguments to the plot command. Common options include:\n\nlw: Determine plot line widths.\nla: Determine plot line's transparency (at la = 0.0 lines are fully transparent, i.e. not visible).\nlinestyle: Determines plot line style.\ncolor: Determines the line colours.\nlegend: Determines the position of the legend/labels.\nlabel: Determines label texts.\nxguide, yguide: Determines x and y axis labels.\ntitle: Determines plot title.\nlegendfontsize, guidefontsize, titlefontsize: Determines the font size of the labels, x and y guides, and title, respectively.\n\nHere, we re-plot our simulations, utilising some of these options (legend = :none is used to disable the legends).\n\nplot(sol; lw = 4, linestyle = :dash, color = :green, xguide = \"Time\", yguide = \"Concentration\", guidefontsize = 14)\n\nNote that, by default, Catalyst uses xguide = \"t\". Here, however, we modify this to xguide = \"Time\". We also note that the color = :green change both lines' colours to green. To set different colours for each line, we provide these as a vector without , in-between elements (in Julia interpreted as a matrix with its first dimension equal to 1):\n\nplot(sol; lw = 4, color = [:green :purple])\n\nA full list of available colours can be found here. A full list of possible plotting options can be found here (look at the list of various plot attributes, e.g. \"Series Attributes\"). if there is some option(s) you intend to use multiple times, you can call the default function using these, in which case they will be used for all subsequent plots. E.g. here:\n\ndefault(framestyle = :box, grid = false)\n\nwe designate a box-style frame, and remove the faint background grid, for all subsequent plots in this tutorial.\n\nA useful option unique to Catalyst (and other DifferentialEquations.jl-based) plots is idxs. Its input is a vector, listing all the species (or quantities) that should be plotted. I.e.\n\nplot(sol; idxs = [:X])\n\ncan be used to plot X only. When only a single argument is given, the vector form is unnecessary (e.g. idxs = :X could have been used instead). If symbolic species representation is used, this can be used to designate any algebraic expression(s) that should be plotted. E.g. here we plot the total concentration of X + Y throughout the simulation:\n\nplot(sol; idxs = brusselator.X + brusselator.Y)","category":"section"},{"location":"model_simulation/simulation_plotting/#simulation_plotting_options_subplots","page":"Simulation Plotting","title":"Multi-plot plots","text":"It is possible to save plots in variables. These can then be used as input to the plot command. Here, the plot command can be used to create plots containing multiple plots (by providing multiple inputs). E.g. here we plot the concentration of X and Y in separate subplots:\n\nplt_X = plot(sol; idxs = [:X])\nplt_Y = plot(sol; idxs = [:Y])\nplot(plt_X, plt_Y)\n\nWhen working with subplots, the layout and size options are typically useful. Here we use layout to put the first plot above the second one, and size to reshape the plot size:\n\nplot(plt_X, plt_Y; layout = (2,1), size = (700,500))","category":"section"},{"location":"model_simulation/simulation_plotting/#simulation_plotting_options_saving","page":"Simulation Plotting","title":"Saving plots","text":"Once a plot has been saved to a variable, the savefig function can be used to save it to a file. Here we save our Brusselator plot simulation (the first argument) to a file called \"saved_plot.png\" (the second argument):\n\nplt = plot(sol)\nsavefig(plt, \"saved_plot.png\")\nrm(\"saved_plot.png\") # hide\n\nThe plot file type is automatically determined from the extension (if none is given, a .png file is created).","category":"section"},{"location":"model_simulation/simulation_plotting/#simulation_plotting_options_phasespace","page":"Simulation Plotting","title":"Phase-space plots","text":"By default, simulations are plotted as species concentrations over time. However, phase space plots are also possible. This is done by designating the axis arguments using the idxs option, but providing them as a tuple. E.g. here we plot our simulation in X-Y space:\n\nplot(sol; idxs = (:X, :Y))","category":"section"},{"location":"steady_state_functionality/examples/nullcline_plotting/#nullcline_plotting","page":"Plotting Nullclines and Steady States in Phase Space","title":"Plotting Nullclines and Steady States in Phase Space","text":"<details><summary><strong>Environment setup and package installation</strong></summary>\n\nThe following code sets up an environment for running the code on this page.\n\nusing Pkg\nPkg.activate(; temp = true) # Creates a temporary environment, which is deleted when the Julia session ends.\nPkg.add(\"Catalyst\")\nPkg.add(\"HomotopyContinuation\")\nPkg.add(\"Plots\")\n\n</details>\n\n\n In this tutorial we will show how to extract a system's steady states and nullclines, and how to plot these in phase space. Generally, while nullclines are not directly needed for most types analysis, plotting these can give some understanding of a system's steady state and stability properties.\n\nFor an ordinary differential equation\n\nbeginaligned\nfracdx_1dt = f_1(x_1 x_2  x_n) \nfracdx_2dt = f_2(x_1 x_2  x_n) \n                vdots\nfracdx_ndt = f_n(x_1 x_2  x_n) \nendaligned\n\nthe i'th nullcline is the surface along which fracdx_idt = 0, i.e. the implicit surface given by f_i(x_1dotsx_n) = 0. Nullclines are frequently used when visualizing the phase-planes of two-dimensional models (as these can be easily plotted).","category":"section"},{"location":"steady_state_functionality/examples/nullcline_plotting/#nullcline_plotting_computation","page":"Plotting Nullclines and Steady States in Phase Space","title":"Computing nullclines and steady states for a bistable switch","text":"For our example we will use a simple bistable switch model, consisting of two species (X and Y) which mutually inhibit each other through repressive Hill functions. \n\nusing Catalyst\nbs_switch = @reaction_network begin\n    hillr(Y, v, K, n), 0 --> X\n    hillr(X, v, K, n), 0 --> Y\n    1.0, (X,Y) --> 0\nend\n\nNext, we compute the steady states using homotopy continuation.\n\nimport HomotopyContinuation\nps = [:v => 1.0, :K => 0.6, :n => 4.0]\nsss = hc_steady_states(bs_switch, ps; show_progress = false)\n\nFinally, we will compute the nullclines. First we create a function which, for species values (XY), returns the evaluation of the model's ODE's right-hand side.\n\nnlprob = NonlinearProblem(bs_switch, [:X => 0.0, :Y => 0.0], ps)\nfunction get_XY(Xval, Yval)\n    prob = Catalyst.remake(nlprob; u0 = [:X => Xval, :Y => Yval])\n    return nlprob.f(prob.u0, prob.p)\nend\nnothing # hide\n\nNext, we plot our nullclines by using Plot.jl's contour function. Here, we plot the 0 contour lines (which corresponds to the nullclines) for both the X and Y nullclines. We will plot the steady states using scatter. We use the steady_state_stability function to compute steady state stabilities (and use this to determine how to plot the steady state markers).\n\nusing Plots\n# Plot the nullclines. Line labels added in separate `plot` commands (due to how the `contour` works).\nplt_mesh = 0:0.02:1.25\ncontour(plt_mesh, plt_mesh, (x,y) -> get_XY(x,y)[1]; levels = [0.0], lw = 7, la = 0.7, color = 1, cbar = false)\ncontour!(plt_mesh, plt_mesh, (x,y) -> get_XY(x,y)[2]; levels = [0.0], lw = 7, la = 0.7, color = 2, cbar = false)\nplot!([]; label = \"dX/dt = 0\", lw = 7, la = 0.7, color = 1)\nplot!([]; label = \"dY/dt = 0\", lw = 7, la = 0.7, color = 2)\n\n# Plot the steady states.\nfor ss in sss\n    color, markershape = if steady_state_stability(ss, bs_switch, ps)\n        :blue, :circle\n    else\n        :red, :star4\n    end\n    scatter!((ss[2], ss[1]); color, markershape, label = \"\", markersize = 10)\nend\nscatter!([], []; color = :blue, markershape = :circle, label = \"Stable stead state\")\nscatter!([], []; color = :red, markershape = :star4, label = \"Unstable stead state\")\n\n# Finishing touches.\nplot!(xguide = \"X\", yguide = \"Y\", xlimit = (0.0, 1.25), ylimit = (0.0, 1.25), legendfontsize = 10, size = (600,600), legend = :topright)\n\nHere we can see how the steady states occur at the nullclines intersections.\n\nnote: Note\nHere we use an inherent Plots function to plot the nullclines. However, there are also specialised packages for these kinds of plots, such as ImplicitPlots.jl.","category":"section"},{"location":"steady_state_functionality/examples/nullcline_plotting/#nullcline_plotting_directions","page":"Plotting Nullclines and Steady States in Phase Space","title":"Plotting system directions in phase space","text":"One useful property of nullclines is that the sign of dXdt will only switch whenever the solution crosses the dXdt=0 nullcline. This means that, within each region defined by the nullclines, the direction of the solution remains constant. Below we use this to, for each such region, plot arrows showing the solution's direction.\n\n# Creates a function for plotting the ODE's direction at a point in phase space.\nfunction plot_xy_arrow!(Xval, Yval)\n    dX, dY = get_XY(Xval, Yval)\n    dX = dX > 0 ? 0.05 : -0.05\n    dY = dY > 0 ? 0.05 : -0.05\n    plot!([Xval, Xval + dX], [Yval, Yval]; color = :black, lw = 2, arrow = :arrow, label = \"\")\n    plot!([Xval, Xval], [Yval, Yval + dY]; color = :black, lw = 2, arrow = :arrow, label = \"\")\nend\n\n# Plots the ODE's direction in phase space at selected positions.\narrow_positions = [(0.25, 0.25), (0.75, 0.75), (0.35, 0.8), (0.8, 0.35), (0.02, 1.1), (1.1, 0.02)]\nforeach(pos -> plot_xy_arrow!(pos...), arrow_positions)\nplot!()\n\nThis also works as a form of simple stability analysis, where we can see how the solution moves away from the unstable steady state, and to the stable ones.","category":"section"},{"location":"model_creation/examples/smoluchowski_coagulation_equation/#smoluchowski_coagulation_equation","page":"Smoluchowski Coagulation Equation","title":"Smoluchowski Coagulation Equation","text":"<details><summary><strong>Environment setup and package installation</strong></summary>\n\nThe following code sets up an environment for running the code on this page.\n\nusing Pkg\nPkg.activate(; temp = true) # Creates a temporary environment, which is deleted when the Julia session ends.\nPkg.add(\"Catalyst\")\nPkg.add(\"JumpProcesses\")\nPkg.add(\"ModelingToolkitBase\")\nPkg.add(\"Plots\")\nPkg.add(\"SpecialFunctions\")\n\n</details>\n\n\n This tutorial shows how to programmatically construct a ReactionSystem corresponding to the chemistry underlying the Smoluchowski coagulation model using ModelingToolkit/Catalyst. A jump process version of the model is then constructed from the ReactionSystem, and compared to the model's analytical solution obtained by the method of Scott (see also 3).\n\nThe Smoluchowski coagulation equation describes a system of reactions in which monomers may collide to form dimers, monomers and dimers may collide to form trimers, and so on. This models a variety of chemical/physical processes, including polymerization and flocculation.\n\nWe begin by importing some necessary packages.\n\nusing ModelingToolkitBase, Catalyst, LinearAlgebra\nusing JumpProcesses\nusing Plots, SpecialFunctions\n\nSuppose the maximum cluster size is N. We assume an initial concentration of monomers, Nₒ, and let uₒ denote the initial number of monomers in the system. We have nr total reactions, and label by V the bulk volume of the system (which plays an important role in the calculation of rate laws since we have bimolecular reactions). Our basic parameters are then\n\n# maximum cluster size\nN = 10\n\n# volume of a monomers in cm³\nVₒ = (4π / 3) * (10e-06 * 100)^3\n\n# initial conc. = (No. of init. monomers) / bulk volume\nNₒ = 1e-06 / Vₒ\n\n# No. of monomers initially\nuₒ = 10000\n\n# Bulk volume of system in cm³\nV = uₒ / Nₒ\nn = floor(Int, N / 2)\n\n# No. of forward reactions\nnr = ((N % 2) == 0) ? (n*(n + 1) - n) : (n*(n + 1))\nnothing #hide\n\nThe Smoluchowski coagulation equation Wikipedia page illustrates the set of possible reactions that can occur. We can easily enumerate the pairs of multimer reactants that can combine when allowing a maximal cluster size of N monomers. We initialize the volumes of the reactant multimers as volᵢ and volⱼ\n\n# possible pairs of reactant multimers\npair = []\nfor i = 2:N\n    halfi = floor(Int, i/2)\n    push!(pair, [(1:halfi)  (i .- (1:halfi))])\nend\npair = vcat(pair...)\nvᵢ = @view pair[:, 1]  # Reactant 1 indices\nvⱼ = @view pair[:, 2]  # Reactant 2 indices\nvolᵢ = Vₒ * vᵢ         # cm⁻³\nvolⱼ = Vₒ * vⱼ         # cm⁻³\nsum_vᵢvⱼ = @. vᵢ + vⱼ  # Product index\nnothing #hide\n\nWe next specify the rates (i.e. kernel) at which reactants collide to form products. For simplicity, we allow a user-selected additive kernel or constant kernel. The constants(B and C) are adopted from Scott's paper 2\n\n# set i to  1 for additive kernel, 2  for constant\ni = 1\nif i == 1\n    B = 1.53e03    # s⁻¹\n\n    # dividing by volume as it is a bimolecular reaction chain\n    kv = @. B * (volᵢ + volⱼ) / V\nelseif i==2\n    C = 1.84e-04    # cm³ s⁻¹\n    kv = fill(C / V, nr)\nend\nnothing #hide\n\nWe'll set the parameters and the initial condition that only monomers are present at t=0 in u₀map.\n\n# k is a vector of the parameters, with values given by the vector kv\n@parameters k[1:nr] = kv\n\n# create the vector of species X_1,...,X_N\nt = default_t()\n@species (X(t))[1:N]\n\n# time-span\nif i == 1\n    tspan = (0.0, 2000.0)\nelseif i == 2\n    tspan = (0.0, 350.0)\nend\n\n # initial condition of monomers\nu₀    = zeros(Int64, N)\nu₀[1] = uₒ\nu₀map = Pair.(collect(X), u₀)   # map species to its initial value\nnothing #hide\n\nHere we generate the reactions programmatically. We systematically create Catalyst Reactions for each possible reaction shown in the figure on Wikipedia. When vᵢ[n] == vⱼ[n], we set the stoichiometric coefficient of the reactant multimer to two.\n\n# vector to store the Reactions in\nrx = []\nfor n = 1:nr\n    # for clusters of the same size, double the rate\n    if (vᵢ[n] == vⱼ[n])\n        push!(rx, Reaction(k[n], [X[vᵢ[n]]], [X[sum_vᵢvⱼ[n]]], [2], [1]))\n    else\n        push!(rx, Reaction(k[n], [X[vᵢ[n]], X[vⱼ[n]]], [X[sum_vᵢvⱼ[n]]],\n                           [1, 1], [1]))\n    end\nend\n@named rs = ReactionSystem(rx, t, collect(X), [k])\nrs = complete(rs)\n\nWe now convert the ReactionSystem into a ModelingToolkitBase jump System, and solve it using Gillespie's direct method. For details on other possible solvers (SSAs), see the DifferentialEquations.jl documentation\n\n# solving the system\njprob = JumpProblem(rs, u₀map, tspan; save_positions = (false, false))\njsol = solve(jprob; saveat = tspan[2] / 30)\nnothing #hide\n\nLets check the results for the first three polymers/cluster sizes. We compare to the analytical solution for this system:\n\n# Results for first three polymers...i.e. monomers, dimers and trimers\nv_res = [1; 2; 3]\n\n# comparison with analytical solution\n# we only plot the stochastic solution at a small number of points\n# to ease distinguishing it from the exact solution\nt   = jsol.t\nsol = zeros(length(v_res), length(t))\nif i == 1\n    ϕ = @. 1 - exp(-B*Nₒ*Vₒ*t)\n    for j in v_res\n        sol[j,:] = @. Nₒ*(1 - ϕ)*(((j*ϕ)^(j-1))/gamma(j+1))*exp(-j*ϕ)\n    end\nelseif i == 2\n    ϕ = @. (C*Nₒ*t)\n    for j in v_res\n        sol[j,:] = @. 4Nₒ*((ϕ^(j-1))/((ϕ + 2)^(j+1)))\n    end\nend\n\n# plotting normalised concentration vs analytical solution\ndefault(lw = 2, xlabel = \"Time (sec)\")\nscatter(ϕ, jsol(t)[1,:] / uₒ, label = \"X1 (monomers)\", markercolor = :blue)\nplot!(ϕ, sol[1,:]/Nₒ, line = (:dot,4,:blue), label=\"Analytical sol--X1\")\n\nscatter!(ϕ, jsol(t)[2,:] / uₒ, label = \"X2 (dimers)\", markercolor = :orange)\nplot!(ϕ, sol[2,:] / Nₒ, line = (:dot, 4, :orange), label = \"Analytical sol--X2\")\n\nscatter!(ϕ, jsol(t)[3,:] / uₒ, label = \"X3 (trimers)\", markercolor = :purple)\nplot!(ϕ, sol[3,:] / Nₒ, line = (:dot, 4, :purple), label = \"Analytical sol--X3\",\n      ylabel = \"Normalized Concentration\")\n\n","category":"section"},{"location":"model_creation/examples/smoluchowski_coagulation_equation/#References","page":"Smoluchowski Coagulation Equation","title":"References","text":"https://en.wikipedia.org/wiki/Smoluchowski_coagulation_equation\nScott, W. T. (1968). Analytic Studies of Cloud Droplet Coalescence I, Journal of Atmospheric Sciences, 25(1), 54-65. Retrieved Feb 18, 2021, from https://journals.ametsoc.org/view/journals/atsc/25/1/1520-0469_1968_025_0054_asocdc_2_0_co_2.xml\nIan J. Laurenzi, John D. Bartels, Scott L. Diamond, A General Algorithm for Exact Simulation of Multicomponent Aggregation Processes, Journal of Computational Physics, Volume 177, Issue 2, 2002, Pages 418-449, ISSN 0021-9991, https://doi.org/10.1006/jcph.2002.7017.","category":"section"},{"location":"network_analysis/crn_theory/#network_analysis_structural_aspects","page":"Chemical Reaction Network Theory","title":"Chemical Reaction Network Theory","text":"<details><summary><strong>Environment setup and package installation</strong></summary>\n\nThe following code sets up an environment for running the code on this page.\n\nusing Pkg\nPkg.activate(; temp = true) # Creates a temporary environment, which is deleted when the Julia session ends.\nPkg.add(\"CairoMakie\")\nPkg.add(\"Catalyst\")\nPkg.add(\"GraphMakie\")\nPkg.add(\"NetworkLayout\")\n\n</details>\n\n<details><summary><strong>Quick-start example</strong></summary>\n\nCatalyst provides a range of function for querying a reaction network model for various network properties. A few examples include:\n\nrn = @reaction_network begin\n    (p,d), 0 <--> X\n    (kB,kD), 2X <--> X2\n    (kB,kD), X2 + X <--> X3\nend\nlinkageclasses(rn)\ndeficiency(rn)\nisreversible(rn)\n\n</details>\n\n\n  \n\nThe systems of ODEs or stochastic chemical kinetics models that arise from chemical reaction networks can often have their steady-state properties (number of steady states, etc.) known in advance, simply by analyzing the graph structure of the network. The subfield of chemistry and math studying this relationship is called Chemical Reaction Network Theory.\n\nIn this tutorial we give a broad overview of chemical reaction network theory, building on the discussion of the structure of  the reaction network ODEs in the previous section. We also introduce several of the Catalyst API functions for network analysis. A complete summary of the exported functions is given in the API section Network Analysis and Representations.\n\nBroadly, results from chemical reaction network theory relate a purely graph-structural property (e.g. deficiency) to dynamical properties of the reaction system (e.g. complex balance). The relevant graph here is the one corresponding to the reaction complex representation of the network, where the nodes represent the reaction complexes and the edges represent reactions. Let us illustrate some of the types of network properties that  Catalyst can determine.\n\nTo begin, consider the following reaction network:\n\nusing Catalyst\nrn = @reaction_network begin\n    (k1,k2), A + B <--> C\n    k3, C --> D+E\n    (k4,k5), D+E <--> F\n    (k6,k7), 2A <--> B+G\n    k8, B+G --> H\n    k9, H --> 2A\nend\n\nwith graph\n\nusing CairoMakie, GraphMakie, NetworkLayout\nplot_complexes(rn)","category":"section"},{"location":"network_analysis/crn_theory/#network_analysis_structural_aspects_linkage","page":"Chemical Reaction Network Theory","title":"Linkage classes and sub-networks of the reaction network","text":"The preceding reaction complex graph shows that rn is composed of two disconnected sub-graphs, one containing the complexes A+B, C, D+E, and F, the other containing the complexes 2A, B + G, and H. These sets, A+B C D+E F and 2A B + GH are called the \"linkage classes\" of the reaction network. The function linkageclasses will calculate these for a given network, returning a vector of the integer indices of reaction complexes participating in each set of linkage-classes. Note, indices of reaction complexes can be determined from the ordering returned by reactioncomplexes.\n\n# we must first calculate the reaction complexes -- they are cached in rn\nreactioncomplexes(rn)\n\n# now we can calculate the linkage classes\nlcs = linkageclasses(rn)\n\nIt can often be convenient to obtain the disconnected sub-networks as distinct ReactionSystems, which are returned by the subnetworks function:\n\nsubnets = subnetworks(rn)\n\n# check the reactions in each subnetwork\nreactions.(subnets)\n\nThe graphs of the reaction complexes in the two sub-networks are then\n\n  plot_complexes(subnets[1])\n\nand,\n\n plot_complexes(subnets[2])","category":"section"},{"location":"network_analysis/crn_theory/#network_analysis_structural_aspects_deficiency","page":"Chemical Reaction Network Theory","title":"Deficiency of the network","text":"A famous theorem in Chemical Reaction Network Theory, the Deficiency Zero Theorem[1], allows us to use knowledge of the net stoichiometry matrix and the linkage classes of a mass action RRE ODE system to draw conclusions about the system's possible steady states. In this section we'll see how Catalyst can calculate a network's deficiency.\n\nThe rank, r, of a reaction network is defined as the dimension of the subspace spanned by the net stoichiometry vectors of the reaction-network[1], i.e. the span of the columns of the net stoichiometry matrix N. It corresponds to the number of independent species in a chemical reaction network. That is, if we calculate the linear conservation laws of a network, and use them to eliminate the dependent species of the network, we will have r independent species remaining. For our current example the conservation laws are given by\n\n# first we calculate the conservation laws -- they are cached in rn\nconservationlaws(rn)\n\n# then we display them as equations for the dependent variables\nconservedequations(rn)\nshow(stdout, MIME\"text/plain\"(), ans) # hide\n\nHere the parameters Γ[i] represent the constants of the three conservation laws, and we see that there are three dependent species that could be eliminated. As\n\nnumspecies(rn)\n\nwe find that there are five independent species. Let's check this is correct:\n\nusing LinearAlgebra\nrank(netstoichmat(rn)) == 5\n\nSo we know that the rank of our reaction network is five. An extended section discussing how to utilise conservation law elimination during chemical reaction network modelling can be found here.\n\nThe deficiency, delta, of a reaction network is defined as\n\ndelta = textrm(number of complexes) - textrm(number of linkage classes) - textrm(rank)\n\nFor our network this is 7 - 2 - 5 = 0, which we can calculate in Catalyst as\n\n# first we calculate the reaction complexes of rn and cache them in rn\nreactioncomplexes(rn)\n\n# then we can calculate the deficiency\nδ = deficiency(rn)\n\nQuoting Feinberg [1]\n\nDeficiency zero networks are ones for which the reaction vectors [i.e. net stoichiometry vectors] are as independent as the partition of complexes into linkage classes will allow.","category":"section"},{"location":"network_analysis/crn_theory/#network_analysis_structural_aspects_reversibility","page":"Chemical Reaction Network Theory","title":"Reversibility of the network","text":"A reaction network is reversible if the \"arrows\" of the reactions are symmetric so that every reaction is accompanied by its reverse reaction. Catalyst's API provides the isreversible function to determine whether a reaction network is reversible. As an example, consider\n\nrn = @reaction_network begin\n  (k1,k2),A <--> B\n  (k3,k4),A + C <--> D\n  (k5,k6),D <--> B+E\n  (k7,k8),B+E <--> A+C\nend\n\n# calculate the set of reaction complexes\nreactioncomplexes(rn)\n\n# test if the system is reversible\nisreversible(rn)\n\nConsider another example,\n\nrn = @reaction_network begin\n  (k1,k2),A <--> B\n  k3, A + C --> D\n  k4, D --> B+E\n  k5, B+E --> A+C\nend\nreactioncomplexes(rn)\nisreversible(rn)\n\nplot_complexes(rn)\n\nIt is evident from the preceding graph that the network is not reversible. However, it satisfies a weaker property in that there is a path from each reaction complex back to itself within its associated subgraph. This is known as weak reversibility. One can test a network for weak reversibility by using the isweaklyreversible function:\n\n# need subnetworks from the reaction network first\nsubnets = subnetworks(rn)\nisweaklyreversible(rn, subnets)\n\nEvery reversible network is also weakly reversible, but not every weakly reversible network is reversible.","category":"section"},{"location":"network_analysis/crn_theory/#network_analysis_structural_aspects_deficiency_zero_theorem","page":"Chemical Reaction Network Theory","title":"Deficiency Zero Theorem","text":"Knowing the deficiency and weak reversibility of a mass action chemical reaction network ODE model allows us to make inferences about the corresponding steady state behavior. Before illustrating how this works for one example, we need one last definition.\n\nRecall that in the matrix-vector representation for the RRE ODEs, the entries, N_m k, of the stoichiometry matrix, N, give the net change in species m due to reaction k. If we let mathbfN_k denote the kth column of this matrix, this vector corresponds to the change in the species state vector, mathbfx(t), due to reaction k, i.e. when reaction k occurs mathbfx(t) to mathbfx(t) + mathbfN_k. Moreover, by integrating the ODE\n\nfracdmathbfxdt = N mathbfv(mathbfx(t)) = sum_k=1^K v_k(mathbfx(t))  mathbfN_k\n\nwe find\n\nmathbfx(t) = mathbfx(0) + sum_k=1^K left(int_0^t v_k(mathbfx)(s)  dsright) mathbfN_k\n\nwhich demonstrates that mathbfx(t) - mathbfx(0) is always given by a linear combination of the stoichiometry vectors, i.e.\n\nmathbfx(t) - mathbfx(0) in operatornamespanmathbfN_k \n\nIn particular, this says that mathbfx(t) lives in the translation of the operatornamespanmathbfN_k  by mathbfx(0) which we write as (mathbfx(0) + operatornamespanmathbfN_k). In fact, since the solution should stay non-negative, if we let barmathbbR_+^M denote the subset of vectors in mathbbR^M with non-negative components, the possible physical values for the solution, mathbfx(t), must be in the set\n\n(mathbfx(0) + operatornamespanmathbfN_k) cap barmathbbR_+^M\n\nThis set is called the stoichiometric compatibility class of mathbfx(t). The key property of stoichiometric compatibility classes is that they are invariant under the RRE ODE's dynamics, i.e. a solution will always remain within the subspace given by the stoichiometric compatibility class. Finally, we note that the positive stoichiometric compatibility class generated by mathbfx(0) is just (mathbfx(0) + operatornamespanmathbfN_k) cap mathbbR_+^M, where mathbbR_+^M denotes the vectors in mathbbR^M with strictly positive components.\n\nWith these definitions we can now see how knowing the deficiency and weak reversibility of the network can tell us about its steady state behavior. Consider the previous example, which we know is weakly reversible. Its deficiency is\n\ndeficiency(rn)\n\nWe also verify that the system is purely mass action (though it is apparent from the network's definition):\n\nall(rx -> ismassaction(rx, rn), reactions(rn))\n\nWe can therefore apply the Deficiency Zero Theorem to draw conclusions about the system's steady state behavior. The Deficiency Zero Theorem (roughly) says that a mass action network with deficiency zero satisfies\n\nIf the network is weakly reversible, then independent of the reaction rate constants the RRE ODEs have exactly one equilibrium solution within each positive stoichiometric compatibility class. That equilibrium is locally asymptotically stable.\nIf the network is not weakly reversible, then the RRE ODEs cannot admit a positive equilibrium solution.\n\nSee [1] for a more precise statement, proof, and additional examples.\n\nWe can therefore conclude that for any initial condition that is positive, and hence in some positive stoichiometric compatibility class, rn will have exactly one equilibrium solution which will be positive and locally asymptotically stable.\n\nAs a final example, consider the following network from [1]\n\ndef0_rn = @reaction_network begin\n  (k1,k2),A <--> 2B\n  (k3,k4), A + C <--> D\n  k5, B+E --> C + D\nend\nreactioncomplexes(def0_rn)\nsubnets = subnetworks(def0_rn)\nisma = all(rx -> ismassaction(rx,def0_rn), reactions(def0_rn))\ndef = deficiency(def0_rn)\niswr = isweaklyreversible(def0_rn, subnets)\nisma,def,iswr\n\nwhich we see is mass action and has deficiency zero, but is not weakly reversible. As such, we can conclude that for any choice of rate constants the RRE ODEs cannot have a positive equilibrium solution.\n\nThere is an API function satisfiesdeficiencyzero that will let us check all these conditions easily:\n\nsatisfiesdeficiencyzero(def0_rn)","category":"section"},{"location":"network_analysis/crn_theory/#Deficiency-One-Theorem","page":"Chemical Reaction Network Theory","title":"Deficiency One Theorem","text":"Very analogous to the deficiency zero theorem is the deficiency one theorem. The deficiency one theorem applies to a network with the following properties: \n\nThe deficiency of each linkage class of the network is at most 1, \nThe sum of the linkage class deficiencies is the total deficiency of the network, and\nEach linkage class has at most one terminal linkage class, which is a linkage class that is 1) strongly connected, and 2) has no outgoing reactions.\n\nFor the set of reactions A to B B to A B to C, A B C is a linkage class, and A B is a strong linkage class (since A is reachable from B and vice versa). However, A B is not a terminal linkage class, because the reaction B to C goes to a complex outside the linkage class. \n\nIf these conditions are met, then the network will have at most one steady state in each  stoichiometric compatibility class for any choice of rate constants and parameters. Unlike  the deficiency zero theorem, networks obeying the deficiency one theorem are not guaranteed  to have stable solutions.\n\nLet's look at an example network.\n\ndef1_network = @reaction_network begin\n    (k1, k2), A <--> 2B\n    k3, C --> 2D\n    k4, 2D --> C + E\n    (k5, k6), C + E <--> E + 2D\nend\nplot_complexes(def1_network)\n\nWe can see from the complex graph that there are two linkage classes, the deficiency of the bottom one is zero and the deficiency of the top one is one. The total deficiency is one:\n\ndeficiency(def1_network)\n\nAnd there is only one terminal linkage class in each, so our network satisfies all three conditions. As in the deficiency zero case, there is the API function satisfiesdeficiencyone for quickly checking these conditions:\n\nsatisfiesdeficiencyone(def1_network)","category":"section"},{"location":"network_analysis/crn_theory/#network_analysis_complex_and_detailed_balance","page":"Chemical Reaction Network Theory","title":"Complex and Detailed Balance","text":"A reaction network's steady state is complex-balanced if the total production of each  complex is zero at the steady state. A reaction network's steady state is detailed balanced  if every reaction is balanced by its reverse reaction at the steady-state (this corresponds to  the usual notion of chemical equilibrium; note that this requires every reaction be reversible). \n\nNote that detailed balance at a given steady state implies complex balance for that steady state, i.e. detailed balance is a stronger property than complex balance.\n\nRemarkably, having just one positive steady state that is complex (detailed) balance implies that  complex (detailed) balance holds for every positive steady state, so we say that a network  is complex (detailed) balanced if any one of its steady states are complex (detailed) balanced.  Additionally, there will be exactly one steady state in every positive stoichiometric  compatibility class, and this steady state is asymptotically stable. (For proofs of these results, please consult Martin Feinberg's Foundations of Chemical Reaction Network Theory[1]). So  knowing that a network is complex balanced is really quite powerful.\n\nLet's check whether the deficiency 0 reaction network that we defined above is complex balanced by providing a set of rates:\n\nrates = Dict([:k1 => 2.4, :k2 => 4., :k3 => 10., :k4 => 5.5, :k5 => 0.4])\niscomplexbalanced(rn, rates)\n\nWe can do a similar check for detailed balance.\n\nisdetailedbalanced(rn, rates)\n\nThe reason that the deficiency zero theorem puts such strong restrictions on the steady state  properties of the reaction network is because it implies that the reaction network will be  complex balanced for any set of rate constants and parameters. The fact that this holds  from a purely structural property of the graph, regardless of kinetics, is what makes it so  useful. But in some cases it might be desirable to check complex balance on its own, as for  higher deficiency networks.","category":"section"},{"location":"network_analysis/crn_theory/#network_analysis_concentration_robustness","page":"Chemical Reaction Network Theory","title":"Concentration Robustness","text":"Certain reaction networks have species that do not change their concentration, regardless of whether the system is perturbed to a different stoichiometric compatibility class. This is a very useful property to have in biological contexts, where it might be important to keep the concentration of a critical species relatively stable in the face of changes in its environment. \n\nDetermining every species with concentration-robustness in a network is in general very difficult. However, there are certain cases where there are sufficient conditions that can be checked relatively easily. One example is for deficiency one networks.\n\nTheorem (a sufficient condition for concentration robustness for deficiency one networks): If there are two non-terminal reaction complexes that differ only in species s, then the system is absolutely concentration robust with respect to s. \n\nThis is the check provided by the API function robustspecies(rn). More general concentration robustness analysis can be done using the forthcoming CatalystNetworkAnalysis package.\n\n","category":"section"},{"location":"network_analysis/crn_theory/#References","page":"Chemical Reaction Network Theory","title":"References","text":"[1]: Feinberg, M. Foundations of Chemical Reaction Network Theory, Applied Mathematical Sciences 202, Springer (2019).","category":"section"},{"location":"network_analysis/crn_theory/#Catalyst.iscomplexbalanced-network_analysis-crn_theory","page":"Chemical Reaction Network Theory","title":"Catalyst.iscomplexbalanced","text":"iscomplexbalanced(rs::ReactionSystem, parametermap; sparse = false)\n\nConstructively compute whether a network will have complex-balanced equilibrium solutions, following the method in van der Schaft et al., 2015.\n\nRequires the specification of values for the parameters/rate constants. Accepts a dictionary, vector, or tuple of parameter-to-value mappings, e.g. [k1 => 1.0, k2 => 2.0,...].\n\n\n\n\n\n","category":"function"},{"location":"network_analysis/crn_theory/#Catalyst.isdetailedbalanced-network_analysis-crn_theory","page":"Chemical Reaction Network Theory","title":"Catalyst.isdetailedbalanced","text":"isdetailedbalanced(rs::ReactionSystem, parametermap; reltol=1e-9, abstol)\n\nConstructively compute whether a kinetic system (a reaction network with a set of rate constants) will admit detailed-balanced equilibrium solutions, using the Wegscheider conditions, Feinberg, 1989. A detailed-balanced solution is one for which the rate of every forward reaction exactly equals its reverse reaction. Accepts a dictionary, vector, or tuple of variable-to-value mappings, e.g. [k1 => 1.0, k2 => 2.0,...].\n\n\n\n\n\n","category":"function"},{"location":"network_analysis/crn_theory/#Catalyst.robustspecies-network_analysis-crn_theory","page":"Chemical Reaction Network Theory","title":"Catalyst.robustspecies","text":"robustspecies(rn::ReactionSystem)\n\nReturn a vector of indices corresponding to species that are concentration robust, i.e. for every positive equilbrium, the concentration of species s will be the same.\n\nNote: This function currently only works for networks of deficiency one, and is not currently guaranteed to return *all* the concentration-robust species in the network. Any species returned by the function will be robust, but this may not include all of them. Use with caution. Support for higher deficiency networks and necessary conditions for robustness will be coming in the future.\n\n\n\n\n\n","category":"function"},{"location":"devdocs/dev_guide/#Catalyst-Developer-Documentation","page":"Developer Documentation","title":"Catalyst Developer Documentation","text":"","category":"section"},{"location":"devdocs/dev_guide/#devdocs_releaseprocess","page":"Developer Documentation","title":"Release Process","text":"Beginning with v15, Catalyst is using a new release process to try to ensure continuing stability of releases. Before making a release one should\n\nCreate a new release branch, i.e. \"release-15.0.0\".\nOn this branch, cap major dependencies to their latest version that works and for which tests pass.\nCaps need to be included in both Project.toml and docs/Project.toml.\nDo not cap the master branch as this can prevent upstream libraries from properly testing against Catalyst, and hide breaking changes that impact Catalyst.\nCheck docs build with the capped dependencies.  Visually verify via checking the artifact in the doc build that the docs actually look ok (since sometimes issues can arise that do not lead to actual errors in the doc CI).\nRelease via the registration issue with the command: @JuliaRegistrator register branch=release-15.0.0, modifying as appropriate for the version you are releasing.\n\nIf there is subsequently a need to increment the version of a dependency, this should be done via a new release that follows the above process, and modifies the patch, minor, or major Catalyst version (as appropriate for the potential impact of the dependency change on Catalyst users). If the dependency being updated is a non-breaking release, and would have automatically been installed by the package resolver had it not been capped, a patch release should be preferred. If the new release branch is branched from master, it needs to ensure Project.toml caps are all ≥ to those listed in the previous Catalyst release branch.","category":"section"},{"location":"devdocs/dev_guide/#devdocs_advice","page":"Developer Documentation","title":"Development advice","text":"","category":"section"},{"location":"devdocs/dev_guide/#devdocs_advice_doc_error_checks","page":"Developer Documentation","title":"Checking doc builds for errors","text":"When updating documentation, Catalyst will run any Julia code provided within example blocks to dynamically create figures and outputs. In addition to automatically creating these for us, it also provides an automatic check that all code in documentation is correct. Here, if any of the documentation code throws an error, the build job will fail. The documentation build job can be found at the bottom of a PRs conversation, here is an example of a failed one:\n\n(Image: Failed builddocs link)\n\nTo check what errors were produced, click on the \"Details\" link of the job. Next, any errors can be found at the bottom of the \"Build and deploy\" section (which should be opened automatically).","category":"section"},{"location":"devdocs/dev_guide/#devdocs_advice_doc_inspection","page":"Developer Documentation","title":"Inspecting the built documentation of a PR or branch","text":"When updating documentation it is typically useful to view the updated documentation in HTML format (which is the format users will see). Here, some errors are much easier to spot in .html format as compared with the raw text files from which these are generated. There are two primary ways to view updated documentation, either by downloading them from the PR or by building the docs locally.\n\nWhenever a PR to Catalyst is created, CI will create a corresponding documenter build job. If the build job passes, you can access the built documentation (which will be the new Catalyst documentation if the PR is merged) through the following steps:\n\nClick on \"Details\" in the documentation build job (at the bottom of the PR conversation tab). \nExpand the \"Upload site as artifact\" section.\nClick on the link at the end (which follows the \"Artifact download URL: \" text).\nThis will download a zip folder containing the documentation. Extract it to a location on your computer and then open the \"index.html\" file.\n\nTo build the Catalyst documentation locally:\n\nNavigate to the \".julia/dev/Catalyst/docs/\" folder, and run the \"make.jl\" file using \">julia –project=. make.jl\". Alternatively, open a Julia session, activate the \"docs\" environment, and run the file using `include(\"make.jl\").\nOpen the \".julia/dev/Catalyst/docs/build/index.html\" file.","category":"section"},{"location":"devdocs/dev_guide/#devdocs_advice_codespellchecker","page":"Developer Documentation","title":"Spellchecking in your code","text":"Especially when writing documentation, but also when writing normal code, it can be useful to have a spellchecker running through your texts. While code can be copied into a spellchecker and checked there (which is still useful to check grammar), it can also be very useful to (for users of VSCode) run the Code Spell Checker extension. This will automatically provide simple spell-checking for code and documentation as you write it.","category":"section"},{"location":"inverse_problems/structural_identifiability/#structural_identifiability","page":"Structural Identifiability Analysis","title":"Structural Identifiability Analysis","text":"<details><summary><strong>Environment setup and package installation</strong></summary>\n\nThe following code sets up an environment for running the code on this page.\n\nusing Pkg\nPkg.activate(; temp = true) # Creates a temporary environment, which is deleted when the Julia session ends.\nPkg.add(\"Catalyst\")\nPkg.add(\"StructuralIdentifiability\")\n\n</details>\n\n<details><summary><strong>Quick-start example</strong></summary>\n\nThe following code provides a brief example of how structural identifiability can be determined for a reaction network model using the StructuralIdentifiability.jl package.\n\nusing Catalyst, StructuralIdentifiability\ngwo = @reaction_network begin\n    (pₘ/(1+P), dₘ), 0 <--> M\n    (pₑ*M,dₑ), 0 <--> E\n    (pₚ*E,dₚ), 0 <--> P\nend\n\n# Asses global identifiability using the `assess_identifiability` function.\nassess_identifiability(gwo; measured_quantities = [:M])\n\n</details>\n\n\n  \n\nDuring parameter fitting, parameter values are inferred from data. Parameter identifiability refers to whether inferring parameter values for a given model is mathematically feasible. Ideally, parameter fitting should always be accompanied with an identifiability analysis of the problem. \n\nIdentifiability can be divided into structural and practical identifiability[1]. Structural identifiability considers only the mathematical model, and which parameters are and are not inherently identifiable due to model structure. Practical identifiability also considers the available data, and determines what system quantities can be inferred from it. In the idealised case of an infinite amount of non-noisy data, practical identifiability converges to structural identifiability. Generally, structural identifiability is assessed before parameters are fitted, while practical identifiability is assessed afterwards.\n\nStructural identifiability (which is what this tutorial considers) can be illustrated by the following differential equation: dx over dt = p1*p2*x(t) where, however much data is collected on x, it is impossible to determine the distinct values of p1 and p2. Hence, these parameters are non-identifiable (however, their product, p1*p2, is identifiable).\n\nCatalyst contains a special extension for carrying out structural identifiability analysis of generated reaction rate equation ODE models using the StructuralIdentifiability.jl package. This enables StructuralIdentifiability's assess_identifiability, assess_local_identifiability, and find_identifiable_functions functions to be called directly on Catalyst ReactionSystems. It also implements specialised routines to make these more efficient when applied to reaction network models (e.g. by improving runtimes). How to use these functions is described in the following tutorial, with StructuralIdentifiability providing a more extensive documentation. \n\nStructural identifiability can be divided into local and global identifiability. If a model quantity is locally identifiable, it means that its true value can be determined down to a finite-number of possible options. This also means that there is some limited region around the quantity's true value where this true value is the only possible value (and hence, within this region, the quantity is fully identifiable). Globally identifiable quantities' values, on the other hand, can be uniquely determined. Again, while identifiability can be confirmed structurally for a quantity, it does not necessarily mean that it is practically identifiable for some given data.\n\nGenerally, there are three types of quantities for which identifiability can be assessed.\n\nParameters (e.g. p1 and p2).\nFull variable trajectories (e.g. x(t)).\nVariable initial conditions (e.g. x(0)). \n\nStructuralIdentifiability currently assesses identifiability for the first two only (however, if x(t) is identifiable, then x(0) will be as well).\n\nnote: Note\nCurrently, the StructuralIdentifiability.jl extension only considers structural identifiability for the ODE generated by the reaction rate equation. It is possible that for the SDE model (generated by the chemical Langevin equation) and the jump model (generated by stochastic chemical kinetics) the identifiability of model quantities is different.","category":"section"},{"location":"inverse_problems/structural_identifiability/#structural_identifiability_gi","page":"Structural Identifiability Analysis","title":"Global identifiability analysis","text":"","category":"section"},{"location":"inverse_problems/structural_identifiability/#structural_identifiability_gi_example","page":"Structural Identifiability Analysis","title":"Basic example","text":"Global identifiability can be assessed using the assess_identifiability function. For each model quantity (parameters and variables), it will assess whether they are:\n\nGlobally identifiable.\nLocally identifiable.\nUnidentifiable.\n\nTo it, we provide our ReactionSystem model and a list of quantities that we are able to measure. Here, we consider a Goodwind oscillator (a simple 3-component model, where the three species M, E, and P are produced and degraded, which may exhibit oscillations)[2]. Let us say that we are able to measure the concentration of M, we then designate this using the measured_quantities argument. We can now assess identifiability in the following way:\n\nusing Catalyst, Logging, StructuralIdentifiability\ngwo = @reaction_network begin\n    (pₘ/(1+P), dₘ), 0 <--> M\n    (pₑ*M,dₑ), 0 <--> E\n    (pₚ*E,dₚ), 0 <--> P\nend\nassess_identifiability(gwo; measured_quantities = [:M], loglevel = Logging.Error)\n\nFrom the output, we find that E(t), pₑ, and pₚ (the trajectory of E, and the production rates of E and P, respectively) are non-identifiable. Next, dₑ and dₚ (the degradation rates of E and P, respectively) are locally identifiable. Finally, P(t), M(t), pₘ, and dₘ (the trajectories of P and M, and the production and degradation rate of M, respectively) are all globally identifiable. We note that we also imported the Logging.jl package, and provided the loglevel = Logging.Error input argument. StructuralIdentifiability functions generally provide a large number of output messages. Hence, we will use this argument (which requires the Logging package) throughout this tutorial to decrease the amount of printed text.\n\nNext, we also assess identifiability in the case where we can measure all three species concentrations:\n\nassess_identifiability(gwo; measured_quantities = [:M, :P, :E], loglevel = Logging.Error)\n\nin which case all species trajectories and parameters become identifiable.","category":"section"},{"location":"inverse_problems/structural_identifiability/#structural_identifiability_gi_known_ps","page":"Structural Identifiability Analysis","title":"Indicating known parameters","text":"In the previous case we assumed that all parameters are unknown, however, this is not necessarily true. If there are parameters with known values, we can supply these using the known_p argument. Providing this additional information might also make other, previously unidentifiable, parameters identifiable. Let us consider the previous example, where we measure the concentration of M only, but now assume we also know the production rate of E (pₑ):\n\nassess_identifiability(gwo; measured_quantities = [:M], known_p = [:pₑ], loglevel = Logging.Error)\n\nNot only does this turn the previously non-identifiable pₑ (globally) identifiable (which is obvious, given that its value is now known), but this additional information improve identifiability for several other network components.\n\nTo, in a similar manner, indicate that certain initial conditions are known is a work in progress. Hopefully this feature should be an available in the near future.","category":"section"},{"location":"inverse_problems/structural_identifiability/#structural_identifiability_gi_nontrivial_mq","page":"Structural Identifiability Analysis","title":"Providing non-trivial measured quantities","text":"Sometimes, ones may not have measurements of species, but rather some combinations of species (or possibly parameters). To account for this, measured_quantities accepts any algebraic expression (and not just single species). To form such expressions, species and parameters have to first be @unpack'ed from the model. Say that we have a model where an enzyme (E) is converted between an active and inactive form, which in turns activates the production of a product, P:\n\nrs = @reaction_network begin\n    (kA,kD), Eᵢ <--> Eₐ\n    (Eₐ, d), 0 <-->P\nend\n\nIf we can measure the total amount of E (=Eᵢ+Eₐ), as well as the amount of P, we can use the following to assess identifiability:\n\n@unpack Eᵢ, Eₐ = rs\nassess_identifiability(rs; measured_quantities = [Eᵢ + Eₐ, :P], loglevel = Logging.Error)\nnothing # hide","category":"section"},{"location":"inverse_problems/structural_identifiability/#structural_identifiability_gi_ftc","page":"Structural Identifiability Analysis","title":"Assessing identifiability for specified quantities only","text":"By default, StructuralIdentifiability assesses identifiability for all parameters and variables. It is, however, possible to designate precisely which quantities you want to check using the funcs_to_check option. This both includes selecting a smaller subset of parameters and variables to check, or defining customised expressions. Let us consider the Goodwind from previously, and say that we would like to check whether the production parameters (pₘ, pₑ, and pₚ) and the total amount of the three species (P + M + E) are identifiable quantities. Here, we would first unpack these (allowing us to form algebraic expressions) and then use the following code:\n\n@unpack pₘ, pₑ, pₚ, M, E, P = gwo\nassess_identifiability(gwo; measured_quantities = [:M], funcs_to_check = [pₘ, pₑ, pₚ, M + E + P], loglevel = Logging.Error)\nnothing # hide","category":"section"},{"location":"inverse_problems/structural_identifiability/#structural_identifiability_gi_probs","page":"Structural Identifiability Analysis","title":"Probability of correctness","text":"The identifiability methods used can, in theory, produce erroneous results. However, it is possible to adjust the lower bound for the probability of correctness using the argument prob_threshold (by default set to 0.99, that is, at least a 99 chance of correctness). We can e.g. increase the bound through:\n\nassess_identifiability(gwo; measured_quantities=[:M], prob_threshold = 0.999, loglevel = Logging.Error)\nnothing # hide\n\ngiving a minimum bound of 999 chance of correctness. In practise, the bounds used by StructuralIdentifiability are very conservative, which means that while the minimum guaranteed probability of correctness in the default case is 99, in practise it is much higher. While increasing the value of prob_threshold increases the certainty of correctness, it will also increase the time required to assess identifiability.","category":"section"},{"location":"inverse_problems/structural_identifiability/#structural_identifiability_lit","page":"Structural Identifiability Analysis","title":"Local identifiability analysis","text":"Local identifiability can be assessed through the assess_local_identifiability function. While this is already determined by assess_identifiability, assessing local identifiability only has the advantage that it is easier to compute. Hence, there might be models where global identifiability analysis fails (or takes a prohibitively long time), where instead assess_local_identifiability can be used. This function takes the same inputs as assess_identifiability and returns, for each quantity, true if it is locally identifiable (or false if it is not). Here, for the Goodwind oscillator, we assesses it for local identifiability only:\n\nassess_local_identifiability(gwo; measured_quantities = [:M], loglevel = Logging.Error)\n\nWe note that the results are consistent with those produced by assess_identifiability (with globally or locally identifiable quantities here all being assessed as at least locally identifiable).","category":"section"},{"location":"inverse_problems/structural_identifiability/#structural_identifiability_identifiabile_funcs","page":"Structural Identifiability Analysis","title":"Finding identifiable functions","text":"Finally, StructuralIdentifiability provides the find_identifiable_functions function. Rather than determining the identifiability of each parameter and unknown of the model, it finds a set of identifiable functions, such as any other identifiable expression of the model can be generated by these. Let us again consider the Goodwind oscillator, using the find_identifiable_functions function we find that identifiability can be reduced to five globally identifiable expressions:\n\nfind_identifiable_functions(gwo; measured_quantities = [:M], loglevel = Logging.Error)\n\nAgain, these results are consistent with those produced by assess_identifiability. There, pₑ and pₚ where found to be globally identifiable. Here, they correspond directly to identifiable expressions. The remaining four parameters (pₘ, dₘ, dₑ, and dₚ) occur as part of more complicated composite expressions.\n\nfind_identifiable_functions tries to simplify its output functions to create nice expressions. The degree to which it does this can be adjusted using the simplify keywords. Using the :weak, :standard (default), and :strong arguments, increased simplification can be forced (at the expense of longer runtime).","category":"section"},{"location":"inverse_problems/structural_identifiability/#structural_identifiability_si_odes","page":"Structural Identifiability Analysis","title":"Creating StructuralIdentifiability compatible ODE models from Catalyst ReactionSystems","text":"While the functionality described above covers the vast majority of analysis that user might want to perform, the StructuralIdentifiability package supports several additional features. While these does not have inherent Catalyst support, we do provide the make_si_ode function to simplify their use. Similar to the previous functions, it takes a ReactionSystem, lists of measured quantities, and known parameter values. The output is a ODE of the standard form supported by StructuralIdentifiability. It can be created using the following syntax:\n\nsi_ode = make_si_ode(gwo; measured_quantities = [:M])\nnothing # hide\n\nand then used as input to various StructuralIdentifiability functions. In the following example we use StructuralIdentifiability's print_for_DAISY function, printing the model as an expression that can be used by the DAISY software for identifiability analysis[3].\n\nprint_for_DAISY(si_ode)\nnothing # hide","category":"section"},{"location":"inverse_problems/structural_identifiability/#structural_identifiability_conslaws","page":"Structural Identifiability Analysis","title":"Notes on systems with conservation laws","text":"Several reaction network models, such as\n\nusing Catalyst, Logging, StructuralIdentifiability # hide\nrs = @reaction_network begin\n    (k1,k2), X1 <--> X2\nend\n\ncontain conservation laws (in this case Γ = X1 + X2, where Γ = X1(0) + X2(0) is a constant). Because the presence of such conservation laws makes structural identifiability analysis prohibitively computationally expensive (for all but the simplest of cases), these are automatically eliminated by Catalyst. This is handled internally, and should not be noticeable to the user. The exception is the make_si_ode function. For each conservation law, its output will have one ODE removed, and instead have a conservation parameter (of the form Γ[i]) added to its equations. This feature can be disabled through the remove_conserved = false option.","category":"section"},{"location":"inverse_problems/structural_identifiability/#structural_identifiability_exp_params","page":"Structural Identifiability Analysis","title":"Systems with exponent parameters","text":"Structural identifiability cannot currently be applied to systems with parameters (or species) in exponents. E.g. this\n\nrn_si_impossible = @reaction_network begin\n    (hill(X,v,K,n),d), 0 <--> X\nend\nassess_identifiability(rn; measured_quantities = [:X])\n\nis currently not possible. Hopefully this will be a supported feature in the future. For now, such expressions will have to be rewritten to not include such exponents. For some cases, e.g. 10^k this is trivial. However, it is also possible generally (but more involved and often includes introducing additional variables). \n\n","category":"section"},{"location":"inverse_problems/structural_identifiability/#structural_identifiability_citation","page":"Structural Identifiability Analysis","title":"Citation","text":"If you use this functionality in your research, please cite the following paper to support the authors of the StructuralIdentifiability package:\n\n@article{structidjl,\n  author  = {Dong, R. and Goodbrake, C. and Harrington, H. and Pogudin G.},\n  title   = {Differential Elimination for Dynamical Models via Projections with Applications to Structural Identifiability},\n  journal = {SIAM Journal on Applied Algebra and Geometry},\n  url     = {https://doi.org/10.1137/22M1469067},\n  year    = {2023}\n  volume  = {7},\n  number  = {1},\n  pages   = {194-235}\n}\n\n","category":"section"},{"location":"inverse_problems/structural_identifiability/#References","page":"Structural Identifiability Analysis","title":"References","text":"[1]: Guillaume H.A. Joseph et al., Introductory overview of identifiability analysis: A guide to evaluating whether you have the right type of data for your modeling purpose, Environmental Modelling & Software (2019).\n\n[2]: Goodwin B.C., Oscillatory Behavior in Enzymatic Control Processes, Advances in Enzyme Regulation (1965).\n\n[3]: Bellu G., et al., DAISY: A new software tool to test global identifiability of biological and physiological systems, Computer Methods and Programs in Biomedicine (2007).","category":"section"},{"location":"model_simulation/sde_simulation_performance/#sde_simulation_performance","page":"Advice for Performant SDE Simulations","title":"Advice for Performant SDE Simulations","text":"<details><summary><strong>Environment setup and package installation</strong></summary>\n\nThe following code sets up an environment for running the code on this page.\n\nusing Pkg\nPkg.activate(; temp = true) # Creates a temporary environment, which is deleted when the Julia session ends.\nPkg.add(\"Catalyst\")\nPkg.add(\"CUDA\")\nPkg.add(\"DiffEqGPU\")\nPkg.add(\"StaticArrays\")\nPkg.add(\"StochasticDiffEq\")\n\n</details>\n\n\n   While there exist relatively straightforward approaches to manage performance for ODE and jump simulations, this is generally not the case for SDE simulations. Below, we briefly describe some options. However, as one starts to investigate these, one quickly reaches what is (or could be) active areas of research.","category":"section"},{"location":"model_simulation/sde_simulation_performance/#sde_simulation_performance_solvers","page":"Advice for Performant SDE Simulations","title":"SDE solver selection","text":"We have previously described how ODE solver selection can impact simulation performance. Again, it can be worthwhile to investigate solver selection's impact on performance for SDE simulations. Throughout this documentation, we generally use the STrapezoid solver as the default choice. However, if the DifferentialEquations package is loaded\n\nusing DifferentialEquations\n\nautomatic SDE solver selection enabled (just like is the case for ODEs by default). Generally, the automatic SDE solver choice enabled by DifferentialEquations is better than just using STrapezoid. Next, if performance is critical, it can be worthwhile to check the list of available SDE solvers to find one with advantageous performance for a given problem. When doing so, it is important to pick a solver compatible with non-diagonal noise and with Ito problems. ","category":"section"},{"location":"model_simulation/sde_simulation_performance/#sde_simulation_performance_jacobian","page":"Advice for Performant SDE Simulations","title":"Options for Jacobian computation","text":"In the section on ODE simulation performance, we describe various options for computing the system Jacobian, and how these could be used to improve performance for implicit solvers. These can be used in tandem with implicit SDE solvers (such as STrapezoid). However, due to additional considerations during SDE simulations, it is much less certain whether these will actually have any impact on performance. So while these options might be worth reading about and trialling, there is no guarantee that they will be beneficial.","category":"section"},{"location":"model_simulation/sde_simulation_performance/#sde_simulation_performance_parallelisation","page":"Advice for Performant SDE Simulations","title":"Parallelisation on CPUs and GPUs","text":"We have previously described how simulation parallelisation can be used to improve performance when multiple ODE simulations are carried out. The same approaches can be used for SDE simulations. Indeed, it is often more relevant for SDEs, as these are often re-simulated using identical simulation conditions (to investigate their typical behaviour across many samples). CPU parallelisation of SDE simulations uses the same approach as ODEs. GPU parallelisation requires some additional considerations, which are described below.","category":"section"},{"location":"model_simulation/sde_simulation_performance/#sde_simulation_performance_parallelisation_GPU","page":"Advice for Performant SDE Simulations","title":"GPU parallelisation of SDE simulations","text":"GPU parallelisation of SDE simulations uses a similar approach as that for ODE simulations. The main differences are that SDE parallelisation requires a GPU SDE solver (like GPUEM) and fixed time stepping.\n\nWe will assume that we are using the CUDA GPU hardware, so we will first load the CUDA.jl backend package, as well as DiffEqGPU:\n\nusing CUDA, DiffEqGPU\n\nWhich backend package you should use depends on your available hardware, with the alternatives being listed here.\n\nNext, we create the SDEProblem which we wish to simulate. Like for ODEs, we ensure that all vectors are static vectors and that all values are Float32s. Here we prepare the parallel simulations of a simple birth-death process.\n\nusing Catalyst, StochasticDiffEq, StaticArrays\nbd_model = @reaction_network begin\n    (p,d), 0 <--> X\nend\n@unpack X, p, d = bd_model\n\nu0 = @SVector [X => 20.0f0]\ntspan = (0.0f0, 10.0f0)\nps = @SVector [p => 10.0f0, d => 1.0f0]\nsprob = SDEProblem(bd_model, u0, tspan, ps)\nnothing # hide\n\nThe SDEProblem is then used to create an EnsembleProblem.\n\neprob = EnsembleProblem(sprob)\nnothing # hide\n\nFinally, we can solve our EnsembleProblem while:\n\nUsing a valid GPU SDE solver (either GPUEM or GPUSIEA).\nDesignating the GPU ensemble method, EnsembleGPUKernel (with the correct GPU backend as input).\nDesignating the number of trajectories we wish to simulate.\nDesignating a fixed time step size.\n\nesol = solve(eprob, GPUEM(), EnsembleGPUKernel(CUDA.CUDABackend()); trajectories = 10000, dt = 0.01)\n\nAbove we parallelise GPU simulations with identical initial conditions and parameter values. However, varying these is also possible.","category":"section"},{"location":"model_simulation/sde_simulation_performance/#sde_simulation_performance_parallelisation_mlmc","page":"Advice for Performant SDE Simulations","title":"Multilevel Monte Carlo","text":"An approach for speeding up parallel stochastic simulations is so-called multilevel Monte Carlo approaches (MLMC). These are used when a stochastic process is simulated repeatedly using identical simulation conditions. Here, instead of performing all simulations using identical tolerance, the ensemble is simulated using a range of tolerances (primarily lower ones, which yields faster simulations). Currently, StochasticDiffEq.jl do not have a native implementation for performing MLMC simulations (this will hopefully be added in the future). However, if high performance of parallel SDE simulations is required, these approaches may be worth investigating.","category":"section"},{"location":"steady_state_functionality/nonlinear_solve/#steady_state_solving","page":"Finding Steady States using NonlinearSolve.jl and SteadyStateDiffEq.jl","title":"Finding Steady States using NonlinearSolve.jl and SteadyStateDiffEq.jl","text":"<details><summary><strong>Environment setup and package installation</strong></summary>\n\nThe following code sets up an environment for running the code on this page.\n\nusing Pkg\nPkg.activate(; temp = true) # Creates a temporary environment, which is deleted when the Julia session ends.\nPkg.add(\"Catalyst\")\nPkg.add(\"NonlinearSolve\")\nPkg.add(\"OrdinaryDiffEqRosenbrock\")\nPkg.add(\"SteadyStateDiffEq\")\n\n</details>\n\n<details><summary><strong>Quick-start example</strong></summary>\n\nThe following code provides a brief example of how a single system steady state can be found using the NonlinearSolve.jl package.\n\nusing Catalyst\nrn = @reaction_network begin \n    (p,d), 0 <--> X\nend\nps = [:p => 2.0, :d => 0.5] # The parameter set for which we want to find a steady state.\nu_guess = [:X => 1.0] # For single steady state models, this value have little impact on the result.\nnlprob = NonlinearProblem(rn, u_guess, ps)\nsteady_state = solve(nlprob)\n\n</details>\n\n\n  \n\nCatalyst ReactionSystem models can be converted to ODEs (through the reaction rate equation). We have previously described how these ODEs' steady states can be found through homotopy continuation. Generally, homotopy continuation (due to its ability to find all of a system's steady states) is the preferred approach. However, Catalyst supports two additional approaches for finding steady states:\n\nThrough solving the nonlinear system produced by setting all ODE differentials to 0[1].\nThrough forward ODE simulation from an initial condition until a steady state has been reached.\n\nWhile these approaches only find a single steady state, they offer two advantages as compared to homotopy continuation:\n\nThey are typically much faster.\nThey can find steady states for models that do not produce multivariate, rational, polynomial systems (which is a requirement for homotopy continuation to work). Examples include models with non-integer hill coefficients.\n\nIn practice, model steady states are found through nonlinear system solving by creating a NonlinearProblem, and through forward ODE simulation by creating a SteadyStateProblem. These are then solved through solvers implemented in the NonlinearSolve.jl, package (with the latter approach also requiring the SteadyStateDiffEq.jl package). This tutorial describes how to find steady states through these two approaches. More extensive descriptions of available solvers and options can be found in NonlinearSolve's documentation.","category":"section"},{"location":"steady_state_functionality/nonlinear_solve/#steady_state_solving_nonlinear","page":"Finding Steady States using NonlinearSolve.jl and SteadyStateDiffEq.jl","title":"Steady state finding through nonlinear solving","text":"Let us consider a simple dimerisation system, where a protein (P) can exist in a monomer and a dimer form. The protein is produced at a constant rate from its mRNA, which is also produced at a constant rate.\n\nusing Catalyst\ndimer_production = @reaction_network begin\n    pₘ, 0 --> mRNA\n    pₚ, mRNA --> mRNA + P\n    (k₁, k₂), 2P <--> P₂\n    d, (mRNA, P, P₂) --> 0\nend\n\nThis system corresponds to the following ODE:\n\nbeginaligned\nfracdmRNAdt = pₘ - d cdot mRNA \nfracdPdt = pₚ cdot mRNA - k₁ cdot P + 2k₂ cdot P₂ - d cdot P \nfracdP₂dt = k₁ cdot P + 2k₂ cdot P₂ \nendaligned\n\nTo find its steady states we need to solve:\n\nbeginaligned\n0 = pₘ - d cdot mRNA \n0 = pₚ cdot mRNA - k₁ cdot P + 2k₂ cdot P₂ - d cdot P \n0 = k₁ cdot P + 2k₂ cdot P₂ \nendaligned\n\nTo solve this problem, we must first designate our parameter values, and also make an initial guess of the solution. Generally, for problems with a single solution (like this one), most arbitrary guesses will work fine (the exception typically being systems with conservation laws). Using these, we can create the NonlinearProblem that we wish to solve.\n\np = [:pₘ => 0.5, :pₚ => 2.0, :k₁ => 5.0, :k₂ => 1.0, :d => 1.0]\nu_guess = [:mRNA => 1.0, :P => 1.0, :P₂ => 1.0]\nnlprob = NonlinearProblem(dimer_production, u_guess, p)\nnothing # hide\n\nFinally, we can solve it using the solve command, returning the steady state solution:\n\nusing NonlinearSolve\nsol = solve(nlprob)\n\nTypically, a good default method is automatically selected for any problem. However, NonlinearSolve does provide a wide range of potential solvers. If we wish to designate one, it can be supplied as a second argument to solve. Here, we use the Newton Trust Region method, and then check that the solution is equal to the previous one.\n\nsol_ntr = solve(nlprob, TrustRegion())\nsol ≈ sol_ntr","category":"section"},{"location":"steady_state_functionality/nonlinear_solve/#steady_state_solving_nonlinear_conservation_laws","page":"Finding Steady States using NonlinearSolve.jl and SteadyStateDiffEq.jl","title":"Systems with conservation laws","text":"As described in the section on homotopy continuation, when finding the steady states of systems with conservation laws, additional considerations have to be taken. E.g. consider the following two-state system:\n\nusing Catalyst, NonlinearSolve # hide\ntwo_state_model = @reaction_network begin\n    (k1,k2), X1 <--> X2\nend\n\nIt has an infinite number of steady states. To make steady state finding possible, information of the system's conserved quantities (here C = X1 + X2) must be provided. Since these can be computed from system initial conditions (u0, i.e. those provided when performing ODE simulations), designating an u0 is often the best way. There are two ways to do this. First, one can perform forward ODE simulation-based steady state finding, using the initial condition as the initial u guess. Alternatively, any conserved quantities can be eliminated when the NonlinearProblem is created. This feature is supported by Catalyst's conservation law finding and elimination feature.\n\nTo eliminate conservation laws we simply provide the remove_conserved = true argument to NonlinearProblem:\n\np = [:k1 => 2.0, :k2 => 3.0]\nu_guess = [:X1 => 3.0, :X2 => 1.0]\nnl_prob = NonlinearProblem(two_state_model, u_guess, p; remove_conserved = true)\nnothing # hide\n\nhere it is important that the quantities used in u_guess correspond to the conserved quantities we wish to use. E.g. here the conserved quantity X1 + X2 = 30 + 10 = 4 holds for the initial condition, and will hence also hold in the computed steady state as well. We can now find the steady states using solve like before:\n\nsol = solve(nl_prob)\n\nWe note that the output only provides a single value. The reason is that the actual system solved only contains a single equation (the other being eliminated with the conserved quantity). To find the values of X1 and X2 we can directly query the solution object for these species' values, using the species themselves as inputs:\n\nsol[[:X1, :X2]]","category":"section"},{"location":"steady_state_functionality/nonlinear_solve/#steady_state_solving_simulation","page":"Finding Steady States using NonlinearSolve.jl and SteadyStateDiffEq.jl","title":"Finding steady states through ODE simulations","text":"The NonlinearProblems generated by Catalyst corresponds to ODEs. A common method of solving these is to simulate the ODE from an initial condition until a steady state is reached. Here we do so for the dimerisation system considered in the previous section. First, we declare our model, initial condition, and parameter values.\n\nusing Catalyst # hide\ndimer_production = @reaction_network begin\n    pₘ, 0 --> mRNA\n    pₚ, mRNA --> mRNA + P\n    (k₁, k₂), 2P <--> P₂\n    d, (mRNA, P, P₂) --> 0\nend\np = [:pₘ => 0.5, :pₚ => 2.0, :k₁ => 5.0, :k₂ => 1.0, :d => 1.0]\nu0 = [:mRNA => 0.1, :P => 0.0, :P₂ => 0.0]\nnothing # hide\n\nNext, we provide these as an input to a SteadyStateProblem\n\nssprob = SteadyStateProblem(dimer_production, u0, p)\nnothing # hide\n\nFinally, we can find the steady states using the solver command. Since SteadyStateProblems are solved through forward ODE simulation, we must load the sublibrary of the OrdinaryDiffEq.jl package that corresponds to the selected ODE solver. Any available ODE solver can be used, however, it has to be encapsulated by the DynamicSS() function. E.g. here we use the Rodas5P solver which is loaded from the OrdinaryDiffEqRosenbrock sublibrary:\n\n(which requires loading the SteadyStateDiffEq package).\n\nusing SteadyStateDiffEq, OrdinaryDiffEqRosenbrock\nsolve(ssprob, DynamicSS(Rodas5P()))\n\nNote that, unlike for nonlinear system solving, u0 is not just an initial guess of the solution, but the initial conditions from which the steady state simulation is carried out. This means that, for a system with multiple steady states, we can determine the steady states associated with specific initial conditions (which is not possible when the nonlinear solving approach is used). This also permits us to easily handle the presence of conservation laws. The forward ODE simulation approach (unlike homotopy continuation and nonlinear solving) cannot find unstable steady states.\n\nGenerally, SteadyStateProblems can be solved using the same options that are available for ODE simulations. E.g. here we designate a specific dt step size:\n\nsolve(ssprob, DynamicSS(Rodas5P()); dt = 0.01)\nnothing # hide\n\nIt is possible to use solve SteadyStateProblems using a nonlinear solver, and NonlinearProblems using forward ODE simulation solvers:\n\nusing NonlinearSolve\nsolve(ssprob, TrustRegion())\nnothing # hide\n\nnlprob = NonlinearProblem(dimer_production, u0, p)\nsolve(nlprob, DynamicSS(Rodas5P()))\nnothing # hide\n\nHowever, especially when the forward ODE simulation approach is used, it is recommended to use the problem type which corresponds to the intended solver.\n\n","category":"section"},{"location":"steady_state_functionality/nonlinear_solve/#nonlinear_solve_citation","page":"Finding Steady States using NonlinearSolve.jl and SteadyStateDiffEq.jl","title":"Citations","text":"If you use this functionality in your research, in addition to Catalyst, please cite the following paper to support the authors of the NonlinearSolve.jl package:\n\n@article{pal2024nonlinearsolve,\n  title={NonlinearSolve. jl: High-Performance and Robust Solvers for Systems of Nonlinear Equations in Julia},\n  author={Pal, Avik and Holtorf, Flemming and Larsson, Axel and Loman, Torkel and Schaefer, Frank and Qu, Qingyu and Edelman, Alan and Rackauckas, Chris and others},\n  journal={arXiv preprint arXiv:2403.16341},\n  year={2024}\n}\n\n","category":"section"},{"location":"steady_state_functionality/nonlinear_solve/#References","page":"Finding Steady States using NonlinearSolve.jl and SteadyStateDiffEq.jl","title":"References","text":"[1]: J. Nocedal, S. J. Wright, Numerical Optimization, Springer (2006).","category":"section"},{"location":"model_simulation/simulation_introduction/#simulation_intro","page":"Model Simulation Introduction","title":"Model Simulation Introduction","text":"<details><summary><strong>Environment setup and package installation</strong></summary>\n\nThe following code sets up an environment for running the code on this page.\n\nusing Pkg\nPkg.activate(; temp = true) # Creates a temporary environment, which is deleted when the Julia session ends.\nPkg.add(\"Catalyst\")\nPkg.add(\"JumpProcesses\")\nPkg.add(\"OrdinaryDiffEqDefault\")\nPkg.add(\"OrdinaryDiffEqRosenbrock\")\nPkg.add(\"OrdinaryDiffEqTsit5\")\nPkg.add(\"Plots\")\nPkg.add(\"StochasticDiffEq\")\n\n</details>\n\n<details><summary><strong>Quick-start example</strong></summary>\n\nThe following code provides brief examples of the three main types of simulation.\n\nusing Catalyst, JumpProcesses, OrdinaryDiffEqDefault, StochasticDiffEq\n\n# First we designate our model, initial condition, time span, and parameter values.\nrn = @reaction_network begin\n    (p,d), 0 <--> X\nend\nu0 = [:X => 2]\ntspan = (0.0, 10.0)\nps = [:p => 5.0, :d => 0.25]\n# Now, for each simulation type, we first bundle these into a `XProblem`. \n# This can be simulated using `solve` and the solution plotted using `plot`.\n\n# ODE simulation.\noprob = ODEProblem(rn, u0, tspan, ps)\nosol = solve(oprob)\nplot(osol)\n\n# SDE simulation.\nsprob = SDEProblem(rn, u0, tspan, ps)\nssol = solve(sprob)\nplot(ssol)\n\n# Jump simulation.\njprob = JumpProblem(rn, u0, tspan, ps)\njsol = solve(jprob)\nplot(jsol)\n\n</details>\n\n\n   Catalyst's core functionality is the creation of chemical reaction network (CRN) models that can be simulated using ODE, SDE, and jump simulations. How such simulations are carried out has already been described in Catalyst's introduction. This page provides a deeper introduction, giving some additional background and introducing various simulation-related options.\n\nHere we will focus on the basics, with other sections of the simulation documentation describing various specialised features, or giving advice on performance. Anyone who plans on using Catalyst's simulation functionality extensively is recommended to also read the documentation on solution plotting, and on how to interact with simulation problems, integrators, and solutions. Anyone with an application for which performance is critical should consider reading the corresponding page on performance advice for ODEs or SDEs.","category":"section"},{"location":"model_simulation/simulation_introduction/#simulation_intro_theory","page":"Model Simulation Introduction","title":"Background to CRN simulations","text":"This section provides some brief theory on CRN simulations. For details on how to carry out these simulations in actual code, please skip to the following sections.\n\nCRNs are defined by a set of species (with the amounts of these determining the system's state during simulations) and a set of reaction events (rules for how the state of the system changes). In real systems, the species amounts are discrete copy-numbers, describing the exact numbers of each species type present in the system (in systems biology this can e.g. be the number of a specific molecule present in a cell). Given rates for these reaction events, stochastic chemical kinetics provides a formula for simulating the system that recreates its real reaction process. During stochastic chemical kinetics simulations, the system's state is defined by discrete copy-numbers (denoting the number of each species present in the system). Next, at the occurrence of individual reaction events, the system's state is updated according to the occurred reaction. The result is a stochastic process. The most well-known approach for simulating stochastic chemical kinetics is Gillespie's algorithm.\n\nIn practice, these jump simulations are computationally expensive. In many cases, copy-numbers are so large that they can be approximated as continuous concentrations, and the time-development of the system as a deterministic process. This creates an ordinary differential equation (ODE), and is the chemical reaction network form most people are most familiar with. The rule for how ODEs are generated from CRNs is called the reaction rate equation (RRE).\n\nHere, the RRE enables fast, approximate, and deterministic simulations of CRNs, while stochastic chemical kinetics enables exact, stochastic, simulations of the true process. An intermediary approach is to use the chemical Langevin equation (CLE) to formulate a stochastic differential equation (SDE). This approximates the system's state as continuous concentrations, but does not assume that its time development is deterministic. Generally, the CLE is used when copy-numbers are large enough that the continuous approximation holds, but not so large that the system's behaviour is deterministic. Generally, the advantage of SDE simulations (compared to jump ones) is that they are faster. Also, since the system state is continuous, interpretation of e.g. stability and steady state results from the deterministic (also continuous) domain is easier for SDEs (however one should be careful when making such interpretations).\n\nThese three different approaches are summed up in the following table:\n\n<style type=\"text/css\">\n.tg  {border-collapse:collapse;border-spacing:0;}\n.tg td{border-color:black;border-style:solid;border-width:1px;font-family:Arial, sans-serif;font-size:14px;\n  overflow:hidden;padding:10px 5px;word-break:normal;}\n.tg th{border-color:black;border-style:solid;border-width:1px;font-family:Arial, sans-serif;font-size:14px;\n  font-weight:normal;overflow:hidden;padding:10px 5px;word-break:normal;}\n.tg .tg-0pky{border-color:inherit;text-align:left;vertical-align:top}\n</style>\n<table class=\"tg\">\n<thead>\n  <tr>\n    <th class=\"tg-0pky\">Interpretation</th>\n    <th class=\"tg-0pky\">Reaction rate equation</th>\n    <th class=\"tg-0pky\">Chemical Langevin equation</th>\n    <th class=\"tg-0pky\">Stochastic chemical kinetics</th>\n  </tr>\n</thead>\n<tbody>\n  <tr>\n    <td class=\"tg-0pky\">Simulation form</td>\n    <td class=\"tg-0pky\">ODE simulations</td>\n    <td class=\"tg-0pky\">SDE simulations</td>\n    <td class=\"tg-0pky\">Jump simulations</td>\n  </tr>\n  <tr>\n    <td class=\"tg-0pky\">Example simulation methods</td>\n    <td class=\"tg-0pky\"><a href=\"https://en.wikipedia.org/wiki/Euler_method\">Euler</a>, <a href=\"https://en.wikipedia.org/wiki/Runge%E2%80%93Kutta_methods\">Runge-Kutta</a></td>\n    <td class=\"tg-0pky\"><a href=\"https://en.wikipedia.org/wiki/Euler%E2%80%93Maruyama_method\">Euler-Maruyama</a>, <a href=\"https://en.wikipedia.org/wiki/Milstein_method\">Milstein</a></td>\n    <td class=\"tg-0pky\"><a href=\"https://en.wikipedia.org/wiki/Gillespie_algorithm\">Gillespie</a>, <a href=\"https://pubmed.ncbi.nlm.nih.gov/16321569/\">Sorting direct</a></td>\n  </tr>\n  <tr>\n    <td class=\"tg-0pky\">Species units</td>\n    <td class=\"tg-0pky\">Concentration</td>\n    <td class=\"tg-0pky\">Concentration</td>\n    <td class=\"tg-0pky\">Copy-numbers</td>\n  </tr>\n  <tr>\n    <td class=\"tg-0pky\">Deterministic/Stochastic</td>\n    <td class=\"tg-0pky\">Deterministic</td>\n    <td class=\"tg-0pky\">Stochastic</td>\n    <td class=\"tg-0pky\">Stochastic</td>\n  </tr>\n  <tr>\n    <td class=\"tg-0pky\">Applicability</td>\n    <td class=\"tg-0pky\">Large species amounts</td>\n    <td class=\"tg-0pky\"><span style=\"font-weight:400;font-style:normal\">Non-small species amounts</span></td>\n    <td class=\"tg-0pky\">Any species amounts</td>\n  </tr>\n  <tr>\n    <td class=\"tg-0pky\">Speed</td>\n    <td class=\"tg-0pky\">Typically fast</td>\n    <td class=\"tg-0pky\">Typically intermediate</td>\n    <td class=\"tg-0pky\">Typically slow</td>\n  </tr>\n  <tr>\n    <td class=\"tg-0pky\">Simulation package</td>\n    <td class=\"tg-0pky\"><a href=\"https://github.com/SciML/OrdinaryDiffEq.jl\">OrdinaryDiffEq.jl</a></td>\n    <td class=\"tg-0pky\"><a href=\"https://github.com/SciML/StochasticDiffEq.jl\">StochasticDiffEq.jl</a></td>\n    <td class=\"tg-0pky\"><a href=\"https://github.com/SciML/JumpProcesses.jl\">JumpProcesses.jl</a></td>\n  </tr>\n</tbody>\n</table>","category":"section"},{"location":"model_simulation/simulation_introduction/#simulation_intro_ODEs","page":"Model Simulation Introduction","title":"Performing (ODE) simulations","text":"The following section gives a (more complete introduction of how to simulate Catalyst models than our previous introduction). This is illustrated using ODE simulations (some ODE-specific options will also be discussed). Later on, we will describe things specific to SDE and jump simulations. All ODE simulations are performed using the OrdinaryDiffEq.jl package, which full documentation can be found here. A dedicated section giving advice on how to optimise ODE simulation performance can be found here\n\nTo perform any simulation, we must first define our model, as well as the simulation's initial conditions, time span, and parameter values. Here we will use a simple two-state model:\n\nusing Catalyst\ntwo_state_model = @reaction_network begin\n    (k1,k2), X1 <--> X2\nend\nu0 = [:X1 => 100.0, :X2 => 200.0]\ntspan = (0.0, 5.0)\nps = [:k1 => 2.0, :k2 => 5.0]\nnothing # hide\n\nTo simulate the model we first bundle these up into an ODEProblem:\n\noprob = ODEProblem(two_state_model, u0, tspan, ps)\nnothing # hide\n\nNext, we can simulate the model (requires loading the OrdinaryDiffEq.jl package). Simulations are performed using the solve function.\n\nusing OrdinaryDiffEqDefault\nsol = solve(oprob)\nnothing # hide\n\nFinally, the result can be plotted using the Plots.jl package's plot function:\n\nusing Plots\nplot(sol)\n\nMore information on how to interact with solution structures is provided here and on how to plot them here.\n\nSome additional considerations:\n\nIf a model without parameters has been declared, only the first three arguments must be provided to ODEProblem.\nWhile the first value of tspan will almost always be 0.0, other starting times (both negative and positive) are possible.","category":"section"},{"location":"model_simulation/simulation_introduction/#simulation_intro_solver_options","page":"Model Simulation Introduction","title":"Designating solvers and solver options","text":"While good defaults are generally selected, OrdinaryDiffEq enables the user to customise simulations through a long range of options that can be provided to the solve function. This includes specifying a solver algorithm, which can be provided as a second argument to solve (if none is provided, a suitable choice is automatically made). E.g. here we specify that the Rodas5P method should be used:\n\nusing OrdinaryDiffEqRosenbrock\nsol = solve(oprob, Rodas5P())\nnothing # hide\n\nA full list of available solvers is provided here, and a discussion on optimal solver choices here.\n\nnote: Note\nUnlike most other libraries, OrdinaryDiffEq is split into multiple libraries. This is due to it implementing a large number of ODE solvers (most of which a user will not use). Splitting the library improves its loading times. At the highest level, there is OrdinaryDiffEq.jl (imported through using OrdinaryDiffEq). This exports all solvers, and is primarily useful if you want to try a wide range of different solvers for a specific problem. Next there is OrdinaryDiffEqDefault.jl (imported through using OrdinaryDiffEq). This exports the automated default solver (which selects a solver for the user). It is likely the best one to use for simple workflows. Then there are multiple solver-specific libraries, such as OrdinaryDiffEqTsit5.jl and OrdinaryDiffEqRosenbrock.jl (a full list can be found here). Each of these exports a specific set of solvers, and are useful if you know in advance which solver you wish to use.\n\nAdditional options can be provided as keyword arguments. E.g. the maxiters arguments determines the maximum number of simulation time steps (before the simulation is terminated). This defaults to 1e5, but can be modified through:\n\nsol = solve(oprob; maxiters = 1e4)\nnothing # hide\n\nHere follows a list of solver options which might be of interest to the user.\n\nadaptive: Toggles adaptive time stepping for valid methods. Default to true.\ndt: For non-adaptive simulations, sets the step size (also sets the initial step size for adaptive methods).\nsaveat: Determines the time points at which the simulation is saved. E.g. for saveat = 2.0 the simulation is saved every second time unit. If not given, the solution is saved after each time step.\nsave_idxs: Provides a vector of species whose values should be saved during the simulation. E.g. for save_idxs = [:X1], only the value of species X1 is saved.\nmaxiters: The maximum number of time steps of the simulation. If this number is reached, the simulation is terminated.\nseed: Sets a seed for stochastic simulations. Stochastic simulations with the same seed generate identical results.\n\nA full list of solver options can be found here.","category":"section"},{"location":"model_simulation/simulation_introduction/#simulation_intro_ODEs_input_forms","page":"Model Simulation Introduction","title":"Alternative problem input forms","text":"Throughout Catalyst's documentation, we typically provide initial condition and parameter values as vectors. However, these can also be provided as tuples:\n\nu0 = (:X1 => 100.0, :X2 => 200.0)\ntspan = (0.0, 5.0)\nps = (:k1 => 2.0, :k2 => 5.0)\noprob = ODEProblem(two_state_model, u0, tspan, ps)\nnothing # hide\n\nor dictionaries:\n\nu0 = Dict([:X1 => 100.0, :X2 => 200.0])\ntspan = (0.0, 5.0)\nps = Dict([:k1 => 2.0, :k2 => 5.0])\noprob = ODEProblem(two_state_model, u0, tspan, ps)\nnothing # hide\n\nThe forms used for u0 and ps does not need to be the same (but can e.g. be a vector and a tuple).\n\nnote: Note\nIt is possible to designate specific types for parameters. When this is done, the tuple form for providing parameter values should be preferred.\n\nThroughout Catalyst's documentation, we typically provide the time span as a tuple. However, if the first time point is 0.0 (which is typically the case), this can be omitted. Here, we supply only the simulation endpoint to our ODEProblem:\n\ntend = 5.0\noprob = ODEProblem(two_state_model, u0, tend, ps)\nnothing # hide","category":"section"},{"location":"model_simulation/simulation_introduction/#simulation_intro_SDEs","page":"Model Simulation Introduction","title":"Performing SDE simulations","text":"Catalyst uses the StochasticDiffEq.jl package to perform SDE simulations. This section provides a brief introduction, with StochasticDiffEq's documentation providing a more extensive description. By default, Catalyst generates SDEs from CRN models using the chemical Langevin equation. A dedicated section giving advice on how to optimise SDE simulation performance can be found here.\n\nSDE simulations are performed in a similar manner to ODE simulations. The only exception is that an SDEProblem is created (rather than an ODEProblem). Furthermore, the StochasticDiffEq.jl package (rather than the OrdinaryDiffEq package) is required for performing simulations. Here we simulate the two-state model for the same parameter set as previously used:\n\nusing Catalyst, StochasticDiffEq, Plots\ntwo_state_model = @reaction_network begin\n    (k1,k2), X1 <--> X2\nend\nu0 = [:X1 => 100.0, :X2 => 200.0]\ntspan = (0.0, 1.0)\nps = [:k1 => 2.0, :k2 => 5.0]\n\nsprob = SDEProblem(two_state_model, u0, tspan, ps)\nsol = solve(sprob, STrapezoid())\nsol = solve(sprob, STrapezoid(); seed = 123) # hide\nplot(sol)\n\nwe can see that while this simulation (unlike the ODE ones) exhibits some fluctuations.\n\nnote: Note\nUnlike for ODE and jump simulations, there are no good heuristics for automatically selecting suitable SDE solvers. Hence, for SDE simulations a solver must be provided. STrapezoid will work for a large number of cases. When this is not the case, however, please check the list of available SDE solvers for a suitable alternative (making sure to select one compatible with non-diagonal noise and the Ito interpretation.","category":"section"},{"location":"model_simulation/simulation_introduction/#simulation_intro_SDEs_pitfalls","page":"Model Simulation Introduction","title":"Common SDE simulation pitfalls","text":"Next, let us reduce species amounts (using remake), thereby also increasing the relative amount of noise, we encounter a problem when the model is simulated:\n\nsprob = remake(sprob; u0 = [:X1 => 0.33, :X2 => 0.66])\nsol = solve(sprob, STrapezoid())\nsol = solve(sprob, STrapezoid(); seed = 1234567) # hide\nnothing # hide\n\nHere, we receive a warning that the simulation was terminated. next, if we plot the solution:\n\nplot(sol)\n\nwe note that the simulation didn't reach the designated final time point (t = 10). In this case we also note that species concentrations are very low (and sometimes, due to the relatively high amount of noise, even negative). This, combined with the early termination, suggests that we are simulating our model for too low species concentration for the assumptions of the CLE to hold. Instead, jump simulations should be used.\n\nNext, let us consider a simulation for another parameter set:\n\nsprob = remake(sprob; u0 = [:X1 => 100.0, :X2 => 200.0], p = [:k1 => 200.0, :k2 => 500.0])\nsol = solve(sprob, STrapezoid())\nnothing # hide\n\nsol = solve(sprob, STrapezoid(); seed = 12345) # hide\nplot(sol)\n\nAgain, the simulation is aborted. This time, however, species concentrations are relatively large, so the CLE might still hold. What has happened this time is that the accuracy of the simulations has not reached its desired threshold. This can be deal with by reducing simulation tolerances:\n\nsol = solve(sprob, STrapezoid(), abstol = 1e-1, reltol = 1e-1)\nsol = solve(sprob, STrapezoid(); seed = 12345, abstol = 1e-1, reltol = 1e-1) # hide\nplot(sol)","category":"section"},{"location":"model_simulation/simulation_introduction/#simulation_intro_SDEs_fixed_dt","page":"Model Simulation Introduction","title":"SDE simulations with fixed time stepping","text":"StochasticDiffEq implements SDE solvers with adaptive time stepping. However, when using a non-adaptive solver (or using the adaptive = false argument to turn adaptive time stepping off for an adaptive solver) a fixed time step dt must be designated. Here we simulate the same SDEProblem which we struggled with previously, but using the non-adaptive EM solver and a fixed dt:\n\nsol = solve(sprob, EM(); dt = 0.001)\nsol = solve(sprob, EM(); dt = 0.001, seed = 1234567) # hide\nplot(sol)\n\nWe note that this approach also enables us to successfully simulate the SDE we previously struggled with.\n\nGenerally, using a smaller fixed dt provides a more exact simulation, but also increases simulation runtime.","category":"section"},{"location":"model_simulation/simulation_introduction/#simulation_intro_SDEs_noise_saling","page":"Model Simulation Introduction","title":"Scaling the noise in the chemical Langevin equation","text":"When using the CLE to generate SDEs from a CRN, it can sometimes be desirable to scale the magnitude of the noise. This can be done by introducing a noise scaling term, with each noise term generated by the CLE being multiplied with this term. A noise scaling term can be set using the @default_noise_scaling option:\n\ntwo_state_model = @reaction_network begin\n    @default_noise_scaling 0.1\n    (k1,k2), X1 <--> X2\nend\nnothing # hide\n\nHere, we set the noise scaling term to 0.1, reducing the noise with a factor 10 (noise scaling terms 10 increase the noise, while terms 10 reduce the noise). If we re-simulate the model using the low-concentration settings used previously, we see that the noise has been reduced (in fact by so much that the model can now be simulated without issues):\n\nu0 = [:X1 => 100.0, :X2 => 200.0]\ntspan = (0.0, 1.0)\nps = [:k1 => 200.0, :k2 => 500.0]\nsprob = SDEProblem(two_state_model, u0, tspan, ps)\nsol = solve(sprob, STrapezoid())\nsol = solve(sprob, STrapezoid(); seed = 123) # hide\nplot(sol)\n\nThe @default_noise_scaling option can take any expression. This can be used to e.g. designate a noise scaling parameter:\n\ntwo_state_model = @reaction_network begin\n    @parameters η\n    @default_noise_scaling η\n    (k1,k2), X1 <--> X2\nend\nnothing # hide\n\nNow we can tune the noise through η's value. E.g. here we remove the noise entirely by setting η = 00 (thereby recreating an ODE simulation's behaviour):\n\nu0 = [:X1 => 0.33, :X2 => 0.66, :η => 0.0]\ntspan = (0.0, 1.0)\nps = [:k1 => 2.0, :k2 => 5.0]\nsprob = SDEProblem(two_state_model, u0, tspan, ps)\nsol = solve(sprob, STrapezoid())\nsol = solve(sprob, STrapezoid(); seed = 123) # hide\nplot(sol)\n\nnote: Note\nAbove, Catalyst is unable to infer that η is a parameter from the @default_noise_scaling η option only. Hence, @parameters η is used to explicitly declare η to be a parameter (as discussed in more detail here).\n\nIt is possible to designate specific noise scaling terms for individual reactions through the noise_scaling reaction metadata. Here, CLE noise terms associated with a specific reaction are multiplied by that reaction's noise scaling term. Here we use this to turn off the noise in the X1 to X2 reaction:\n\ntwo_state_model = @reaction_network begin\n    k1, X1 --> X2, [noise_scaling = 0.0]\n    k2, X2 --> X1\nend\nnothing # hide\n\nIf the @default_noise_scaling option is used, that term is only applied to reactions without noise_scaling metadata.\n\nWhile the @default_noise_scaling option is unavailable for programmatically created models, the set_default_noise_scaling function can be used to achieve a similar effect.","category":"section"},{"location":"model_simulation/simulation_introduction/#simulation_intro_jumps","page":"Model Simulation Introduction","title":"Performing jump simulations using stochastic chemical kinetics","text":"Catalyst uses the JumpProcesses.jl package to perform jump simulations. This section provides a brief introduction, with JumpProcesses's documentation providing a more extensive description.\n\nJump simulations are performed using so-called JumpProblems. In this example, we first declare our two-state model and its initial conditions, time span, and parameter values.\n\nusing Catalyst, JumpProcesses, Plots\ntwo_state_model = @reaction_network begin\n    (k1,k2), X1 <--> X2\nend\nu0 = [:X1 => 5, :X2 => 10]\ntspan = (0.0, 5.0)\nps = [:k1 => 2.0, :k2 => 5.0]\nnothing # hide\n\nnote: Note\nSince jump simulations typically simulate the integer copy-numbers of each species present in the system, we designate our initial conditions for jump simulations as integers. Decimal-numbered initial conditions (and thus jump simulations) are, however, also possible. While ODE and SDE simulations accept integer initial conditions, these will be converted to decimal numbers.\n\nNext, we use these as inputs to a JumpProblem:\n\njprob = JumpProblem(two_state_model, u0, tspan, ps)\nnothing # hide\n\nThe JumpProblem can now be simulated using solve (just like any other problem type).\n\nsol = solve(jprob)\nnothing # hide\n\nIf we plot the solution we can see how the system's state does not change continuously, but instead in discrete jumps (due to the occurrence of the individual reactions of the system).\n\nusing Plots\nplot(sol)","category":"section"},{"location":"model_simulation/simulation_introduction/#simulation_intro_jumps_solver_designation","page":"Model Simulation Introduction","title":"Designating aggregators and simulation methods for jump simulations","text":"Jump simulations (just like ODEs and SDEs) are performed using stochastic simulation algorithms (SSAs) to generate exact samples of the underlying jump process. In JumpProcesses.jl and Catalyst, we call SSAs aggregators. These methods determine the time until, and type of, the next reaction in a system. A separate time-stepping method is then used to actually step from one reaction instance to the next.\n\nSeveral different aggregators are available (a full list is provided here). To designate a specific one, we use JumpProblem's aggregator argument. E.g. to designate that the sorting direct method (SortingDirect) should be used, use:\n\njprob = JumpProblem(two_state_model, u0, tspan, ps; aggregator =  SortingDirect())\nnothing # hide\n\nEspecially for large systems, the choice of aggregator can dramatically impact simulation performance.","category":"section"},{"location":"model_simulation/simulation_introduction/#simulation_intro_jumps_variableratejumps","page":"Model Simulation Introduction","title":"Jump simulations where some rate depends on time","text":"For some models, the rate terms of reactions may explicitly depend on time. E.g. consider the following circadian clock (inspired) model, where the production rate of some protein (P) depends on a sinusoid function:\n\ncircadian_model = @reaction_network begin\n    A*(sin(2π*f*t - ϕ)+1)/2, 0 --> P\n    d, P --> 0\nend\n\nThis type of model will generate so called variable rate jumps (VariableRateJumps in JumpProcesses.jl). Such models can be simulated in Catalyst too, but note that now a method for time-stepping the solver must be provided to solve. Here ODE solvers should be given as they are used to handle integrating the explicitly time-dependent propensities for problems with variable rates, i.e. the proceeding example can be solved like\n\nusing OrdinaryDiffEqTsit5\nu0map = [:P => 0]\npmap = [:f => 1.0, :A => 2.0, :ϕ => 0.0, :d => 1.0]\ntspan = (0.0, 24.0)\njprob = JumpProblem(circadian_model, u0map, tspan, pmap)\nsol = solve(jprob, Tsit5())  # use the Tsit5 ODE solver to time-step\nplot(sol; idxs = :P, lw = 2)\n\n","category":"section"},{"location":"model_simulation/simulation_introduction/#simulation_intro_citation","page":"Model Simulation Introduction","title":"Citation","text":"When you simulate Catalyst models in your research, please cite the corresponding paper(s) to support the simulation package authors. For ODE simulations:\n\n@article{DifferentialEquations.jl-2017,\n author = {Rackauckas, Christopher and Nie, Qing},\n doi = {10.5334/jors.151},\n journal = {The Journal of Open Research Software},\n keywords = {Applied Mathematics},\n note = {Exported from https://app.dimensions.ai on 2019/05/05},\n number = {1},\n pages = {},\n title = {DifferentialEquations.jl – A Performant and Feature-Rich Ecosystem for Solving Differential Equations in Julia},\n url = {https://app.dimensions.ai/details/publication/pub.1085583166 and http://openresearchsoftware.metajnl.com/articles/10.5334/jors.151/galley/245/download/},\n volume = {5},\n year = {2017}\n}\n\nFor SDE simulations:\n\n@article{rackauckas2017adaptive,\n  title={Adaptive methods for stochastic differential equations via natural embeddings and rejection sampling with memory},\n  author={Rackauckas, Christopher and Nie, Qing},\n  journal={Discrete and continuous dynamical systems. Series B},\n  volume={22},\n  number={7},\n  pages={2731},\n  year={2017},\n  publisher={NIH Public Access}\n}\n\nFor jump simulations:\n\n@misc{2022JumpProcesses,\n  author       = {Isaacson, S. A. and Ilin, V. and Rackauckas, C. V.},\n  title        = {{JumpProcesses.jl}},\n  howpublished = {\\url{https://github.com/SciML/JumpProcesses.jl/}},\n  year         = {2022}\n}\n@misc{zagatti_extending_2023,\n\ttitle = {Extending {JumpProcess}.jl for fast point process simulation with time-varying intensities},\n\turl = {http://arxiv.org/abs/2306.06992},\n\tdoi = {10.48550/arXiv.2306.06992},\n\tpublisher = {arXiv},\n\tauthor = {Zagatti, Guilherme Augusto and Isaacson, Samuel A. and Rackauckas, Christopher and Ilin, Vasily and Ng, See-Kiong and Bressan, Stéphane},\n\tyear = {2023},\n}","category":"section"},{"location":"model_creation/examples/basic_CRN_library/#basic_CRN_library","page":"Library of Basic Chemical Reaction Network Models","title":"Library of Basic Chemical Reaction Network Models","text":"<details><summary><strong>Environment setup and package installation</strong></summary>\n\nThe following code sets up an environment for running the code on this page.\n\nusing Pkg\nPkg.activate(\".\")\nPkg.add(\"Catalyst\")\nPkg.add(\"OrdinaryDiffEqDefault\")\nPkg.add(\"StochasticDiffEq\")\nPkg.add(\"JumpProcesses\")\nPkg.add(\"Plots\")\n\n</details>\n\n\n   Below we will present various simple and established chemical reaction network (CRN) models. Each model is given some brief background, implemented using the @reaction_network DSL, and basic simulations are performed.","category":"section"},{"location":"model_creation/examples/basic_CRN_library/#basic_CRN_library_bd","page":"Library of Basic Chemical Reaction Network Models","title":"Birth-death process","text":"The birth-death process is one of the simplest possible CRN models. It consists of a single component (X) which is both produced and degraded at linear rates:\n\nusing Catalyst\nbd_process = @reaction_network begin\n    (p,d), ∅ <--> X\nend\n\nNext, we define simulation conditions. Note that the initial condition is integer-valued (more natural than decimal numbers for jump simulations).\n\nu0 = [:X => 1]\ntspan = (0.0, 10.0)\nps = [:p => 1.0, :d => 0.2]\nnothing # hide\n\nWe can now simulate our model using all three interpretations. First, we perform a reaction rate equation-based ODE simulation:\n\nusing OrdinaryDiffEqDefault\noprob = ODEProblem(bd_process, u0, tspan, ps)\nosol = solve(oprob)\nnothing # hide\n\nNext, a chemical Langevin equation-based SDE simulation:\n\nusing StochasticDiffEq\nsprob = SDEProblem(bd_process, u0, tspan, ps)\nssol = solve(sprob, STrapezoid())\nssol = solve(sprob, STrapezoid(); seed = 12) # hide\nnothing # hide\n\nNext, a stochastic chemical kinetics-based jump simulation:\n\nusing JumpProcesses\njprob = JumpProblem(bd_process, u0, tspan, ps)\njsol = solve(jprob)\njsol = solve(jprob; seed = 12) # hide\nnothing # hide\n\nFinally, we plot the results:\n\nusing Plots\noplt = plot(osol; title = \"Reaction rate equation (ODE)\")\nsplt = plot(ssol; title = \"Chemical Langevin equation (SDE)\")\njplt = plot(jsol; title = \"Stochastic chemical kinetics (Jump)\")\nplot(oplt, splt, jplt; lw = 3, size=(800,700), layout = (3,1))","category":"section"},{"location":"model_creation/examples/basic_CRN_library/#basic_CRN_library_two_states","page":"Library of Basic Chemical Reaction Network Models","title":"Two-state model","text":"The two-state model describes a component (here called X) which can exist in two different forms (here called X₁ and X₂). It switches between these forms at linear rates. First, we simulate the model using both ODEs and SDEs:\n\nusing Catalyst, OrdinaryDiffEqDefault, StochasticDiffEq, Plots\ntwo_state_model = @reaction_network begin\n    (k₁,k₂), X₁ <--> X₂\nend\n\nu0 = [:X₁ => 50.0, :X₂ => 50.0]\ntspan = (0.0, 1.0)\nps = [:k₁ => 2.0, :k₂ => 3.0]\n\noprob = ODEProblem(two_state_model, u0, tspan, ps)\nosol = solve(oprob)\noplt = plot(osol; title = \"Reaction rate equation (ODE)\")\n\nsprob = SDEProblem(two_state_model, u0, tspan, ps)\nssol = solve(sprob, STrapezoid())\nssol = solve(sprob, STrapezoid(); seed = 12) # hide\nsplt = plot(ssol; title = \"Chemical Langevin equation (SDE)\")\n\nplot(oplt, splt; lw = 3, size = (800,550), layout = (2,1))\n\nWhat is interesting about this model is that it has a conserved quantity, where X₁ + X₂ remains constant throughout the simulation (both in deterministic and stochastic cases). We can show this by instead plotting this conserved quantity.\n\n@unpack X₁, X₂ = two_state_model\noplt = plot(osol; idxs = X₁ + X₂, title = \"Reaction rate equation (ODE)\")\nsplt = plot(ssol; idxs = X₁ + X₂, title = \"Chemical Langevin equation (SDE)\")\nplot(oplt, splt; lw = 3, ylimit = (99,101), size = (800,450), layout = (2,1))","category":"section"},{"location":"model_creation/examples/basic_CRN_library/#basic_CRN_library_mm","page":"Library of Basic Chemical Reaction Network Models","title":"Michaelis-Menten enzyme kinetics","text":"Michaelis-Menten enzyme kinetics is a simple description of an enzyme (E) transforming a substrate (S) into a product (P). Under certain assumptions, it can be simplified to a single function (a Michaelis-Menten function) and used as a reaction rate. Here we instead present the full system model:\n\nusing Catalyst\nmm_system = @reaction_network begin\n    kB, S + E --> SE\n    kD, SE --> S + E\n    kP, SE --> P + E\nend\n\nNext, we perform ODE, SDE, and jump simulations of the model:\n\nu0 = [:S => 301, :E => 100, :SE => 0, :P => 0]\ntspan = (0., 100.)\nps = [:kB => 0.00166, :kD => 0.0001, :kP => 0.1]\n\nusing OrdinaryDiffEqDefault\noprob = ODEProblem(mm_system, u0, tspan, ps)\nosol  = solve(oprob)\n\nusing StochasticDiffEq\nsprob = SDEProblem(mm_system, u0, tspan, ps)\nssol = solve(sprob, STrapezoid())\nssol = solve(sprob, STrapezoid(); seed = 12) # hide\n\nusing JumpProcesses\njprob = JumpProblem(mm_system, u0, tspan, ps)\njsol = solve(jprob)\njsol = solve(jprob; seed = 12) # hide\n\nusing Plots\noplt = plot(osol; title = \"Reaction rate equation (ODE)\")\nsplt = plot(ssol; title = \"Chemical Langevin equation (SDE)\")\njplt = plot(jsol; title = \"Stochastic chemical kinetics (Jump)\")\nplot(oplt, splt, jplt; lw = 2, size=(800,800), layout = (3,1))\nfullplt = plot(oplt, splt, jplt; lw = 2, size=(800,800), layout = (3,1), fmt = :png,\n    dpi = 200, bottom_margin = 3Plots.Measures.mm) # hide\nCatalyst.PNG(fullplt) # hide","category":"section"},{"location":"model_creation/examples/basic_CRN_library/#basic_CRN_library_sir","page":"Library of Basic Chemical Reaction Network Models","title":"SIR infection model","text":"The SIR model is the simplest model of the spread of an infectious disease. While the real system is very different from the chemical and cellular processes typically modelled with CRNs, it (and several other epidemiological systems) can be modelled using the same CRN formalism. The SIR model consists of three species: susceptible (S), infected (I), and removed (R) individuals, and two reaction events: infection and recovery.\n\nusing Catalyst\nsir_model = @reaction_network begin\n    α, S + I --> 2I\n    β, I --> R\nend\n\nFirst, we perform a deterministic ODE simulation:\n\nusing OrdinaryDiffEqDefault, Plots\nu0 = [:S => 99, :I => 1, :R => 0]\ntspan = (0.0, 500.0)\nps = [:α => 0.001, :β => 0.01]\n\n# Solve ODEs.\noprob = ODEProblem(sir_model, u0, tspan, ps)\nosol = solve(oprob)\nplot(osol; title = \"Reaction rate equation (ODE)\", size=(800,350))\n\nNext, we perform 3 different Jump simulations. Note that for the stochastic model, the occurrence of an outbreak is not certain. Rather, there is a possibility that it fizzles out without a noteworthy peak.\n\nusing JumpProcesses\njprob = JumpProblem(sir_model, u0, tspan, ps)\n\njsol1 = solve(jprob)\njsol2 = solve(jprob)\njsol3 = solve(jprob)\njsol1 = solve(jprob, seed = 1) # hide\njsol2 = solve(jprob, seed = 2) # hide\njsol3 = solve(jprob, seed = 3) # hide\n\njplt1 = plot(jsol1; title = \"Outbreak\")\njplt2 = plot(jsol2; title = \"Outbreak\")\njplt3 = plot(jsol3; title = \"No outbreak\")\nplot(jplt1, jplt2, jplt3; lw = 3, size=(800,700), layout = (3,1))\nfullplt = plot(jplt1, jplt2, jplt3; lw = 3, size=(800,700), layout = (3,1), fmt = :png, dpi = 200) # hide\nCatalyst.PNG(fullplt) # hide","category":"section"},{"location":"model_creation/examples/basic_CRN_library/#basic_CRN_library_cc","page":"Library of Basic Chemical Reaction Network Models","title":"Chemical cross-coupling","text":"In chemistry, cross-coupling is when a catalyst combines two substrates to form a product. In this example, the catalyst (C) first binds one substrate (S₁) to form an intermediary complex (S₁C). Next, the complex binds the second substrate (S₂) to form another complex (CP). Finally, the catalyst releases the now-formed product (P). This system is an extended version of the Michaelis-Menten system presented earlier.\n\nusing Catalyst\ncc_system = @reaction_network begin\n    k₁, S₁ + C --> S₁C\n    k₂, S₁C + S₂ --> CP\n    k₃, CP --> C + P\nend\n\nBelow, we perform a simple deterministic ODE simulation of the system. Next, we plot both:\n\nThe concentration of the substrates and the product.\nThe concentration of the catalyst and the intermediaries.\n\nIn two separate plots.\n\nusing OrdinaryDiffEqDefault, Plots\nu0 = [:S₁ => 1.0, :C => 0.05, :S₂ => 1.2, :S₁C => 0.0, :CP => 0.0, :P => 0.0]\ntspan = (0., 15.)\nps = [:k₁ => 5.0, :k₂ => 5.0, :k₃ => 100.0]\n\n# solve ODEs\noprob = ODEProblem(cc_system, u0, tspan, ps)\nosol  = solve(oprob)\n\nplt1 = plot(osol; idxs = [:S₁, :S₂, :P], title = \"Substrate and product dynamics\")\nplt2 = plot(osol; idxs = [:C, :S₁C, :CP], title = \"Catalyst and intermediaries dynamics\")\nplot(plt1, plt2; lw = 3, size = (800,600), layout = (2,1))","category":"section"},{"location":"model_creation/examples/basic_CRN_library/#basic_CRN_library_wilhelm","page":"Library of Basic Chemical Reaction Network Models","title":"The Wilhelm model","text":"The Wilhelm model was introduced in Wilhelm (2009) as the smallest CRN model (with constant rates) that exhibits bistability.\n\nusing Catalyst\nwilhelm_model = @reaction_network begin\n    k1, Y --> 2X\n    k2, 2X --> X + Y\n    k3, X + Y --> Y\n    k4, X --> 0\nend\n\nWe can simulate the model for two different initial conditions, demonstrating the existence of two different stable steady states.\n\nusing OrdinaryDiffEqDefault, Plots\nu0_1 = [:X => 1.5, :Y => 0.5]\nu0_2 = [:X => 2.5, :Y => 0.5]\ntspan = (0., 10.)\nps = [:k1 => 8.0, :k2 => 2.0, :k3 => 1.0, :k4 => 1.5]\n\noprob1 = ODEProblem(wilhelm_model, u0_1, tspan, ps)\noprob2 = ODEProblem(wilhelm_model, u0_2, tspan, ps)\nosol1 = solve(oprob1)\nosol2 = solve(oprob2)\nplot(osol1; lw = 4, idxs = :X, label = \"X(0) = 1.5\")\nplot!(osol2; lw = 4, idxs = :X, label = \"X(0) = 2.5\", yguide = \"X\", size = (800,350))\nplot!(bottom_margin = 3Plots.Measures.mm) # hide","category":"section"},{"location":"model_creation/examples/basic_CRN_library/#basic_CRN_library_self_activation","page":"Library of Basic Chemical Reaction Network Models","title":"Simple self-activation loop","text":"The simplest self-activation loop consists of a single species (here called X) which activates its own production. If its production rate is modelled as a Hill function with n1, the system may exhibit bistability.\n\nusing Catalyst\nsa_loop = @reaction_network begin\n    v₀ + hill(X,v,K,n), ∅ --> X\n    d, X --> ∅\nend\n\nA simple example of such a loop is a transcription factor which activates its own gene. Here, v₀ represents a basic transcription rate (leakage) in the absence of the transcription factor.\n\nWe simulate the self-activation loop from a single initial condition using both deterministic (ODE) and stochastic (jump) simulations. We note that while the deterministic simulation reaches a single steady state, the stochastic one switches between two different states.\n\nusing JumpProcesses, OrdinaryDiffEqDefault, Plots\nu0 = [:X => 4]\ntspan = (0.0, 1000.0)\nps = [:v₀ => 0.1, :v => 2.0, :K => 10.0, :n => 2, :d => 0.1]\n\noprob = ODEProblem(sa_loop, u0, tspan, ps)\nosol = solve(oprob)\n\njprob = JumpProblem(sa_loop, u0, tspan, ps)\njsol = solve(jprob)\njsol = solve(jprob, seed = 12) # hide\n\nfplt = plot(osol; lw = 3, label = \"Reaction rate equation (ODE)\")\nplot!(fplt, jsol; lw = 3, label = \"Stochastic chemical kinetics (Jump)\", yguide = \"X\", size = (800,350))\nplot!(fplt, bottom_margin = 3Plots.Measures.mm, left_margin=3Plots.Measures.mm) # hide\nplot!(fplt; fmt = :png, dpi = 200) # hide\nCatalyst.PNG(fplt) # hide","category":"section"},{"location":"model_creation/examples/basic_CRN_library/#basic_CRN_library_brusselator","page":"Library of Basic Chemical Reaction Network Models","title":"The Brusselator","text":"The Brusselator is a well-known (theoretical) CRN model able to produce oscillations (its name is a portmanteau of \"Brussels\" and \"oscillator\").\n\nusing Catalyst\nbrusselator = @reaction_network begin\n    A, ∅ --> X\n    1, 2X + Y --> 3X\n    B, X --> Y\n    1, X --> ∅\nend\n\nIt is generally known to (for reaction rate equation-based ODE simulations) produce oscillations when B  1 + A^2. However, this result is based on models generated when combinatorial adjustment of rates is not performed. Since Catalyst automatically perform these adjustments, and one reaction contains a stoichiometric constant 1, the threshold will be different. Here, we trial two different values of B. In both cases, B  1 + A^2, however, in the second case the system can generate oscillations.\n\nusing OrdinaryDiffEqDefault, Plots\nu0 = [:X => 1.0, :Y => 1.0]\ntspan = (0., 50.)\nps1 = [:A => 1.0, :B => 1.0]\nps2 = [:A => 1.0, :B => 1.8]\n\noprob1 = ODEProblem(brusselator, u0, tspan, ps1)\noprob2 = ODEProblem(brusselator, u0, tspan, ps2)\nosol1  = solve(oprob1)\nosol2  = solve(oprob2)\noplt1 = plot(osol1; title = \"No Oscillation\")\noplt2 = plot(osol2; title = \"Oscillation\")\n\nplot(oplt1, oplt2; lw = 3, size = (800,600), layout = (2,1))","category":"section"},{"location":"model_creation/examples/basic_CRN_library/#basic_CRN_library_repressilator","page":"Library of Basic Chemical Reaction Network Models","title":"The Repressilator","text":"The Repressilator was introduced in Elowitz & Leibler (2000) as a simple system that can generate oscillations (most notably, they demonstrated this both in a model and in a synthetic in vivo implementation in Escherichia col). It consists of three genes, repressing each other in a cycle. Here, we will implement it using three species (X, Y, and Z) whose production rates are (repressing) Hill functions.\n\nusing Catalyst\nrepressilator = @reaction_network begin\n    hillr(Z,v,K,n), ∅ --> X\n    hillr(X,v,K,n), ∅ --> Y\n    hillr(Y,v,K,n), ∅ --> Z\n    d, (X, Y, Z) --> ∅\nend\n\nWhether the Repressilator oscillates or not depends on its parameter values. Here, we will perform deterministic (ODE) simulations for two different values of K, showing that it oscillates for one value and not the other one. Next, we will perform stochastic (SDE) simulations for both K values, showing that the stochastic model can sustain oscillations in both cases. This is an example of the phenomena of noise-induced oscillation.\n\nusing OrdinaryDiffEqDefault, StochasticDiffEq, Plots\nu0 = [:X => 50.0, :Y => 15.0, :Z => 15.0]\ntspan = (0., 200.)\nps1 = [:v => 10.0, :K => 20.0, :n => 3, :d => 0.1]\nps2 = [:v => 10.0, :K => 50.0, :n => 3, :d => 0.1]\n\noprob1 = ODEProblem(repressilator, u0, tspan, ps1)\noprob2 = ODEProblem(repressilator, u0, tspan, ps2)\nosol1  = solve(oprob1)\nosol2  = solve(oprob2)\noplt1 = plot(osol1; title = \"Oscillation (ODE, K = 20)\")\noplt2 = plot(osol2; title = \"No oscillation (ODE, K = 50)\")\n\nsprob1 = SDEProblem(repressilator, u0, tspan, ps1)\nsprob2 = SDEProblem(repressilator, u0, tspan, ps2)\nssol1  = solve(sprob1, STrapezoid())\nssol2  = solve(sprob2, STrapezoid())\nssol1  = solve(sprob1, STrapezoid(); seed = 1) # hide\nssol2  = solve(sprob2, STrapezoid(); seed = 100) # hide\nsplt1 = plot(ssol1; title = \"Oscillation (SDE, K = 20)\")\nsplt2 = plot(ssol2; title = \"Oscillation (SDE, K = 50)\")\n\nplot(oplt1, oplt2, splt1, splt2; lw = 2, layout = (2,2), size = (800,600))","category":"section"},{"location":"model_creation/examples/basic_CRN_library/#basic_CRN_library_wr","page":"Library of Basic Chemical Reaction Network Models","title":"The Willamowski–Rössler model","text":"The Willamowski–Rössler model was introduced in Willamowski & Rössler (1979) as an example of a simple CRN model which exhibits chaotic behaviours. This means that small changes in initial conditions can produce relatively large changes in the system's trajectory.\n\nusing Catalyst\nwr_model = @reaction_network begin\n    k1, 2X --> 3X\n    k2, X --> 2X\n    k3, Z + 2X --> 2Z\n    k4, Y + X --> 2Y\n    k5, Y --> ∅\n    k6, 2Z --> ∅\n    k7, Z --> ∅\nend\n\nHere we simulate the model for a single initial condition, showing both time-state space and phase space how it reaches a strange attractor.\n\nusing OrdinaryDiffEqDefault, Plots\nu0 = [:X => 1.5, :Y => 1.5, :Z => 1.5]\ntspan = (0.0, 50.0)\np = [:k1 => 2.1, :k2 => 0.7, :k3 => 2.9, :k4 => 1.1, :k5 => 1.0, :k6 => 0.5, :k7 => 2.7]\noprob = ODEProblem(wr_model, u0, tspan, p)\nsol = solve(oprob)\n\nplt1 = plot(sol; title = \"Time-state space\")\nplt2 = plot(sol; idxs = (:X, :Y, :Z), title = \"Phase space\")\nplot(plt1, plt2; layout = (1,2), size = (800,400))\nplot!(bottom_margin = 3Plots.Measures.mm) # hide","category":"section"},{"location":"faqs/#FAQs","page":"FAQs","title":"FAQs","text":"","category":"section"},{"location":"faqs/#How-to-index-solution-objects-using-symbolic-variables-and-observables?","page":"FAQs","title":"How to index solution objects using symbolic variables and observables?","text":"One can directly use symbolic variables to index into SciML solution objects. Moreover, observables can also be evaluated in this way. For example, consider the system\n\nusing Catalyst, OrdinaryDiffEqTsit5, Plots\nrn = @reaction_network ABtoC begin\n  (k₊,k₋), A + B <--> C\nend\nnothing    # hide\n\nLet's convert it to a system of ODEs, using the conservation laws of the system to eliminate two of the species:\n\nosys = ode_model(rn; remove_conserved = true)\nosys = complete(osys)\n\nNotice the resulting ODE system has just one ODE, while algebraic observables have been added for the two removed species (in terms of the conservation law constants, Γ[1] and Γ[2])\n\nobserved(osys)\n\nLet's solve the system and see how to index the solution using our symbolic variables\n\nsim_cond = [\n    osys.A => 1.0, osys.B => 2.0, osys.C => 0.0,\n    osys.k₊ => 1.0, osys.k₋ => 1.0    \n]\noprob = ODEProblem(osys, sim_cond, (0.0, 10.0))\nsol = solve(oprob, Tsit5())\n\nSuppose we want to plot just species C, without having to know its integer index in the unknown vector. We can do this using the symbolic variable C, which we can get at in several ways\n\nsol[osys.C]\n\nor\n\n@unpack C = osys\nsol[C]\n\nTo evaluate C at specific times and plot it we can just do\n\nt = range(0.0, 10.0, length = 101)\nplot(sol(t, idxs = C), label = \"C(t)\", xlabel = \"t\")\n\nIf we want to get multiple variables we can just do\n\n@unpack A, B = osys\nsol(t, idxs = [A, B])\n\nPlotting multiple variables using the SciML plot recipe can be achieved like\n\nplot(sol; idxs = [A, B])","category":"section"},{"location":"faqs/#faq_combinatoric_ratelaws","page":"FAQs","title":"How to disable rescaling of reaction rates in rate laws?","text":"As explained in the Reaction rate laws used in simulations section, for a reaction such as k, 2X --> 0, the generated rate law will rescale the rate constant, giving k*X^2/2 instead of k*X^2 for ODEs and k*X*(X-1)/2 instead of k*X*(X-1) for jumps. This can be disabled when directly converting a ReactionSystem. If rn is a generated ReactionSystem, we can do\n\nosys = ode_model(rn; combinatoric_ratelaws = false)\n\nDisabling these rescalings should work for all conversions of ReactionSystems to other ModelingToolkitBase.AbstractSystems.\n\nWhen creating a ReactionSystem using the DSL, combinatoric rate laws can be disabled (for  the created system, and all systems derived from it) using the @combinatoric_ratelaws option (providing false as its only input):\n\nrn = @reaction_network begin\n    @combinatoric_ratelaws false\n    k, 2X --> 0\nend\nnothing # hide","category":"section"},{"location":"faqs/#How-to-use-non-integer-stoichiometric-coefficients?","page":"FAQs","title":"How to use non-integer stoichiometric coefficients?","text":"using Catalyst\nrn = @reaction_network begin\n  k, 2.5*A --> 3*B\nend\n\nor directly via\n\nt = default_t()\n@parameters k b\n@species A(t) B(t) C(t) D(t)\nrx1 = Reaction(k, [B,C], [B,D], [2.5,1], [3.5, 2.5])\nrx2 = Reaction(2*k, [B], [D], [1], [2.5])\nrx3 = Reaction(2*k, [B], [D], [2.5], [2])\n@named mixedsys = ReactionSystem([rx1, rx2, rx3], t, [A, B, C, D], [k, b])\nmixedsys = complete(mixedsys)\nosys = ode_model(mixedsys; combinatoric_ratelaws = false)\nosys = complete(osys)\n\nNote, when using ode_model(mixedsys; combinatoric_ratelaws=false) the combinatoric_ratelaws=false parameter must be passed. This is also true when calling ODEProblem(mixedsys,...; combinatoric_ratelaws=false). As described above, this disables Catalyst's standard rescaling of reaction rates when generating reaction rate laws, see also the Reaction rate laws used in simulations section. Leaving this keyword out for systems with floating point stoichiometry will give an error message.\n\nFor a more extensive documentation of using non-integer stoichiometric coefficients, please see the Symbolic Stochiometries section.","category":"section"},{"location":"faqs/#How-to-set-default-values-for-initial-conditions-and-parameters?","page":"FAQs","title":"How to set default values for initial conditions and parameters?","text":"How to set defaults when using the @reaction_network macro is described in more detail here. There are several ways to do this. Using the DSL, one can use the @species and @parameters options:\n\nusing Catalyst\nsir = @reaction_network sir begin\n    @species S(t)=999.0 I(t)=1.0 R(t)=0.0\n    @parameters β=1e-4 ν=0.01\n    β, S + I --> 2I\n    ν, I --> R\nend\nshow(stdout, MIME\"text/plain\"(), sir) # hide\n\nWhen directly constructing a ReactionSystem, we can set the symbolic values to have the desired default values, and this will automatically be propagated through to the equation solvers:\n\nusing Catalyst, Plots, OrdinaryDiffEqTsit5\nt = default_t()\n@parameters β=1e-4 ν=.01\n@species S(t)=999.0 I(t)=1.0 R(t)=0.0\nrx1 = Reaction(β, [S, I], [I], [1,1], [2])\nrx2 = Reaction(ν, [I], [R])\n@named sir = ReactionSystem([rx1, rx2], t)\nsir = complete(sir)\noprob = ODEProblem(sir, [], (0.0, 250.0))\nsol = solve(oprob, Tsit5())\nplot(sol)\n\nOne can also build a mapping from symbolic parameter/species to value/initial condition and pass these to the ReactionSystem via the initial_conditions keyword argument:\n\n@parameters β ν\n@species S(t) I(t) R(t)\nrx1 = Reaction(β, [S,I], [I], [1,1], [2])\nrx2 = Reaction(ν, [I], [R])\nics = [β => 1e-4, ν => .01, S => 999.0, I => 1.0, R => 0.0]\n@named sir = ReactionSystem([rx1, rx2], t; initial_conditions = ics)\nnothing # hide","category":"section"},{"location":"faqs/#How-to-specify-initial-conditions-and-parameters-values-for-ODEProblem-and-other-problem-types?","page":"FAQs","title":"How to specify initial conditions and parameters values for ODEProblem and other problem types?","text":"To explicitly pass initial conditions and parameters we can use mappings from Julia Symbols corresponding to each variable/parameter to their values, or from ModelingToolkitBase symbolic variables/parameters to their values. Using Symbols we have\n\nusing Catalyst, OrdinaryDiffEqTsit5\nrn = @reaction_network begin\n    α, S + I --> 2I\n    β, I --> R\nend\nu0 = [:S => 999.0, :I => 1.0, :R => 0.0]\np  = (:α => 1e-4, :β => .01)\nop1  = ODEProblem(rn, u0, (0.0, 250.0), p)\nnothing  # hide\n\nwhile using ModelingToolkit symbolic variables we have\n\nt = default_t()\nu0 = [rn.S => 999.0, rn.I => 1.0, rn.R => 0.0]\np  = (rn.α => 1e-4, rn.β => .01)\nop2  = ODEProblem(rn, u0, (0.0, 250.0), p)\nnothing  # hide","category":"section"},{"location":"faqs/#How-to-include-non-reaction-terms-in-equations-for-a-chemical-species?","page":"FAQs","title":"How to include non-reaction terms in equations for a chemical species?","text":"One method to add non-reaction terms into an ODE or algebraic equation for a chemical species is to add a new (non-species) unknown variable that represents those terms, let it be the rate of zero order reaction, and add a constraint equation. I.e., to add a force of (1 + sin(t)) to dAdt in a system with the reaction k, A --> 0, we can do\n\nusing Catalyst\nt = default_t()\n@variables f(t)\nrx1 = @reaction k, A --> 0\nrx2 = @reaction $f, 0 --> A\neq = f ~ (1 + sin(t))\n@named rs = ReactionSystem([rx1, rx2, eq], t)\nrs = complete(rs)\nosys = ode_model(rs)\n\nIn the final ODE model, f can be eliminated by using ModelingToolkitBase.mtkcompile\n\nosyss = mtkcompile(osys)\nfull_equations(osyss)","category":"section"},{"location":"faqs/#How-to-modify-generated-ODEs?","page":"FAQs","title":"How to modify generated ODEs?","text":"Conversion to ModelingToolkitBase.System allows the possibility to modify the system with further terms that are difficult to encode as a chemical reaction or a constraint equation. For example, an alternative method to the previous question for adding a forcing function, 1 + sin(t), to the ODE for dA/dt is\n\nusing Catalyst\nrn = @reaction_network begin\n    k, A --> 0\nend\nosys = ode_model(rn)\ndAdteq = equations(osys)[1]\nt      = ModelingToolkitBase.get_iv(osys)\ndAdteq = Equation(dAdteq.lhs, dAdteq.rhs + 1 + sin(t))\n\n# create a new ODE System with the modified equation\n@named osys2  = System([dAdteq], t)","category":"section"},{"location":"faqs/#How-to-override-mass-action-kinetics-rate-laws?","page":"FAQs","title":"How to override mass action kinetics rate laws?","text":"While generally one wants the reaction rate law to use the law of mass action, so the reaction\n\nusing Catalyst\nusing Latexify # hide\nrn = @reaction_network begin\n    k, X --> ∅\nend\node_model(rn)\nlatexify(rn; form = :ode, math_delimiters = true) # hide\n\noccurs at the (ODE) rate dXdt = -kX, it is possible to override this by using any of the following non-filled arrows when declaring the reaction: <=, ⇐, ⟽, =>, ⇒, ⟾, ⇔, ⟺. This means that the reaction\n\nrn = @reaction_network begin\n    k, X => ∅\nend\node_model(rn)\nlatexify(rn; form = :ode, math_delimiters = true) # hide\n\nwill occur at rate dXdt = -k (which might become a problem since X will be degraded at a constant rate even when very small or equal to 0).\n\nNote, stoichiometric coefficients are still included, i.e. the reaction\n\nrn = @reaction_network begin\n    k, 2*X ⇒ ∅\nend\node_model(rn)\nlatexify(rn; form = :ode, math_delimiters = true) # hide\n\nhas rate dXdt = -2 k.","category":"section"},{"location":"faqs/#user_functions","page":"FAQs","title":"How to specify user-defined functions as reaction rates?","text":"The reaction network DSL can \"see\" user-defined functions that work with ModelingToolkitBase. e.g., this is should work\n\nusing Catalyst\nmyHill(x) = 2*x^3/(x^3+1.5^3)\nrn = @reaction_network begin\n    myHill(X), ∅ --> X\nend\n\nIn some cases, it may be necessary or desirable to register functions with Symbolics.jl before their use in Catalyst, see the discussion here.","category":"section"},{"location":"faqs/#faq_dsl_sym_inference","page":"FAQs","title":"How does the Catalyst DSL (@reaction_network) infer what different symbols represent?","text":"When declaring a model using the Catalyst DSL, e.g.\n\nusing Catalyst\nrn = @reaction_network begin\n (p,d), 0 <--> X\nend\n\nCatalyst can automatically infer that X is a species and p and d are parameters. In total, Catalyst can infer the following quantities:\n\nSpecies (from reaction reactants).\nParameters (from reaction rates and stoichiometries).\n(non-species) Variables (from the @equations option).\nDifferential functions (from the @equations option).\nObservables (from the @observables option).\nCompound species (from the @compounds option).\n\nInference of species, variables, and parameters follows the following steps:\n\nEvery symbol explicitly declared using some option (typically @species, @variables, and @parameters, although other options such as @brownian also are possible) are assigned to the corresponding category.\nEvery symbol not declared in (1) that occurs as a reaction reactant is inferred as a species.\nEvery symbol not declared in (1) or (2) that occurs in an expression provided after @equations is inferred as a variable.\nEvery symbol not declared in (1), (2), or (3) that occurs either as a reaction rate or stoichiometric coefficient is inferred to be a parameter.\nIf the symbol D is not declared in (1), (2), (3), or (4), and occurs on the form D(...) within the @equations option, D is inferred to be the differential with respect to the default independent variable (typically t).\n\nHere, in \n\nusing Catalyst\nrn = @reaction_network begin\n    @parameters p1\n    @equations V ~ X + p1\n X + V + p1 + p2, 0 --> X\nend\n\np is first set as a parameter (as it is explicitly declared as such). Next, X is inferred as a species. Next, V is inferred as a variable. Finally, p2 is inferred as a parameter.\n\nNext, if any expression D(...) (where ... can be anything) is encountered within the @equations option, D is inferred to be the differential with respect to the default independent variable (typically t). Note that using D in this way, while also using it in another form (e.g. in a reaction rate) will produce an error.\n\nAny symbol used as the left-hand side within the @observables option is inferred to be an observable. These are by default assumed to be variables. It is possible to simultaneously explicitly declare an observable using the @species or @variables options (in the former case, the observable will be treated as a species instead). Using observables within most other expressions (e.g. as a reactant) will produce an error.\n\nAny symbol declared as a compound using the @compound option is automatically inferred to be a system species.\n\nSymbols occurring within other expressions will not be inferred as anything. These must either occur in one of the forms described above (which enables Catalyst to infer what they are) or be explicitly declared. E.g. having a parameter which only occurs in an event:\n\nusing Catalyst\nrn_error = @reaction_network begin\n    @discrete_events 1.0 => [X => X + Xadd] \n d, X --> 0\nend\n\nis not permitted. E.g. here Xadd must be explicitly declared as a parameter using @parameters:\n\nusing Catalyst\nrn = @reaction_network begin\n    @parameters Xadd\n    @discrete_events 1.0 => [X => X + Xadd] \n d, X --> 0\nend\n\nIt is possible to turn off all inference (requiring all symbols to be declared using @parameters, @species, @variables, or some other option) through the @require_declaration option.","category":"section"},{"location":"faqs/#faq_require_declaration","page":"FAQs","title":"How can I turn off automatic inferring of species and parameters when using the DSL?","text":"This option can be set using the @require_declaration option inside @reaction_network. In this case all the species, parameters, and variables in the system must be pre-declared using one of the @species, @parameters, or @variables macros. For more information about what is inferred automatically and not, please see the section on @require_declaration.\n\nusing Catalyst\nrn = @reaction_network begin\n    @require_declaration\n    @species A(t) B(t)\n    @parameters k1 k2\n    (k1, k2), A <--> B\nend","category":"section"},{"location":"model_simulation/examples/interactive_brusselator_simulation/#interactive_brusselator","page":"Interactive Simulation and Plotting","title":"Interactive Simulation and Plotting","text":"<details><summary><strong>Environment setup and package installation</strong></summary>\n\nThe following code sets up an environment for running the code on this page.\n\nusing Pkg\nPkg.activate(; temp = true) # Creates a temporary environment, which is deleted when the Julia session ends.\nPkg.add(\"CairoMakie\")\nPkg.add(\"Catalyst\")\nPkg.add(\"GLMakie\")\nPkg.add(\"OrdinaryDiffEqTsit5\")\n\n</details>\n\n\n Catalyst can utilize the GLMakie.jl package for creating interactive visualizations of your reaction network dynamics. This tutorial provides a step-by-step guide to creating an interactive visualization of the Brusselator model, building upon the basic Brusselator example.","category":"section"},{"location":"model_simulation/examples/interactive_brusselator_simulation/#setup_brusselator","page":"Interactive Simulation and Plotting","title":"Setting up the Brusselator model","text":"Let's again use the oscillating Brusselator model, extending the basic simulation plotting workflow we saw earlier. \n\nusing Catalyst\nusing OrdinaryDiffEqTsit5\n\nusing GLMakie\n\nusing CairoMakie # hide\n\n# Define the Brusselator model\nbrusselator = @reaction_network begin\n    A, ∅ → X\n    1, 2X + Y → 3X\n    B, X → Y\n    1, X → ∅\nend\n\n# Initial parameter values and conditions\np = [:A => 1.0, :B => 4.0]\nu0 = [:X => 1.0, :Y => 0.0]\ntspan = (0.0, 50.0)\n\noprob = ODEProblem(brusselator, u0, tspan, p)\n\n# Function to solve the ODE\nfunction solve_brusselator(A, B, X0, Y0, prob = oprob)\n    p = [:A => A, :B => B]\n    u0 = [:X => X0, :Y => Y0]\n    newprob = remake(prob, p=p, u0=u0)\n    solve(newprob, Tsit5(), saveat = 0.1) \nend\n\nThis code sets up our Brusselator model using Catalyst.jl's @reaction_network macro. We also define initial parameters, initial conditions, create an ODEProblem, and define a function to solve the ODE with given parameters.  Setting saveat = 0.1 in the call to solve ensures the solution is saved with the desired temporal frequency we want for our later plots.\n\nnote: Note\nBe sure to set saveat to a value that is appropriate for your system; otherwise, the size of the solution can change during interactivity, which will cause dimension mismatch errors once we add our interactive elements.","category":"section"},{"location":"model_simulation/examples/interactive_brusselator_simulation/#basic_static_plotting","page":"Interactive Simulation and Plotting","title":"Basic static plotting","text":"Let's start by creating a basic plot of our Brusselator model:\n\n# Create the main figure\nfig = Figure(size = (800, 600), fontsize = 18);\n\n# Create an axis for the plot\nax = Axis(fig[1, 1], \n    title = \"Brusselator Model\", \n    xlabel = \"Time\", \n    ylabel = \"Concentration\")\n\n# Solve the ODE\nsol = solve_brusselator(1.0, 4.0, 1.0, 0.0)\n\n# Plot the solution\nlines!(ax, sol.t, sol[:X], label = \"X\", color = :blue, linewidth = 3)\nlines!(ax, sol.t, sol[:Y], label = \"Y\", color = :red, linewidth = 3)\n\n# Add a legend\naxislegend(ax, position = :rt)\n\n# Display the figure\nfig\n\nThe plot shows the concentrations of species X and Y over time. Notice the oscillatory behavior characteristic of the Brusselator model.","category":"section"},{"location":"model_simulation/examples/interactive_brusselator_simulation/#adding_interactivity","page":"Interactive Simulation and Plotting","title":"Adding interactivity","text":"Now, let's add interactivity to our plot using Observables and sliders. We'll build this up step by step.","category":"section"},{"location":"model_simulation/examples/interactive_brusselator_simulation/#creating_observables","page":"Interactive Simulation and Plotting","title":"Creating Observables","text":"Observables are a key concept in reactive programming and are central to how Makie.jl creates interactive visualizations. You can read more about them here.\n\n# Create observables for parameters and initial conditions\nA = Observable(1.0)\nB = Observable(4.0)\nX0 = Observable(1.0)\nY0 = Observable(0.0)\n\nAn Observable is a container for a value that can change over time. When the value changes, any dependent computations are automatically updated.","category":"section"},{"location":"model_simulation/examples/interactive_brusselator_simulation/#adding_sliders","page":"Interactive Simulation and Plotting","title":"Adding sliders and connecting to Observables","text":"Let's add sliders that will control our Observables:\n\n# Create the main figure\nfig = Figure(size = (800, 600), fontsize = 18);\n\n# Create layout for plot and sliders\nplot_layout = fig[1, 1] = GridLayout()\nslider_layout = fig[2, 1] = GridLayout()\n\n# Create sliders\nslider_A = Slider(slider_layout[1, 1], range = 0.0:0.01:5.0, startvalue = to_value(A)) # to_value(A) unwraps the Observable to a value\nslider_B = Slider(slider_layout[2, 1], range = 0.0:0.01:5.0, startvalue = to_value(B))\nslider_X0 = Slider(slider_layout[3, 1], range = 0.0:0.01:5.0, startvalue = to_value(X0))\nslider_Y0 = Slider(slider_layout[4, 1], range = 0.0:0.01:5.0, startvalue = to_value(Y0))\n\n# Add labels for sliders\nLabel(slider_layout[1, 1, Left()], \"A\")\nLabel(slider_layout[2, 1, Left()], \"B\")\nLabel(slider_layout[3, 1, Left()], \"X₀\")\nLabel(slider_layout[4, 1, Left()], \"Y₀\")\n\n# Connect the values of the sliders to the observables\nconnect!(A, slider_A.value)\nconnect!(B, slider_B.value)\nconnect!(X0, slider_X0.value)\nconnect!(Y0, slider_Y0.value)\n\nThese sliders allow us to interactively change the parameters A and B, as well as the initial conditions X₀ and Y₀.","category":"section"},{"location":"model_simulation/examples/interactive_brusselator_simulation/#reactive_plot","page":"Interactive Simulation and Plotting","title":"Creating a reactive plot","text":"Now, let's create a plot that reacts to changes in our sliders:\n\n# Create an axis for the plot\nax = Axis(plot_layout[1, 1], \n    title = \"Brusselator Model\", \n    xlabel = \"Time\", \n    ylabel = \"Concentration\")\n\n# Create an observable for the solution\n# The `@lift` macro is used to create an Observable that depends on the observables `A`, `B`, `X0`, and `Y0`, and automatically updates when any of these observables change\nsolution = @lift(solve_brusselator($A, $B, $X0, $Y0))\n\n# Plot the solution\n# We don't use the ODESolution plot recipe here, as you've seen in the previous examples where only the solution and an `idxs` argument was passed to the plot method, because we are passing in an Observable wrapping the solution\nlines!(ax, lift(sol -> sol.t, solution), lift(sol -> sol[:X], solution), label = \"X\", color = :blue, linewidth = 3) # `lift` can either be used as a function or a macro\nlines!(ax, lift(sol -> sol.t, solution), lift(sol -> sol[:Y], solution), label = \"Y\", color = :red, linewidth = 3)\n\n# Add a legend\naxislegend(ax, position = :rt)\n\n# Display the figure\nfig\n\n(Image: Interactive Brusselator Plot)\n\nThis plot will now update in real-time as you move the sliders, allowing for interactive exploration of the Brusselator's behavior under different conditions. (Note the figure above is not interactive, but for illustrative purposes to show what you should see locally.)","category":"section"},{"location":"model_simulation/examples/interactive_brusselator_simulation/#adding_phase_plot","page":"Interactive Simulation and Plotting","title":"Adding a phase plot","text":"To gain more insight into the system's behavior, let's enhance our visualization by adding a phase plot, along with some other improvements:\n\n# Create the main figure\nfig = Figure(size = (1200, 800), fontsize = 18);\n\n# Create main layout: plots on top, sliders at bottom\nplot_grid = fig[1, 1] = GridLayout()\nslider_grid = fig[2, 1] = GridLayout()\n\n# Create sub-grids for plots\ntime_plot = plot_grid[1, 1] = GridLayout()\nphase_plot = plot_grid[1, 2] = GridLayout()\n\n# Create axes for the time series plot and phase plot\nax_time = Axis(time_plot[1, 1], \n    title = \"Brusselator Model - Time Series\", \n    xlabel = \"Time\", \n    ylabel = \"Concentration\")\n\nax_phase = Axis(phase_plot[1, 1], \n    title = \"Brusselator Model - Phase Plot\", \n    xlabel = \"X\", \n    ylabel = \"Y\")\n\n# Create sub-grids for sliders\nparam_grid = slider_grid[1, 1] = GridLayout()\nic_grid = slider_grid[1, 2] = GridLayout()\n\n# Create observables for parameters and initial conditions\nA = Observable{Float64}(1.0) # We can specify the type of the Observable value, which can help with type stability and performance\nB = Observable{Float64}(4.0)\nX0 = Observable{Float64}(1.0)\nY0 = Observable{Float64}(0.0)\n\n# Create sliders with labels and group titles\nLabel(param_grid[1, 1:2], \"Parameters\", fontsize = 22)\nslider_A = Slider(param_grid[2, 2], range = 0.0:0.01:5.0, startvalue = to_value(A))\nslider_B = Slider(param_grid[3, 2], range = 0.0:0.01:5.0, startvalue = to_value(B))\nLabel(param_grid[2, 1], \"A\")\nLabel(param_grid[3, 1], \"B\")\n\nLabel(ic_grid[1, 1:2], \"Initial Conditions\", fontsize = 22)\nslider_X0 = Slider(ic_grid[2, 2], range = 0.0:0.01:5.0, startvalue = to_value(X0))\nslider_Y0 = Slider(ic_grid[3, 2], range = 0.0:0.01:5.0, startvalue = to_value(Y0))\nLabel(ic_grid[2, 1], \"X₀\")\nLabel(ic_grid[3, 1], \"Y₀\")\n\n# Connect sliders to observables\nconnect!(A, slider_A.value)\nconnect!(B, slider_B.value)\nconnect!(X0, slider_X0.value)\nconnect!(Y0, slider_Y0.value)\n\n# Create an observable for the solution. \nsolution = @lift(solve_brusselator($A, $B, $X0, $Y0))\n\n# Plot the time series\nlines!(ax_time, lift(sol -> sol.t, solution), lift(sol -> sol[:X], solution), label = \"X\", color = :blue, linewidth = 3) \nlines!(ax_time, lift(sol -> sol.t, solution), lift(sol -> sol[:Y], solution), label = \"Y\", color = :red, linewidth = 3)\n\n# Plot the phase plot\nphase_plot_obj = lines!(ax_phase, lift(sol -> sol[:X], solution), lift(sol -> sol[:Y], solution), \n                        color = lift(sol -> sol.t, solution), colormap = :viridis)\n\n# Add a colorbar for the phase plot\nColorbar(phase_plot[1, 2], phase_plot_obj, label = \"Time\")\n\n# Add legends\naxislegend(ax_time, position = :rt)\n\n# Adjust layout to your liking\ncolgap!(plot_grid, 20)\nrowgap!(fig.layout, 20)\ncolgap!(param_grid, 10)\ncolgap!(ic_grid, 10)\n\n# Display the figure\nfig\n\nThis will create a visualization with both time series and phase plots:\n\n(Image: Interactive Brusselator Plot with Time Series and Phase Plot)","category":"section"},{"location":"model_simulation/examples/interactive_brusselator_simulation/#common_makie_plotting_options","page":"Interactive Simulation and Plotting","title":"Common plotting options","text":"Various plotting options can be provided as optional arguments to the lines! command. Common options include:\n\nlinewidth or lw: Determine plot line widths.\nlinestyle: Determines plot line style.\ncolor: Determines the line colors.\nlabel: Determines label texts displayed in the legend.\n\nFor example:\n\nlines!(ax_time, lift(sol -> sol.t, solution), lift(sol -> sol[:X], solution), \n       label = \"X\", color = :green, linewidth = 2, linestyle = :dash)","category":"section"},{"location":"model_simulation/examples/interactive_brusselator_simulation/#extending_interactive_visualization","page":"Interactive Simulation and Plotting","title":"Extending the interactive visualization","text":"You can further extend this visualization by:\n\nAdding other interactive elements, such as buttons or dropdown menus to control different aspects of the simulation or visualization.\nAdding additional axes to the plot, such as plotting the derivatives of the species.\nColor coding the slider and slider labels to match the plot colors.","category":"section"},{"location":"model_creation/events/#events","page":"Modelling Events","title":"Modelling Events","text":"<details><summary><strong>Environment setup and package installation</strong></summary>\n\nThe following code sets up an environment for running the code on this page.\n\nusing Pkg\nPkg.activate(; temp = true) # Creates a temporary environment, which is deleted when the Julia session ends.\nPkg.add(\"Catalyst\")\nPkg.add(\"OrdinaryDiffEqTsit5\")\nPkg.add(\"Plots\")\n\n</details>\n\n<details><summary><strong>Quick-start example</strong></summary>\n\nIn this quick-start example we show how to add a \"continuous event\" directly to a model declared through the DSL.\n\nusing Catalyst, OrdinaryDiffEqDefault, Plots\n\n# The `@continuous_events option is used to declare events.\n# When the condition on the event left hold, the effect on the right is triggered.\n# In this case, whenever `X` reached `1.0`, `X` is assigned the value `2.0`.\nrn = @reaction_network begin\n    @continuous_events [X ~ 1.0] => [X => 2]\n    d, X --> 0\nend\n\n# The model can be simulated using normal syntax.\nu0 = [:X => 2.0]\nps = [:d => 1.0]\noprob = ODEProblem(rn, u0, 10.0, ps)\nsol = solve(oprob)\nplot(sol)\n\nIn addition to continuous events, three different types of \"discrete events\" also exists. Furthermore, events can be added to programmatically created models through a slightly different syntax. Finally, callbacks (which are similar to events) can be added at the time of simulation.\n\n</details>\n\n\n   In many applications one needs to model events that can occur when a set condition is reached, such as providing a drug treatment at a specified time, or turning off production of cells once the population reaches a given level. Catalyst supports the event representation provided by ModelingToolkit, see here, allowing for both continuous and discrete events.\n\nIn this tutorial we'll illustrate how to make use of events. We will continue on the model  of a cell with volume V(t) that grows at a rate lambda that was considered in the coupled equations tutorial.","category":"section"},{"location":"model_creation/events/#events_base_example","page":"Modelling Events","title":"Adding events","text":"In the model presented there, it is unrealistic in assuming the cell will grow exponentially forever. Let's modify it such that the cell divides in half every time its volume reaches a size of 2. We also assume we lose half of the protein upon division. Note, we will only keep track of one cell, and hence follow a specific  lineage of the system. To do this we can create a continuous event using the ModelingToolkitBase symbolic event interface and attach it to our system. Please see the associated ModelingToolkitBase tutorial for more details on the types of events that can be represented symbolically. A lower-level approach for creating events via the DifferentialEquations.jl callback interface is illustrated here tutorial.\n\nusing Catalyst, OrdinaryDiffEqTsit5, Plots\n\nrn = @reaction_network growing_cell begin\n    # the growth rate\n    @parameters λ = 1.0\n\n    # assume there is no protein initially\n    @species P(t) = 0.0\n\n    # set the initial volume to 1.0\n    @variables V(t) = 1.0\n\n    # the reactions\n    V, 0 --> P\n    1.0, P --> 0\n\n    # the coupled ODE for V(t)\n    @equations begin\n        D(V) ~ λ * V\n    end\n\n    # every 1.0 time unit we half the volume of the cell and the number of proteins\n    @continuous_events begin \n        [V ~ 2.0] => [V => V/2, P => P/2]\n    end\nend\n\nWe can now create and simulate our model\n\noprob = ODEProblem(rn, [], (0.0, 10.0))\nsol = solve(oprob, Tsit5())\nplot(sol)\nCatalyst.PNG(plot(sol; fmt = :png, dpi = 200)) # hide\n\nWe can also model discrete events. Here at a time switch_time we will set the parameter k_on to be zero:\n\nrn = @reaction_network param_off_ex begin\n    @parameters switch_time\n    k_on, A --> B\n    k_off, B --> A\n    \n    @discrete_events begin\n        (t == switch_time) => [k_on => 0.0]\n    end\nend\n\nu0 = [:A => 10.0, :B => 0.0]\ntspan = (0.0, 4.0)\np = [:k_on => 100.0, :switch_time => 2.0, :k_off => 10.0]\noprob = ODEProblem(rn, u0, tspan, p)\nsol = solve(oprob, Tsit5(); tstops = 2.0)\nplot(sol)\n\nNote that for discrete events we need to set a stop time via tstops so that the ODE solver can step exactly to the specific time of our event. In the previous example we just manually set the numeric value of the parameter in the tstops kwarg to solve, however, it can often be convenient to instead get the value of the parameter from oprob and pass this numeric value. This helps ensure consistency between the value passed via p and/or symbolic defaults and what we pass as a tstop to solve. We can do this as\n\noprob = ODEProblem(rn, u0, tspan, p)\nswitch_time_val = oprob.ps[:switch_time]\nsol = solve(oprob, Tsit5(); tstops = switch_time_val)\nplot(sol)\n\nFor a detailed discussion on how to directly use the lower-level but more flexible DifferentialEquations.jl event/callback interface, see the tutorial on event handling using callbacks.","category":"section"},{"location":"model_creation/events/#events_symbolic","page":"Modelling Events","title":"Adding events via the symbolic interface","text":"Let's repeat the previous two models using the symbolic interface. We first create our equations and unknowns/species again\n\nusing Catalyst, OrdinaryDiffEqTsit5, Plots\nt = default_t()\nD = default_time_deriv()\n\n@parameters λ = 1.0\n@variables V(t) = 1.0\n@species P(t) = 0.0\neq = D(V) ~ λ * V\nrx1 = @reaction $V, 0 --> $P\nrx2 = @reaction 1.0, $P --> 0\n\nBefore creating our ReactionSystem we make the event.\n\n# every 1.0 time unit we half the volume of the cell and the number of proteins\ncontinuous_events = [V ~ 2.0] => [V ~ Pre(V)/2, P ~ Pre(P)/2]\n\nWe can now create and simulate our model\n\n@named rs = ReactionSystem([rx1, rx2, eq], t; continuous_events)\nrs = complete(rs)\n\noprob = ODEProblem(rs, [], (0.0, 10.0))\nsol = solve(oprob, Tsit5())\nplot(sol)\nCatalyst.PNG(plot(sol; fmt = :png, dpi = 200)) # hide\n\nWe can again also model discrete events. Similar to our example with continuous events, we start by creating reaction equations, parameters, variables, and unknowns.\n\nt = default_t()\n@parameters switch_time k_off\n@discretes k_on(t)\n@species A(t) B(t)\n\nrxs = [(@reaction $k_on, A --> B), (@reaction k_off, B --> A)]\n\nNow we add an event such that at time t (switch_time), k_on is set to zero.\n\ndiscrete_events = ModelingToolkitBase.SymbolicDiscreteCallback((t == switch_time) => [k_on ~ 0.0]; discrete_parameters = [k_on])\n\nu0 = [:A => 10.0, :B => 0.0]\ntspan = (0.0, 4.0)\np = [k_on => 100.0, switch_time => 2.0, k_off => 10.0]\n\nSimulating our model,\n\n@named rs2 = ReactionSystem(rxs, t, [A, B], [k_on, k_off, switch_time]; discrete_events)\nrs2 = complete(rs2)\n\noprob = ODEProblem(rs2, u0, tspan, p)\nswitch_time_val = oprob.ps[:switch_time]\nsol = solve(oprob, Tsit5(); tstops = switch_time_val)\nplot(sol)","category":"section"},{"location":"model_creation/parametric_stoichiometry/#parametric_stoichiometry","page":"Symbolic Stoichiometries","title":"Symbolic Stoichiometries","text":"<details><summary><strong>Environment setup and package installation</strong></summary>\n\nThe following code sets up an environment for running the code on this page.\n\nusing Pkg\nPkg.activate(; temp = true) # Creates a temporary environment, which is deleted when the Julia session ends.\nPkg.add(\"Catalyst\")\nPkg.add(\"Distributions\")\nPkg.add(\"JumpProcesses\")\nPkg.add(\"Latexify\")\nPkg.add(\"ModelingToolkitBase\")\nPkg.add(\"OrdinaryDiffEqTsit5\")\nPkg.add(\"Plots\")\n\n</details>\n\n<details><summary><strong>Quick-start example</strong></summary>\n\nReaction stoichiometric coefficients can be parameters, which values are then designated after model creation. In the following example we designate n as a parametric stoichiometry.\n\nusing Catalyst, OrdinaryDiffEqDefault, Plots\nrn = @reaction_network begin\n    (kB,kD), n*X <--> Xn\nend\nu0 = [:X => 10.0, :Xn => 1.0]\nps = (:kB => 0.2, :kD => 1.0, :n => 3)\noprob = ODEProblem(rn, u0, 10.0, ps)\nsol = solve(oprob)\nplot(sol)\n\n</details>\n\n\n   Catalyst supports stoichiometric coefficients that involve parameters, species, or even general expressions. In this tutorial we show several examples of how to use symbolic stoichiometries, and discuss several caveats to be aware of.","category":"section"},{"location":"model_creation/parametric_stoichiometry/#Using-symbolic-stoichiometry","page":"Symbolic Stoichiometries","title":"Using symbolic stoichiometry","text":"Let's first consider a simple reversible reaction where the number of reactants is a parameter, and the number of products is the product of two parameters.\n\nusing Catalyst, Latexify, OrdinaryDiffEqTsit5, ModelingToolkitBase, Plots\nrevsys = @reaction_network revsys begin\n    @parameters m::Int64 n::Int64\n    k₊, m*A --> (m*n)*B\n    k₋, B --> A\nend\nreactions(revsys)\n\nNotice, as described in the Reaction rate laws used in simulations section, the default rate laws involve factorials in the stoichiometric coefficients. For this reason we explicitly specify m and n as integers (as otherwise ModelingToolkit will implicitly assume they are floating point).\n\nAs always the @reaction_network macro defaults to setting all symbols neither used as a reaction substrate nor a product to be parameters. Hence, in this example we have two species (A and B) and four parameters (k₊, k₋, m, and n). In addition, the stoichiometry is applied to the rightmost symbol in a given term, i.e. in the first equation the substrate A has stoichiometry m and the product B has stoichiometry m*n. For example, in\n\nrn = @reaction_network begin\n    k, A*C --> 2B\n    end\nreactions(rn)\n\nwe see two species, (B,C), with A treated as a parameter representing the stoichiometric coefficient of C, i.e.\n\nrx = reactions(rn)[1]\nrx.substrates[1],rx.substoich[1]\n\nWe could have equivalently specified our systems directly via the Catalyst API. For example, for revsys we would could use\n\nt = default_t()\n@parameters k₊ k₋ m::Int n::Int\n@species A(t), B(t)\nrxs = [Reaction(k₊, [A], [B], [m], [m*n]),\n       Reaction(k₋, [B], [A])]\nrevsys2 = ReactionSystem(rxs,t; name=:revsys)\nrevsys2 == revsys\n\nor\n\nrxs2 = [(@reaction k₊, $m*A --> ($m*$n)*B),\n        (@reaction k₋, B --> A)]\nrevsys3 = ReactionSystem(rxs2,t; name=:revsys)\nrevsys3 == revsys\n\nHere we interpolate in the pre-declared m and n symbolic variables using $m and $n to ensure the parameter is known to be integer-valued. The @reaction macro again assumes all symbols are parameters except the substrates or reactants (i.e. A and B). For example, in @reaction k, F*A + 2(H*G+B) --> D, the substrates are (A,G,B) with stoichiometries (F,2*H,2).\n\nLet's now convert revsys to ODEs and look at the resulting equations:\n\nosys = ode_model(revsys)\nosys = complete(osys)\nequations(osys)\nshow(stdout, MIME\"text/plain\"(), equations(osys)) # hide\n\nSpecifying the parameter and initial condition values,\n\nsim_cond = (revsys.k₊ => 1.0, revsys.k₋ => 1.0, revsys.m => 2, revsys.n => 2,\n    revsys.A => 1.0, revsys.B => 1.0)\noprob = ODEProblem(osys, sim_cond, (0.0, 1.0))\nnothing # hide\n\nwe can now solve and plot the system\n\nsol = solve(oprob, Tsit5())\nplot(sol)\n\nAn alternative approach to avoid the issues of using mixed floating point and integer variables is to disable the rescaling of rate laws as described in Reaction rate laws used in simulations section. This requires passing the combinatoric_ratelaws=false keyword to convert or to ODEProblem (if directly building the problem from a ReactionSystem instead of first converting to an ODE System). For the previous example this gives the following (different) system of ODEs where we now let m and n be floating point valued parameters (the default):\n\nrevsys = @reaction_network revsys begin\n    k₊, m*A --> (m*n)*B\n    k₋, B --> A\nend\nosys = ode_model(revsys; combinatoric_ratelaws = false)\nosys = complete(osys)\nequations(osys)\nshow(stdout, MIME\"text/plain\"(), equations(osys)) # hide\n\nSince we no longer have factorial functions appearing, our example will now run with m and n treated as floating point parameters:\n\nsim_cond = (revsys.k₊ => 1.0, revsys.k₋ => 1.0, revsys.m => 2.0, revsys.n => 2.0,\n    revsys.A => 1.0, revsys.B => 1.0)\noprob = ODEProblem(osys, sim_cond, (0.0, 1.0))\nsol = solve(oprob, Tsit5())\nplot(sol)","category":"section"},{"location":"model_creation/parametric_stoichiometry/#Gene-expression-with-randomly-produced-amounts-of-protein","page":"Symbolic Stoichiometries","title":"Gene expression with randomly produced amounts of protein","text":"As a second example, let's build the negative feedback model from MomentClosure.jl that involves a bursty reaction that produces a random amount of protein.\n\nIn our model G₋ will denote the repressed state, and G₊ the active state where the gene can transcribe. P will denote the protein product of the gene. We will assume that proteins are produced in bursts that produce m proteins, where m is a (shifted) geometric random variable with mean b. To define m we must register the Distributions.Geometric distribution from Distributions.jl with Symbolics.jl, after which we can use it in symbolic expressions:\n\nusing Distributions: Geometric\n@register_symbolic Geometric(b)\n@parameters b\nm = rand(Geometric(1/b)) + 1\nnothing # hide\n\nNote, as we require the shifted geometric distribution, we add one to Distributions.jl's Geometric random variable (which includes zero).\n\nWe can now define our model\n\nburstyrn = @reaction_network burstyrn begin\n    k₊, G₋ --> G₊\n    k₋*P^2, G₊ --> G₋\n    kₚ, G₊ --> G₊ + $m*P\n    γₚ, P --> ∅\nend\nreactions(burstyrn)\nshow(stdout, MIME\"text/plain\"(), reactions(burstyrn)) # hide\n\nThe parameter b does not need to be explicitly declared in the @reaction_network macro as it is detected when the expression rand(Geometric(1/b)) + 1 is substituted for m.\n\nWe next convert our network to a jump process representation\n\nusing JumpProcesses\njsys = jump_model(burstyrn; combinatoric_ratelaws = false)\njsys = complete(jsys)\njumps(jsys)\nshow(stdout, MIME\"text/plain\"(), jumps(jsys)) # hide\n\nNotice, the equations of jsys have three MassActionJumps for the first three reactions, and one ConstantRateJump for the last reaction. If we examine the ConstantRateJump more closely we can see the generated rate and affect! functions for the bursty reaction that makes protein\n\njumps(jsys)[4].rate\nshow(stdout, MIME\"text/plain\"(), jumps(jsys)[4].rate) # hide\n\njumps(jsys)[4].affect!\nshow(stdout, MIME\"text/plain\"(), jumps(jsys)[4].affect!) # hide\n\nFinally, we can now simulate our jump System\n\npmean = 200\nbval = 70\nγₚval = 1\nk₋val = 0.001\nk₊val = 0.05\nkₚval = pmean * γₚval * (k₋val * pmean^2 + k₊val) / (k₊val * bval)\nsim_cond = [\n    :k₊ => k₊val, :k₋ => k₋val, :kₚ => kₚval, :γₚ => γₚval, :b => bval,\n    :G₊ => 1, :G₋ => 0, :P => 1\n]\ntspan = (0., 6.0)   # time interval to solve over\njprob = JumpProblem(jsys, sim_cond, tspan)\nsol = solve(jprob)\nplot(sol.t, sol[jsys.P], legend = false, xlabel = \"time\", ylabel = \"P(t)\")\n\nTo double check our results are consistent with MomentClosure.jl, let's calculate and plot the average amount of protein (which is also plotted in the MomentClosure.jl tutorial).\n\nt = default_t()\nfunction getmean(jprob, Nsims, tv)\n    Pmean = zeros(length(tv))\n    @variables P(t)\n    for n in 1:Nsims\n        sol = solve(jprob)\n        Pmean .+= sol(tv, idxs=P)\n    end\n    Pmean ./= Nsims\nend\ntv = range(tspan[1],tspan[2],step=.1)\npsim_mean = getmean(jprob, 20000, tv)\nplot(tv, psim_mean; ylabel = \"average of P(t)\", xlabel = \"time\",\n                    xlim = (0.0,6.0), legend = false)\n\nComparing, we see similar averages for P(t).","category":"section"},{"location":"#doc_index","page":"Home","title":"Catalyst.jl for Reaction Network Modeling","text":"Catalyst.jl is a symbolic modeling package for analysis and high-performance simulation of chemical reaction networks. Catalyst defines symbolic ReactionSystems, which can be created programmatically or easily specified using Catalyst's domain-specific language (DSL). Leveraging ModelingToolkitBase.jl and Symbolics.jl, Catalyst enables large-scale simulations through auto-vectorization and parallelism. Symbolic ReactionSystems can be used to generate ModelingToolkit-based models, allowing the easy simulation and parameter estimation of mass action ODE models, Chemical Langevin SDE models, stochastic chemical kinetics jump process models, and more. Generated models can be used with solvers throughout the broader Julia and SciML ecosystems, including higher-level SciML packages (e.g. for sensitivity analysis, parameter estimation, machine learning applications, etc).","category":"section"},{"location":"#doc_index_features","page":"Home","title":"Features","text":"","category":"section"},{"location":"#doc_index_features_catalyst","page":"Home","title":"Features of Catalyst","text":"The Catalyst DSL provides a simple and readable format for manually specifying reaction network models using chemical reaction notation.\nCatalyst ReactionSystems provides a symbolic representation of reaction networks, built on ModelingToolkitBase.jl and Symbolics.jl.\nThe Catalyst.jl API provides functionality for building networks programmatically and for composing multiple networks together.\nLeveraging ModelingToolkitBase, generated models can be converted to symbolic reaction rate equation ODE models, symbolic Chemical Langevin Equation models, and symbolic stochastic chemical kinetics (jump process) models. These can be simulated using any DifferentialEquations.jl ODE/SDE/jump solver, and can be used within EnsembleProblems for carrying out parallelized parameter sweeps and statistical sampling. Plot recipes are available for visualization of all solutions.\nNon-integer (e.g. Float64) stoichiometric coefficients are supported for generating ODE models, and symbolic expressions for stoichiometric coefficients are supported for all system types.\nA network analysis suite permits the computation of linkage classes, deficiencies, reversibility, and other network properties.\nConservation laws can be detected and utilized to reduce system sizes, and to generate non-singular Jacobians (e.g. during conversion to ODEs, SDEs, and steady state equations).\nCatalyst reaction network models can be coupled with differential and algebraic equations (which are then incorporated during conversion to ODEs, SDEs, and steady state equations).\nModels can be coupled with events that affect the system and its state during simulations.\nBy leveraging ModelingToolkit, users have a variety of options for generating optimized system representations to use in solvers. These include construction of dense or sparse Jacobians, multithreading or parallelization of generated derivative functions, automatic classification of reactions into optimized jump types for Gillespie type simulations, automatic construction of dependency graphs for jump systems, and more.\nSymbolics.jl symbolic expressions and Julia Exprs can be obtained for all rate laws and functions determining the deterministic and stochastic terms within resulting ODE, SDE, or jump models.\nSteady states (and their stabilities) can be computed for model ODE representations.","category":"section"},{"location":"#doc_index_features_composed","page":"Home","title":"Features of Catalyst composing with other packages","text":"OrdinaryDiffEq.jl Can be used to numerically solve generated reaction rate equation ODE models.\nStochasticDiffEq.jl can be used to numerically solve generated Chemical Langevin Equation SDE models.\nJumpProcesses.jl can be used to numerically sample generated Stochastic Chemical Kinetics Jump Process models.\nSupport for parallelization of all simulations, including parallelization of ODE and SDE simulations on GPUs using DiffEqGPU.jl.\nLatexify can be used to generate LaTeX expressions corresponding to generated mathematical models or the underlying set of reactions.\nGraphMakie recipes are provided that can be used to generate and visualize reaction network graphs \nModel steady states can be computed through homotopy continuation using HomotopyContinuation.jl (which can find all steady states of systems with multiple ones), by forward ODE simulations using SteadyStateDiffEq.jl, or by numerically solving steady-state nonlinear equations using NonlinearSolve.jl.\nBifurcationKit.jl can be used to compute bifurcation diagrams of model steady states (including finding periodic orbits).\nDynamicalSystems.jl can be used to compute model basins of attraction, Lyapunov spectrums, and other dynamical system properties.\nOptimization.jl and PEtab.jl can all be used to fit model parameters to data.\nGlobalSensitivity.jl can be used to perform global sensitivity analysis of model behaviors.\nSciMLSensitivity.jl can be used to compute local sensitivities of functions containing forward model simulations.\nStructuralIdentifiability.jl can be used to perform structural identifiability analysis.","category":"section"},{"location":"#doc_index_features_other_packages","page":"Home","title":"Features of packages built upon Catalyst","text":"Catalyst ReactionSystems can be imported from SBML files via SBMLImporter.jl and SBMLToolkit.jl, and from BioNetGen .net files and various stoichiometric matrix network representations using ReactionNetworkImporters.jl.\nMomentClosure.jl allows generation of symbolic ModelingToolkitBase ODE Systems that represent moment closure approximations to moments of the Chemical Master Equation, from reaction networks defined in Catalyst.\nFiniteStateProjection.jl allows the construction and numerical solution of Chemical Master Equation models from reaction networks defined in Catalyst.\nDelaySSAToolkit.jl can augment Catalyst reaction network models with delays, and can simulate the resulting stochastic chemical kinetics with delays models.","category":"section"},{"location":"#doc_index_documentation","page":"Home","title":"How to read this documentation","text":"The Catalyst documentation is separated into sections describing Catalyst's various features. Where appropriate, some sections will also give advice on best practices for various modeling workflows, and provide links with further reading. Each section also contains a set of relevant example workflows. Finally, the API section contains a list of all functions exported by Catalyst (as well as descriptions of them and their inputs and outputs).\n\nNew users are recommended to start with either the Introduction to Catalyst and Julia for New Julia users or Introduction to Catalyst sections (depending on whether they are familiar with Julia programming or not). This should be enough to carry out many basic Catalyst workflows.\n\nThis documentation contains code which is dynamically run whenever it is built. If you copy the code and run it in your Julia environment it should work. The exact Julia environment that is used in this documentation can be found here.\n\nFor most code blocks in this documentation, the output of the last line of code is printed at the of the block, e.g.\n\n1 + 2\n\nand\n\nusing Catalyst # hide\n@reaction_network begin\n    (p,d), 0 <--> X\nend\n\nHowever, in some situations (e.g. when output is extensive, or irrelevant to what is currently being described) we have disabled this, e.g. like here:\n\n1 + 2\nnothing # hide\n\nand here:\n\n@reaction_network begin\n    (p,d), 0 <--> X\nend\nnothing # hide","category":"section"},{"location":"#doc_index_installation","page":"Home","title":"Installation","text":"Catalyst is an officially registered Julia package, which can be installed through the Julia package manager:\n\nusing Pkg\nPkg.add(\"Catalyst\")\n\nMany Catalyst features require the installation of additional packages. E.g. for ODE-solving and simulation plotting\n\nPkg.add(\"OrdinaryDiffEqDefault\")\nPkg.add(\"Plots\")\n\nis also needed.\n\nIt is strongly recommended to install and use Catalyst in its own environment with the minimal set of needed packages. For example, to install Catalyst and Plots in a new environment named catalyst_project (saved in the current directory) one can say\n\nPkg.activate(\"catalyst_project\")\nPkg.add(\"Catalyst\")\nPkg.add(\"Plots\")\n\nIf one would rather just create a temporary environment that is not saved when exiting Julia you can say\n\nPkg.activate(; temp = true)\nPkg.add(\"Catalyst\")\nPkg.add(\"Plots\")\n\nAfter installation, we suggest running \n\nPkg.status(\"Catalyst\")\n\nto confirm that the latest version of Catalyst was installed (and not an older version).  If you have installed into a new environment this should always be the case. However, if you  installed into an existing environment, such as the default Julia global environment, the presence of incompatible versions of other pre-installed packages could lead to older versions of Catalyst  being installed. In this case we again recommend creating a new environment and installing Catalyst  there to obtain the latest version.\n\nA more thorough guide for setting up Catalyst and installing Julia packages can be found here.","category":"section"},{"location":"#doc_index_example","page":"Home","title":"Illustrative example","text":"","category":"section"},{"location":"#doc_index_example_ode","page":"Home","title":"Deterministic ODE simulation of Michaelis-Menten enzyme kinetics","text":"Here we show a simple example where a model is created using the Catalyst DSL, and then simulated as an ordinary differential equation.\n\n# Fetch required packages.\nusing Catalyst, OrdinaryDiffEqDefault, Plots\n\n# Create model.\nmodel = @reaction_network begin\n    kB, S + E --> SE\n    kD, SE --> S + E\n    kP, SE --> P + E\nend\n\n# Create an ODE that can be simulated.\nu0 = [:S => 50.0, :E => 10.0, :SE => 0.0, :P => 0.0]\ntspan = (0., 200.)\nps = [:kB => 0.01, :kD => 0.1, :kP => 0.1]\node = ODEProblem(model, u0, tspan, ps)\n\n# Simulate ODE and plot results.\nsol = solve(ode)\nplot(sol; lw = 5)","category":"section"},{"location":"#doc_index_example_jump","page":"Home","title":"Stochastic jump simulations","text":"The same model can be used as input to other types of simulations. E.g. here we instead generate and simulate a stochastic chemical kinetics jump process model.\n\n# Create and simulate a jump process (here using Gillespie's direct algorithm).\n# The initial conditions are now integers as we track exact populations for each species.\nusing JumpProcesses\nu0_integers = [:S => 50, :E => 10, :SE => 0, :P => 0]\njprob = JumpProblem(model, u0_integers, tspan, ps)\njump_sol = solve(jprob)\njump_sol = solve(jprob; seed = 1234) # hide\nplot(jump_sol; lw = 2)","category":"section"},{"location":"#doc_index_elaborate_example","page":"Home","title":"More elaborate example","text":"In the above example, we used basic Catalyst workflows to simulate a simple model. Here we instead show how various Catalyst features can compose to create a much more advanced model. Our model describes how the volume of a cell (V) is affected by a growth factor (G). The growth factor only promotes growth while in its phosphorylated form (G^P). The phosphorylation of G (G to G^P) is promoted by sunlight, which is modeled as the cyclic sinusoid k_a (sin(t) + 1). When the cell reaches a critical volume (V_m) it undergoes cell division. First, we  declare our model:\n\nusing Catalyst\ncell_model = @reaction_network begin\n    @parameters Vₘ g\n    @equations begin\n        D(V) ~ g*Gᴾ\n    end\n    @continuous_events begin\n        [V ~ Vₘ] => [V => V/2]\n    end\n    kₚ*(sin(t)+1)/V, G --> Gᴾ\n    kᵢ/V, Gᴾ --> G\nend\n\nWe now study the system as a Chemical Langevin Dynamics SDE model, which can be generated as follows\n\nu0 = [:V => 25.0, :G => 50.0, :Gᴾ => 0.0]\ntspan = (0.0, 20.0)\nps = [:Vₘ => 50.0, :g => 0.3, :kₚ => 100.0, :kᵢ => 60.0]\nsprob = SDEProblem(cell_model, u0, tspan, ps)\n\nThis problem encodes the following stochastic differential equation model:\n\nbeginalign*\ndG(t) =  - left( frack_p(sin(t)+1)V(t) G(t) + frack_iV(t) G^P(t) right) dt - sqrtfrack_p (sin(t)+1)V(t) G(t)  dW_1(t) + sqrtfrack_iV(t) G^P(t)  dW_2(t) \ndG^P(t) = left( frack_p(sin(t)+1)V(t) G(t) - frack_iV(t) G^P(t) right) dt + sqrtfrack_p (sin(t)+1)V(t) G(t)  dW_1(t) - sqrtfrack_iV(t) G^P(t)  dW_2(t) \ndV(t) = left(g  G^P(t)right) dt\nendalign*\n\nwhere the dW_1(t) and dW_2(t) terms represent independent Brownian Motions, encoding the noise added by the Chemical Langevin Equation. Finally, we can simulate and plot the results.\n\nusing StochasticDiffEq, Plots\nsol = solve(sprob, EM(); dt = 0.05)\nsol = solve(sprob, EM(); dt = 0.05, seed = 1234) # hide\nplot(sol; xguide = \"Time (au)\", lw = 2)","category":"section"},{"location":"#doc_index_help","page":"Home","title":"Getting Help","text":"Catalyst developers are active on the Julia Discourse and the Julia Slack channels #sciml-bridged and #sciml-sysbio. For bugs or feature requests, open an issue.","category":"section"},{"location":"#doc_index_citation","page":"Home","title":"Supporting and Citing Catalyst.jl","text":"The software in this ecosystem was developed as part of academic research. If you would like to help support it, please star the repository as such metrics may help us secure funding in the future. If you use Catalyst as part of your research, teaching, or other activities, we would be grateful if you could cite our work:\n\n@article{CatalystPLOSCompBio2023,\n doi = {10.1371/journal.pcbi.1011530},\n author = {Loman, Torkel E. AND Ma, Yingbo AND Ilin, Vasily AND Gowda, Shashi AND Korsbo, Niklas AND Yewale, Nikhil AND Rackauckas, Chris AND Isaacson, Samuel A.},\n journal = {PLOS Computational Biology},\n publisher = {Public Library of Science},\n title = {Catalyst: Fast and flexible modeling of reaction networks},\n year = {2023},\n month = {10},\n volume = {19},\n url = {https://doi.org/10.1371/journal.pcbi.1011530},\n pages = {1-19},\n number = {10},\n}","category":"section"},{"location":"#doc_index_reproducibility","page":"Home","title":"Reproducibility","text":"<details><summary>The documentation of this SciML package was built using these direct dependencies,</summary>\n\nusing Pkg # hide\nPkg.status() # hide\n\n</details>\n\n<details><summary>and using this machine and Julia version.</summary>\n\nusing InteractiveUtils # hide\nversioninfo() # hide\n\n</details>","category":"section"},{"location":"spatial_modelling/discrete_space_simulation_plotting/#dspace_simulation_plotting","page":"Plotting Spatial Simulations","title":"Plotting Spatial Simulations","text":"To aid the investigation of spatial simulations we have implemented several helper functions for creating plots and animation of these simulations. The development of these functions are currently under development. Currently, Catalyst exports three different such functions:\n\ndspace_plot: Which plots a discrete space simulation at a specified time point.\ndspace_animation: Which creates an animation of a discrete space simulation across the entire time series.\ndspace_kymograph: Which, for a 1d Cartesian or masked discrete space based simulation, creates a kymograph of the simulation across time and space.\n\nThe first two functions can be applied to graph and 1d and 2d Cartesian and masked space based simulations, while dspace_kymograph can only be applied to 1d Cartesian and masked space-based simulations. Currently, there is no functionality for plotting simulations based on 3d Cartesian and masked spaces. Here we will demonstrate all plotting functions using ODE simulations, but they work equally well for jump simulations.\n\nnote: Note\nThe plotting interfaces used for non-spatial Catalyst simulations have seen lots of work to ensure high quality plots. However, the corresponding functions for spatial simulations are primarily intended to aid the user to investigate their simulation results. Thus, they might not be fully suitable for e.g. creating publication-quality graphics. If you are using these functions, please let us now. This helps inform us whether continued development of spatial modelling features is worthwhile. \n\nnote: Note\nTo create animations we use Makie.jl, which is an alternative plotting package to Plots.jl (which is typically the preferred plotting package within the context of Catalyst). Generally, Makie is good at creating animations, hence we use it here (however, it is also a popular competitor to Plots.jl for general-purpose plotting). \n\nwarning: Warning\nThese plotting interfaces are a work in progress. Hence, they and their interfaces may see more change that Catalyst features typically do. This include the possibility of breaking changes without breaking releases to Catalyst.","category":"section"},{"location":"spatial_modelling/discrete_space_simulation_plotting/#dspace_simulation_plotting_1d_grids","page":"Plotting Spatial Simulations","title":"Animation and plotting of 1d Cartesian or masked discrete space simulations","text":"Let us consider a spatial simulation on a 1d Cartesian grid space:\n\nusing Catalyst, OrdinaryDiffEqDefault\ntwo_state_model = @reaction_network begin\n    (k1,k2), X1 <--> X2\nend\ndiffusion_rx = @transport_reaction D X1\nspace = CartesianGrid(10)\ndsrs = DiscreteSpaceReactionSystem(two_state_model, [diffusion_rx], space)\n\nX1_0 = ones(10)\nX1_0[5] = 10.0\nu0 = [:X1 => X1_0, :X2 => 1.0]\ntspan = (0.0, 5.0)\nps = [:k1 => 1.0, :k2 => 2.0, :D => 0.5]\noprob = ODEProblem(dsrs, u0, tspan, ps)\nsol = solve(oprob)\nnothing # hide\n\nTo plot the simulation at a specific time point we use the dspace_plot function. In addition to the simulation we wish to plot, it takes the species we wish to plot and the DiscreteSpaceReactionSystem which generated the simulation as arguments. It also takes the time point at which we wish to plot the simulation as an additional argument (if not provided, the simulation will be plotted at the final time point). To use the dspace_plot function (or any other of Catalyst's spatial plotting functions) we also need to load the CairoMakie package (here, import CairoMakie is enough, and using CairoMakie is not required).\n\nimport CairoMakie\ndspace_plot(sol, :X1, dsrs; t = 2.0)\n\nIf we instead wish to create an animation of our solution across the entire simulation, we can use the dspace_animation function. This takes a fourth required argument, a file to which the animation is saved.\n\ndspace_animation(sol, :X1, dsrs, \"dspace_simulation_1d.mp4\")\n\n(Image: ) Since we animate the solution across the entire simulation, we do not need to provide a t value. However, there are some additional (optional) arguments we might wish to provide:\n\nnframes = 200: The number of frames in the animation (these are evenly samples across the simulation).\nframerate = 20: The frame rate of the animation.\n\nFinally, we can display a kymograph of our simulation across the full time span using dspace_kymograph.\n\ndspace_kymograph(sol, :X1, dsrs)\n\nHere, we require neither a filename nor a t to be provided. However, the nframes argument can still be used to determine how frequently (in time) we wish to sample our simulation.\n\nFor more information of either function, and additional optional arguments, please read their corresponding api sections (dspace_plot, dspace_animation, and dspace_kymograph).","category":"section"},{"location":"spatial_modelling/discrete_space_simulation_plotting/#dspace_simulation_plotting_2d_grids","page":"Plotting Spatial Simulations","title":"Animation and plotting of 2d Cartesian or masked space simulations","text":"Two-dimensional discrete space simulations can be plotted in the same manner as one-dimensional ones. However, instead of displaying a species's value as a line plot, it is displayed as a heatmap. E.g. here we simulate a spatial Brusselator model and display the value of X at a designated time point.\n\nusing Catalyst, OrdinaryDiffEqBDF\nbrusselator = @reaction_network begin\n    A, ∅ --> X\n    1, 2X + Y --> 3X\n    B, X --> Y\n    1, X --> ∅\nend\ndiffusion_rx = @transport_reaction D X\nspace = CartesianGrid((20,20))\ndsrs = DiscreteSpaceReactionSystem(brusselator, [diffusion_rx], space)\n\nu0 = [:X => rand(20, 20), :Y => 10.0]\ntspan = (0.0, 20.0)\nps = [:A => 1.0, :B => 4.0, :D => 0.2]\noprob = ODEProblem(dsrs, u0, tspan, ps; jac = true, sparse = true)\nsol = solve(oprob, FBDF())\n\nimport CairoMakie\ndspace_plot(sol, :X, dsrs; t = 18.0)\n\nHere, the intensity of the colour scale describes X's value in each compartment. The color scale is normalised across the entire simulation (not the plotted time step).\n\nAn animation of the solution can be created in a similar manner as for the one-dimensional case:\n\ndspace_animation(sol, :X, dsrs, \"dspace_simulation_2d.mp4\")\n\n(Image: )\n\nAgain, please check the API pages for the dspace_plot and dspace_animation functions to see more details of their various options.","category":"section"},{"location":"spatial_modelling/discrete_space_simulation_plotting/#dspace_simulation_plotting_graphs","page":"Plotting Spatial Simulations","title":"Animation and plotting of graph space simulations","text":"Finally, we consider space simulations on graph spaces. We first simulate a simple birth-death process on a (6-node cyclic) graph space.\n\nusing Catalyst, Graphs, OrdinaryDiffEqDefault\nrs = @reaction_network begin\n    (p,d), 0 <--> X\nend\ndiffusion_rx = @transport_reaction D X\nspace = cycle_graph(6)\ndsrs = DiscreteSpaceReactionSystem(rs, [diffusion_rx], space)\n\nu0 = [:X => rand(6)]\ntspan = (0.0, 1.0)\nps = [:p => 1.0, :d => 0.5, :D => 0.1]\noprob = ODEProblem(dsrs, u0, tspan, ps)\nosol = solve(oprob)\nnothing # hide\n\nAs for Cartesian and masked space-based simulations we can plot this simulation using the dspace_plot function. In this case, however, we need to first load both CairoMakie and the GraphMakie packages.\n\nimport CairoMakie, GraphMakie\ndspace_plot(osol, :X, dsrs)\n\nHere we provide no t argument, and hence plot the solution at the final time point. As for 2d simulations, a colour scale described X's concentration in each compartment. Some arguments, are not relevant when plotting simulations on Cartesian and masked spaces, but which are relevant to graph spaces are:\n\nnode_size: Sets the size of each compartment in the plot.\nlayout: A vector of x and y values which can be used to determine the position of each compartment in the plot.\n\nE.g. here we manually set where each compartment should be plotted:\n\nisq2 = 1/sqrt(2)\nlayout = [(0, 0), (isq2, isq2), (isq2, 1 + isq2), (0, 1 + 2isq2), (-isq2, 1 + isq2), (-isq2, isq2)]\ndspace_plot(osol, :X, dsrs; layout)\n\nFinally, animations of graph space simulation work similarly to 2d ones, but accept the additional arguments relevant to plotting graphs.","category":"section"},{"location":"spatial_modelling/discrete_space_simulation_plotting/#dspace_simulation_plotting_notes","page":"Plotting Spatial Simulations","title":"Final notes","text":"If you are using these interfaces, but there is some feature that is missing, you might wish to consider modifying the original code. This can be found here, from which you can copy any code you need to make your own plotting interfaces. If you do so, please provide any feedback by raising an issue on the Catalyst GitHub page. As mentioned, these plotting interfaces are a work in progress, and input from users is valuable to improve them further. Many of Makie's plotting arguments, even those not described here, are handled by these functions, and someone familiar with the package should be able to use these to customise the plots further.\n\nIt should also be noted that these interfaces has note been optimised for performance, and the generation of an animation can often surpass 1 second for larger models. Again, this can likely be improved, and if performance is an problem do raise an issue, in which case an additional effort can be made to improve performance.","category":"section"},{"location":"model_simulation/examples/periodic_events_simulation/#periodic_event_simulation_example","page":"Modelling a Periodic Event During ODE and Jump Simulations","title":"Modelling a Periodic Event During ODE and Jump Simulations","text":"<details><summary><strong>Environment setup and package installation</strong></summary>\n\nThe following code sets up an environment for running the code on this page.\n\nusing Pkg\nPkg.activate(; temp = true) # Creates a temporary environment, which is deleted when the Julia session ends.\nPkg.add(\"Catalyst\")\nPkg.add(\"JumpProcesses\")\nPkg.add(\"OrdinaryDiffEqDefault\")\nPkg.add(\"Plots\")\n\n</details>\n\n\n   This tutorial will describe how to simulate systems with periodic events in ODEs and jump simulations (SDEs use identical syntax). We will consider a model with a circadian rhythm, where a parameter represents the level of light. While outdoor light varies smoothly, in experimental settings a lamp is often simply turned on/off every 12 hours. Here we will model this toggling of the light using a periodic event that is triggered every 12 hours.","category":"section"},{"location":"model_simulation/examples/periodic_events_simulation/#periodic_event_simulation_example_ode","page":"Modelling a Periodic Event During ODE and Jump Simulations","title":"Modelling a circadian periodic event in an ODE simulation","text":"We will consider a simple circadian model, consisting of a single protein (X), which is phosphorylated (X to Xᴾ) in the presence of light (l). Here, the light parameter can either be 0 (night) or 1 (day). We can model this using a simple periodic event which switches the value of l every 12 hours (here, % is the remainder operator).\n\nusing Catalyst\ncircadian_model = @reaction_network begin\n    @discrete_events begin\n        12 => [l => (l + 1)%2]\n    end\n (kₚ*l,kᵢ), X <--> Xᴾ\nend\n\nWe can now simulate this model, observing how a 24-hour cycle is reached\n\nusing OrdinaryDiffEqDefault, Plots\nu0 = [:X => 150.0, :Xᴾ => 50.0]\nps = [:kₚ => 0.1, :kᵢ => 0.1, :l => 1.0]\ntspan = (0.0, 100.0)\noprob = ODEProblem(circadian_model, u0, tspan, ps)\nsol = solve(oprob)\nplot(sol)","category":"section"},{"location":"model_simulation/examples/periodic_events_simulation/#periodic_event_simulation_example_ode-2","page":"Modelling a Periodic Event During ODE and Jump Simulations","title":"Modelling a circadian periodic event in a jump simulation","text":"We can define periodic events in an identical manner for jump simulations. Let's reuse our previously defined network, but now simulate it as a stochastic chemical kinetics jump process model\n\nusing JumpProcesses\nu0 = [:X => 150, :Xᴾ => 50]  # define u0 as integers now\njprob = JumpProblem(circadian_model, u0, tspan, ps)\njsol = solve(jprob)\nplot(jsol)\nCatalyst.PNG(plot(jsol; fmt = :png, dpi = 200)) # hide\n\nSometimes we might want to setup a more complicated condition than simply that our event occurs at a fixed time or with a fixed frequency. For example, suppose we want to skip the first event at time t = 12 and then have the event be periodic after that point every 12 units of time. We can do so using a more general discrete callback as follows\n\ncircadian_model = @reaction_network begin\n    @discrete_events begin\n        ((t % 12 == 0) & (t > 12)) => [l => (l + 1)%2]\n    end\n    (kₚ*l,kᵢ), X <--> Xᴾ\nend\n\nHere our condition ((t % 12 == 0) & (t > 12)) determines when the event occurs, evaluating to true when t is a multiple of 12 that is also larger than 12. We now finish specifying our model\n\nusing JumpProcesses\nu0 = [:X => 150, :Xᴾ => 50]\nps = [:kₚ => 0.1, :kᵢ => 0.1, :l => 1.0]\ntspan = (0.0, 100.0)\njprob = JumpProblem(circadian_model, u0, tspan, ps)\nnothing # hide\n\nNext, if we simulate our model, we note that the events do not seem to be triggered\n\nsol = solve(jprob)\nplot(sol)\nCatalyst.PNG(plot(sol; fmt = :png, dpi = 200)) # hide\n\nThe reason is that general discrete callbacks' conditions are only checked at the end of each simulation time step, and the probability that these will coincide with the callback's trigger times (t = 12 24 36 ) is infinitely small. Hence, we must directly instruct our simulation to stop at these time points. We can do this using the tstops argument:\n\ntstops = range(12.0, tspan[2]; step = 12.0)\nsol = solve(jprob; tstops)\nplot(sol)\nCatalyst.PNG(plot(sol; fmt = :png, dpi = 200)) # hide","category":"section"},{"location":"model_simulation/examples/periodic_events_simulation/#periodic_event_simulation_plotting_light","page":"Modelling a Periodic Event During ODE and Jump Simulations","title":"Plotting the light level","text":"Parameters which values change throughout the simulation can be plotted directly using the normal interface.\n\ncircadian_model = @reaction_network begin\n    @discrete_events begin\n        12 => [l => (l + 1)%2]\n    end\n (kₚ*l,kᵢ), X <--> Xᴾ\nend\nu0 = [:X => 150.0, :Xᴾ => 50.0]\nps = [:kₚ => 0.1, :kᵢ => 0.1, :l => 1.0]\noprob = ODEProblem(circadian_model, u0, (0.0, 100.0), ps)\nsol = solve(oprob)\nnothing # hide\n\nHere, if we directly plot l's value, it will be too small (compared to X and Xᴾ to be discernible). We instead @unpack our variables, and then plot a re-scaled version:\n\n@unpack X, Xᴾ, l = circadian_model\nplot(sol; idxs = [X, Xᴾ, 150*l], labels = [\"X\" \"Xᴾ\" \"Light amplitude\"])","category":"section"},{"location":"model_simulation/ode_simulation_performance/#ode_simulation_performance","page":"Advice for Performant ODE Simulations","title":"Advice for Performant ODE Simulations","text":"<details><summary><strong>Environment setup and package installation</strong></summary>\n\nThe following code sets up an environment for running the code on this page.\n\nusing Pkg\nPkg.activate(; temp = true) # Creates a temporary environment, which is deleted when the Julia session ends.\nPkg.add(\"BenchmarkTools\")\nPkg.add(\"Catalyst\")\nPkg.add(\"CUDA\")\nPkg.add(\"DiffEqGPU\")\nPkg.add(\"IncompleteLU\")\nPkg.add(\"OrdinaryDiffEqBDF\")\nPkg.add(\"OrdinaryDiffEqRosenbrock\")\nPkg.add(\"OrdinaryDiffEqTsit5\")\nPkg.add(\"OrdinaryDiffEqVerner\")\nPkg.add(\"Plots\")\nPkg.add(\"StaticArrays\")\n\n</details>\n\n\n   We have previously described how to perform ODE simulations of chemical reaction network (CRN) models. These simulations are typically fast and require little additional consideration. However, when a model is simulated many times (e.g. as a part of solving an inverse problem), or is very large, simulation run times may become noticeable. Here we will give some advice on how to improve performance for these cases [1].\n\nGenerally, there are few good ways to, before a simulation, determine the best options. Hence, while we below provide several options, if you face an application for which reducing run time is critical (e.g. if you need to simulate the same ODE many times), it might be required to manually trial these various options to see which yields the best performance (BenchmarkTools.jl's @btime macro is useful for this purpose). It should be noted that the default options typically perform well, and it is primarily for large models where investigating alternative options is worthwhile. All ODE simulations of Catalyst models are performed using the OrdinaryDiffEq.jl package, which documentation provides additional advice on performance.\n\nGenerally, this short checklist provides a quick guide for dealing with ODE performance:\n\nIf performance is not critical, use the default solver choice and do not worry further about the issue.\nIf improved performance would be useful, read about solver selection (both in this tutorial and OrdinaryDiffEq's documentation) and then try a few different solvers to find one with good performance.\nIf you have a large ODE (approximately 100 variables or more), try the various options for efficient Jacobian computation (noting that some are non-trivial to use, and should only be investigated if truly required).\nIf you plan to simulate your ODE many times, try parallelise it on CPUs or GPUs (with preference for the former, which is easier to use).","category":"section"},{"location":"model_simulation/ode_simulation_performance/#ode_simulation_performance_stiffness","page":"Advice for Performant ODE Simulations","title":"Regarding stiff and non-stiff problems and solvers","text":"Generally, ODE problems can be categorised into stiff ODEs and non-stiff ODEs. This categorisation is important due to stiff ODEs requiring specialised solvers. A common cause of failure to simulate an ODE is the use of a non-stiff solver for a stiff problem. There is no exact way to determine whether a given ODE is stiff or not, however, systems with several different time scales (e.g. a CRN with both slow and fast reactions) typically generate stiff ODEs.\n\nHere we simulate the (stiff) Brusselator model using the Tsit5 solver (which is designed for non-stiff ODEs):\n\nusing Catalyst, OrdinaryDiffEqTsit5, Plots\n\nbrusselator = @reaction_network begin\n    A, ∅ --> X\n    1, 2X + Y --> 3X\n    B, X --> Y\n    1, X --> ∅\nend\n\nu0 = [:X => 1.0, :Y => 0.0]\ntspan = (0.0, 20.0)\nps = [:A => 400.0, :B => 2000.0]\noprob = ODEProblem(brusselator, u0, tspan, ps)\n\nsol1 = solve(oprob, Tsit5())\nplot(sol1)\nCatalyst.PNG(plot(sol1; fmt = :png, dpi = 200)) # hide\n\nWe get a warning, indicating that the simulation was terminated. Furthermore, the resulting plot ends at t  12, meaning that the simulation was not completed (as the simulation's endpoint is t = 20). Indeed, we can confirm this by checking the return code of the solution object:\n\nsol1.retcode\n\nNext, we instead try the Rodas5P solver (which is designed for stiff problems):\n\nusing OrdinaryDiffEqRosenbrock\nsol2 = solve(oprob, Rodas5P())\nplot(sol2)\n\nThis time the simulation was successfully completed, which can be confirmed by checking the return code:\n\nsol2.retcode\n\nGenerally, ODE solvers can be divided into explicit and implicit solvers. Roughly, explicit solvers are better for non-stiff problems, with implicit solvers being required for stiff problems. While we could use implicit solvers for all problems (to guarantee successful simulations irrespective of stiffness), these are typically less performant (as compared to the explicit solvers) on equations that are non-stiff. An important property of implicit solvers is that they require the computation of a Jacobian as part of their routine. This means that the various options for efficient Jacobian computation described later in this tutorial are only relevant to implicit solvers.\n\nFinally, we should note that stiffness is not tied to the model equations only. If we change the parameter values of our previous Brusselator model to [:A => 1.0, :B => 4.0], the non-stiff Tsit5 solver can successfully simulate it.","category":"section"},{"location":"model_simulation/ode_simulation_performance/#ode_simulation_performance_solvers","page":"Advice for Performant ODE Simulations","title":"ODE solver selection","text":"OrdinaryDiffEq implements an unusually large number of ODE solvers, with the performance of the simulation heavily depending on which one is chosen. These are provided as the second argument to the solve command, e.g. here we use the Tsit5 solver to simulate a simple birth-death process:\n\nusing Catalyst, OrdinaryDiffEqTsit5\n\nbd_model = @reaction_network begin\n    (p,d), 0 <--> X\nend\nu0 = [:X => 0.1]\ntspan = (0.0, 1.0)\nps = [:p => 1.0, :d => 0.2]\n\noprob = ODEProblem(bd_model, u0, tspan, ps)\nsolve(oprob, Tsit5())\nnothing # hide\n\nIf no solver argument is provided to solve, and the OrdinaryDiffEqDefault sub-library or meta OrdinaryDiffEq library are loaded, then one is automatically selected:\n\nusing OrdinaryDiffEqDefault\nsolve(oprob)\nnothing # hide\n\nWhile the default choice is typically enough for most single simulations, if performance is important, it can be worthwhile exploring the available solvers to find one that is especially suited for the given problem. A complete list of possible ODE solvers, with advice on optimal selection, can be found here. This section will give some general advice.\n\nThe most important part of solver selection is to select one appropriate for the problem's stiffness. Generally, the Tsit5 solver is good for non-stiff problems, and Rodas5P for stiff problems. For large stiff problems (with many species), FBDF can be a good choice. We can illustrate the impact of these choices by simulating our birth-death process using the Tsit5, Vern7 (an explicit solver yielding low error in the solution), Rodas5P, and FBDF solvers (benchmarking their respective performance using BenchmarkTools.jl):\n\nusing BenchmarkTools \nusing OrdinaryDiffEqTsit5, OrdinaryDiffEqRosenbrock, OrdinaryDiffEqVerner, OrdinaryDiffEqBDF\n@btime solve(oprob, Tsit5())\n@btime solve(oprob, Vern7())\n@btime solve(oprob, Rodas5P())\n@btime solve(oprob, FBDF())\n\nIf you perform the above benchmarks on your machine, and check the results, you will note that the fastest solver is several times faster than the slowest one (FBDF, which is a poor choice for this ODE).","category":"section"},{"location":"model_simulation/ode_simulation_performance/#ode_simulation_performance_error","page":"Advice for Performant ODE Simulations","title":"Simulation error, tolerance, and solver selection","text":"Numerical ODE simulations approximate an ODEs' continuous solutions as discrete vectors. This introduces errors in the computed solution. The magnitude of these errors can be controlled by setting solver tolerances. By reducing the tolerance, solution errors will be reduced, however, this will also increase simulation run times. The (absolute and relative) tolerance of a solver can be tuned through the abstol and reltol arguments. Here we see how run time increases with larger tolerances:\n\n@btime solve(oprob, Tsit5(); abstol=1e-6, reltol=1e-6)\n@btime solve(oprob, Tsit5(); abstol=1e-12, reltol=1e-12)\n\nIt should be noted, however, that the result of the second simulation is a lot more accurate. Thus, ODE solver performance cannot be determined solely from run time, but is a composite of run time and error. Benchmarks comparing solver performance (by plotting the run time as a function of the error) for various CRN models can be found in the SciMLBenchmarks repository.\n\nGenerally, whether solution error is a consideration depends on the application. If you want to compute the trajectory of an expensive space probe as it is sent from Earth, to slingshot Jupiter, and then reach Pluto a few years later, ensuring a minimal error will be essential. However, if you want to simulate a simple CRN to determine whether it oscillates for a given parameter set, a small error will not constitute a problem. An important aspect with regard to error is that it affects the selection of the optimal solver. E.g. if tolerance is low (generating larger errors) the Rosenbrock23 method performs well for small, stiff, problems (again, more details can be found in OrdinaryDiffEq's documentation).","category":"section"},{"location":"model_simulation/ode_simulation_performance/#ode_simulation_performance_jacobian","page":"Advice for Performant ODE Simulations","title":"Jacobian computation options for implicit solvers","text":"As previously mentioned, implicit ODE solvers require the computation of the system's Jacobian. The reason is (roughly) that, in each time step, these solvers need to solve a non-linear equation to find the simulation's value at the next timestep (unlike explicit solvers, which compute the value at the next time step directly). Typically this is done using the Newton-Raphson method, which requires the Jacobian. Especially for large systems, this can be computationally expensive (and a potential strain on available memory), in which case one might consider various Jacobian-computation options (as described below). A throughout tutorial on simulating a large, stiff, ODE can be found here.","category":"section"},{"location":"model_simulation/ode_simulation_performance/#ode_simulation_performance_symbolic_jacobian","page":"Advice for Performant ODE Simulations","title":"Building the Jacobian symbolically","text":"By default, OrdinaryDiffEq computes the Jacobian using automatic differentiation (however, using finite differences is also possible). Since Catalyst builds its ODEs symbolically, it is able to compute an analytic Jacobian symbolically. Typically, this is only advantageous when you are also using a sparse Jacobian.\n\nTo use this option, simply set jac = true when constructing an ODEProblem:\n\nusing Catalyst, OrdinaryDiffEqDefault\n\nbrusselator = @reaction_network begin\n    A, ∅ --> X\n    1, 2X + Y --> 3X\n    B, X --> Y\n    1, X --> ∅\nend\nu0 = [:X => 1.0, :Y => 0.0]\ntspan = (0.0, 20.0)\nps = [:A => 10.0, :B => 40.0]\n\noprob = ODEProblem(brusselator, u0, tspan, ps; jac = true)\nnothing # hide","category":"section"},{"location":"model_simulation/ode_simulation_performance/#ode_simulation_performance_sparse_jacobian","page":"Advice for Performant ODE Simulations","title":"Using a sparse Jacobian","text":"For a system with n variables, the Jacobian will be an ntimes n matrix. This means that, as n becomes large, the Jacobian can become very large, potentially causing a significant strain on memory. In these cases, most Jacobian entries are typically 0. This means that a sparse Jacobian (rather than a dense one, which is the default) can be advantageous. To designate sparse Jacobian usage, simply provide the sparse = true option when constructing an ODEProblem:\n\noprob = ODEProblem(brusselator, u0, tspan, ps; sparse = true)\nnothing # hide","category":"section"},{"location":"model_simulation/ode_simulation_performance/#ode_simulation_performance_symbolic_jacobian_linear_solver","page":"Advice for Performant ODE Simulations","title":"Linear solver selection","text":"When implicit solvers use e.g. the Newton-Raphson method to (at each simulation time step) solve a (typically non-linear) equation, they actually solve a linearised version of this equation. For this, they use a linear solver, the choice of which can impact performance. To specify one, we use the linsolve option (given to the solver function, not the solve command). E.g. to use the KLUFactorization linear solver (which requires loading the LinearSolve.jl package) we run\n\nusing LinearSolve, OrdinaryDiffEqRosenbrock\nsolve(oprob, Rodas5P(linsolve = KLUFactorization()))\nnothing # hide\n\nA full list of potential linear solvers can be found here. Typically, the default choice performs well.\n\nA unique approach to the linear solvers is to use a matrix-free Newton-Krylov method. These do not actually compute the Jacobian, but rather the effect of multiplying it with a vector. They are typically advantageous for large systems (with large Jacobians), and can be designated using the KrylovJL_GMRES linear solver:\n\nsolve(oprob, Rodas5P(linsolve = KrylovJL_GMRES()))\nnothing # hide\n\nSince these methods do not depend on a Jacobian, certain Jacobian options (such as computing it symbolically) are irrelevant to them.","category":"section"},{"location":"model_simulation/ode_simulation_performance/#ode_simulation_performance_preconditioners","page":"Advice for Performant ODE Simulations","title":"Designating preconditioners for Jacobian-free linear solvers","text":"When an implicit method solves a linear equation through an (iterative) matrix-free Newton-Krylov method, the rate of convergence depends on the numerical properties of the matrix defining the linear system. To speed up convergence, a preconditioner can be applied to both sides of the linear equation, attempting to create an equation that converges faster. Preconditioners are only relevant when using matrix-free Newton-Krylov methods.\n\nIn practice, preconditioners are implemented as functions with a specific set of arguments. How to implement these is non-trivial, and we recommend reading OrdinaryDiffEq's documentation pages here and here. In this example, we will define an Incomplete LU preconditioner (which requires the IncompleteLU.jl package):\n\nusing IncompleteLU\nfunction incompletelu(W, du, u, p, t, newW, Plprev, Prprev, solverdata)\n    if newW === nothing || newW\n        Pl = ilu(convert(AbstractMatrix, W), τ = 50.0)\n    else\n        Pl = Plprev\n    end\n    Pl, nothing\nend\nnothing # hide\n\nNext, incompletelu can be supplied to our solver using the precs argument:\n\nsolve(oprob, Rodas5P(linsolve = KrylovJL_GMRES(), precs = incompletelu, concrete_jac = true))\nnothing # hide\n\nFinally, we note that when using preconditioners with a matrix-free method (like KrylovJL_GMRES, which is also the only case when these are relevant), the concrete_jac = true argument is required.\n\nGenerally, the use of preconditioners is only recommended for advanced users who are familiar with the concepts. However, for large systems, if performance is essential, they can be worth looking into.","category":"section"},{"location":"model_simulation/ode_simulation_performance/#ode_simulation_performance_conservation_laws","page":"Advice for Performant ODE Simulations","title":"Elimination of system conservation laws","text":"Previously, we have described how Catalyst, when it generates ODEs, is able to detect and eliminate conserved quantities. In certain cases, doing this can improve performance. E.g. in the following example we will eliminate the single conserved quantity in a two-state model. This results in a differential algebraic equation with a single differential equation and a single algebraic equation (as opposed to two differential equations). However, as the algebraic equation is fully determined by the ODE solution, Catalyst moves it to be an observable and our new system therefore only contains one ODE that must be solved numerically. Conservation laws can be eliminated by providing the remove_conserved = true option to ODEProblem:\n\nusing Catalyst, OrdinaryDiffEqTsit5\n\n# Declare model.\nrs = @reaction_network begin\n    (k₁,k₂), X₁ <--> X₂\nend\n\n# Generate and simulate and ODE (while eliminating any conserved quantities).\nu0 = [:X₁ => 2.0, :X₂ => 3.0]\nps = [:k₁ => 1.0, :k₂ => 2.0]\noprob = ODEProblem(rs, u0, (0.0, 10.0), ps; remove_conserved = true)\nsol = solve(oprob)\nnothing # hide\n\nConservation law elimination is not expected to ever impact performance negatively; it simply results in a (possibly) lower-dimensional system of ODEs to solve. However, eliminating conserved species may have minimal performance benefits; it is model-dependent whether elimination results in faster ODE solving times and/or increased solution accuracy.","category":"section"},{"location":"model_simulation/ode_simulation_performance/#ode_simulation_performance_parallelisation","page":"Advice for Performant ODE Simulations","title":"Parallelisation on CPUs and GPUs","text":"Whenever an ODE is simulated a large number of times (e.g. when investigating its behaviour for different parameter values), the best way to improve performance is to parallelise the simulation over multiple processing units. Indeed, an advantage of the Julia programming language is that it was designed after the advent of parallel computing, making it well-suited for this task. Roughly, parallelisation can be divided into parallelisation on CPUs and on GPUs. CPU parallelisation is most straightforward, while GPU parallelisation requires specialised ODE solvers (which Catalyst have access to).\n\nBoth CPU and GPU parallelisation require first building an EnsembleProblem (which defines the simulations you wish to perform) and then supplying this with the correct parallelisation options. EnsembleProblems have previously been introduced in Catalyst's documentation (but in the context of convenient bundling of similar simulations, rather than to improve performance), with a more throughout description being found in OrdinaryDiffEq's documentation. Finally, a general documentation of parallel computing in Julia is available here.","category":"section"},{"location":"model_simulation/ode_simulation_performance/#ode_simulation_performance_parallelisation_CPU","page":"Advice for Performant ODE Simulations","title":"CPU parallelisation","text":"For this example (and the one for GPUs), we will consider a modified Michaelis-Menten enzyme kinetics model, which describes an enzyme (E) that converts a substrate (S) to a product (P):\n\nusing Catalyst\nmm_model = @reaction_network begin\n    kB, S + E --> SE\n    kD, SE --> S + E\n    kP, SE --> P + E\n    d, S --> ∅\nend\n\nThe model can be simulated, showing how P is produced from S:\n\nusing OrdinaryDiffEqTsit5, Plots\nu0 = [:S => 1.0, :E => 1.0, :SE => 0.0, :P => 0.0]\ntspan = (0.0, 50.0)\nps = [:kB => 1.0, :kD => 0.1, :kP => 0.5, :d => 0.1]\noprob = ODEProblem(mm_model, u0, tspan, ps)\nsol = solve(oprob, Tsit5())\nplot(sol)\n\nDue to the degradation of S, if the production rate is not high enough, the total amount of P produced is reduced. For this tutorial, we will investigate this effect for a range of values of kP. This will require a large number of simulations (for various kP values), which we will parallelise on CPUs (this section) and GPUs (next section).\n\nTo parallelise our simulations, we first need to create an EnsembleProblem. These describe which simulations we wish to perform. The input to this is:\n\nThe ODEProblem corresponds to the model simulation (SDEProblem and JumpProblems can also be supplied, enabling the parallelisation of these problem types).\nA function, prob_func, describing how to modify the problem for each simulation. If we wish to simulate the same, unmodified problem, in each simulation (primarily relevant for stochastic simulations), this argument is not required.\n\nHere, prob_func takes 3 arguments:\n\nprob: The problem that it modifies at the start of each individual run (which will be the same as EnsembleProblem's first argument).\ni: The index of the specific simulation (in the array of all simulations that are performed).\nrepeat: The repeat of a specific simulation in the array. We will not use this option in this example, however, it is discussed in more detail here.\n\nand output the ODEProblem simulated in the i'th simulation.\n\nLet us assume that we wish to simulate our model 100 times, for kP = 001 002  099 10. We define our prob_func using remake:\n\nfunction prob_func(prob, i, repeat)\n    return remake(prob; p = [:kP => 0.01*i])\nend\nnothing # hide\n\nNext, we can create our EnsembleProblem:\n\neprob = EnsembleProblem(oprob; prob_func)\nnothing # hide\n\nWe can now solve our ODEProblem using the same syntax we would normally use to solve the original ODEProblem, with the exception that an additional argument, trajectories, is required (which denotes how many simulations should be performed).\n\nesol = solve(eprob, Tsit5(); trajectories = 100)\nnothing # hide\n\nTo access the i'th solution we use esol.u[i]. To e.g. plot the 47'th solution we use:\n\nplot(esol.u[47])\n\nTo extract the amount of P produced in each simulation, and plot this against the corresponding kP value, we can use:\n\nplot(0.01:0.01:1.0, map(sol -> sol[:P][end], esol.u), xguide = \"kP\", yguide = \"P produced\", label=\"\")\nplot!(left_margin = 3Plots.Measures.mm) # hide\n\nAbove, we have simply used EnsembleProblem as a convenient interface to run a large number of similar simulations. However, these problems have the advantage that they allow the passing of an ensemble algorithm to the solve command, which describes a strategy for parallelising the simulations. By default, EnsembleThreads is used. This parallelises the simulations using multithreading (parallelisation within a single process), which is typically advantageous for small problems on shared memory devices. An alternative is EnsembleDistributed which instead parallelises the simulations using multiprocessing (parallelisation across multiple processes). To do this, we simply supply this additional solver to the solve command:\n\nesol = solve(eprob, Tsit5(), EnsembleDistributed(); trajectories = 100)\n\nTo utilise multiple processes, you must first give Julia access to these. You can check how many processes are available using the nprocs (which requires the Distributed.jl package):\n\nusing Distributed\nnprocs()\n\nNext, more processes can be added using addprocs. E.g. here we add an additional 4 processes:\n\naddprocs(4)\n\nPowerful personal computers and HPC clusters typically have a large number of available additional processes that can be added to improve performance.\n\nWhile EnsembleThreads and EnsembleDistributed cover the main cases, additional ensemble algorithms exist. A more throughout description of these can be found here.\n\nFinally, it should be noted that OrdinaryDiffEq, if additional processes are available, automatically parallelises the linear solve part of implicit simulations. It is thus possible to see performance improvements from adding additional processes when running single simulations, not only multiple parallelised ones (this effect is primarily noticeable for large systems with many species).","category":"section"},{"location":"model_simulation/ode_simulation_performance/#ode_simulation_performance_parallelisation_GPU","page":"Advice for Performant ODE Simulations","title":"GPU parallelisation","text":"GPUs are different from CPUs in that they are much more restricted in what computations they can carry out. However, unlike CPUs, they are typically available in far larger numbers. Their original purpose is for rendering graphics (which typically involves solving a large number of very simple computations, something CPUs with their few, but powerful, cores are unsuited for). Recently, they have also started to be applied to other problems, such as ODE simulations. Generally, GPU parallelisation is only worthwhile when you have a very large number of parallel simulations to run (and access to good GPU resources, either locally or on a cluster).\n\nGenerally, we can parallelise EnsembleProblems across several GPUs in a very similar manner to how we parallelised them across several CPUs, but by using a different ensemble algorithm (such as EnsembleGPUArray). However, there are some additional requirements:\n\nGPU parallelisation requires the DiffEqGPU.jl package.\nDepending on which GPU hardware is used, a specific back-end package has to be installed and imported (e.g. CUDA for NVIDIA's GPUs or Metal for Apple's).\nFor some cases, we must use a special ODE solver supporting simulations on GPUs.\n\nFurthermore (while not required) to receive good performance, we should also make the following adaptations:\n\nBy default, Julia's decimal numbers are implemented as Float64s, however, using Float32s is advantageous on GPUs. Ideally, all initial conditions and parameter values should be specified using these.\nWe should designate all our vectors (i.e. initial conditions and parameter values) as static vectors.\n\nWe will assume that we are using the CUDA GPU hardware, so we will first load the CUDA.jl backend package, as well as DiffEqGPU:\n\nusing CUDA, DiffEqGPU\n\nWhich backend package you should use depends on your available hardware, with the alternatives being listed here.\n\nNext, we declare our model and ODEProblem. However, we make all values Float64 (by appending f0 to them) and all vectors static (by adding @SVector before their declaration, something which requires the StaticArrays package).\n\nusing Catalyst, OrdinaryDiffEqDefault, StaticArrays\n\nmm_model = @reaction_network begin\n    kB, S + E --> SE\n    kD, SE --> S + E\n    kP, SE --> P + E\n    d, S --> ∅\nend\n@unpack S, E, SE, P, kB, kD, kP, d = mm_model\n\nusing OrdinaryDiffEqDefault, Plots\nu0 = @SVector [S => 1.0f0, E => 1.0f0, SE => 0.0f0, P => 0.0f0]\ntspan = (0.0f0, 50.0f0)\np = @SVector [kB => 1.0f0, kD => 0.1f0, kP => 0.5f0, d => 0.1f0]\noprob = ODEProblem(mm_model, u0, tspan, p)\nnothing # hide\n\nWhen we declare our prob_func and EnsembleProblem we need to ensure that the updated ODEProblem uses Float32:\n\nfunction prob_func(prob, i, repeat)\n    return remake(prob; p = [:kP => 0.0001f0*i])\nend\neprob = EnsembleProblem(oprob; prob_func = prob_func)\nnothing # hide\n\nHere have we increased the number of simulations to 10,000, since this is a more appropriate number for GPU parallelisation (as compared to the 100 simulations we performed in our CPU example).\n\nnote: Note\nCurrently, declaration of static vectors requires symbolic, rather than symbol, form for species and parameters. Hence, we here first @unpack these before constructing u0 and ps using @SVector.\n\nWe can now simulate our model using a GPU-based ensemble algorithm. Currently, two such algorithms are available, EnsembleGPUArray and EnsembleGPUKernel. Their differences are that\n\nOnly EnsembleGPUKernel requires arrays to be static arrays (although it is still advantageous for EnsembleGPUArray).\nWhile EnsembleGPUArray can use standard ODE solvers, EnsembleGPUKernel requires specialised versions (such as GPUTsit5). A list of available such solvers can be found here.\n\nGenerally, it is recommended to use EnsembleGPUArray for large models (that have at least 100 variables), and EnsembleGPUKernel for smaller ones. Here we simulate our model using both approaches (noting that EnsembleGPUKernel requires GPUTsit5):\n\nesol1 = solve(eprob, Tsit5(), EnsembleGPUArray(CUDA.CUDABackend()); trajectories = 10000)\nesol2 = solve(eprob, GPUTsit5(), EnsembleGPUKernel(CUDA.CUDABackend()); trajectories = 10000)\n\nNote that we have to provide the CUDA.CUDABackend() argument to our ensemble algorithms (to designate our GPU backend, in this case, CUDA).\n\nJust like OrdinaryDiffEq is able to utilise parallel CPU processes to speed up the linear solve part of ODE simulations, GPUs can also be used. More details on this can be found here. This is only recommended when ODEs are very large (at least 1,000 species), which is typically not the case for CRNs.\n\nFor more information on differential equation simulations on GPUs in Julia, please read DiffEqGPU's documentation. Furthermore, if performance is critical, this tutorial provides information on how to redesign your simulation code to make it more suitable for GPU simulations.\n\n","category":"section"},{"location":"model_simulation/ode_simulation_performance/#Citation","page":"Advice for Performant ODE Simulations","title":"Citation","text":"If you use GPU simulations in your research, please cite the following paper to support the authors of the DiffEqGPU package:\n\n@article{utkarsh2024automated,\n  title={Automated translation and accelerated solving of differential equations on multiple GPU platforms},\n  author={Utkarsh, Utkarsh and Churavy, Valentin and Ma, Yingbo and Besard, Tim and Srisuma, Prakitr and Gymnich, Tim and Gerlach, Adam R and Edelman, Alan and Barbastathis, George and Braatz, Richard D and others},\n  journal={Computer Methods in Applied Mechanics and Engineering},\n  volume={419},\n  pages={116591},\n  year={2024},\n  publisher={Elsevier}\n}\n\n","category":"section"},{"location":"model_simulation/ode_simulation_performance/#References","page":"Advice for Performant ODE Simulations","title":"References","text":"[1]: E. Hairer, G. Wanner, Solving Ordinary Differential Equations II, Springer (1996).","category":"section"},{"location":"steady_state_functionality/bifurcation_diagrams/#bifurcation_diagrams","page":"Bifurcation Diagrams","title":"Bifurcation Diagrams","text":"Bifurcation diagrams describe how, for a dynamical system, the quantity and type of its steady states change as a parameter is varied[1]. When using Catalyst-generated models, these can be computed with the BifurcationKit.jl package. Catalyst provides a simple interface for creating BifurcationKit compatible BifurcationProblems from ReactionSystems.\n\nThis tutorial briefly introduces how to use Catalyst with BifurcationKit through basic examples, with BifurcationKit.jl providing a more extensive documentation. Especially for more complicated systems, where careful tuning of algorithm options might be required, reading the BifurcationKit documentation is recommended. Finally, BifurcationKit provides many additional features not described here, including computation of periodic orbits, tracking of bifurcation points along secondary parameters, and bifurcation computations for PDEs.","category":"section"},{"location":"steady_state_functionality/bifurcation_diagrams/#bifurcation_diagrams_basic_example","page":"Bifurcation Diagrams","title":"Basic example","text":"For this example, we will use a modified version of the Wilhelm model (which demonstrates a bistable switch as the parameter k1 is varied). We declare the model using Catalyst:\n\nusing Catalyst\nwilhelm_model = @reaction_network begin\n    k1, Y --> 2X\n    k2, 2X --> X + Y\n    k3, X + Y --> Y\n    k4, X --> 0\n    k5, 0 --> X\nend\n\nNext we will create a BifurcationProblem. In addition to the ReactionSystem, we need to provide:\n\nThe bifurcation parameter (the parameter which is varied in the bifurcation diagram).\nA full model parameter set. This includes the values of all non-bifurcation parameters, but also a value for the bifurcation parameter (which corresponds to the point in parameter space from which the computation of the bifurcation diagram starts).\nAn initial guess of the steady state values of the system at the provided parameter set. Using this point as a starting guess for root finding, BifurcationKit calculates an initial steady state from which to compute the bifurcation diagram. Hence, this guess does not need to be very exact (but may be important if the system exhibits multistability for the initial parameter set).\nThe species or statistic we wish to plot on the y-axis of the bifurcation diagram (customised properties can be recorded and plotted using the record_from_solution argument).\n\nWe combine all this information to form a BifurcationProblem:\n\nusing BifurcationKit\nbif_par = :k1\nu_guess = [:X => 5.0, :Y => 2.0]\np_start = [:k1 => 8.0, :k2 => 1.0, :k3 => 1.0, :k4 => 1.5, :k5 => 1.25]\nplot_var = :X\nbprob = BifurcationProblem(wilhelm_model, u_guess, p_start, bif_par; plot_var)\nnothing # hide\n\nBifurcationKit computes bifurcation diagrams using the bifurcationdiagram function. From an initial point in the diagram, it tracks the solution (using a continuation algorithm) until the entire diagram is computed (BifurcationKit's continuation can be used for other purposes, however, this tutorial focuses on bifurcation diagram computation). The continuation settings are provided in a ContinuationPar structure. In this example, we will only specify two settings (p_min and p_max), which sets the minimum and maximum values over which the bifurcation parameter is varied, respectively. We wish to compute a bifurcation diagram over the interval (20200), and will use the following settings:\n\np_span = (2.0, 20.0)\nopts_br = ContinuationPar(p_min = p_span[1], p_max = p_span[2])\nnothing # hide\n\nFinally, we compute our bifurcation diagram using:\n\nbif_dia = bifurcationdiagram(bprob, PALC(), 2, opts_br; bothside = true)\nnothing # hide\n\nWhere PALC() designates that we wish to use the pseudo-arclength continuation method to track our solution. The third argument (2) designates the maximum number of recursions when branches of branches are computed (branches appear as continuation encounters certain bifurcation points). For diagrams with highly branched structures (rare for many common small chemical reaction networks) this input is important. Finally, bothside = true designates that we wish to perform continuation on both sides of the initial point (which is typically the case). \n\nWe can plot our bifurcation diagram using the Plots.jl package:\n\nusing Plots\nplot(bif_dia; xguide = bif_par, yguide = plot_var, branchlabel = \"Steady state concentration\")\n\nHere, the steady state concentration of X is shown as a function of k1's value. Stable steady states are shown with thick lines, unstable ones with thin lines. The two fold bifurcation points are marked with \"bp\".","category":"section"},{"location":"steady_state_functionality/bifurcation_diagrams/#bifurcation_diagrams_continuationpar","page":"Bifurcation Diagrams","title":"Additional ContinuationPar options","text":"Most of the options required by the bifurcationdiagram function are provided through the ContinuationPar structure. For full details, please read the BifurcationKit documentation. However, a few common options, and how they affect the continuation computation, are described here:\n\np_min and p_max: Set the interval over which the bifurcation diagram is computed (with the continuation stopping if it reaches these bounds).\ndsmin and dsmax: The minimum and maximum length of the continuation steps (in the bifurcation parameter's value).\nds: The initial length of the continuation steps. This is especially important when bothside = true is not used, as the sign of ds determines the direction from the initial point in which the continuation will proceed.\nmax_steps: The maximum number of continuation steps. If a bifurcation diagram looks incomplete, try increasing this value.\nnewton_options: Options for the Newton's method that BifurcationKit uses to find steady states. This can be created using NewtonPar(tol = 1e-9, max_iterations = 100) which here sets the tolerance (to 1e-9) and the maximum number of newton iterations (to 100).\n\nThe previous bifurcation diagram can be computed, with these various options specified, in the following way:\n\np_span = (2.0, 20.0)\nnewton_options = NewtonPar(tol = 1e-9, max_iterations = 20)\nopts_br = ContinuationPar(; p_min = p_span[1], p_max = p_span[2], ds = 0.005, \n                          dsmin = 0.001, dsmax = 0.01, max_steps = 1000, newton_options)\nbif_dia = bifurcationdiagram(bprob, PALC(), 2, opts_br; bothside = true)\nnothing # hide\n\n(In this case, however, these additional settings have no significant effect on the result)","category":"section"},{"location":"steady_state_functionality/bifurcation_diagrams/#bifurcation_diagrams_disjoint_branches","page":"Bifurcation Diagrams","title":"Bifurcation diagrams with disjoint branches","text":"Let's consider the previous case, but instead compute the bifurcation diagram over the interval (80150):\n\np_span = (8.0, 15.0)\nopts_br = ContinuationPar(p_min = p_span[1], p_max = p_span[2])\nbif_dia = bifurcationdiagram(bprob, PALC(), 2, opts_br; bothside = true)\nplot(bif_dia; ylimit = (0.0, 17.0), xguide = bif_par, yguide = plot_var, branchlabel = \"Steady state concentration\")\n\nHere, in the bistable region, we only see a single branch. The reason is that the continuation algorithm starts at our initial guess (here made at k1 = 80 for (XY) = (5020)) and tracks the diagram from there. However, with the upper bound set at k1 = 150 the bifurcation diagram has a disjoint branch structure, preventing the full diagram from being computed by continuation alone. In this case it could be solved by increasing the bound from k1 = 150, however, this is not possible in all cases. Here we will describe a more general solution, the use of deflation to find multiple steady states.\n\nWe will not describe the details of how deflation works here (instead, please read BifurcationKit's documentation). In practise, in involves first creating a deflation continuation algorithm (defalg). This one, in turn, requires us to create a deflation operator and a perturbation function.\n\ndeflation_operator = DeflationOperator(2, 0.001, [zeros(numspecies(wilhelm_model))])\nperturb_solution(x, _, _) = (x  .+ 0.1 .* rand(length(x)))\ndefalg = DefCont(; deflation_operator, perturb_solution)\nnothing # hide\n\nNext we compute our bifurcation diagram using defalg (instead of PALC() as previous) as our method. We also use a range of additional continuation parameter options to ensure a smooth tracking of the solution. Here we have \n\nnewton_options = NewtonPar(max_iterations = 10)\ncont_par = ContinuationPar(; p_min = p_span[1], p_max = p_span[2], ds = 0.001, max_steps = 10000, newton_options)\nbif_dia = continuation(bprob, defalg, cont_par; plot = false, verbosity = 0)\nnothing # hide\n\nFinally we can plot the diagram, noting that (unlike previously) we are able to compute the full diagram.\n\nplot(bif_dia; xguide = bif_par, yguide = plot_var, ylimit = (0.0, 17.0), color = 1)","category":"section"},{"location":"steady_state_functionality/bifurcation_diagrams/#bifurcation_diagrams_cons_laws","page":"Bifurcation Diagrams","title":"Systems with conservation laws","text":"Some systems are under-determined at steady state, so that for a given parameter set they have an infinite number of possible steady state solutions, preventing bifurcation diagrams from being computed. Similar to when we compute steady states for fixed parameter values, we can utilise Catalyst's ability to detect and eliminate conservation laws to resolve this issue. This requires us to provide information of the species concentrations at which we wish to compute the bifurcation diagram (to determine the values of conserved quantities). These are provided to the BifurcationProblem using the u0 argument.\n\nTo illustrate this, we will create a simple model of a kinase that is produced and degraded (at rates p and d). The kinase facilitates the phosphorylation of a protein (X), which is dephosphorylated at a constant rate. For this system, we will compute a bifurcation diagram, showing how the concentration of the phosphorylated protein (Xp) depends on the degradation rate of the kinase (d). We will set the total amount of protein (X + Xp) to 10.\n\nusing BifurcationKit, Catalyst, Plots\nkinase_model = @reaction_network begin\n    (p, d), 0 <--> K\n    (K*kP,kD), X <--> Xp\nend\n\nu_guess = [:K => 1.0, :X => 1.0, :Xp => 1.0]\np_start = [:p => 1.0, :d => 0.5, :kP => 2.0, :kD => 5.0]\nu0 = [:X => 1.0, :Xp => 0.0]\nbprob = BifurcationProblem(kinase_model, u_guess, p_start, :d; plot_var = :Xp, u0)\n\np_span = (0.1, 10.0)\nopts_br = ContinuationPar(p_min = p_span[1], p_max = p_span[2])\nbif_dia = bifurcationdiagram(bprob, PALC(), 2, opts_br; bothside = true)\nplot(bif_dia; xguide = \"d\", yguide = \"Xp\")\n\nThis bifurcation diagram does not contain any interesting features (such as bifurcation points), and only shows how the steady state concentration of Xp is reduced as d increases. \n\nFinally, for additional clarity, we reiterate the purpose of the two u arguments used:\n\nu_guess: A guess of the initial steady states (which BifurcationKit uses to find its starting point). Typically, most trivial guesses work (e.g. setting all species concentrations to 1.0). u_guess does not have to fulfil the conserved concentrations provided in u0.\nu0: Used to compute the concentrations of any conserved quantities (e.g. in our example X + Xp = 10). Technically, values are only required for species that are involved in conservation laws (in our case we do not need to provide a value for K). However, sometimes determining which species are actually involved in conservation laws can be difficult, and it might be easier to simply provide concentrations for all species.\n\n","category":"section"},{"location":"steady_state_functionality/bifurcation_diagrams/#bifurcationkit_periodic_orbits_citation","page":"Bifurcation Diagrams","title":"Citation","text":"If you use BifurcationKit.jl for your work, we ask that you cite the following paper!! Open source development strongly depends on this. It is referenced on HAL-Inria with bibtex entry CITATION.bib.\n\n","category":"section"},{"location":"steady_state_functionality/bifurcation_diagrams/#References","page":"Bifurcation Diagrams","title":"References","text":"[1]: Yuri A. Kuznetsov, Elements of Applied Bifurcation Theory, Springer (2023).","category":"section"},{"location":"model_creation/examples/noise_modelling_approaches/#noise_modelling_approaches","page":"Approaches for modelling system noise","title":"Approaches for modelling system noise","text":"<details><summary><strong>Environment setup and package installation</strong></summary>\n\nThe following code sets up an environment for running the code on this page.\n\nusing Pkg\nPkg.activate(; temp = true) # Creates a temporary environment, which is deleted when the Julia session ends.\nPkg.add(\"Catalyst\")\nPkg.add(\"DataInterpolations\")\nPkg.add(\"Distributions\")\nPkg.add(\"OrdinaryDiffEqDefault\")\nPkg.add(\"Plots\")\nPkg.add(\"StochasticDiffEq\")\n\n</details>\n\n\n   Catalyst's primary tools for modelling stochasticity include the creation of SDEProblems or JumpProblems from reaction network models. However, other approaches for incorporating model noise exist, some of which will be discussed here. We will first consider intrinsic and extrinsic noise. These are well-established terms, both of which we will describe below (however, to our knowledge, no generally agreed-upon definition of these terms exists)[1]. Finally, we will demonstrate a third approach, the utilisation of a noisy input process to an otherwise deterministic system. This approach is infrequently used, however, as it is encountered in the literature, we will demonstrate it here as well. \n\nWe note that these approaches can all be combined. E.g. an intrinsic noise model (using an SDE) can be combined with extrinsic noise (using randomised parameter values), while also feeding a noisy input process into the system.\n\nnote: Note\nHere we use intrinsic and extrinsic noise as descriptions of two of our modelling approaches. It should be noted that while these are established terminologies for noisy biological systems[1], our use of these terms to describe different approaches for modelling noise is only inspired by this terminology, and nothing that is established in the field.  Please consider the references for more information on intrinsic and extrinsic noise.","category":"section"},{"location":"model_creation/examples/noise_modelling_approaches/#noise_modelling_approaches_model_intro","page":"Approaches for modelling system noise","title":"The repressilator model","text":"For this tutorial we will use the oscillating repressilator model.\n\nusing Catalyst\nrepressilator = @reaction_network begin\n    hillr(Z,v,K,n), ∅ --> X\n    hillr(X,v,K,n), ∅ --> Y\n    hillr(Y,v,K,n), ∅ --> Z\n    d, (X, Y, Z) --> ∅\nend","category":"section"},{"location":"model_creation/examples/noise_modelling_approaches/#noise_modelling_approaches_model_intrinsic","page":"Approaches for modelling system noise","title":"Using intrinsic noise","text":"Generally, intrinsic noise is randomness inherent to a system itself. This means that it cannot be controlled for, or filtered out by, experimental settings. Low-copy number cellular systems, were reaction occurs due to the encounters of molecules due to random diffusion, is an example of intrinsic noise. In practise, this can be modelled exactly through SDE (chemical Langevin equations) or jump (stochastic chemical kinetics) simulations. \n\nIn Catalyst, intrinsic noise is accounted for whenever an SDEProblem or JumpProblem is created and simulated. Here we will model intrinsic noise through SDEs, which means creating an SDEProblem using the standard approach.\n\nu0 = [:X => 45.0, :Y => 20.0, :Z => 20.0]\ntend = 200.0\nps = [:v => 10.0, :K => 20.0, :n => 3, :d => 0.1]\nsprob = SDEProblem(repressilator, u0, tend, ps)\nnothing # hide\n\nNext, to illustrate the system's noisiness, we will perform multiple simulations. We do this by creating an EnsembleProblem. From it, we perform, and plot, 4 simulations.\n\nusing StochasticDiffEq, Plots\neprob_intrinsic = EnsembleProblem(sprob)\nsol_intrinsic = solve(eprob_intrinsic, ImplicitEM(); trajectories = 4)\nplot(sol_intrinsic; idxs = :X)\n\nHere, each simulation is performed from the same system using the same settings. Despite this, due to the noise, the individual trajectories are different.","category":"section"},{"location":"model_creation/examples/noise_modelling_approaches/#noise_modelling_approaches_model_extrinsic","page":"Approaches for modelling system noise","title":"Using extrinsic noise","text":"Next, we consider extrinsic noise. This is randomness caused by stochasticity external to, yet affecting, a system. Examples could be different bacteria experiencing different microenvironments or cells being in different parts of the cell cycle. This is noise which (in theory) can be controlled for experimentally (e.g. by ensuring a uniform environment). Whenever a specific source of noise is intrinsic and extrinsic to a system may depend on how one defines the system itself (this is a reason why giving an exact definition of these terms is difficult).\n\nIn Catalyst we can model extrinsic noise by letting the model parameters be probability distributions. Here, at the beginning of each simulation, random parameter values are drawn from their distributions. Let us imagine that our repressilator circuit was inserted into a bacterial population. Here, while each bacteria would have the same circuit, their individual number of e.g. ribosomes (which will be random) might affect the production rates (which while constant within each bacteria, might differ between the individuals).\n\nAgain we will perform ensemble simulation. Instead of creating an SDEProblem, we will create an ODEProblem, as well as a problem function which draws random parameter values for each simulation. Here we have implemented the parameter's probability distributions as normal distributions using the Distributions.jl package.\n\nusing Distributions\np_dists = Dict([:v => Normal(10.0, 2.0), :K => Normal(20.0, 5.0), :n => Normal(3, 0.2), :d => Normal(0.1, 0.02)])\nfunction prob_func(prob, i, repeat)\n    p = [par => rand(p_dists[par]) for par in keys(p_dists)]\n    return remake(prob; p)\nend\nnothing # hide\n\nNext, we again perform 4 simulations. While the individual trajectories are performed using deterministic simulations, the randomised parameter values create heterogeneity across the ensemble.\n\nusing OrdinaryDiffEqDefault\noprob = ODEProblem(repressilator, u0, tend, ps)\neprob_extrinsic = EnsembleProblem(oprob; prob_func)\nsol_extrinsic = solve(eprob_extrinsic; trajectories = 4)\nplot(sol_extrinsic; idxs = :X)\n\nWe note that a similar approach can be used to also randomise the initial conditions. In a very detailed model, the parameter values could fluctuate during a single simulation, something which could be implemented using the approach from the next section.","category":"section"},{"location":"model_creation/examples/noise_modelling_approaches/#noise_modelling_approaches_model_input_noise","page":"Approaches for modelling system noise","title":"Using a noisy input process","text":"Finally, we will consider the case where we have a deterministic system, but which is exposed to a noisy input process. One example could be a light sensitive system, where the amount of experienced sunlight is stochastic due to e.g. variable cloud cover. Practically, this can be considered as extrinsic noise, however, we will generate the noise using a different approach from in the previous section. Here, we pre-simulate a random process in time, which we then feed into the system as a functional, time-dependent, parameter. A more detailed description of functional parameters can be found here.\n\nWe assume that our repressilator has an input, which corresponds to the K value that controls X's production. First we create a function, make_K_series, which creates a randomised time series representing K's value over time. \n\nusing DataInterpolations\nfunction make_K_series(; K_mean = 20.0, n = 500, θ = 0.01)\n    t_samples = range(0.0, stop = tend, length = n)\n    K_series = fill(K_mean, n)\n    for i = 2:n\n        K_series[i] = K_series[i-1] + (rand() - 0.5) - θ*(K_series[i-1] - K_mean)\n    end\n    return LinearInterpolation(K_series, t_samples)\nend\nplot(make_K_series())\n\nNext, we create an updated repressilator model, where the input K value is modelled as a time-dependent parameter.\n\n@parameters (K_in::typeof(make_K_series()))(..)\nK_in = K_in(default_t())\nrepressilator_Kin = @reaction_network begin\n    hillr(Z,v,$K_in,n), ∅ --> X\n    hillr(X,v,K,n), ∅ --> Y\n    hillr(Y,v,K,n), ∅ --> Z\n    d, (X, Y, Z) --> ∅\nend\nnothing # hide\n\nFinally, we will again perform ensemble simulations of our model. This time, at the beginning of each simulation, we will use make_K_series to generate a new K, and set this as the K_in parameter's value.\n\nfunction prob_func_Kin(prob, i, repeat)\n    p = [ps; :K_in => make_K_series()]    \n    return remake(prob; p)\nend\noprob = ODEProblem(repressilator_Kin, u0, tend, [ps; :K_in => make_K_series()])\neprob_inputnoise = EnsembleProblem(oprob; prob_func = prob_func_Kin)\nsol_inputnoise = solve(eprob_inputnoise; trajectories = 4)\nplot(sol_inputnoise; idxs = :X)\n\nLike in the previous two cases, this generates heterogeneous trajectories across our ensemble.","category":"section"},{"location":"model_creation/examples/noise_modelling_approaches/#noise_modelling_approaches_model_noisy_oscillation_mean","page":"Approaches for modelling system noise","title":"Investigating the mean of noisy oscillations","text":"Finally, we will observe an interesting phenomenon for ensembles of stochastic oscillators. First, we create ensemble simulations with a larger number of trajectories.\n\nsol_intrinsic = solve(eprob_intrinsic, ImplicitEM(); trajectories = 200)\nsol_extrinsic = solve(eprob_extrinsic; trajectories = 200)\nnothing # hide\n\nNext, we can use the EnsembleSummary interface to plot each ensemble's mean activity (as well as 5% and 95% quantiles) over time:\n\ne_summary_intrinsic = EnsembleAnalysis.EnsembleSummary(sol_intrinsic, 0.0:1.0:tend)\ne_summary_extrinsic = EnsembleAnalysis.EnsembleSummary(sol_extrinsic, 0.0:1.0:tend)\nplot(e_summary_intrinsic; label = \"Intrinsic noise\", idxs = 1)\nplot!(e_summary_extrinsic; label = \"Extrinsic noise\", idxs = 1)\n\nHere we can see that, over time, the systems' mean X activity reaches a constant level around 30.\n\nThis is a well-known phenomenon (especially in circadian biology[2]). Here, as stochastic oscillators evolve from a common initial condition the mean behaves as a damped oscillator. This can be caused by two different phenomena:\n\nThe individual trajectories are themselves damped.\nThe individual trajectories's phases get de-synchronised.\n\nHowever, if we only observe the mean behaviour (and not the single trajectories), we cannot know which of these cases we are encountering. Here, by checking the single-trajectory plots from the previous sections, we note that this is due to trajectory de-synchronisation. Stochastic oscillators have often been cited as a reason for the importance to study cellular systems at the single-cell level, and not just in bulk.\n\n","category":"section"},{"location":"model_creation/examples/noise_modelling_approaches/#noise_modelling_approaches_references","page":"Approaches for modelling system noise","title":"References","text":"[1]: Michael B. Elowitz, Arnold J. Levine, Eric D. Siggia, Peter S. Swain, Stochastic Gene Expression in a Single Cell, Science (2002).\n\n[2]: Qiong Yang, Bernardo F. Pando, Guogang Dong, Susan S. Golden, Alexander van Oudenaarden, Circadian Gating of the Cell Cycle Revealed in Single Cyanobacterial Cells, Science (2010).","category":"section"},{"location":"inverse_problems/behaviour_optimisation/#behaviour_optimisation","page":"Optimization for Non-data Fitting Purposes","title":"Optimization for Non-data Fitting Purposes","text":"<details><summary><strong>Environment setup and package installation</strong></summary>\n\nThe following code sets up an environment for running the code on this page.\n\nusing Pkg\nPkg.activate(; temp = true) # Creates a temporary environment, which is deleted when the Julia session ends.\nPkg.add(\"Catalyst\")\nPkg.add(\"OptimizationBase\")\nPkg.add(\"OptimizationBBO\")\nPkg.add(\"OrdinaryDiffEqDefault\")\nPkg.add(\"Plots\")\n\n</details>\n\n\n   In previous tutorials we have described how to use PEtab.jl and Optimization.jl for parameter fitting. This involves solving an optimisation problem (to find the parameter set yielding the best model-to-data fit). There are, however, other situations that require solving optimisation problems. Typically, these involve the creation of a custom objective function, which minimizer can then be found using Optimization.jl. In this tutorial we will describe this process, demonstrating how parameter space can be searched to find values that achieve a desired system behaviour. Many options used here are described in more detail in the tutorial on using Optimization.jl for parameter fitting. A more throughout description of how to solve these problems is provided by Optimization.jl's documentation and the literature[1].","category":"section"},{"location":"inverse_problems/behaviour_optimisation/#behaviour_optimisation_IFFL_example","page":"Optimization for Non-data Fitting Purposes","title":"Maximising the pulse amplitude of an incoherent feed forward loop","text":"Incoherent feedforward loops (network motifs where a single component both activates and deactivates a downstream component) are able to generate pulses in response to step inputs[2]. In this tutorial we will consider such an incoherent feedforward loop, attempting to generate a system with as prominent a response pulse as possible.\n\nOur model consists of 3 species: X (the input node), Y (an intermediary), and Z (the output node). In it, X activates the production of both Y and Z, with Y also deactivating Z. When X is activated, there will be a brief time window where Y is still inactive, and Z is activated. However, as Y becomes active, it will turn Z off. This creates a pulse of Z activity. To trigger the system, we create an event, which increases the production rate of X (pX) by a factor of 10 at time t = 10.\n\nusing Catalyst\nincoherent_feed_forward = @reaction_network begin\n    @discrete_events [10.0] => [pX => 10*pX]\n    pX, 0 --> X\n    pY*X, 0 --> Y\n    pZ*X/Y, 0 --> Z\n    1.0, (X,Y,Z) --> 0\nend\n\nTo demonstrate this pulsing behaviour we will simulate the system for an example parameter set. We select an initial condition (u0) so the system begins in a steady state.\n\nusing OrdinaryDiffEqDefault, Plots\nexample_p = [:pX => 0.1, :pY => 1.0, :pZ => 1.0]\ntspan = (0.0, 50.0)\nexample_u0 = [:X => 0.1, :Y => 0.1, :Z => 1.0]\n\noprob = ODEProblem(incoherent_feed_forward, example_u0, tspan, example_p)\nsol = solve(oprob)\nplot(sol)\n\nHere we note that, while X and Y reach new steady state levels in response to the increase in pX, Z resumes to its initial concentration after the pulse.\n\nWe will now attempt to find the parameter set (pXpYpZ) which maximises the response pulse amplitude (defined by the maximum activity of Z subtracted by its steady state activity). To do this, we create a custom objective function:\n\nfunction pulse_amplitude(p, _)\n    p = Dict([:pX => p[1], :pY => p[2], :pZ => p[2]])\n    u0 = [:X => p[:pX], :Y => p[:pX]*p[:pY], :Z => p[:pZ]/p[:pY]^2]\n    oprob_local = remake(oprob; u0, p)\n    sol = solve(oprob_local; verbose = false, maxiters = 10000)\n    SciMLBase.successful_retcode(sol) || return Inf\n    return -(maximum(sol[:Z]) - sol[:Z][1])\nend\nnothing # hide\n\nThis objective function takes two arguments (a parameter value p, and an additional one which we will ignore but is discussed in a note here). It first calculates the new initial steady state concentration for the given parameter set. Next, it creates an updated ODEProblem using the steady state as initial conditions and the, to the objective function provided, input parameter set.  Finally, Optimization.jl finds the function's minimum value, so to find the maximum relative pulse amplitude, we make our objective function return the negative pulse amplitude.\n\nAs described in our tutorial on parameter fitting using Optimization.jl we use remake, verbose = false, maxiters = 10000, and a check on the simulations return code, all providing various advantages to the optimisation procedure (as explained in that tutorial).\n\nJust like for parameter fitting, we create an OptimizationProblem using our objective function, and some initial guess of the parameter values. We also set upper and lower bounds for each parameter using the lb and ub optional arguments (in this case limiting each parameter's value to the interval (01100)).\n\nusing OptimizationBase\ninitial_guess = [1.0, 1.0, 1.0]\nopt_prob = OptimizationProblem(pulse_amplitude, initial_guess; lb = [1e-1, 1e-1, 1e-1], ub = [1e1, 1e1, 1e1])\nnothing # hide\n\nnote: Note\nAs described in a previous section on Optimization.jl, OptimizationProblems do not support setting parameter values using maps. We must instead set initial_guess values using a vector. Next, in the first line of our objective function, we reshape the parameter values to the common form used across Catalyst (e.g. [:pX => p[1], :pY => p[2], :pZ => p[2]], however, here we use a dictionary to easier compute the steady state initial condition). We also note that the order used in this array corresponds to the order we give each parameter's bounds in lb and ub, and the order in which their values occur in the output solution.\n\nAs previously described, Optimization.jl supports a wide range of optimisation algorithms. Here we use one from BlackBoxOptim.jl:\n\nusing OptimizationBBO\nopt_sol = solve(opt_prob, BBO_adaptive_de_rand_1_bin_radiuslimited())\nnothing # hide\n\nFinally, we plot a simulation using the found parameter set (stored in opt_sol.u):\n\nps_res = Dict([:pX => opt_sol.u[1], :pY => opt_sol.u[2], :pZ => opt_sol.u[2]])\nu0_res = [:X => ps_res[:pX], :Y => ps_res[:pX]*ps_res[:pY], :Z => ps_res[:pZ]/ps_res[:pY]^2]\noprob_res = remake(oprob; u0 = u0_res, p = ps_res)\nsol_res = solve(oprob_res)\nplot(sol_res; idxs = :Z)\n\nFor this model, it turns out that Z's maximum pulse amplitude is equal to twice its steady state concentration. Hence, the maximisation of its pulse amplitude is equivalent to maximising its steady state concentration.","category":"section"},{"location":"inverse_problems/behaviour_optimisation/#behaviour_optimisation_options","page":"Optimization for Non-data Fitting Purposes","title":"Other optimisation options","text":"How to use Optimization.jl is discussed in more detail in this tutorial. This includes options such as using automatic differentiation, setting constraints, and setting optimisation solver options. Finally, it discusses the advantages of carrying out the fitting in logarithmic space, something which can be advantageous for the problem described above as well.\n\n","category":"section"},{"location":"inverse_problems/behaviour_optimisation/#structural_identifiability_citation","page":"Optimization for Non-data Fitting Purposes","title":"Citation","text":"If you use this functionality in your research, please cite the following paper to support the authors of the Optimization.jl package:\n\n@software{vaibhav_kumar_dixit_2023_7738525,\n\tauthor = {Vaibhav Kumar Dixit and Christopher Rackauckas},\n\tmonth = mar,\n\tpublisher = {Zenodo},\n\ttitle = {Optimization.jl: A Unified Optimization Package},\n\tversion = {v3.12.1},\n\tdoi = {10.5281/zenodo.7738525},\n  \turl = {https://doi.org/10.5281/zenodo.7738525},\n\tyear = 2023\n}\n\n","category":"section"},{"location":"inverse_problems/behaviour_optimisation/#References","page":"Optimization for Non-data Fitting Purposes","title":"References","text":"[1]: Mykel J. Kochenderfer, Tim A. Wheeler Algorithms for Optimization, The MIT Press (2019).\n\n[2]: Lea Goentoro, Oren Shoval, Marc W Kirschner, Uri Alon The incoherent feedforward loop can provide fold-change detection in gene regulation, Molecular Cell (2009).","category":"section"},{"location":"steady_state_functionality/examples/bifurcationkit_codim2/#bifurcationkit_codim2","page":"Tracking Bifurcation Point w.r.t. Secondary Parameters using BifurcationKit.jl","title":"Tracking Bifurcation Point w.r.t. Secondary Parameters using BifurcationKit.jl","text":"<details><summary><strong>Environment setup and package installation</strong></summary>\n\nThe following code sets up an environment for running the code on this page.\n\nusing Pkg\nPkg.activate(; temp = true) # Creates a temporary environment, which is deleted when the Julia session ends.\nPkg.add(\"BifurcationKit\")\nPkg.add(\"Catalyst\")\nPkg.add(\"OrdinaryDiffEqDefault\")\nPkg.add(\"Plots\")\n\n</details>\n\n\n  \n\nPreviously, we have shown how to compute bifurcation diagrams using BifurcationKit.jl. In this example, we will show how, after computing the initial diagram, we can track how the position of a bifurcation point moves as a secondary parameter is changed (so-called codimensional 2 bifurcation analysis). More information on how to track bifurcation points along secondary parameters can be found in the BifurcationKit documentation.","category":"section"},{"location":"steady_state_functionality/examples/bifurcationkit_codim2/#bifurcationkit_codim2_bifdia","page":"Tracking Bifurcation Point w.r.t. Secondary Parameters using BifurcationKit.jl","title":"Computing the bifurcation diagram for the Repressilator","text":"We will first compute the bifurcation diagram, using the same approach as in the corresponding tutorial. For this example, we will use the oscillating Repressilator model.\n\nusing Catalyst\nrepressilator = @reaction_network begin\n    hillr(Z,v,K,n), ∅ --> X\n    hillr(X,v,K,n), ∅ --> Y\n    hillr(Y,v,K,n), ∅ --> Z\n    d, (X, Y, Z) --> ∅\nend\n\nNext, we create a BifurcationProblem for our model. We will compute the bifurcation diagram with respect to the parameter v, and plot the species X in the diagram.\n\nusing BifurcationKit\nbif_par = :v\nu_guess = [:X => 20.0, :Y => 15.0, :Z => 15.0]\np_start = [:v => 10.0, :K => 15.0, :n => 3, :d => 0.2]\nplot_var = :X\nbprob = BifurcationProblem(repressilator, u_guess, p_start, bif_par; plot_var)\nnothing # hide\n\nWe compute the bifurcation diagram using the bifurcationdiagram function. We will compute it across the interval v in (020).\n\nv_span = (0.0, 20.0)\nopts_br = ContinuationPar(p_min = v_span[1], p_max = v_span[2])\nbifdia = bifurcationdiagram(bprob, PALC(), 3, opts_br; bothside = true)\nnothing # hide\n\nFinally, we plot the bifurcation diagram.\n\nusing Plots\nplot(bifdia; xguide = bif_par, yguide = plot_var, branchlabel = \"Continuation of steady state w.r.t. v\",\n    linewidthstable = 6, linewidthunstable = 3, markersize = 5)\n\nWe note that for small values of v the system's single steady state is stable (where the line is thicker). After a Hopf bifurcation (the red point), the state turns unstable (where the line is thinner). For chemical reaction networks (which mostly are well-behaved) a single unstable steady state typically corresponds to an oscillation. We can confirm that the system oscillates in the unstable region (while it reaches a stable steady state in the stable region) using simulations:\n\nusing OrdinaryDiffEqDefault\np_nosc = [:v => 5.0, :K => 15.0, :n => 3, :d => 0.2]\np_osc = [:v => 15.0, :K => 15.0, :n => 3, :d => 0.2]\nprob_nosc = ODEProblem(repressilator, u_guess, 80.0, p_nosc)\nprob_osc = ODEProblem(repressilator, u_guess, 80.0, p_osc)\nsol_nosc = OrdinaryDiffEqDefault.solve(prob_nosc)\nsol_osc = OrdinaryDiffEqDefault.solve(prob_osc)\nplot(plot(sol_nosc; title = \"v = 5\"), plot(sol_osc; title = \"v = 15\"), size = (1000,400), lw = 4)","category":"section"},{"location":"steady_state_functionality/examples/bifurcationkit_codim2/#bifurcationkit_codim2_2ndpar_cont","page":"Tracking Bifurcation Point w.r.t. Secondary Parameters using BifurcationKit.jl","title":"Tracking the bifurcation point w.r.t. a second parameter","text":"Next, we will investigate how the Hopf bifurcation point moves (in v-X space) as a second parameter (K) is changed. To do this we will use BifurcationKit.jl's continuation function (the bifurcationdiagram function, which we previously have used, works by calling continuation recursively). We will call it on the Hopf bifurcation point. First we need to retrieve some indexes that are required to make Catalyst (which primarily indexes through symbols) work with BifurcationKit (which primarily indexes through numbers). A smoother interface for this will hopefully be added in the future.\n\nK_idx = findfirst(isequal(repressilator.K), parameters(complete(ss_ode_model(repressilator))))\nv_idx = findfirst(isequal(repressilator.v), parameters(complete(ss_ode_model(repressilator))))\nK_sym = Symbol(:p, K_idx)\nv_sym = Symbol(:p, v_idx)\nnothing # hide\n\nNext we designate the interval over which we wish to change K, and from this create a new ContinuationPar.\n\nK_span = (0.01, 27.0)\nopts_br_2 = ContinuationPar(p_min = K_span[1], p_max = K_span[2])\nnothing # hide\n\nNow we can compute the continuation of the Hopf bifurcation. First we must extract a branch from our bifurcation diagram (using get_branch), as continuation cannot work on bifurcation diagrams directly (our bifurcation diagram consists of a single branch, which we here extract). Next, we can call continuation on it, designating that we wish to perform continuation from the second point on the branch (which corresponds to the Hopf bifurcation, the first point is one of the branch's two endpoints).\n\nbranch = get_branch(bifdia, ()).γ\ncont_hopf = continuation(branch, 2, (@optic _[K_idx]), opts_br_2; start_with_eigen = true, bothside = true)\nnothing # hide\n\nWe can now plot how the position of the bifurcation point changes with K. Here, we use vars = (v_sym, :x) to designate that we wish to plot (across the continuation branch) the plotting variable (X, which we designated when we created our BifurcationProblem) against the first parameter (v).\n\nplot(bifdia; branchlabel = \"Continuation of steady state w.r.t. v, (K = 15)\", linewidthstable = 6, \n    linewidthunstable = 3, markersize = 5)\nplot!(cont_hopf; vars = (v_sym, :x), xlimit = v_span, xguide = bif_par, yguide = plot_var, \n    branchlabel = \"Continuation of Hopf bifurcation w.r.t. K\")\n\nIn this case we cannot see directly which part of the K continuation branch corresponds to low values, however, for low K the Hopf bifurcation occurs for much lower values of v (and corresponds to lower steady state values of X). We can check this by e.g. re-computing the Hopf branch for K_span = (0.01, 20.0) and see that the rightmost part of the branch is shortened. \n\nWe can confirm that the new line corresponds to the Hopf Bifurcation point by recomputing the initial bifurcation diagram, but for a lower K value.\n\np_start_2 = [:v => 10.0, :K => 10.0, :n => 3, :d => 0.2]\nbprob_2 = BifurcationProblem(repressilator, u_guess, p_start_2, bif_par; plot_var)\nbifdia_2 = bifurcationdiagram(bprob_2, PALC(), 3, opts_br; bothside = true)\nplot!(bifdia_2; xguide = bif_par, yguide = plot_var, branchlabel = \"Continuation of steady state w.r.t. v, (K = 10)\",\n    linewidthstable = 6, linewidthunstable = 3, markersize = 5)\n\nHere we see that the Hopf bifurcation point of the new diagram also lies on the Hopf continuation line.\n\nFinally, we have already noted that the Hopf bifurcation splits parameter space into one part where the system oscillates and one where it doesn't. Previously we plotted the Hopf continuation in v-X space, however, it is also possible to plot it in v-K space using the vars = (v_sym, K_sym) argument:\n\nxlimit = extrema(getfield.(cont_hopf.branch, v_sym))\nylimit = extrema(getfield.(cont_hopf.branch, K_sym))\nplot(cont_hopf; vars = (v_sym, K_sym), xlimit, ylimit, branchlabel = \"Hopf bifurcation\", \n    xguide = \"v\", yguide = \"K\", lw = 6)\n\nNext, we colour parameter space according to whether the steady state is stable (blue) or unstable (red). We also mark two sample values (one in each region).\n\nsample1 = (15.0, 10.0)\nsample2 = (5.0, 15.0)\nplot(cont_hopf; vars = (v_sym, K_sym), fillrange = ylimit[2], xguide = \"v\", yguide = \"K\")\nplot!(cont_hopf; vars = (v_sym, K_sym), fillrange = ylimit[1], xlimit, ylimit)\nscatter!(sample1; label = \"Oscillatory parameter set\", markersize = 7)\nscatter!(sample2; label = \"Non-oscillatory parameter set\", markersize = 7)\n\nFinally, we can perform one simulation using each of the parameter samples, confirming that one corresponds to an oscillation, while the other one does not.\n\nps_nosc = [:v => sample2[1], :K => sample2[2], :n => 3, :d => 0.2]\nps_osc = [:v => sample1[1], :K => sample1[2], :n => 3, :d => 0.2]\noprob_nosc = ODEProblem(repressilator, u_guess, 100.0, ps_nosc)\noprob_osc = ODEProblem(repressilator, u_guess, 100.0, ps_osc)\nsol_nosc = OrdinaryDiffEqDefault.solve(oprob_nosc)\nsol_osc = OrdinaryDiffEqDefault.solve(oprob_osc)\nplot(plot(sol_nosc; title = \"No oscillation\"), plot(sol_osc; title = \"Oscillation\"); size = (1000, 400), lw = 4)\n\n","category":"section"},{"location":"steady_state_functionality/examples/bifurcationkit_codim2/#bifurcationkit_periodic_orbits_citation","page":"Tracking Bifurcation Point w.r.t. Secondary Parameters using BifurcationKit.jl","title":"Citation","text":"If you use BifurcationKit.jl for your work, we ask that you cite the following paper!! Open source development strongly depends on this. It is referenced on HAL-Inria with bibtex entry CITATION.bib.","category":"section"},{"location":"model_creation/chemistry_related_functionality/#chemistry_functionality","page":"Chemistry-related Functionality","title":"Chemistry-related Functionality","text":"<details><summary><strong>Environment setup and package installation</strong></summary>\n\nThe following code sets up an environment for running the code on this page.\n\nusing Pkg\nPkg.activate(; temp = true) # Creates a temporary environment, which is deleted when the Julia session ends.\nPkg.add(\"Catalyst\")\n\n</details>\n\n\n While Catalyst has primarily been designed around the modelling of biological systems, reaction network models are also common in chemistry. This section describes two types of functionality, that while of general interest, should be especially useful in the modelling of chemical systems.\n\nThe @compound option, which enables the user to designate that a specific species is composed of certain subspecies.\nThe balance_reaction function, which enables the user to balance a reaction so the same number of components occur on both sides.","category":"section"},{"location":"model_creation/chemistry_related_functionality/#chemistry_functionality_compounds","page":"Chemistry-related Functionality","title":"Modelling with compound species","text":"","category":"section"},{"location":"model_creation/chemistry_related_functionality/#chemistry_functionality_compounds_programmatic","page":"Chemistry-related Functionality","title":"Creating compound species programmatically","text":"We will first show how to create compound species through programmatic model construction, and then demonstrate using the DSL. To create a compound species, use the @compound macro, first designating the compound, followed by its components (and their stoichiometries). In this example, we will create a CO₂ molecule, consisting of one C atom and two O atoms. First, we create species corresponding to the components:\n\nusing Catalyst\nt = default_t()\n@species C(t) O(t)\nnothing # hide\n\nNext, we create the CO2 compound species:\n\n@compound CO2 ~ C + 2O\n\nHere, the compound is the first argument to the macro, followed by its component (with the left-hand and right-hand sides separated by a ~ sign). While non-compound species (such as C and O) have their independent variable (in this case t) designated, independent variables are generally not designated for compounds (these are instead directly inferred from their components). Components with non-unitary stoichiometries have this value written before the component (generally, the rules for designating the components of a compound are identical to those of designating the substrates or products of a reaction). The created compound, CO2, is also a species, and can be used wherever e.g. C can be used:\n\nisspecies(CO2)\n\nIn its metadata, however, is stored information of its components, which can be retrieved using the components (returning a vector of its component species) and coefficients (returning a vector with each component's stoichiometry) functions:\n\ncomponents(CO2)\n\ncoefficients(CO2)\n\nAlternatively, we can retrieve the components and their stoichiometric coefficients as a single vector using the component_coefficients function:\n\ncomponent_coefficients(CO2)\n\nFinally, it is possible to check whether a species is a compound using the iscompound function:\n\niscompound(CO2)\n\nCompound components that are also compounds are allowed, e.g. we can create a carbonic acid compound (H₂CO₃) that consists of CO₂ and H₂O:\n\n@species H(t)\n@compound H2O ~ 2H + O\n@compound H2CO3 ~ CO2 + H2O\n\nWhen multiple compounds are created, they can be created simultaneously using the @compounds macro, e.g. the previous code-block can be re-written as:\n\n@species H(t)\n@compounds begin\n    H2O ~ 2H + O\n    H2CO3 ~ CO2 + H2O\nend","category":"section"},{"location":"model_creation/chemistry_related_functionality/#chemistry_functionality_compounds_DSL","page":"Chemistry-related Functionality","title":"Creating compound species within the DSL","text":"It is also possible to declare species as compound species within the @reaction_network DSL, using the @compounds options:\n\nrn = @reaction_network begin\n    @species C(t) H(t) O(t)\n    @compounds begin\n        C2O ~ C + 2O\n        H2O ~ 2H + O\n        H2CO3 ~ CO2 + H2O\n    end\n    (k1,k2), H2O + CO2 <--> H2CO3\nend\n\nWhen creating compound species using the DSL, it is important to note that every component must be known to the system as a species, either by being declared using the @species or @compound options, or by appearing in a reaction. E.g. the following is not valid\n\nrn = @reaction_network begin\n    @compounds begin\n        C2O ~ C + 2O\n        H2O ~ 2H + O\n        H2CO3 ~ CO2 + H2O\n    end\n    (k1,k2), H2O+ CO2 <--> H2CO3\nend\n\nas the components C, H, and O are not declared as species anywhere. Please also note that only @compounds can be used as an option in the DSL, not @compound.","category":"section"},{"location":"model_creation/chemistry_related_functionality/#chemistry_functionality_compounds_metadata","page":"Chemistry-related Functionality","title":"Designating metadata and default values for compounds","text":"Just like for normal species, it is possible to designate metadata and default values for compounds. Metadata is provided after the compound name, but separated from it by a ,:\n\n@compound (CO2, [unit=\"mol\"]) ~ C + 2O\nnothing # hide\n\nDefault values are designated using =, and provided directly after the compound name.:\n\n@compound (CO2 = 2.0) ~ C + 2O\nnothing # hide\n\nIf both default values and meta data are provided, the metadata is provided after the default value:\n\n@compound (CO2 = 2.0, [unit=\"mol\"]) ~ C + 2O\nnothing # hide\n\nIn all of these cases, the left-hand side must be enclosed within ().","category":"section"},{"location":"model_creation/chemistry_related_functionality/#chemistry_functionality_compounds_mult_ivs","page":"Chemistry-related Functionality","title":"Compounds with multiple independent variables","text":"While we generally do not need to specify independent variables for compound, if the components (together) have more than one independent variable, this must be done:\n\nt = default_t()\n@parameters s\n@species N(s) O(t) \n@compound NO2(t,s) ~ N + 2O\n\nHere, NO2 depend both on a spatial independent variable (s) and a time one (t). This is required since, while multiple independent variables can be inferred, their internal order cannot (and must hence be provided by the user).","category":"section"},{"location":"model_creation/chemistry_related_functionality/#Balancing-chemical-reactions","page":"Chemistry-related Functionality","title":"Balancing chemical reactions","text":"One use of defining a species as a compound is that they can be used to balance reactions so that the number of components are the same on both sides. Catalyst provides the balance_reaction function, which takes a reaction, and returns a balanced version. E.g. let us consider a reaction when carbon dioxide is formed from carbon and oxide C + O --> CO2. Here, balance_reaction enables us to find coefficients creating a balanced reaction (in this case, where the number of carbon and oxygen atoms are the same on both sides). To demonstrate, we first created the unbalanced reactions:\n\nrx = @reaction k, C + O --> $CO2\n\nHere, the reaction rate (k) is not involved in the reaction balancing. We use interpolation for CO2, ensuring that the CO2 used in the reaction is the same one we previously defined as a compound of C and O. Next, we call the balance_reaction function\n\nbalance_reaction(rx)\n\nwhich correctly finds the (rather trivial) solution C + 2O --> CO2. Here we note that balance_reaction actually returns a vector. The reason is that, in some cases, the reaction balancing problem does not have a single obvious solution. Typically, a single solution is the obvious candidate (in which case this is the vector's only element). However, when this is not the case, the vector instead contain several reactions (from which a balanced reaction cab be generated).\n\nLet us consider a more elaborate example, the reaction between ammonia (NH₃) and oxygen (O₂) to form nitrogen monoxide (NO) and water (H₂O). Let us first create the components and the unbalanced reaction:\n\nusing Catalyst # hide\nt = default_t()\n@species N(t) H(t) O(t) \n@compounds begin\n    NH3 ~ N + 3H\n    O2 ~ 2O\n    NO ~ N + O\n    H2O ~ 2H + O\nend\nunbalanced_reaction = @reaction k, $NH3 + $O2 --> $NO + $H2O\n\nWe can now create a balanced version (where the amount of H, N, and O is the same on both sides):\n\nbalanced_reaction = balance_reaction(unbalanced_reaction)[1]\n\nReactions declared as a part of a ReactionSystem (e.g. using the DSL) can be retrieved for balancing using the reactions function. Please note that balancing these will not mutate the ReactionSystem, but a new reaction system will need to be created using the balanced reactions.\n\nnote: Note\nReaction balancing is currently not supported for reactions involving compounds of compounds.","category":"section"},{"location":"model_creation/chemistry_related_functionality/#Balancing-full-systems","page":"Chemistry-related Functionality","title":"Balancing full systems","text":"It is possible to balance all the reactions of a reaction system simultaneously using the balance_system function. Here, the output is a new system, where all reactions are balanced. E.g. We can use it to balance this system of methane formation/combustion:\n\nrs = @reaction_network begin\n    @species C(t) O(t) H(t)\n    @compounds begin\n        H2(t) ~ 2H\n        CH4(t) ~ C + 4H\n        O2(t) ~ 2O\n        CO2(t) ~ C + 2O\n        H2O(t) ~ 2H + O\n    end\n    1.0, C + H2 --> CH4\n    2.0, CH4 + O2 --> CO2 + H2O\nend\nrs_balanced = balance_system(rs)\n\nExcept for the modified reaction stoichiometries, the new system is identical to the previous one.","category":"section"},{"location":"spatial_modelling/spatial_ode_simulations/#spatial_dspace_ode_simulations","page":"Spatial ODE simulations","title":"Spatial ODE simulations","text":"Our introduction to spatial space simulations has already provided an extensive description of how to simulate DiscreteSpaceReactionSystems using ODEs. Further tutorials have also shown how to retrieve values from simulations and or how to plot them. Here we will build on this, primarily discussing strategies for increasing ODE simulation performance. This is especially important for spatial simulations, as these typically are more computationally demanding as compared to non-spatial ones. While focusing on non-spatial simulations, this ODE performance tutorial is also be useful to read.","category":"section"},{"location":"spatial_modelling/spatial_ode_simulations/#spatial_dspace_ode_simulations_solvers","page":"Spatial ODE simulations","title":"Solver selection for spatial ODE simulations","text":"Previously we have described how to select ODE solvers, and how this can impact simulation performance. This is especially relevant for spatial simulations. For stiff problems, FBDF is a good first solver to try. For non-stiff problems, ROCK2 is instead a good first alternative. However, it is still worthwhile to explore a range of alternative solvers.","category":"section"},{"location":"spatial_modelling/spatial_ode_simulations/#spatial_dspace_ode_simulations_jacobians","page":"Spatial ODE simulations","title":"Jacobian options for spatial ODE simulations","text":"We have previously described how, when implicit solvers are used to solve stiff ODEs, the strategy for computing the system Jacobian is important. This is especially the case for spatial simulations, where the Jacobian often is large and highly sparse. Catalyst implements special methods for spatial Jacobians. To utilise these, provide the jac = true argument to your ODEProblem when it is created (if jac = false, which is the default, automatic differentiation will be used for Jacobian computation). Here we simulate a Brusselator while designating to use Catalyst's computed Jacobian:\n\nusing Catalyst, OrdinaryDiffEqBDF\nbrusselator = @reaction_network begin\n    A, ∅ --> X\n    1, 2X + Y --> 3X\n    B, X --> Y\n    1, X --> ∅\nend\ndiffusion_rx = @transport_reaction D X\nspace = CartesianGrid((20,20))\ndsrs = DiscreteSpaceReactionSystem(brusselator, [diffusion_rx], space)\n\nu0 = [:X => rand(20, 20), :Y => 10.0]\ntspan = (0.0, 1.0)\nps = [:A => 1.0, :B => 4.0, :D => 0.2]\noprob = ODEProblem(dsrs, u0, tspan, ps; jac = true)\nsol = solve(oprob, FBDF())\nnothing # hide\n\nFor large systems, building a dense Jacobian can be problematic, in which case a sparse Jacobian can be designated using sparse = true: \n\noprob = ODEProblem(dsrs, u0, tspan, ps; jac = true, sparse = true)\nsol = solve(oprob, FBDF())\nnothing # hide\n\nIt is possible to use sparse = true while jac = false, in which case a sparse Jacobian is computed using automatic differentiation. ","category":"section"},{"location":"spatial_modelling/discrete_space_reaction_systems/#spatial_dspace_modelling_intro","page":"Introduction to Spatial Modelling with Catalyst","title":"Introduction to Spatial Modelling with Catalyst","text":"Catalyst supports the expansion of non-spatial ReactionSystems (created using e.g. the @reaction_network DSL) to spatial domains. Spatial simulation of Catalyst models is a work in progress. Currently, the following is supported:\n\nSpatial ODE and Jump simulations.\nDiscrete spatial domains.\nConstant-rate transportation reactions (species moving spatially at constant rates). \n\nFeatures for which support is planned in future updates include:\n\nModels on continuous domains with automatic discretisation (these models can already be simulated if the user provides a discretisation).\nSDE simulations.\nTransport reactions with non-constant rates as well as more general spatial reactions.\n\nThis tutorial introduces spatial modelling on discrete domains. It describes the basics of creating and simulating such models. To do so, it uses ODE simulations as examples. Additional tutorials provide further details on how to interact with spatial simulation structures and plot spatial simulations, and also provide further details on ODE and jump simulations, respectively.","category":"section"},{"location":"spatial_modelling/discrete_space_reaction_systems/#spatial_dspace_modelling_intro_example","page":"Introduction to Spatial Modelling with Catalyst","title":"Basic example of a spatial simulation on a discrete domain","text":"To perform discrete-space spatial simulations, the user must first define a DiscreteSpaceReactionSystem. These combine:\n\nA (non-spatial) ReactionSystem(@ref) model (created using standard Catalyst syntax).\nA vector of spatial reactions, describing how species can move spatially across the domain.\nA space defining the spatial domain's compartments and how they are connected.\n\nHere, as an example, we will simulate a spatial two-state model. To do so, we first define our (non-spatial) model, the spatial reactions, and the space. These are then bundled into a DiscreteSpaceReactionSystem.\n\nusing Catalyst\ntwo_state_model = @reaction_network begin\n    (k1,k2), X1 <--> X2\nend\ndiffusion_rx = @transport_reaction D X1\nspace = CartesianGrid((5,5))\ndsrs = DiscreteSpaceReactionSystem(two_state_model, [diffusion_rx], space)\nnothing # hide\n\nThis model contains:\n\nA single spatial reaction, a transport reaction where X1 moves at constant rate D between adjacent compartments.\nA 2d Cartesian grid of 5x5 compartments to simulate our model on.\n\nMore details on spatial reactions are available here. In addition to Cartesian grid spaces (in 1, 2, and 3 dimensions), masked and unstructured (graph) spaces are also supported. The different discrete space types described in more detail here.\n\nOnce created, DiscreteSpaceReactionSystems can be used as input to various problem types, which then can be simulated using the same syntax as non-spatial models. Here, we prepare an ODE simulation by creating an ODEProblem:\n\nu0 = [:X1 => rand(5, 5), :X2 => 2.0]\ntspan = (0.0, 10.0)\nps = [:k1 => 1.0, :k2 => 2.0, :D => 0.2]\noprob = ODEProblem(dsrs, u0, tspan, ps)\nnothing # hide\n\nIn this example we used non-uniform values for X1's initial condition, but uniform values for the remaining initial condition and parameter values. More details of uniform and non-uniform initial conditions and parameter values are provided here. We also note that the diffusion reaction introduces a new parameter, D (determining X1's diffusion rate), whose value must be designated in the parameter vector. \n\nWe can now simulate our model:\n\nusing OrdinaryDiffEqDefault\nsol = solve(oprob)\nnothing # hide\n\nWe note that simulations of spatial models are often computationally expensive. Advice on the performance of spatial ODE simulations is provided here. \n\nFinally, we can access \"X1's value across the simulation using\n\nspat_getu(sol, :X1, dsrs)\n\nand plot the simulation using\n\nimport CairoMakie\ndspace_animation(sol, :X1, dsrs, \"dspace_simulation.mp4\")\n\n(Image: ) More information on how to retrieve values from spatial simulations can be found here, and for plotting them, here. Finally, a list of functions for querying DiscreteSpaceReactionSystems for various properties can be found here.","category":"section"},{"location":"spatial_modelling/discrete_space_reaction_systems/#spatial_dspace_modelling_intro_spatial_reactions","page":"Introduction to Spatial Modelling with Catalyst","title":"Spatial reactions","text":"Spatial reactions describe reaction events which involve species across two connected compartments. Currently, only so-called transportation reactions are supported. These consist of:\n\nA rate at which the reaction occurs. As for non-spatial reactions, this can be any expression. However, currently, it may only consist of parameters and other constants. \nA single species which is transported from one compartment to an adjacent one.\n\nAt the occurrence of a transport reaction, the specific species moves to the adjacent compartment. Many common spatial models can be represented using transport reactions only. These can model phenomena such as diffusion or constant flux. A transportation reaction can be created using the @transportation_reaction macro. E.g. above we used\n\nusing Catalyst # hide\ndiffusion_rx = @transport_reaction D X1\nnothing # hide\n\nto create a reaction where species X moves at a constant rate D between adjacent compartments (in the ODE this creates terms Dcdot X1_i, where X1_i is the concentration of X1 in compartment i). Transport reactions may have rates depending on several parameters. E.g. to model a system with two species X1 and X2, where both species are transported at a rate which depends both on the species, but also on some non-uniform parameter which is unique to each connection (e.g. representing the area connecting two cells in a tissue) we could do:\n\ndr_X1 = @transport_reaction D1*a X1\ndr_X2 = @transport_reaction D2*a X2\nnothing # hide\n\nnote: Note\nAny species which occurs is occurs in a transport reaction that is used to construct a DiscreteSpaceReactionSystem must also occur in the corresponding non-spatial ReactionSystem.","category":"section"},{"location":"spatial_modelling/discrete_space_reaction_systems/#spatial_dspace_modelling_intro_spatial_reactions_programmatic","page":"Introduction to Spatial Modelling with Catalyst","title":"Creating transport reactions programmatically","text":"If models are created programmatically it is also possible to create transportation reactions programmatically. To do so, use the TransportReaction constructor, providing first the rate and then the transported species:\n\n@variables t\n@species X1(t) X2(t)\n@parameters k1 k2 D [edgeparameter=true]\ntr_X = TransportReaction(D, X1)\nnothing # hide\n\nNote that in this example, we specifically designate D as an edge parameter.","category":"section"},{"location":"spatial_modelling/discrete_space_reaction_systems/#spatial_dspace_modelling_intro_dspaces","page":"Introduction to Spatial Modelling with Catalyst","title":"Defining discrete spatial domains","text":"Discrete spatial domains can represent:\n\nSystems which are composed of a (finite number of) compartments, where each compartment can be considered well-mixed (e.g. can be modelled non-spatially) and where (potentially) species can move between adjacent compartments. Tissues, where each compartment corresponds to a biological cell, are examples of such systems.\nSystems that are continuous in nature, but have been approximated as a discrete domain. Future Catalyst updates will include the ability for the definition, and automatic discretisation, of continuous domains. Currently, however, the user has to perform this discretisation themselves.\n\nCatalyst supports three distinct types of discrete spaces:\n\nCartesian spaces. These are grids where each grid point corresponds to a compartment. Spatial transportation is permitted between adjacent compartments.\nMasked spaces. In these grids, only a subset of the grid point actually corresponds to compartments. Spatial transportation is permitted between adjacent compartments.\nUnstructured (or graph) spaces. These are defined by graphs, where vertices correspond to compartments and edges connect adjacent compartments.\n\nHere, Cartesian spaces are a subset of the masked spaces, which are a subset of the unstructured spaces. If possible, it is advantageous to use as narrow a space definition as possible (this may both improve simulation performance and simplify syntax). Cartesian and masked spaces can be defined as one, two, and three-dimensional. By default, these spaces assume that diagonally neighbouring compartments are non-adjacent (do not permit direct movement of species in between themselves). To change this, provide the diagonally_adjacent = true argument to your DiscreteSpaceReactionSystem when it is created.","category":"section"},{"location":"spatial_modelling/discrete_space_reaction_systems/#spatial_dspace_modelling_intro_dspaces_cartesian","page":"Introduction to Spatial Modelling with Catalyst","title":"Defining Cartesian spaces","text":"A Cartesian space is defined using the CartesianGrid function, which takes a single argument. For a 1d grid, simply provide the length of the grid as a single argument:\n\nusing Catalyst # hide\ncgrid_1d = CartesianGrid(5)\nnothing # hide\n\nFor 2d and 3d grids, we instead provide a Tuple with the length of the grid in each dimension:\n\ncgrid_2d = CartesianGrid((3, 9))\ncgrid_3d = CartesianGrid((2, 4, 8))\nnothing # hide","category":"section"},{"location":"spatial_modelling/discrete_space_reaction_systems/#spatial_dspace_modelling_intro_dspaces_masked","page":"Introduction to Spatial Modelling with Catalyst","title":"Defining masked spaces","text":"Masked spaces are defined through 1d, 2d, or 3d Boolean arrays. Each position in the array is true if it corresponds to a compartment, and false if it does not. E.g. to define a 1d grid corresponding to two disjoint sets, each with 3 compartments, use:\n\nrgrid_1d = [true, true, true, false, true, true, true]\nnothing # hide\n\nTo define a 2d grid corresponding to the shape of an (laying) \"8\", we can use:\n\nrgrid_2d = [\n    true  true  true  true  true;\n    true  false true  false true;\n    true  true  true  true  true\n]\nnothing # hide\n\nFinally, a 4x5x6 3d grid of randomly distributed compartments can be created using:\n\nrgrid_3d = rand([true, false], 4, 5, 6)\nnothing # hide","category":"section"},{"location":"spatial_modelling/discrete_space_reaction_systems/#spatial_dspace_modelling_intro_dspaces_graph","page":"Introduction to Spatial Modelling with Catalyst","title":"Defining unstructured spaces","text":"To define unstructured spaces, we must first import the Graphs.jl package. Next, we can either use some pre-defined formula for building graphs, or build a graph from scratch. Here we create a cyclic graph (where each compartment is connected to exactly two other compartments):\n\nusing Graphs\ncycle_graph(7)\nnothing # hide\n\nSince graphs can represent any network of connected compartments, they do not have dimensions (like Cartesian or masked spaces). Another feature of graph spaces is that they can have non-symmetric connections, i.e. pairs of compartments where spatial movement of species is only permitted in one direction (in practice, this can be done for Cartesian and masked spaces as well, by defining non-uniform spatial rates and setting them to zero in one direction). This can be done by using a directed graph as input. E.g. here we define a directed cyclic graph, where movement is only allowed in one direction of the cycle:\n\ncycle_digraph(7)\nnothing # hide","category":"section"},{"location":"spatial_modelling/discrete_space_reaction_systems/#spatial_dspace_modelling_intro_simulation_inputs","page":"Introduction to Spatial Modelling with Catalyst","title":"Non-uniform initial conditions and parameter values","text":"For spatial models, initial conditions and parameter values are provided similarly as for non-spatial models. Wherever a single value is provided, it is used uniformly across the space. E.g. if we, for our previous two-state model, set\n\nu0 = [:X1 => 1.0, :X2 => 2.0]\nps = [:k1 => 1.0, :k2 => 2.0, :D => 0.2]\nnothing # hide\n\nThe initial condition will be 10 for X1 across compartments, and 20 for X2. Furthermore, for each simulation, in each compartment, the value of 10 will be used for k1 and 20 for k2. Finally, the transportation rate of X1 (set by the parameter D) will be 02 across all connections. To set non-uniform values, non-scalar values must be provided in the map. How to do this depends on how the space was defined. Furthermore, some parameters that are part of spatial reactions may have their value tied to connections between compartments, rather than compartments (we call these edge parameters). These are handled slightly differently. How to designate parameters as either edge parameters or compartment parameters is described here.\n\nBelow we describe how to set non-uniform values in the various cases.","category":"section"},{"location":"spatial_modelling/discrete_space_reaction_systems/#spatial_dspace_modelling_intro_simulation_inputs_cartesian","page":"Introduction to Spatial Modelling with Catalyst","title":"Non-uniform compartment values for Cartesian spaces","text":"To provide non-uniform values across a Cartesian space, simply provide the values in an array of the same dimension and size as the Cartesian space. E.g. for a 5x10 Cartesian space:\n\nusing Catalyst # hide\nccart_dspaces = CartesianGrid((5, 10))\nnothing # hide\n\nrandom values (uniformly distributed between 0 and 1) can be provided using\n\n[:X1 => rand(5, 10), :X2 => 10.0]\nnothing # hide\n\nNon-uniform values for parameters (which values are tied to compartments) are provided similarly.","category":"section"},{"location":"spatial_modelling/discrete_space_reaction_systems/#spatial_dspace_modelling_intro_simulation_inputs_masked","page":"Introduction to Spatial Modelling with Catalyst","title":"Non-uniform compartment values for masked spaces","text":"Non-uniform values for masked spaces are provided in the same manner as for Cartesian spaces (however, values at coordinates that do not hold compartments are ignored). E.g. To provide random values for a masked space contained within a 5x10 Cartesian space we can again set:\n\n[:X1 => rand(5, 10), :X2 => 10.0]\nnothing # hide\n\nIf we want, it is also possible to provide the values as a sparse array with values only in the coordinates that corresponds to compartments.","category":"section"},{"location":"spatial_modelling/discrete_space_reaction_systems/#spatial_dspace_modelling_intro_simulation_inputs_graphs","page":"Introduction to Spatial Modelling with Catalyst","title":"Non-uniform compartment values for unstructured spaces","text":"In graphs (which are used to represent unstructured spaces) each vertex (i.e. compartment) has a specific index. To set non-uniform values for unstructured spaces, provide a vector where the i'th value corresponds to the value in the compartment with index i in the graph. E.g. for a graph with 5 vertices, where we want X to be zero in all compartments bar one (where it is 10) we use:\n\n[:X1 => [0.0, 0.0, 0.0, 0.0, 1.0], :X2 => 10.0]\nnothing # hide","category":"section"},{"location":"spatial_modelling/discrete_space_reaction_systems/#spatial_dspace_modelling_intro_simulation_inputs_edge_parameters","page":"Introduction to Spatial Modelling with Catalyst","title":"Non-uniform values for edge-parameters","text":"Adjacent compartments are connected by edges (with which compartments are connected by edges being defined by the space). For unstructured spaces, it is possible (if a directed graph was used) to have edges from one compartment to another, but not in the opposite direction. For a space with N compartments, edge values are set by a NxN matrix, where value (ij) corresponds to the parameter's values in the edge going from compartment i to compartment j. This matrix can be either sparse or non-sparse. In the latter cases, values corresponding to non-existing edges are ignored. \n\nLet's consider a 1d Cartesian space with 4 compartments. Here, an edge parameter's values are provided in a 4x4 matrix. For the Brusselator model described previously, D's value was tied to edges. If we wish to set the value of D to various values between 01 and 04 we can do:\n\nps = [:k1 => 1.0, :k2 => 2.0, \n      :D => [\n        0.0 0.1 0.0 0.0;\n        0.1 0.0 0.2 0.0;\n        0.0 0.2 0.0 0.3;\n        0.0 0.0 0.3 0.0]\n]\nnothing # hide\n\nHere, the value at index ij corresponds to D's value in the edge from compartment i to compartment j. 0.0 is used for elements that do not correspond to an edge. The make_edge_p_values and make_directed_edge_values provide convenient interfaces for generating non-uniform edge parameter values.","category":"section"},{"location":"spatial_modelling/discrete_space_reaction_systems/#spatial_dspace_modelling_intro_simulation_edge_parameters","page":"Introduction to Spatial Modelling with Catalyst","title":"Edge parameters and compartment parameters","text":"Parameters can be divided into edge parameters and compartment parameters (initial condition values are always tied to compartments). Here, edge parameters have their values tied to edges, while compartment parameters have their values tied to compartments. All parameters that are part of the rates (or stoichiometries) of non-spatial reactions must be compartment parameters. Parameters that are part of spatial reactions can be either compartment parameters or edge parameters. When a spatial reaction's rate is computed, edge parameters fetch their values for from the edge of the transition, and compartment parameters from the compartment from which the edge originates.\n\nWhen a DiscreteSpaceReactionSystem is created, its parameters is the union of all parameters occurring in the (non-spatial) ReactionSystem and in all spatial reactions. By default, parameters occurring only in spatial reactions are considered edge parameters (and if they occur in the non-spatial ReactionSystem they are considered compartment parameters). It is, however, possible to designate a parameter specifically as an edge parameter (or not), by using the edgeparameter metadata. E.g. to designate that D (when declared in a non-spatial ReactionSystem using the DSL) is an edge parameter, not a compartment parameter, we use:\n\nusing Catalyst # hide\ntwo_state_model = @reaction_network begin\n    @parameters D [edgeparameter=true]\n    (k1,k2), X1 <--> X2\nend\nnothing # hide\n\nTo learn the compartment and edge parameters of a DiscreteSpaceReactionSystem, the vertex_parameters and edge_parameters functions can be used:\n\ntwo_state_model = @reaction_network begin\n    (k1,k2), X1 <--> X2\nend\ndiffusion_rx = @transport_reaction D X1\nspace = CartesianGrid((20,20))\ndsrs = DiscreteSpaceReactionSystem(two_state_model, [diffusion_rx], space)\nedge_parameters(dsrs)","category":"section"},{"location":"spatial_modelling/discrete_space_reaction_systems/#spatial_dspace_modelling_intro_limitations","page":"Introduction to Spatial Modelling with Catalyst","title":"Spatial modelling limitations","text":"Many features which are supported for non-spatial ReactionSystems are currently unsupported for DiscreteSpaceReactionSystems. This includes observables, algebraic and differential equations, hierarchical models, and events. It is possible that these features will be supported in the future. Furthermore, removal of conserved quantities is not supported when creating spatial ODEProblems.\n\nIf you are using Catalyst's features for spatial modelling, please give us feedback on how we can improve these features. Additionally, just letting us know that you use these features is useful, as it helps inform us whether continued development of spatial modelling features is worthwhile. ","category":"section"},{"location":"model_simulation/simulation_structure_interfacing/#simulation_structure_interfacing","page":"Interfacing Problems, Integrators, and Solutions","title":"Interfacing Problems, Integrators, and Solutions","text":"<details><summary><strong>Environment setup and package installation</strong></summary>\n\nThe following code sets up an environment for running the code on this page.\n\nusing Pkg\nPkg.activate(; temp = true) # Creates a temporary environment, which is deleted when the Julia session ends.\nPkg.add(\"Catalyst\")\nPkg.add(\"OrdinaryDiffEqDefault\")\nPkg.add(\"SymbolicIndexingInterface\")\n\n</details>\n\n<details><summary><strong>Quick-start example</strong></summary>\n\nVariable and species values stored in e.g. ODEProblems and solution objects can be easily accessed, e.g. as in the following simple example:\n\nusing Catalyst, OrdinaryDiffEqDefault\n\n# First we create a `ODEProblem` and find its solution.\nrs = @reaction_network begin\n    (p,d), 0 <--> X\nend\noprob = ODEProblem(bd_model, [:X => 0.5], 5.0, [:p => 1.0, :d => 0.2])\nsol = solve(oprob)\n\n# Accessing species values:\noprob[:X] # Returns initial condition value of `X`.\nsol[:X] # Returns `X`'s value across the simulation.\nsol(0:5; idxs = :X) # Returns `X`'s value sampled at time points 0, 1, 2, 3, 4, 5.\n\n# Accessing parameter values requires adding `.ps` to the call:\noprob.ps[:p]\nsol.ps[:p]\n\n# Problems can have their values updated using the `remake` function:\noprob_new = remake(oprob; u0 = [:X => 0.1], p = [:d => 2.0])\n# Undesignated values uses the values of input problem.\n\n</details>\n\n\n   When simulating a model, one begins with creating a problem. Next, a simulation is performed on the problem, during which the simulation's state is recorded through an integrator. Finally, the simulation output is returned as a solution. This tutorial describes how to access (or modify) the state (or parameter) values of problem, integrator, and solution structures.\n\nGenerally, when we have a structure simulation_struct and want to interface with the unknown (or parameter) x, we use simulation_struct[:x] to access the value. For situations where a value is accessed (or changed) a large number of times, it can improve performance to first create a specialised getter/setter function.","category":"section"},{"location":"model_simulation/simulation_structure_interfacing/#simulation_structure_interfacing_problems","page":"Interfacing Problems, Integrators, and Solutions","title":"Interfacing problem objects","text":"We begin by demonstrating how we can interface with problem objects. First, we create an ODEProblem representation of a chemical cross-coupling model (where a catalyst, C, couples two substrates, S₁ and S₂, to form a product, P).\n\nusing Catalyst\ncc_system = @reaction_network begin\n    k₁, S₁ + C --> S₁C\n    k₂, S₁C + S₂ --> CP\n    k₃, CP --> C + P\nend\n\nu0 = [:S₁ => 1.0, :C => 0.05, :S₂ => 1.2, :S₁C => 0.0, :CP => 0.0, :P => 0.0]\ntspan = (0., 10.0)\nps = [:k₁ => 5.0, :k₂ => 5.0, :k₃ => 100.0] \noprob = ODEProblem(cc_system, u0, tspan, ps)\nnothing    # hide\n\nWe can find a species's (or variable's) initial condition value by simply indexing with the species of interest as input. Here we check the initial condition value of C:\n\noprob[:C]\n\nAn almost identical notation can be used for parameters, however, here we use oprob.ps (rather than oprob): with the notation being identical for parameters:\n\noprob.ps[:k₁]\n\nTo retrieve several species initial condition (or parameter) values, simply give a vector input. Here we check the values of the two substrates (S₁ and S₂):\n\noprob[[:S₁, :S₂]]\n\nA problem's time span can be accessed through the tspan field:\n\noprob.tspan\n\nnote: Note\nHere we have used an ODEProblemto demonstrate all interfacing functionality. However, identical workflows work for the other problem types.","category":"section"},{"location":"model_simulation/simulation_structure_interfacing/#simulation_structure_interfacing_problems_remake","page":"Interfacing Problems, Integrators, and Solutions","title":"Remaking problems using the remake function","text":"To modify a problem, the remake function should be used. It takes an already created problem, and returns a new, updated, one (the input problem is unchanged). The remake function takes the following inputs:\n\nThe problem that it remakes.\n(optionally) u0: A vector with initial conditions that should be updated. The vector takes the same form as normal initial condition vectors, but does not need to be complete (in which case only a subset of the initial conditions are updated).\n(optionally) tspan: An updated time span (using the same format as time spans normally are given in).\n(optionally) p: A vector with parameters that should be updated. The vector takes the same form as normal parameter vectors, but does not need to be complete (in which case only a subset of the parameters are updated).\n\nHere we modify our problem to increase the initial condition concentrations of the two substrates (S₁ and S₂), and also confirm that the new problem is different from the old (unchanged) one:\n\nusing OrdinaryDiffEqDefault\noprob_new = remake(oprob; u0 = [:S₁ => 5.0, :S₂ => 2.5])\noprob_new != oprob\n\nHere, we instead use remake to simultaneously update all three fields:\n\noprob_new_2 = remake(oprob; u0 = [:C => 0.2], tspan = (0.0, 20.0), p = [:k₁ => 2.0, :k₂ => 2.0])\nnothing # hide\n\nTypically, when using remake to update a problem, the common workflow is to overwrite the old one with the output. E.g. to set the value of k₁ to 5.0 in oprob, you would do:\n\noprob = remake(oprob; p = [:k₁ => 5.0])\nnothing # hide","category":"section"},{"location":"model_simulation/simulation_structure_interfacing/#simulation_structure_interfacing_integrators","page":"Interfacing Problems, Integrators, and Solutions","title":"Interfacing integrator objects","text":"During a simulation, the solution is stored in an integrator object. Here, we will describe how to interface with these. The almost exclusive circumstance when integrator-interfacing is relevant is when simulation events are implemented through callbacks. However, to demonstrate integrator indexing in this tutorial, we will create one through the init function (while circumstances where one might want to use init function exist, since integrators are automatically created during simulations, these are rare).\n\nintegrator = init(oprob)\nnothing # hide\n\nWe can interface with our integrator using an identical syntax as was used for problems. The primary exception is that there is no remake function for integrators. Instead, we can update species and parameter values using normal indexing. Here we update, and then check the values of, first the species C and then the parameter k₁:\n\nintegrator[:C] = 0.0\nintegrator[:C]\n\nor a parameter:\n\nintegrator.ps[:k₂] = 1.0\nintegrator.ps[:k₂]\n\nNote that here, species-interfacing yields (or changes) a simulation's current value for a species, not its initial condition.\n\nIf you are interfacing with jump simulation integrators, you must always call reset_aggregated_jumps!(integrator) afterwards.","category":"section"},{"location":"model_simulation/simulation_structure_interfacing/#simulation_structure_interfacing_solutions","page":"Interfacing Problems, Integrators, and Solutions","title":"Interfacing solution objects","text":"Finally, we consider solution objects. First, we simulate our problem:\n\nsol = solve(oprob)\nnothing # hide\n\nNext, we can access the simulation's values using the same notation as previously. When we access a species's, its values across the full simulation is returned as a vector:\n\nsol[:P]\n\nParameter values can also be accessed (however, here we only get a single value):\n\nsol.ps[:k₃]\n\nUnlike for problems and integrators, species or parameter values of solutions cannot be changed. \n\nA vector with the time values for all simulation time steps can be retrieved using\n\nsol.t\n\nTo find simulation values at a specific time point, simply use this time point as input to your solution object (treating it as a function). I.e. here we get our simulation's values at time t = 10\n\nsol(1.0)\n\nThis works whenever the simulations actually stopped at time t = 10 (if not, an interpolated value is returned). To get the simulation's values for a specific subset of species, we can use the idxs optional argument. I.e. here we get the value of C at time t = 10\n\nsol(1.0; idxs = [:C])","category":"section"},{"location":"model_simulation/simulation_structure_interfacing/#simulation_structure_interfacing_functions","page":"Interfacing Problems, Integrators, and Solutions","title":"Interfacing using specialised getter/setter functions","text":"Internally, species and parameter values are stored in vectors. Whenever e.g. oprob[:C] is called, Julia must first find which index in the storage vector C is stored in. Next, its value can be retrieved. If oprob[:C] is called a large number of times, this index must be found in each call. If a large number of such accesses are carried out, and performance is essential, it can be worthwhile to pre-compute a function to carry this out.\n\nThere exist four different functions, all implemented by the SymbolicIndexingInterface.jl, each returning a function for performing a specific type of interfacing:\n\ngetu: For accessing species values.\ngetp: For accessing parameter values.\nsetu: For changing species values.\nsetp: For changing parameter values.\n\nFor each species (or parameter) we wish to interface with, a new interfacing function must be created. Here we first creates a function for retrieving the value of C, and then use it for this purpose:\n\nusing SymbolicIndexingInterface\nget_C = getu(oprob, :C)\nget_C(oprob)\n\nHere, getu (as well as getp, setu, and setp) first takes the structure we wish to interface with, and then the target quantity. When using setu and setp, in the second step, we must also provide the update value:\n\nset_C = setu(oprob, :C)\nset_C(oprob, 0.2)\nget_C(oprob)\n\nLike when indexing-based interfacing is used, these functions also work with vectors:\n\nget_S = getu(oprob, [:S₁, :S₂])\nget_S(oprob)","category":"section"},{"location":"model_simulation/simulation_structure_interfacing/#simulation_structure_interfacing_symbolic_representation","page":"Interfacing Problems, Integrators, and Solutions","title":"Interfacing using symbolic representations","text":"When e.g. programmatic modelling is used, species and parameters can be represented as symbolic variables. These can be used to index a problem, just like symbol-based representations can. Here we create a simple two-state model programmatically, and use its symbolic variables to check, and update, an initial condition:\n\nusing Catalyst, OrdinaryDiffEqDefault\nt = default_t()\n@species X1(t) X2(t)\n@parameters k1 k2\nrxs = [\n    Reaction(k1, [X1], [X2]),\n    Reaction(k2, [X2], [X1])\n]\n@named two_state_model = ReactionSystem(rxs, t)\ntwo_state_model = complete(two_state_model)\n\nu0 = [X1 => 2.0, X2 => 0.0]\ntspan = (0.0, 1.0)\nps = [k1 => 1.0, k2 => 2.0]\noprob = ODEProblem(two_state_model, u0, tspan, ps)\n\noprob = remake(oprob; u0 = [X1 => 5.0])\noprob[X1]\n\nSymbolic variables can be used to access or update species or parameters for all the cases when Symbols can (including when using remake or e.g. getu).\n\nAn advantage when quantities are represented as symbolic variables is that symbolic expressions can be formed and used to index a structure. E.g. here we check the combined initial concentration of X (X1 + X2) in our two-state problem:\n\noprob[X1 + X2]\n\nJust like symbolic variables can be used to directly interface with a structure, symbolic variables stored in ReactionSystem models can be used:\n\noprob[two_state_model.X1 + two_state_model.X2]\n\nThis can be used to form symbolic expressions using model quantities when a model has been created using the DSL (as an alternative to @unpack). Alternatively, creating an observable, and then interface using its Symbol representation, is also possible.\n\nwarning: Warning\nWhen accessing a simulation structure using symbolic variables from a ReactionSystem model, such as rn.A for rn a ReactionSystem and A a species within it, ensure that the model is complete.","category":"section"},{"location":"model_creation/conservation_laws/#conservation_laws","page":"Working with Conservation Laws","title":"Working with Conservation Laws","text":"<details><summary><strong>Environment setup and package installation</strong></summary>\n\nThe following code sets up an environment for running the code on this page.\n\nusing Pkg\nPkg.activate(; temp = true) # Creates a temporary environment, which is deleted when the Julia session ends.\nPkg.add(\"Catalyst\")\nPkg.add(\"Latexify\")\nPkg.add(\"OrdinaryDiffEqDefault\")\nPkg.add(\"Plots\")\n\n</details>\n\n\n Catalyst can detect, and eliminate for differential-equation based models, conserved quantities, i.e. linear combinations of species which are conserved via their chemistry. This functionality is both automatically utilised by Catalyst (e.g. to remove singular Jacobians during steady state computations), but is also available for users to utilise directly (e.g. to potentially improve simulation performance).\n\nTo illustrate conserved quantities, let us consider the following two-state model:\n\nusing Catalyst\nrs = @reaction_network begin\n (k₁,k₂), X₁ <--> X₂\nend\n\nIf we simulate it, we note that while the concentrations of X₁ and X₂ change throughout the simulation, the total concentration of X (= X₁ + X₂) is constant:\n\nusing OrdinaryDiffEqDefault, Plots\nu0 = [:X₁ => 80.0, :X₂ => 20.0]\nps = [:k₁ => 10.0, :k₂ => 2.0]\noprob = ODEProblem(rs, u0, (0.0, 1.0), ps)\nsol = solve(oprob)\nplot(sol; idxs = [rs.X₁, rs.X₂, rs.X₁ + rs.X₂], label = [\"X₁\" \"X₂\" \"X₁ + X₂ (a conserved quantity)\"])\n\nThis makes sense, as while X is converted between two different forms (X₁ and X₂), it is neither produced nor degraded. That is, it is a conserved quantity. Next, if we consider the ODE that our model generates:\n\nusing Latexify\nlatexify(rs; form = :ode)\nlatexify(rs; form = :ode, math_delimiters = true) # hide\n\nwe note that it essentially generates the same equation twice (i.e. fracdX₁(t)dt = -fracdX₂(t)dt). By designating our conserved quantity X₁ + X₂ = Γ, we can rewrite our differential equation model as a differential-algebraic equation (with a single differential equation and a single algebraic equation):\n\nfracdX₁(t)dt = - k₁X₁(t) + k₂(-X₁(t) + Γ) \nX₂(t) = -X₁(t) + Γ\n\nUsing Catalyst, it is possible to detect any such conserved quantities and eliminate them from the system. Here, when we convert our ReactionSystem to an ODE System, we provide the remove_conserved = true argument to instruct Catalyst to perform this elimination:\n\nosys = ode_model(rs; remove_conserved = true)\nlatexify(Catalyst.system_to_reactionsystem(osys; disable_forbidden_symbol_check = true); math_delimiters = true) # hide\n\nWe note that the output system only contains a single (differential) equation and can hence be solved with an ODE solver. The second (algebraic) equation is stored as an observable, and can be retrieved using the observed function:\n\nobserved(osys)\n\nUsing the unknowns function we can confirm that the ODE only has a single unknown variable:\n\nunknowns(osys)\n\nNext, using parameters we note that an additional (vector) parameter, Γ has been added to the system:\n\nparameters(osys)\n\nHere, Catalyst encodes all conserved quantities in a single, vector-valued, parameter called Γ.\n\nPractically, the remove_conserved = true argument can be provided when a ReactionSystem is converted to an ODEProblem:\n\nusing OrdinaryDiffEqDefault, Plots\nu0 = [:X₁ => 80.0, :X₂ => 20.0]\nps = [:k₁ => 10.0, :k₂ => 2.0]\noprob = ODEProblem(rs, u0, (0.0, 1.0), ps; remove_conserved = true)\nnothing # hide\n\nHere, while Γ becomes a parameter of the new system, it has a default value equal to the corresponding conservation law. Hence, its value is computed from the initial condition [:X₁ => 80.0, :X₂ => 20.0], and does not need to be provided in the parameter vector. Next, we can simulate and plot our model using normal syntax:\n\nsol = solve(oprob)\nplot(sol)\n\nnote: Note\nAny species eliminated using remove_conserved = true will not be plotted by default. However, it can be added to the plot using the idxs plotting option. E.g. here we would use plot(sol; idxs = [:X₁, :X₂]) to plot both species.\n\nWhile X₂ is an observable (and not unknown) of the ODE, we can access it just like if remove_conserved = true had not been used:\n\nsol[:X₂]\n\nnote: Note\nGenerally, remove_conserved = true should not change any modelling workflows. I.e. anything that works without this option should also work when an ODEProblem is created using remove_conserved = true.\n\nnote: Note\nThe remove_conserved = true option is available when creating SDEProblems, NonlinearProblems, and SteadyStateProblems (and their corresponding systems). However, it cannot be used when creating JumpProblems.","category":"section"},{"location":"model_creation/conservation_laws/#conservation_laws_accessors","page":"Working with Conservation Laws","title":"Conservation law accessor functions","text":"For any given ReactionSystem model, we can use conservationlaw_constants to compute all of a system's conserved quantities:\n\nconservationlaw_constants(rs)\n\nNext, the conservedequations function can be used to compute the algebraic equations produced when a system's conserved quantities are eliminated:\n\nconservedequations(rs)\n\nFinally, the conservationlaws function yields a m times n matrix, where n is a system's number of species, m its number of conservation laws, and element (ij) corresponds to the copy number of the jth species that occurs in the ith conserved quantity:\n\nconservationlaws(rs)\n\nI.e. in this case we have a single conserved quantity, which contains a single copy each of the system's two species.","category":"section"},{"location":"model_creation/conservation_laws/#conservation_laws_prob_updating","page":"Working with Conservation Laws","title":"Updating conservation law values directly","text":"Previously we noted that conservation law elimination adds the conservation law as a system parameter (named Γ). E.g. here we find it among the parameters of the ODE model generated by the previously used two-state system:\n\nparameters(ode_model(rs; remove_conserved = true))\n\nAn advantage of this is that we can set conserved quantities's values directly in simulations. Here we simulate the model while specifying that X₁ + X₂ = 100 (and while also specifying an initial condition for X₁, but none for X₂).\n\nu0 = [:X₁ => 6.0]\nps = [:k₁ => 1.0, :k₂ => 2.0, :Γ => [10.0]]\noprob = ODEProblem(rs, u0, 1.0, ps; remove_conserved = true)\nsol = solve(oprob)\nnothing # hide\n\nGenerally, for each conservation law, one can omit specifying either the conservation law constant, or one initial condition (whichever quantity is missing will be computed from the other ones). \n\nnote: Note\nAs previously mentioned, the conservation law parameter Γ is a vector-valued parameter with one value for each conservation law. That is why we above provide its value as a vector (:Γ => [5.0]). If there had been multiple conservation laws, we would provide the :Γ vector with one value for each of them (e.g. :Γ => [10.0, 15.0]).\n\nwarn: Warn\nIf you specify the value of a conservation law parameter, you must not specify the value of all species of that conservation law (this can result in an error). Instead, the value of exactly one species must be left unspecified.\n\nJust like when we create a problem, if we update the species (or conservation law parameter) values of oprob, the remaining ones will be recomputed to generate an accurate conservation law.\n\nu0 = [:X₁ => 6.0, :X₂ => 4.0]\nps = [:k₁ => 1.0, :k₂ => 2.0]\noprob = ODEProblem(rs, u0, 10.0, ps; remove_conserved = true)\noprob = remake(oprob; u0 = [:X₁ => 16.0])\n\nwarn: Warn\nIf a problem with conservation laws have had values related to that conservation law updated using remake, the values stored within the problem will no longer appear correct. I.e. oprob[:X₂] will not necessarily yield the correct value. The correct values are, however, computed correctly during solve, and it is only when checking the content of the problem that erroneous values appears.\n\nIt is also possible to update the value of Γ. Here, as X₂ is the species eliminated by the conservation law (which can be checked using conservedequations), X₂'s value will be modified to ensure that Γ's new value is correct. This, however, also requires designating X₂ = nothing\n\noprob = remake(oprob; u0 = [:X₂ => nothing], p = [:Γ => [30.0]] )\noprob[:X₂]\n\nGenerally, for a conservation law where we have:\n\nOne (or more) non-eliminated species.\nOne eliminated species.\nA conservation law parameter.\n\nIf the value of the conservation law parameter Γ's value has never been specified, its value will be updated to accommodate changes in species values. If the conservation law parameter (Γ)'s value has been specified (either when the ODEProblem was created, or using remake), then the eliminated species's value will be updated to accommodate changes in the conservation law parameter or non-eliminated species's values. Furthermore, in this case, the value of the eliminated species cannot be updated. \n\nwarn: Warn\nWhen updating the values of problems with conservation laws it is additionally important to use remake (as opposed to direct indexing, e.g. setting oprob[:X₁] = 16.0).","category":"section"},{"location":"model_creation/conservation_laws/#conservation_laws_prob_updating_symvar","page":"Working with Conservation Laws","title":"Extracting the conservation law parameter's symbolic variable","text":"Catalyst represents its models using the Symbolics.jl computer algebraic system, something which allows the user to form symbolic expressions of model components. If you wish to extract and use the symbolic variable corresponding to a model's conserved quantities, you can use the following syntax:\n\nΓ = Catalyst.get_networkproperties(rs).conservedconst","category":"section"},{"location":"model_creation/examples/hodgkin_huxley_equation/#hodgkin_huxley_equation","page":"Hodgkin-Huxley Equation","title":"Hodgkin-Huxley Equation","text":"<details><summary><strong>Environment setup and package installation</strong></summary>\n\nThe following code sets up an environment for running the code on this page.\n\nusing Pkg\nPkg.activate(; temp = true) # Creates a temporary environment, which is deleted when the Julia session ends.\nPkg.add(\"Catalyst\")\nPkg.add(\"ModelingToolkitBase\")\nPkg.add(\"NonlinearSolveFirstOrder\")\nPkg.add(\"OrdinaryDiffEqRosenbrock\")\nPkg.add(\"Plots\")\n\n</details>\n\n\n This tutorial shows how to construct a Catalyst ReactionSystem that includes a coupled ODE, corresponding to the Hodgkin–Huxley model for an excitable cell. The Hodgkin–Huxley model is a mathematical model that describes how action potentials in neurons are initiated and propagated. It is a continuous-time dynamical system given by a coupled system of nonlinear differential equations that model the electrical characteristics of excitable cells such as neurons and muscle cells.\n\nWe'll present two different approaches for constructing the model. The first will show how it can be built entirely within a single DSL model, while the second illustrates another work flow, showing how separate models containing the chemistry and the dynamics of the transmembrane potential can be combined into a complete model.\n\nWe begin by importing some necessary packages:\n\nusing ModelingToolkitBase, Catalyst, NonlinearSolveFirstOrder, Plots, OrdinaryDiffEqRosenbrock","category":"section"},{"location":"model_creation/examples/hodgkin_huxley_equation/#Building-the-model-via-the-Catalyst-DSL","page":"Hodgkin-Huxley Equation","title":"Building the model via the Catalyst DSL","text":"Let's build a simple Hodgkin-Huxley model for a single neuron, with the voltage, V(t), included as a coupled ODE. We first specify the transition rates for three gating variables, m(t), n(t) and h(t).\n\ns xleftrightarrowbeta_s(V(t))alpha_s(V(t)) s quad s in mnh\n\nHere each gating variable is used in determining the fraction of active (i.e. open) or inactive (m = 1 - m, n = 1 -n, h = 1 - h) sodium (m and h) and potassium (n) channels.\n\nThe transition rate functions, which depend on the voltage, V(t), are given by\n\nfunction αₘ(V)\n    theta = (V + 45) / 10\n    ifelse(theta == 0.0, 1.0, theta/(1 - exp(-theta)))\nend\nβₘ(V) = 4*exp(-(V + 70)/18)\n\nαₕ(V) = .07 * exp(-(V + 70)/20)\nβₕ(V) = 1/(1 + exp(-(V + 40)/10))\n\nfunction αₙ(V)\n    theta = (V + 60) / 10\n    ifelse(theta == 0.0, .1, .1*theta / (1 - exp(-theta)))\nend\nβₙ(V) = .125 * exp(-(V + 70)/80)\nnothing # hide\n\nWe also declare a function to represent an applied current in our model, which we will use to perturb the system and create action potentials. \n\nIapp(t,I₀) = I₀ * sin(2*pi*t/30)^2\n\nWe now declare a ReactionSystem that encompasses the Hodgkin-Huxley model. Note, we will also give the (default) values for our parameters as part of constructing the model to avoid having to specify them later on via parameter maps.\n\nhhmodel = @reaction_network hhmodel begin\n    @parameters begin\n        C = 1.0 \n        ḡNa = 120.0 \n        ḡK = 36.0 \n        ḡL = .3 \n        ENa = 45.0 \n        EK = -82.0 \n        EL = -59.0 \n        I₀ = 0.0\n    end\n\n    @variables V(t)\n\n    (αₙ(V), βₙ(V)), n′ <--> n\n    (αₘ(V), βₘ(V)), m′ <--> m\n    (αₕ(V), βₕ(V)), h′ <--> h\n    \n    @equations begin\n        D(V) ~ -1/C * (ḡK*n^4*(V-EK) + ḡNa*m^3*h*(V-ENa) + ḡL*(V-EL)) + Iapp(t,I₀)\n    end\nend\n\nFor now we turn off the applied current by setting its amplitude, I₀, to zero.\n\nhhmodel is now a ReactionSystem that is coupled to an internal constraint ODE for dVdt. Let's now solve to steady-state, as we can then use these resting values as an initial condition before applying a current to create an action potential.\n\ntspan = (0.0, 50.0)\nu₀ = [:V => -70, :m => 0.0, :h => 0.0, :n => 0.0,\n\t  :m′ => 1.0, :n′ => 1.0, :h′ => 1.0]\noprob = ODEProblem(hhmodel, u₀, tspan)\nhhsssol = solve(oprob, Rosenbrock23())\nnothing # hide\n\nFrom the artificial initial condition we specified, the solution approaches the physiological steady-state via firing one action potential:\n\nplot(hhsssol, idxs = hhmodel.V)\n\nWe now save this steady-state to use as the initial condition for simulating how a resting neuron responds to an applied current. We save the steady-state values as a mapping from the symbolic variables to their steady-states that we can later use as an initial condition:\n\nu_ss = unknowns(hhmodel) .=> hhsssol(tspan[2], idxs = unknowns(hhmodel))\nnothing # hide\n\nFinally, starting from this resting state let's solve the system when the amplitude of the stimulus is non-zero and see if we get action potentials\n\ntspan = (0.0, 50.0)\n@unpack V,I₀ = hhmodel\noprob = ODEProblem(hhmodel, u_ss, tspan, [I₀ => 10.0])\nsol = solve(oprob)\nplot(sol, idxs = V, legend = :outerright)\n\nWe observe three action potentials due to the steady applied current.","category":"section"},{"location":"model_creation/examples/hodgkin_huxley_equation/#Building-the-model-via-composition-of-separate-systems-for-the-ion-channel-and-transmembrane-voltage-dynamics","page":"Hodgkin-Huxley Equation","title":"Building the model via composition of separate systems for the ion channel and transmembrane voltage dynamics","text":"As an illustration of how one can construct models from individual components, we now separately construct and compose the model components.\n\nWe start by defining systems to model each ionic current. Note we now use @network_component instead of @reaction_network as we want the models to be composable and not marked as finalized.\n\nIKmodel = @network_component IKmodel begin\n    @parameters ḡK = 36.0 EK = -82.0 \n    @variables V(t) Iₖ(t)\n    (αₙ(V), βₙ(V)), n′ <--> n\n    @equations Iₖ ~ ḡK*n^4*(V-EK)\nend\n\nINamodel = @network_component INamodel begin\n    @parameters ḡNa = 120.0 ENa = 45.0 \n    @variables V(t) Iₙₐ(t)\n    (αₘ(V), βₘ(V)), m′ <--> m\n    (αₕ(V), βₕ(V)), h′ <--> h\n    @equations Iₙₐ ~ ḡNa*m^3*h*(V-ENa) \nend\n\nILmodel = @network_component ILmodel begin\n    @parameters ḡL = .3 EL = -59.0 \n    @variables V(t) Iₗ(t)\n    @equations Iₗ ~ ḡL*(V-EL)\nend\nnothing # hide\n\nWe next define the voltage dynamics with unspecified values for the currents\n\nhhmodel2 = @network_component hhmodel2 begin\n    @parameters C = 1.0 I₀ = 0.0\n    @variables V(t) Iₖ(t) Iₙₐ(t) Iₗ(t)\n    @equations D(V) ~ -1/C * (Iₖ + Iₙₐ + Iₗ) + Iapp(t,I₀)\nend\nnothing # hide\n\nFinally, we extend the hhmodel with the systems defining the ion channel currents\n\n@named hhmodel2 = extend(IKmodel, hhmodel2)\n@named hhmodel2 = extend(INamodel, hhmodel2)\n@named hhmodel2 = extend(ILmodel, hhmodel2)\nhhmodel2 = complete(hhmodel2)\n\nLet's again solve the system starting from the previously calculated resting state, using the same applied current as above (to verify we get the same figure). Note, we now run structural_simplify from ModelingToolkit to eliminate the algebraic equations for the ionic currents when constructing the ODEProblem:\n\n@unpack I₀,V = hhmodel2\noprob = ODEProblem(hhmodel2, u_ss, tspan, [I₀ => 10.0]; structural_simplify = true)\nsol = solve(oprob)\nplot(sol, idxs = V, legend = :outerright)\n\nWe observe the same solutions as from our original model.","category":"section"},{"location":"model_creation/compositional_modeling/#compositional_modeling","page":"Compositional Modeling of Reaction Systems","title":"Compositional Modeling of Reaction Systems","text":"<details><summary><strong>Environment setup and package installation</strong></summary>\n\nThe following code sets up an environment for running the code on this page.\n\nusing Pkg\nPkg.activate(; temp = true) # Creates a temporary environment, which is deleted when the Julia session ends.\nPkg.add(\"Catalyst\")\nPkg.add(\"GraphRecipes\")\nPkg.add(\"Plots\")\n\n</details>\n\n\n Catalyst supports the construction of models in a compositional fashion, based on ModelingToolkit's subsystem functionality. In this tutorial we'll see how we can construct the earlier repressilator model by composing together three identically repressed genes, and how to use compositional modeling to create compartments.","category":"section"},{"location":"model_creation/compositional_modeling/#completeness_note","page":"Compositional Modeling of Reaction Systems","title":"A note on completeness","text":"Catalyst ReactionSystem can either be complete or incomplete. When created using the @reaction_network DSL they are created as complete. Here, only complete ReactionSystems can be used to create the various problem types (e.g. ODEProblem). However, only incomplete ReactionSystems can be composed using the features described below. Hence, for compositional modeling, ReactionSystem must be created as incomplete, and later set to complete before simulation.\n\nTo create a ReactionSystems for use in compositional modeling via the DSL, simply use the @network_component macro instead of @reaction_network:\n\nusing Catalyst\ndegradation_component = @network_component begin\n  d, X --> 0\nend\n\nAlternatively one can just build the ReactionSystem via the symbolic interface.\n\n@parameters d\nt = default_t()\n@species X(t)\nrx = Reaction(d, [X], nothing)\n@named degradation_component = ReactionSystem([rx], t)\n\nWe can test whether a system is complete using the ModelingToolkit.iscomplete function:\n\nModelingToolkitBase.iscomplete(degradation_component)\n\nTo mark a system as complete, after which it should be considered as representing a finalized model, use the complete function\n\ndegradation_component_complete = complete(degradation_component)\nModelingToolkitBase.iscomplete(degradation_component_complete)","category":"section"},{"location":"model_creation/compositional_modeling/#Compositional-modeling-tooling","page":"Compositional Modeling of Reaction Systems","title":"Compositional modeling tooling","text":"Catalyst supports two ModelingToolkit interfaces for composing multiple ReactionSystems together into a full model. The first mechanism allows for extending an existing system by merging in a second system via the extend command\n\nusing Catalyst\nbasern = @network_component rn1 begin\n  k, A + B --> C\nend\nnewrn = @network_component rn2 begin\n  r, C --> A + B\nend\n@named rn = extend(newrn, basern)\n\nHere we extended basern with newrn giving a system with all the reactions. Note, if a name is not specified via @named or the name keyword then rn will have the same name as newrn.\n\nThe second main compositional modeling tool is the use of subsystems. Suppose we now add to basern two subsystems, newrn and newestrn, we get a different result:\n\nnewestrn = @network_component rn3 begin\n  v, A + D --> 2D\nend\n@named rn = compose(basern, [newrn, newestrn])\n\nHere we have created a new ReactionSystem that adds newrn and newestrn as subsystems of basern. The variables and parameters in the sub-systems are considered distinct from those in other systems, and so are namespaced (i.e. prefaced) by the name of the system they come from.\n\nWe can see the subsystems of a given system by\n\nModelingToolkitBase.get_systems(rn)\n\nThey naturally form a tree-like structure\n\nusing Plots, GraphRecipes\nplot(TreePlot(rn), method=:tree, fontsize=12, nodeshape=:ellipse)\n\n(Image: rn network with subsystems)\n\nWe could also have directly constructed rn using the same reaction as in basern as\n\nt = default_t()\n@parameters k\n@species A(t), B(t), C(t)\nrxs = [Reaction(k, [A,B], [C])]\n@named rn = ReactionSystem(rxs, t; systems = [newrn, newestrn])\n\nCatalyst provides several different accessors for getting information from a single system, or all systems in the tree. To get the species, parameters, and reactions only within a given system (i.e. ignoring subsystems), we can use\n\nCatalyst.get_species(rn)\n\nCatalyst.get_ps(rn)\n\nCatalyst.get_rxs(rn)\n\nTo see all the species, parameters and reactions in the tree we can use\n\nspecies(rn)   # or unknowns(rn)\n\nparameters(rn)\n\nreactions(rn)   # or equations(rn)\n\nIf we want to collapse rn down to a single system with no subsystems we can use\n\nflatrn = Catalyst.flatten(rn)\n\nwhere\n\nModelingToolkitBase.get_systems(flatrn)\n\nMore about ModelingToolkit's interface for compositional modeling can be found in the ModelingToolkit docs.","category":"section"},{"location":"model_creation/compositional_modeling/#Compositional-model-of-the-repressilator","page":"Compositional Modeling of Reaction Systems","title":"Compositional model of the repressilator","text":"Let's apply the tooling we've just seen to create the repressilator in a more modular fashion. We start by defining a function that creates a negatively repressed gene, taking the repressor as input\n\nfunction repressed_gene(; R, name)\n  @network_component $name begin\n    hillr($R,α,K,n), ∅ --> m\n    (δ,γ), m <--> ∅\n    β, m --> m + P\n    μ, P --> ∅\n  end\nend\nnothing # hide\n\nHere we assume the user will pass in the repressor species as a ModelingToolkit variable, and specify a name for the network. We use Catalyst's interpolation ability to substitute the value of these variables into the DSL (see Interpolation of Julia Variables). To make the repressilator we now make three genes, and then compose them together\n\nt = default_t()\n@species G3₊P(t)\n@named G1 = repressed_gene(; R=ParentScope(G3₊P))\n@named G2 = repressed_gene(; R=ParentScope(G1.P))\n@named G3 = repressed_gene(; R=ParentScope(G2.P))\n@named repressilator = ReactionSystem(t; systems=[G1,G2,G3])\n\nNotice, in this system each gene is a child node in the system graph of the repressilator\n\nplot(TreePlot(repressilator), method=:tree, fontsize=12, nodeshape=:ellipse)\n\n(Image: repressilator tree plot)\n\nIn building the repressilator we needed to use two new features. First, we needed to create a symbolic variable that corresponds to the protein produced by the third gene before we created the corresponding system. We did this via @variables G3₊P(t). We also needed to set the scope where each repressor lived. Here ParentScope(G3₊P), ParentScope(G1.P), and ParentScope(G2.P) signal Catalyst that these variables will come from parallel systems in the tree that have the same parent as the system being constructed (in this case the top-level repressilator system).","category":"section"},{"location":"model_creation/compositional_modeling/#Compartment-based-models","page":"Compositional Modeling of Reaction Systems","title":"Compartment-based models","text":"Finally, let's see how we can make a compartment-based model. Let's create a simple eukaryotic gene expression model with negative feedback by protein dimers. Transcription and gene inhibition by the protein dimer occurs in the nucleus, translation and dimerization occur in the cytosol, and nuclear import and export reactions couple the two compartments. We'll include volume parameters for the nucleus and cytosol, and assume we are working with species having units of number of molecules. Rate constants will have their common concentration units, i.e. if V denotes the volume of a compartment then\n\nReaction Type Example Rate Constant Units Effective rate constant (units of per time)\nZero order varnothing oversetalphato A concentration / time alpha V\nFirst order A oversetbetato B (time)⁻¹ beta\nSecond order A + B oversetgammato C (concentration × time)⁻¹ gammaV\n\nIn our model we'll therefore add the conversions of the last column to properly account for compartment volumes:\n\n# transcription and regulation\nnuc = @network_component nuc begin\n  α, G --> G + M\n  (κ₊/V,κ₋), D + G <--> DG\nend\n\n# translation and dimerization\ncyto = @network_component cyto begin\n  β, M --> M + P\n  (k₊/V,k₋), 2P <--> D\n  σ, P --> 0\n  μ, M --> 0\nend\n\n# export reactions,\n# γ,δ=probability per time to be exported/imported\nmodel = @network_component model begin\n  γ, $(nuc.M) --> $(cyto.M)\n  δ, $(cyto.D) --> $(nuc.D)\nend\n@named model = compose(model, [nuc, cyto])\n\nA graph of the resulting network is\n\nGraph(model)\n\n(Image: graph of gene regulation model)","category":"section"},{"location":"model_creation/dsl_basics/#dsl_description","page":"The Catalyst DSL - Introduction","title":"The Catalyst DSL - Introduction","text":"<details><summary><strong>Environment setup and package installation</strong></summary>\n\nThe following code sets up an environment for running the code on this page.\n\nusing Pkg\nPkg.activate(; temp = true) # Creates a temporary environment, which is deleted when the Julia session ends.\nPkg.add(\"Catalyst\")\nPkg.add(\"Latexify\")\n\n</details>\n\n\n   In the introduction to Catalyst we described how the @reaction_network macro can be used to create chemical reaction network (CRN) models. This macro enables a so-called domain-specific language (DSL) for creating CRN models. This tutorial will give a basic introduction on how to create Catalyst models using this macro (from now onwards called \"the Catalyst DSL\"). A follow-up tutorial will describe some of the DSL's more advanced features.\n\nThe Catalyst DSL generates a ReactionSystem (the julia structure Catalyst uses to represent CRN models). These can be created through alternative methods (e.g. programmatically or compositionally). Previous and following tutorials describe how to simulate models once they have been created using the DSL. This tutorial will solely focus on model creation.\n\nBefore we begin, we will first load the Catalyst package (which is required to run the code).\n\nusing Catalyst","category":"section"},{"location":"model_creation/dsl_basics/#dsl_description_quick_start","page":"The Catalyst DSL - Introduction","title":"Quick-start summary","text":"The DSL is initiated through the @reaction_network macro, which is followed by one line for each reaction. Each reaction consists of a rate, followed lists first of the substrates and next of the products. E.g. a Michaelis-Menten enzyme kinetics system can be written as\n\nrn = @reaction_network begin\n    (kB,kD), S + E <--> SE\n    kP, SE --> P + E\nend\n\nHere, <--> is used to create a bi-directional reaction (with forward rate kP and backward rate kD). Next, the model (stored in the variable rn) can be used as input to various types of simulations.","category":"section"},{"location":"model_creation/dsl_basics/#dsl_description_basic_syntax","page":"The Catalyst DSL - Introduction","title":"Basic syntax","text":"The basic syntax of the DSL is\n\nusing Catalyst # hide\nrn = @reaction_network begin\n    2.0, X --> Y\n    1.0, Y --> X\nend\n\nHere, you start with @reaction_network begin, next list all of the model's reactions, and finish with end. Each reaction consists of\n\nA rate.\nA (potentially empty) set of substrates.\nA (potentially empty) set of products.\n\nEach reaction line declares, in order, the rate, the substrate(s), and the product(s). The rate is separated from the substrate(s) by a ,, and the substrate(s) from the production by a --> (other arrows, however, are also possible). In the above example, our model consists of two reactions. In the first one, X (the single substrate) becomes Y (the single product) at rate 2.0. In the second reaction, Y becomes X at rate 1.0.\n\nFinally, rn = is used to store the model in the variable rn (a normal Julia variable, which does not need to be called rn).","category":"section"},{"location":"model_creation/dsl_basics/#dsl_description_parameters_basics","page":"The Catalyst DSL - Introduction","title":"Defining parameters and species in the DSL","text":"Typically, the rates are not constants, but rather parameters (which values can be set e.g. at the beginning of each simulation). To set parametric rates, simply use whichever symbol you wish to represent your parameter with. E.g. to set the above rates to a and b, we use:\n\nrn = @reaction_network begin\n    a, X --> Y\n    b, Y --> X\nend\n\nHere we have used single-character symbols to designate all species and parameters. Multi-character symbols, however, are also permitted. E.g. we could call the rates kX and kY:\n\nrn = @reaction_network begin\n    kX, X --> Y\n    kY, Y --> X\nend\n\nGenerally, anything that is a permitted Julia variable name can be used to designate a species or parameter in Catalyst.","category":"section"},{"location":"model_creation/dsl_basics/#dsl_description_reactions","page":"The Catalyst DSL - Introduction","title":"Different types of reactions","text":"","category":"section"},{"location":"model_creation/dsl_basics/#dsl_description_reactions_multiples","page":"The Catalyst DSL - Introduction","title":"Reactions with multiple substrates or products","text":"Previously, our reactions have had a single substrate and a single product. However, reactions with multiple substrates and/or products are possible. Here, all the substrates (or products) are listed and separated by a +. E.g. to create a model where X and Y bind (at rate kB) to form XY (which then can dissociate, at rate kD, to form XY) we use:\n\nrn = @reaction_network begin\n    kB, X + Y --> XY\n    kD, XY --> X + Y\nend\n\nReactions can have any number of substrates and products, and their names do not need to have any relationship to each other, as demonstrated by the following mock model:\n\nrn = @reaction_network begin\n    k, X + Y + Z --> A + B + C + D\nend","category":"section"},{"location":"model_creation/dsl_basics/#dsl_description_reactions_degradation_and_production","page":"The Catalyst DSL - Introduction","title":"Reactions with degradation or production","text":"Some reactions have no products, in which case the substrate(s) are degraded (i.e. removed from the system). To denote this, set the reaction's right-hand side to 0. Similarly, some reactions have no substrates, in which case the product(s) are produced (i.e. added to the system). This is denoted by setting the left-hand side to 0. E.g. to create a model where a single species X is both created (in the first reaction) and degraded (in a second reaction), we use:\n\nrn = @reaction_network begin\n    p, 0 --> X\n    d, X --> 0\nend","category":"section"},{"location":"model_creation/dsl_basics/#dsl_description_reactions_stoichiometries","page":"The Catalyst DSL - Introduction","title":"Reactions with non-unitary stoichiometries","text":"Reactions may include multiple copies of the same reactant (i.e. a substrate or a product). To specify this, the reactant is preceded by a number indicating its number of copies (also called the reactant's stoichiometry). E.g. to create a model where two copies of X dimerise to form X2 (which then dissociate back to two X copies) we use:\n\nrn = @reaction_network begin\n    kB, 2X --> X2\n    kD, X2 --> 2X\nend\n\nReactants whose stoichiometries are not defined are assumed to have stoichiometry 1. Any integer number can be used, furthermore, decimal numbers and parameters can also be used as stoichiometries. A discussion of non-unitary (i.e. not equal to 1) stoichiometries affecting the created model can be found here.\n\nStoichiometries can be combined with () to define them for multiple reactants. Here, the following (mock) model declares the same reaction twice, both with and without this notation:\n\nrn = @reaction_network begin\n    k, 2X + 3(Y + 2Z) --> 5(V + W)\n    k, 2X + 3Y + 6Z --> 5V + 5W\nend","category":"section"},{"location":"model_creation/dsl_basics/#dsl_description_reaction_bundling","page":"The Catalyst DSL - Introduction","title":"Bundling of similar reactions","text":"","category":"section"},{"location":"model_creation/dsl_basics/#dsl_description_reaction_bundling_reversible","page":"The Catalyst DSL - Introduction","title":"Bi-directional (or reversible) reactions","text":"As is the case for the following two-state model:\n\nrn_bidir = @reaction_network begin\n    k1, X1 --> X2\n    k2, X2 --> X1\nend\n\nit is common that reactions occur in both directions (so-called bi-directional reactions). Here, it is possible to bundle the reactions into a single line by using the <--> arrow. When we do this, the rate term must include two separate rates (one for each direction, these are enclosed by a () and separated by a ,). I.e. the two-state model can be declared using:\n\nrn_bidir = @reaction_network begin\n    (k1,k2), X1 <--> X2\nend\n\nHere, the first rate (k1) denotes the forward rate and the second rate (k2) the backwards rate.\n\nCatalyst also permits writing pure backwards reactions. These use identical syntax to forward reactions, but with the <-- arrow:\n\nrn_ytox = @reaction_network begin\n    k, X <-- Y\nend\n\nHere, the substrate(s) are on the right-hand side and the product(s) are on the left-hand side. Hence, the above model can be written identically using:\n\nrn_ytox = @reaction_network begin\n    k, Y --> X\nend\n\nGenerally, using forward reactions is clearer than backwards ones, with the latter typically never being used.","category":"section"},{"location":"model_creation/dsl_basics/#dsl_description_reaction_bundling_similar","page":"The Catalyst DSL - Introduction","title":"Bundling similar reactions on a single line","text":"There exist additional situations where models contain similar reactions (e.g. systems where all system components degrade at identical rates). Reactions which share either rates, substrates, or products can be bundled into a single line. Here, the parts which are different for the reactions are written using (,) (containing one separate expression for each reaction). E.g., let us consider the following model where species X and Y both degrade at the rate d:\n\nrn_deg = @reaction_network begin\n    d, X --> 0\n    d, Y --> 0\nend\n\nThese share both their rates (d) and products (0), however, the substrates are different (X and Y). Hence, the reactions can be bundled into a single line using the common rate and product expression while providing separate substrate expressions:\n\nrn_deg = @reaction_network begin\n    d, (X,Y) --> 0\nend\n\nThis declaration of the model is identical to the previous one. Reactions can share any subset of the rate, substrate, and product expression (the cases where they share all or none, however, do not make sense to use). I.e. if the two reactions also have different degradation rates:\n\nrn_deg2 = @reaction_network begin\n    dX, X --> 0\n    dY, Y --> 0\nend\n\nThis can be represented using:\n\nrn_deg2 = @reaction_network begin\n    (dX,dY), (X,Y) --> 0\nend\n\nIt is possible to use bundling for any number of reactions. E.g. in the following model we bundle the conversion of a species X between its various forms (where all reactions use the same rate k):\n\nrn = @reaction_network begin\n    k, (X0,X1,X2,X3) --> (X1,X2,X3,X4)\nend\n\nIt is possible to combine bundling with bi-directional reactions. In this case, the rate is first split into the forward and backwards rates. These may then (or may not) indicate several rates. We exemplify this using the two following two (identical) networks, created with and without bundling.\n\nrn_sp = @reaction_network begin\n    kf, S --> P1\n    kf, S --> P2\n    kb_1, P1 --> S\n    kb_2, P2 --> S\nend\n\nrn_sp = @reaction_network begin\n    (kf, (kb_1, kb_2)), S <--> (P1,P2)\nend\n\nLike when we designated stoichiometries, reaction bundling can be applied very generally to create some truly complicated reactions:\n\nrn = @reaction_network begin\n    ((pX, pY, pZ),d), (0, Y0, Z0) <--> (X, Y, Z1+Z2)\nend\n\nHowever, like for the above model, bundling reactions too zealously can reduce (rather than improve) a model's readability.\n\nThe one exception to reaction bundling is that we do not permit the user to provide multiple rates but only set one set each for the substrates and products. I.e.\n\nrn_erroneous = @reaction_network begin\n    (k1,k2), X --> Y\nend\n\nis not permitted (due to this notation's similarity to a bidirectional reaction). However, if multiples are provided for substrates and/or products, like (k1,k2), (X1,X2) --> Y, then bundling works.","category":"section"},{"location":"model_creation/dsl_basics/#dsl_description_nonconstant_rates","page":"The Catalyst DSL - Introduction","title":"Non-constant reaction rates","text":"So far we have assumed that all reaction rates are constant (being either a number of a parameter). Non-constant rates that depend on one (or several) species are also possible. More generally, the rate can be any valid expression of parameters and species.\n\nLet us consider a model with an activator (A, which degraded at a constant rate) and a protein (P). The production rate of P depends both on A and a parameter (kP). We model this through:\n\nrn_ap = @reaction_network begin\n    d, A --> 0\n    kP*A, 0 --> P\nend\n\nHere, P's production rate will be reduced as A decays. We can print the ODE this model produces with Latexify:\n\nusing Latexify\nlatexify(rn_ap; form = :ode)\nlatexify(rn_ap; form = :ode, math_delimiters = true) # hide\n\nIn this case, we can generate an equivalent model by instead adding A as both a substrate and a product to P's production reaction:\n\nrn_ap_alt = @reaction_network begin\n    d, A --> 0\n    kp, A --> A + P\nend\n\nWe can confirm that this generates the same ODE:\n\nlatexify(rn_ap_alt; form = :ode)\nlatexify(rn_ap_alt; form = :ode, math_delimiters = true) # hide\n\nHere, while these models will generate identical ODE, SDE, and jump simulations, the chemical reaction network models themselves are not equivalent. Generally, as pointed out in the two notes below, using the second form is preferable.\n\nwarning: Warning\nWhile rn_ap and rn_ap_alt will generate equivalent simulations, for jump simulations, the first model will have reduced performance as it generates a less performant representation of the system in JumpProcesses. It is generally recommended to write pure mass action reactions such that there is just a single constant within the rate constant expression for optimal performance of jump process simulations.\n\ndanger: Danger\nCatalyst automatically infers whether quantities appearing in the DSL are species or parameters (as described here). Generally, anything that does not appear as a reactant is inferred to be a parameter. This means that if you want to model a reaction activated by a species (e.g. kp*A, 0 --> P), but that species does not occur as a reactant, it will be interpreted as a parameter. This can be handled by manually declaring the system species.\n\nAbove we used a simple example where the rate was the product of a species and a parameter. However, any valid Julia expression of parameters, species, and values can be used. E.g the following is a valid model:\n\nrn = @reaction_network begin\n    2.0 + X^2, 0 --> X + Y\n    k1 + k2^k3, X --> ∅\n    pi * X/(sqrt(2) + Y), Y → ∅\nend","category":"section"},{"location":"model_creation/dsl_basics/#dsl_description_nonconstant_rates_functions","page":"The Catalyst DSL - Introduction","title":"Using functions in rates","text":"It is possible for the rate to contain Julia functions. These can either be functions from Julia's standard library:\n\nrn = @reaction_network begin\n    d, A --> 0\n    kp*sqrt(A), 0 --> P\nend\n\nor ones defined by the user:\n\ncustom_function(p1, p2, X) = (p1 + X) / (p2 + X)\nrn = @reaction_network begin\n    d, A --> 0\n    custom_function(k1,k2,E), 0 --> P\nend","category":"section"},{"location":"model_creation/dsl_basics/#dsl_description_nonconstant_rates_available_functions","page":"The Catalyst DSL - Introduction","title":"Pre-defined functions","text":"Two functions frequently used within systems biology are the Michaelis-Menten and Hill functions. These are pre-defined in Catalyst and can be called using mm(X,v,K) and hill(X,v,K,n). E.g. a self-activation loop where X activates its own production through a Hill function can be created using:\n\nrn = @reaction_network begin\n    hill(X,v,K,n), 0 --> P\n    d, X --> 0\nend\n\nCatalyst comes with the following predefined functions:\n\nThe Michaelis-Menten function: mm(XvK) = v * X(X + K).\nThe repressive Michaelis-Menten function: mmr(XvK) = v * K(X + K).\nThe Hill function: hill(XvKn) = v * (X^n)(X^n + K^n).\nThe repressive Hill function: hillr(XvKn) = v * (K^n)(X^n + K^n).\nThe activating/repressive Hill function: hillar(XYvKn) = v * (X^n)(X^n + Y^n + K^n).","category":"section"},{"location":"model_creation/dsl_basics/#dsl_description_nonconstant_rates_function_registration","page":"The Catalyst DSL - Introduction","title":"Registration of non-algebraic functions","text":"Previously we showed how user-defined functions can be used in rates directly. For functions containing more complicated syntax (e.g. for loops or if statements), we must add an additional step: registering it using the @register_symbolic macro. Below we define a non-standard function of one variable. Next, we register it using @register_symbolic, after which we can use it within the DSL.\n\nweirdfunc(x) = round(x) + 2.0\n@register_symbolic weirdfunc(X)\nrn = @reaction_network begin\n    weirdfunc(X), 0 --> X\n    d, X --> 0\nend","category":"section"},{"location":"model_creation/dsl_basics/#dsl_description_nonconstant_rates_time","page":"The Catalyst DSL - Introduction","title":"Time-dependant rates","text":"Previously we have assumed that the rates are independent of the time variable, t. However, time-dependent reactions are also possible. Here, simply use t to represent the time variable. E.g., to create a production/degradation model where the production rate decays as time progresses, we can use:\n\nrn = @reaction_network begin\n    kp/(1 + t), 0 --> P\n    d, P --> 0\nend\n\nLike previously, t can be part of any valid expression. E.g. to create a reaction with a cyclic rate (e.g. to represent a circadian system) we can use:\n\nrn = @reaction_network begin\n    A*(sin(2π*f*t - ϕ)+1)/2, 0 --> P\n    d, P --> 0\nend\n\nwarning: Warning\nModels with explicit time-dependent rates require additional steps to correctly convert to stochastic chemical kinetics jump process representations. See here for guidance on manually creating such representations. Enabling Catalyst to handle this seamlessly is work in progress.","category":"section"},{"location":"model_creation/dsl_basics/#dsl_description_stoichiometries","page":"The Catalyst DSL - Introduction","title":"Non-standard stoichiometries","text":"","category":"section"},{"location":"model_creation/dsl_basics/#dsl_description_stoichiometries_decimal","page":"The Catalyst DSL - Introduction","title":"Non-integer stoichiometries","text":"Previously all stoichiometric constants have been integer numbers, however, decimal numbers are also permitted. Here we create a birth-death model where each production reaction produces 1.5 units of X:\n\nrn = @reaction_network begin\n    p, 0 --> 1.5X\n    d, X --> 0\nend\n\nIt is also possible to have non-integer stoichiometric coefficients for substrates. However, in this case the combinatoric_ratelaw = false option must be used. We note that non-integer stoichiometric coefficients do not make sense in most fields, however, this feature is available for use for models where it does make sense.","category":"section"},{"location":"model_creation/dsl_basics/#dsl_description_stoichiometries_parameters","page":"The Catalyst DSL - Introduction","title":"Parametric stoichiometries","text":"It is possible for stoichiometric coefficients to be parameters. E.g. here we create a generic polymerisation system where n copies of X bind to form Xn:\n\nrn = @reaction_network begin\n    @parameters n::Int64 \n    (kB,kD), n*X <--> Xn\nend\n\nNow we can designate the value of n through a parameter when we e.g. create an ODEProblem:\n\nu0 = [:X => 5.0, :Xn => 1.0]\nps = [:kB => 1.0, :kD => 0.1, :n => 4]\noprob = ODEProblem(rn, u0, (0.0, 1.0), ps)\nnothing # hide\n\nnote: Note\nUsing non-integer stoichiometric coefficients is possible. However, this requires both using the combinatoric_ratelaw = false option, and also explicitly designating the stoichiometric parameter as a Float64 using the approach described here.","category":"section"},{"location":"model_creation/dsl_basics/#dsl_description_symbols","page":"The Catalyst DSL - Introduction","title":"Using special symbols","text":"Julia permits any Unicode characters to be used in variable names, thus Catalyst can use these as well. Below we describe some cases where this can be useful. No functionality is, however, tied to this.","category":"section"},{"location":"model_creation/dsl_basics/#dsl_description_symbols_empty_set","page":"The Catalyst DSL - Introduction","title":"Using ∅ in degradation/production reactions","text":"Previously, we described how 0 could be used to create degradation or production reactions. Catalyst permits the user to instead use the ∅ symbol. E.g. the production/degradation system can alternatively be written as:\n\nrn_pd = @reaction_network begin\n    p, ∅ --> X\n    d, X --> ∅\nend","category":"section"},{"location":"model_creation/dsl_basics/#dsl_description_symbols_arrows","page":"The Catalyst DSL - Introduction","title":"Using special arrow symbols","text":"Catalyst uses -->, <-->, and <-- to denote forward, bi-directional, and backwards reactions, respectively. Several unicode representations of these arrows are available. Here,\n\n>, →, ↣, ↦, ⇾, ⟶, ⟼, ⥟, ⥟, ⇀, and ⇁ can be used to represent forward reactions.\n↔, ⟷, ⇄, ⇆, ⇌, ⇋, , and ⇔ can be used to represent bi-directional reactions.\n<, ←, ↢, ↤, ⇽, ⟵, ⟻, ⥚, ⥞, ↼, , and ↽ can be used to represent backwards reactions.\n\nE.g. the production/degradation system can alternatively be written as:\n\nrn_pd = @reaction_network begin\n    p, ∅ → X\n    d, X → ∅\nend","category":"section"},{"location":"model_creation/dsl_basics/#dsl_description_symbols_special","page":"The Catalyst DSL - Introduction","title":"Using special symbols to denote species or parameters","text":"A range of possible characters are available which can be incorporated into species and parameter names. This includes, but is not limited to:\n\nGreek letters (e.g α, σ, τ, and Ω).\nSuperscript and subscript characters (to create e.g. k₁, k₂, Xₐ, and Xᴾ).\nNon-latin, non-greek, letters (e.g. ä, Д, س, and א).\nOther symbols (e.g. £, ℂ, ▲, and ♠).\n\nAn example of how this can be used to create a neat-looking model can be found in Schwall et al. (2021) where it was used to model a sigma factor V circuit in the bacteria Bacillus subtilis:\n\nσᵛ_model = @reaction_network begin\n    v₀ + hill(σᵛ,v,K,n), ∅ → σᵛ + A\n    kdeg, (σᵛ, A, Aσᵛ) → ∅\n    (kB,kD), A + σᵛ ↔ Aσᵛ\n    L, Aσᵛ → σᵛ\nend\nnothing # hide\n\nThis functionality can also be used to create less serious models:\n\nrn = @reaction_network begin\n    🍦, 😢 --> 😃\nend\n\nIt should be noted that the following symbols are not permitted to be used as species or parameter names:\n\npi and π (used in Julia to denote 3.1415926535897...).\nℯ (used in Julia to denote Euler's constant).\nt (used to denote the time variable).\n∅ (used for production/degradation reactions).\nim (used in Julia to represent complex numbers).\nnothing (used in Julia to denote nothing).\nΓ (used by Catalyst to represent conserved quantities).","category":"section"},{"location":"introduction_to_catalyst/math_models_intro/#math_models_in_catalyst","page":"Mathematical Models Catalyst can Generate","title":"Mathematical Models Catalyst can Generate","text":"We now describe the types of mathematical models that Catalyst can generate from chemical reaction networks (CRNs), corresponding to reaction rate equation (RRE) ordinary differential equation (ODE) models, Chemical Langevin equation (CLE) stochastic differential equation (SDE) models, and stochastic chemical kinetics (jump process) models. For each we show the abstract representations for the models that Catalyst can support, along with concrete examples. Note that we restrict ourselves to models involving only chemical reactions, and do not consider more general models that Catalyst can support such as coupling in non-reaction ODEs, algebraic equations, or events. Please see the broader documentation for more details on how Catalyst supports such functionality.\n\nnote: Note\nThis documentation assumes you have already read the Introduction to Catalyst tutorial.","category":"section"},{"location":"introduction_to_catalyst/math_models_intro/#General-Chemical-Reaction-Notation","page":"Mathematical Models Catalyst can Generate","title":"General Chemical Reaction Notation","text":"Suppose we have a reaction network with K reactions and M species, with the species labeled by S_1, S_2, dots, S_M. We denote by\n\nmathbfX(t) = beginpmatrix X_1(t)  vdots  X_M(t)) endpmatrix\n\nthe state vector for the amount of each species, i.e. X_m(t) represents the amount of species S_m at time t. This could be either a concentration or a count (i.e. \"number of molecules\" units), but for consistency between modeling representations we will assume the latter in the remainder of this introduction.\n\nThe kth chemical reaction is given by\n\nalpha_1^k S_1 + alpha_2^k S_2 + dots alpha_M^k S_M to beta_1^k S_1 + beta_2^k S_2 + dots beta_M^k S_M\n\nwith alpha^k = (alpha_1^kdotsalpha_M^k) its substrate stoichiometry vector, beta^k = (beta_1^kdotsbeta_M^k) its product stoichiometry vector, and nu^k = beta^k - alpha^k its net stoichiometry vector. nu^k corresponds to the change in mathbfX(t) when reaction k occurs, i.e. mathbfX(t) to mathbfX(t) + nu^k. Along with the stoichiometry vectors, we assume each reaction has a reaction rate law (ODEs/SDEs) or propensity (jump process) function, a_k(mathbfX(t)t).\n\nAs explained in the Catalyst introduction, for a mass action reaction where the preceding reaction has a fixed rate constant, k, this function would be the rate law\n\na_k(mathbfX(t)) = k prod_m=1^M frac(X_m(t))^alpha_m^kalpha_m^k\n\nfor RRE ODE and CLE SDE models, and the propensity function\n\na_k(mathbfX(t)) = k prod_m=1^M fracX_m(t) (X_m(t)-1) dots (X_m(t)-alpha_m^k+1)alpha_m^k\n\nfor stochastic chemical kinetics jump process models.","category":"section"},{"location":"introduction_to_catalyst/math_models_intro/#Rate-Law-vs.-Propensity-Example:","page":"Mathematical Models Catalyst can Generate","title":"Rate Law vs. Propensity Example:","text":"For the reaction 2A + B oversetkto 3 C we would have\n\nmathbfX(t) = (A(t) B(t) C(t))\n\nwith alpha_1 = 2, alpha_2 = 1, alpha_3 = 0, beta_1 = 0, beta_2 = 0, beta_3 = 3, nu_1 = -2, nu_2 = -1, and nu_3 = 3. For an ODE/SDE model we would have the rate law\n\na(mathbfX(t)) = frack2 A^2 B\n\nwhile for a jump process model we would have the propensity function\n\na(mathbfX(t)) = frack2 A (A-1) B\n\nNote, if the combinatoric factors are already included in one's rate constants, the implicit rescaling of rate constants can be disabled through use of the combinatoric_ratelaws = false argument to ode_model (or similar) or whatever Problem is being generated, i.e.\n\nrn = @reaction_network ...\nosys = ode_model(rn; combinatoric_ratelaws = false)\noprob = ODEProblem(osys, ...)\n\n# or\noprob = ODEProblem(rn, ...; combinatoric_ratelaws = false)\n\nIn this case our ODE/SDE rate law would be\n\na(mathbfX(t)) = k A^2 B\n\nwhile the jump process propensity function is\n\na(mathbfX(t)) = k A (A-1) B\n\nOne can also specify during system construction that by default combinatoric scalings should be disabled, i.e.\n\nusing Catalyst\nusing Latexify # hide\nrn = @reaction_network begin\n    @combinatoric_ratelaws false\n    k, 3A + 2B --> A + 3D\nend\nosys = ode_model(rn)\nlatexify(rn; form = :ode, math_delimiters = true) # hide","category":"section"},{"location":"introduction_to_catalyst/math_models_intro/#math_models_in_catalyst_rre_odes","page":"Mathematical Models Catalyst can Generate","title":"Reaction Rate Equation (RRE) ODE Models","text":"The RRE ODE models Catalyst creates for a general system correspond to the coupled system of ODEs given by\n\nfracd X_mdt =sum_k=1^K nu_m^k a_k(mathbfX(t)t) quad m = 1dotsM\n\nThese models can be generated by creating ODEProblems from Catalyst ReactionSystems, and solved using the solvers in OrdinaryDiffEq.jl. Similarly, creating NonlinearProblems or SteadyStateProblems will generate the coupled algebraic system of steady-state equations associated with a RRE ODE model, i.e.\n\n0 =sum_k=1^K nu_m^k a_k(barmathbfX) quad m = 1dotsM\n\nfor a steady-state barmathbfX. Note, here we have assumed the rate laws are autonomous so that the equations are well-defined.","category":"section"},{"location":"introduction_to_catalyst/math_models_intro/#RRE-ODE-Example","page":"Mathematical Models Catalyst can Generate","title":"RRE ODE Example","text":"Let's see the generated ODEs for the following network\n\nusing Catalyst, ModelingToolkitBase, Latexify\nrn = @reaction_network begin\n    k₁, 2A + B --> 3C\n    k₂, A --> 0\n    k₃, 0 --> A\nend\nosys = ode_model(rn)\nlatexify(rn; form = :ode, math_delimiters = true) # hide\n\nLikewise, the following drops the combinatoric scaling factors, giving unscaled ODEs\n\nosys = ode_model(rn; combinatoric_ratelaws = false)\nlatexify(Catalyst.system_to_reactionsystem(osys); math_delimiters = true) # hide","category":"section"},{"location":"introduction_to_catalyst/math_models_intro/#math_models_in_catalyst_cle_sdes","page":"Mathematical Models Catalyst can Generate","title":"Chemical Langevin Equation (CLE) SDE Models","text":"The CLE SDE models Catalyst creates for a general system correspond to the coupled system of SDEs given by\n\nd X_m = sum_k=1^K nu_m^k a_k(mathbfX(t)t) dt + sum_k=1^K nu_m^k sqrta_k(mathbfX(t)t) dW_k(t) quad m = 1dotsM\n\nwhere each W_k(t) represents an independent, standard Brownian Motion. Realizations of these processes can be generated by creating SDEProblems from Catalyst ReactionSystems, and sampling the processes using the solvers in StochasticDiffEq.jl.","category":"section"},{"location":"introduction_to_catalyst/math_models_intro/#CLE-SDE-Example","page":"Mathematical Models Catalyst can Generate","title":"CLE SDE Example","text":"Consider the same network as above,\n\nrn = @reaction_network begin\n    k₁, 2A + B --> 3C\n    k₂, A --> 0\n    k₃, 0 --> A\nend\n\nWe obtain the CLE SDEs\n\nbeginalign\ndA(t) = left(- k_1 A^2 B - k_2 A + k_3 right) dt\n - 2 sqrttfrack_12 A^2 B  dW_1(t)  - sqrtk_2 A  dW_2(t) + sqrtk_3  dW_3(t)\n\ndB(t) = - frack_12 A^2 B  dt - sqrtfrack_12 A^2 B  dW_1(t) \ndC(t) = frac32 k_1 A^2 B  dt + 3 sqrtfrack_12 A^2 B  dW_1(t)\nendalign","category":"section"},{"location":"introduction_to_catalyst/math_models_intro/#math_models_in_catalyst_sck_jumps","page":"Mathematical Models Catalyst can Generate","title":"Stochastic Chemical Kinetics Jump Process Models","text":"The stochastic chemical kinetics jump process models Catalyst creates for a general system correspond to the coupled system of jump processes, in the time change representation, given by\n\nX_m(t) = X_m(0) + sum_k=1^K nu_m^k Y_kleft( int_0^t a_k(mathbfX(s^-)s)  ds right) quad m = 1dotsM\n\nHere each Y_k(t) denotes an independent unit rate Poisson counting process with Y_k(0) = 0, which counts the number of times the kth reaction has occurred up to time t. Realizations of these processes can be generated by creating JumpProblems from Catalyst ReactionSystems, and sampling the processes using the solvers in JumpProcesses.jl.\n\nLet P(mathbfxt) = operatornameProbmathbfX(t) = mathbfx represent the probability the state of the system, mathbfX(t), has the concrete value mathbfx at time t. The forward equation, i.e. Chemical Master Equation (CME), associated with mathbfX(t) is then the coupled system of ODEs over all possible values for mathbfx given by\n\nfracdPdt(mathbfxt) = sum_k=1^k left a_k(mathbfx - nu^kt) P(mathbfx - nu^kt) - a_k(mathbfxt) P(mathbfxt) right\n\nWhile Catalyst does not currently support generating and solving for P(mathbfxt), for sufficiently small models the FiniteStateProjection.jl package can be used to generate and solve such models directly from Catalyst ReactionSystems.","category":"section"},{"location":"introduction_to_catalyst/math_models_intro/#Stochastic-Chemical-Kinetics-Jump-Process-Example","page":"Mathematical Models Catalyst can Generate","title":"Stochastic Chemical Kinetics Jump Process Example","text":"Consider the same network as above,\n\nrn = @reaction_network begin\n    k₁, 2A + B --> 3C\n    k₂, A --> 0\n    k₃, 0 --> A\nend\n\nThe time change process representation would be\n\nbeginalign*\nA(t) = A(0) - 2 Y_1left( frack_12 int_0^t A(s^-)(A(s^-)-1) B(s^-)  ds right) - Y_2 left( k_2 int_0^t A(s^-)  ds right) + Y_3 left( k_3 t right) \nB(t) = B(0) - Y_1left( frack_12 int_0^t A(s^-)(A(s^-)-1) B(s^-)  ds right) \nC(t) = C(0) + 3 Y_1left( frack_12 int_0^t A(s^-)(A(s^-)-1) B(s^-)  ds right)\nendalign*\n\nwhile the CME would be the coupled (infinite) system of ODEs over all realizable values of the non-negative integers a, b, and c given by\n\nbeginalign*\nfracdPdt(abct) = lefttfrack_12 (a+2) (a+1) (b+1) P(a+2b+1c-3t) - tfrack_12 a (a-1) b P(abct)right \nphantom= + leftk_2 (a+1) P(a+1bct) - k_2 a P(abct)right \nphantom= + leftk_3 P(a-1bct) - k_3 P(abct)right\nendalign*\n\nIf we initially have A(0) = a_0, B(0) = b_0, and C(0) = c_0 then we would have one ODE for each of possible state (abc) where a in 01dots (i.e. a can be any non-negative integer), b in 01dotsb_0, and c in c_0 c_0 + 1dots c_0 + 3 b_0. Other initial conditions would lead to different possible ranges for a, b, and c.","category":"section"},{"location":"api/network_analysis_api/#api_network_analysis","page":"Network analysis and representations","title":"Network analysis and representations","text":"Note, currently API functions for network analysis and conservation law analysis do not work with constant species (which are generated by SBML, and can be declared in Catalyst as well).\n\nFor more information about these functions, please see the sections of the docs on network ODE representation and chemical reaction network theory.","category":"section"},{"location":"api/network_analysis_api/#Catalyst.conservationlaws","page":"Network analysis and representations","title":"Catalyst.conservationlaws","text":"conservationlaws(netstoichmat::AbstractMatrix)::Matrix\n\nGiven the net stoichiometry matrix of a reaction system, computes a matrix of conservation laws, each represented as a row in the output.\n\n\n\n\n\nconservationlaws(rs::ReactionSystem)\n\nReturn the conservation law matrix of the given ReactionSystem, calculating it if it is not already stored within the system, or returning an alias to it.\n\nNotes:\n\nThe first time being called it is calculated and cached in rn, subsequent calls should be fast.\n\n\n\n\n\n","category":"function"},{"location":"api/network_analysis_api/#Catalyst.conservedquantities","page":"Network analysis and representations","title":"Catalyst.conservedquantities","text":"conservedquantities(state, cons_laws)\n\nCompute conserved quantities for a system with the given conservation laws.\n\n\n\n\n\n","category":"function"},{"location":"api/network_analysis_api/#Catalyst.conservedequations","page":"Network analysis and representations","title":"Catalyst.conservedequations","text":"conservedequations(rn::ReactionSystem)\n\nCalculate symbolic equations from conservation laws, writing dependent variables as functions of independent variables and the conservation law constants.\n\nNotes:\n\nCaches the resulting equations in rn, so will be fast on subsequent calls.\n\nExamples:\n\nrn = @reaction_network begin\n    k, A + B --> C\n    k2, C --> A + B\n    end\nconservedequations(rn)\n\ngives\n\n2-element Vector{Equation}:\n B(t) ~ A(t) + Γ[1]\n C(t) ~ Γ[2] - A(t)\n\n\n\n\n\n","category":"function"},{"location":"api/network_analysis_api/#Catalyst.conservationlaw_constants","page":"Network analysis and representations","title":"Catalyst.conservationlaw_constants","text":"conservationlaw_constants(rn::ReactionSystem)\n\nCalculate symbolic equations from conservation laws, writing the conservation law constants in terms of the dependent and independent variables.\n\nNotes:\n\nCaches the resulting equations in rn, so will be fast on subsequent calls.\n\nExamples:\n\nrn = @reaction_network begin\n    k, A + B --> C\n    k2, C --> A + B\n    end\nconservationlaw_constants(rn)\n\ngives\n\n2-element Vector{Equation}:\n Γ[1] ~ B(t) - A(t)\n Γ[2] ~ A(t) + C(t)\n\n\n\n\n\n","category":"function"},{"location":"api/network_analysis_api/#Catalyst.ReactionComplexElement","page":"Network analysis and representations","title":"Catalyst.ReactionComplexElement","text":"struct ReactionComplexElement{T}\n\nOne reaction complex element\n\nFields\n\nspeciesid: The integer id of the species representing this element.\nspeciesstoich: The stoichiometric coefficient of this species.\n\n\n\n\n\n","category":"type"},{"location":"api/network_analysis_api/#Catalyst.ReactionComplex","page":"Network analysis and representations","title":"Catalyst.ReactionComplex","text":"struct ReactionComplex{V<:Integer} <: AbstractArray{Catalyst.ReactionComplexElement{V<:Integer}, 1}\n\nOne reaction complex.\n\nFields\n\nspeciesids: The integer ids of all species participating in this complex.\nspeciesstoichs: The stoichiometric coefficients of all species participating in this complex.\n\n\n\n\n\n","category":"type"},{"location":"api/network_analysis_api/#Catalyst.reactioncomplexmap","page":"Network analysis and representations","title":"Catalyst.reactioncomplexmap","text":"reactioncomplexmap(rn::ReactionSystem)\n\nFind each ReactionComplex within the specified system, constructing a mapping from the complex to vectors that indicate which reactions it appears in as substrates and products.\n\nNotes:\n\nEach ReactionComplex is mapped to a vector of pairs, with each pair having the form reactionidx => ± 1, where -1 indicates the complex appears as a substrate and +1 as a product in the reaction with integer label reactionidx.\nConstant species are ignored as part of a complex. i.e. if species A is constant then the reaction A + B --> C + D is considered to consist of the complexes B and C + D. Likewise A --> B would be treated as the same as 0 --> B.\n\n\n\n\n\n","category":"function"},{"location":"api/network_analysis_api/#Catalyst.reactioncomplexes","page":"Network analysis and representations","title":"Catalyst.reactioncomplexes","text":"reactioncomplexes(network::ReactionSystem; sparse=false)\n\nCalculate the reaction complexes and complex incidence matrix for the given ReactionSystem.\n\nNotes:\n\nreturns a pair of a vector of ReactionComplexs and the complex incidence matrix.\nAn empty ReactionComplex denotes the null (∅) state (from reactions like ∅ -> A or A -> ∅).\nConstant species are ignored in generating a reaction complex. i.e. if A is constant then A –> B consists of the complexes ∅ and B.\nThe complex incidence matrix, B, is number of complexes by number of reactions with\n\nB_i j = begincases\n-1 textif the ith complex is the substrate of the jth reaction\n1 textif the ith complex is the product of the jth reaction\n0 textotherwise\nendcases\n\nSet sparse=true for a sparse matrix representation of the incidence matrix\n\n\n\n\n\n","category":"function"},{"location":"api/network_analysis_api/#Catalyst.incidencemat","page":"Network analysis and representations","title":"Catalyst.incidencemat","text":"incidencemat(rn::ReactionSystem; sparse=false)\n\nCalculate the incidence matrix of rn, see reactioncomplexes.\n\nNotes:\n\nIs cached in rn so that future calls, assuming the same sparsity, will also be fast.\n\n\n\n\n\n","category":"function"},{"location":"api/network_analysis_api/#Catalyst.incidencematgraph","page":"Network analysis and representations","title":"Catalyst.incidencematgraph","text":"incidencematgraph(rn::ReactionSystem)\n\nConstruct a directed simple graph where nodes correspond to reaction complexes and directed edges to reactions converting between two complexes.\n\nFor example,\n\nsir = @reaction_network SIR begin\n    β, S + I --> 2I\n    ν, I --> R\nend\nincidencematgraph(sir)\n\n\n\n\n\n","category":"function"},{"location":"api/network_analysis_api/#Catalyst.complexstoichmat","page":"Network analysis and representations","title":"Catalyst.complexstoichmat","text":"complexstoichmat(network::ReactionSystem; sparse=false)\n\nGiven a ReactionSystem and vector of reaction complexes, return a matrix with positive entries of size number of species by number of complexes, where the non-zero positive entries in the kth column denote stoichiometric coefficients of the species participating in the kth reaction complex.\n\nNotes:\n\nSet sparse=true for a sparse matrix representation\n\n\n\n\n\n","category":"function"},{"location":"api/network_analysis_api/#Catalyst.complexoutgoingmat","page":"Network analysis and representations","title":"Catalyst.complexoutgoingmat","text":"complexoutgoingmat(network::ReactionSystem; sparse=false)\n\nGiven a ReactionSystem and complex incidence matrix, B, return a matrix of size num of complexes by num of reactions that identifies substrate complexes.\n\nNotes:\n\nThe complex outgoing matrix, Delta, is defined by\n\nDelta_i j = begincases\n    = 0    textif  B_i j = 1 \n    = B_i j textotherwise\nendcases\n\nSet sparse=true for a sparse matrix representation\n\n\n\n\n\n","category":"function"},{"location":"api/network_analysis_api/#Catalyst.fluxmat","page":"Network analysis and representations","title":"Catalyst.fluxmat","text":"fluxmat(rn::ReactionSystem, pmap = Dict(); sparse=false)\n\nReturn an r×c matrix K such that, if complex j is the substrate complex of reaction i, then K_ij = k, the rate constant for this reaction. Mostly a helper function for the network Laplacian, laplacianmat. Has the useful property that fracdxdt = S*K*Φ(x), where S is the netstoichmat or net stoichiometry matrix and Φ(x) is the massactionvector.     Returns a symbolic matrix by default, but will return a numerical matrix if rate constants are specified as a Tuple, Vector, or Dict of symbol-value pairs via pmap.\n\nWarning: Unlike other Catalyst functions, the fluxmat function will return a Matrix{Num} in the symbolic case. This is to allow easier computation of the matrix decomposition of the ODEs, and to ensure that multiplying the sparse form of the matrix will work.\n\n\n\n\n\n","category":"function"},{"location":"api/network_analysis_api/#Catalyst.adjacencymat","page":"Network analysis and representations","title":"Catalyst.adjacencymat","text":"adjacencymat(rs::ReactionSystem, pmap = Dict(); sparse = false)\n\nGiven a reaction system with n complexes, outputs an n-by-n matrix where R_{ij} is the rate\nconstant of the reaction between complex i and complex j. Accepts a dictionary, vector, or tuple\nof variable-to-value mappings, e.g. [k1 => 1.0, k2 => 2.0,...].\n\nEquivalent to the adjacency matrix of the weighted graph whose weights are the\nreaction rates.\nReturns a symbolic matrix by default, but will return a numerical matrix if parameter values are specified via pmap.\n\nThe difference between this matrix and [`fluxmat`](@ref) is quite subtle. The non-zero entries to both matrices are the rate constants. However:\n    - In `fluxmat`, the rows represent complexes and columns represent reactions, and an entry (c, r) is non-zero if reaction r's source complex is c.\n    - In `adjacencymat`, the rows and columns both represent complexes, and an entry (c1, c2) is non-zero if there is a reaction c1 --> c2.\n\n\n\n\n\n","category":"function"},{"location":"api/network_analysis_api/#Catalyst.laplacianmat","page":"Network analysis and representations","title":"Catalyst.laplacianmat","text":"laplacianmat(rn::ReactionSystem, pmap=Dict(); sparse=false)\n\nReturn the negative of the graph Laplacian of the reaction network. The ODE system of a chemical reaction network can be factorized as fracdxdt = Y A_k Φ(x), where Y is the complexstoichmat and A_k is the negative of the graph Laplacian, and Φ is the massactionvector. A_k is an n-by-n matrix, where n is the number of complexes, where A_ij = k_ij if a reaction exists between the two complexes and 0 otherwise.\n\nReturns a symbolic matrix by default, but will return a numerical matrix if parameter values are specified via pmap.\n\nWarning: Unlike other Catalyst functions, the laplacianmat function will return a Matrix{Num} in the symbolic case. This is to allow easier computation of the matrix decomposition of the ODEs, and to ensure that multiplying the sparse form of the matrix will work.\n\n\n\n\n\n","category":"function"},{"location":"api/network_analysis_api/#Catalyst.massactionvector","page":"Network analysis and representations","title":"Catalyst.massactionvector","text":"massactionvector(rn::ReactionSystem, scmap = Dict(); combinatoric_ratelaws = true)\n\nReturn the vector whose entries correspond to the \"mass action products\" of each complex. For example, given the complex A + B, the corresponding entry of the vector would be A*B, and for the complex 2X + Y, the corresponding entry would be X^2*Y. The ODE system of a chemical reaction network can be factorized as racdxdt = Y A_k Φ(x), where Y is the complexstoichmat and A_k is the negative of the laplacianmat. This utility returns Φ(x).\n\nReturns a symbolic vector by default, but will return a numerical vector if species concentrations are specified as a tuple, vector, or dictionary via scmap.\n\nIf the combinatoric_ratelaws option is set, will include prefactors for that (see introduction to Catalyst's rate laws. Will default to the default for the system.\n\nWarning: Unlike other Catalyst functions, the massactionvector function will return a Vector{Num} in the symbolic case. This is to allow easier computation of the matrix decomposition of the ODEs.\n\n\n\n\n\n","category":"function"},{"location":"api/network_analysis_api/#Catalyst.linkageclasses","page":"Network analysis and representations","title":"Catalyst.linkageclasses","text":"linkageclasses(rn::ReactionSystem)\n\nGiven the incidence graph of a reaction network, return a vector of the connected components of the graph (i.e. sub-groups of reaction complexes that are connected in the incidence graph).\n\nFor example,\n\nsir = @reaction_network SIR begin\n    β, S + I --> 2I\n    ν, I --> R\nend\nlinkageclasses(sir)\n\ngives\n\n2-element Vector{Vector{Int64}}:\n [1, 2]\n [3, 4]\n\n\n\n\n\n","category":"function"},{"location":"api/network_analysis_api/#Catalyst.deficiency","page":"Network analysis and representations","title":"Catalyst.deficiency","text":"deficiency(rn::ReactionSystem)\n\nCalculate the deficiency of a reaction network.\n\nHere the deficiency, delta, of a network with n reaction complexes, ell linkage classes and a rank s stoichiometric matrix is\n\ndelta = n - ell - s\n\nFor example,\n\nsir = @reaction_network SIR begin\n    β, S + I --> 2I\n    ν, I --> R\nend\nδ = deficiency(sir)\n\n\n\n\n\n","category":"function"},{"location":"api/network_analysis_api/#Catalyst.linkagedeficiencies","page":"Network analysis and representations","title":"Catalyst.linkagedeficiencies","text":"linkagedeficiencies(network::ReactionSystem)\n\nCalculates the deficiency of each sub-reaction network within network.\n\nFor example,\n\nsir = @reaction_network SIR begin\n    β, S + I --> 2I\n    ν, I --> R\nend\nlinkage_deficiencies = linkagedeficiencies(sir)\n\n\n\n\n\n","category":"function"},{"location":"api/network_analysis_api/#Catalyst.satisfiesdeficiencyone","page":"Network analysis and representations","title":"Catalyst.satisfiesdeficiencyone","text":"satisfiesdeficiencyone(rn::ReactionSystem)\n\nCheck if a reaction network obeys the conditions of the deficiency one theorem, which ensures that there is only one equilibrium for every positive stoichiometric compatibility class.\n\n\n\n\n\n","category":"function"},{"location":"api/network_analysis_api/#Catalyst.satisfiesdeficiencyzero","page":"Network analysis and representations","title":"Catalyst.satisfiesdeficiencyzero","text":"satisfiesdeficiencyzero(rn::ReactionSystem)\n\nCheck if a reaction network obeys the conditions of the deficiency zero theorem, which ensures that there is only one equilibrium for every positive stoichiometric compatibility class, this equilibrium is asymptotically stable, and this equilibrium is complex balanced.\n\n\n\n\n\n","category":"function"},{"location":"api/network_analysis_api/#Catalyst.subnetworks","page":"Network analysis and representations","title":"Catalyst.subnetworks","text":"subnetworks(rn::ReactionSystem)\n\nFind subnetworks corresponding to each linkage class of the reaction network.\n\nFor example,\n\nsir = @reaction_network SIR begin\n    β, S + I --> 2I\n    ν, I --> R\nend\nsubnetworks(sir)\n\n\n\n\n\n","category":"function"},{"location":"api/network_analysis_api/#Catalyst.isreversible","page":"Network analysis and representations","title":"Catalyst.isreversible","text":"isreversible(rn::ReactionSystem)\n\nGiven a reaction network, returns if the network is reversible or not.\n\nFor example,\n\nsir = @reaction_network SIR begin\n    β, S + I --> 2I\n    ν, I --> R\nend\nisreversible(sir)\n\n\n\n\n\n","category":"function"},{"location":"api/network_analysis_api/#Catalyst.isweaklyreversible","page":"Network analysis and representations","title":"Catalyst.isweaklyreversible","text":"isweaklyreversible(rn::ReactionSystem, subnetworks)\n\nDetermine if the reaction network with the given subnetworks is weakly reversible or not.\n\nFor example,\n\nsir = @reaction_network SIR begin\n    β, S + I --> 2I\n    ν, I --> R\nend\nsubnets = subnetworks(rn)\nisweaklyreversible(rn, subnets)\n\n\n\n\n\n","category":"function"},{"location":"api/network_analysis_api/#Catalyst.iscomplexbalanced","page":"Network analysis and representations","title":"Catalyst.iscomplexbalanced","text":"iscomplexbalanced(rs::ReactionSystem, parametermap; sparse = false)\n\nConstructively compute whether a network will have complex-balanced equilibrium solutions, following the method in van der Schaft et al., 2015.\n\nRequires the specification of values for the parameters/rate constants. Accepts a dictionary, vector, or tuple of parameter-to-value mappings, e.g. [k1 => 1.0, k2 => 2.0,...].\n\n\n\n\n\n","category":"function"},{"location":"api/network_analysis_api/#Catalyst.isdetailedbalanced","page":"Network analysis and representations","title":"Catalyst.isdetailedbalanced","text":"isdetailedbalanced(rs::ReactionSystem, parametermap; reltol=1e-9, abstol)\n\nConstructively compute whether a kinetic system (a reaction network with a set of rate constants) will admit detailed-balanced equilibrium solutions, using the Wegscheider conditions, Feinberg, 1989. A detailed-balanced solution is one for which the rate of every forward reaction exactly equals its reverse reaction. Accepts a dictionary, vector, or tuple of variable-to-value mappings, e.g. [k1 => 1.0, k2 => 2.0,...].\n\n\n\n\n\n","category":"function"},{"location":"api/network_analysis_api/#Catalyst.robustspecies","page":"Network analysis and representations","title":"Catalyst.robustspecies","text":"robustspecies(rn::ReactionSystem)\n\nReturn a vector of indices corresponding to species that are concentration robust, i.e. for every positive equilbrium, the concentration of species s will be the same.\n\nNote: This function currently only works for networks of deficiency one, and is not currently guaranteed to return *all* the concentration-robust species in the network. Any species returned by the function will be robust, but this may not include all of them. Use with caution. Support for higher deficiency networks and necessary conditions for robustness will be coming in the future.\n\n\n\n\n\n","category":"function"},{"location":"api/network_analysis_api/#Catalyst.reset_networkproperties!","page":"Network analysis and representations","title":"Catalyst.reset_networkproperties!","text":"reset_networkproperties!(rn::ReactionSystem)\n\nClears the cache of various properties (like the netstoichiometry matrix). Use if such properties need to be recalculated for some reason.\n\n\n\n\n\n","category":"function"},{"location":"introduction_to_catalyst/catalyst_for_new_julia_users/#catalyst_for_new_julia_users","page":"Introduction to Catalyst and Julia for New Julia users","title":"Introduction to Catalyst and Julia for New Julia users","text":"The Catalyst tool for the modelling of chemical reaction networks is based in the Julia programming language[1][2]. While experience in Julia programming is advantageous for using Catalyst, it is not necessary for accessing most of its basic features. This tutorial serves as an introduction to Catalyst for those unfamiliar with Julia, while also introducing some basic Julia concepts. Anyone who plans on using Catalyst extensively is recommended to familiarise oneself more thoroughly with the Julia programming language. A collection of resources for learning Julia can be found here, and a full documentation is available here. A more practical (but also extensive) guide to Julia programming can be found here.\n\nJulia can be downloaded here. Generally, it is recommended to use the juliaup tool to install and update Julia. Furthermore, Visual Studio Code is a good IDE with extensive Julia support, and a good default choice.\n\nUsers who are already familiar with Julia can skip to the Introduction to Catalyst tutorial.","category":"section"},{"location":"introduction_to_catalyst/catalyst_for_new_julia_users/#Basic-Julia-usage","page":"Introduction to Catalyst and Julia for New Julia users","title":"Basic Julia usage","text":"On the surface, Julia has many similarities to languages like MATLAB, Python, and R.\n\nValues can be assigned to variables through = sign. Values (possibly stored in variables) can be used for most basic computations.\n\nlength = 2.0\nwidth = 4.0\narea = length * width\n\nFunctions take one or more inputs (enclosed by ()) and return some output. E.g. the min function returns the minimum of two values.\n\nmin(1.0, 3.0)\n\nJulia has a specific help mode, which can be queried for information about any function (including those defined by Catalyst).\n\nEach Julia variable has a specific type, designating what type of value it contains. While not directly required to use Catalyst, this is useful to be aware of. To learn the type of a specific variable, use the typeof function. More information about types can be found here.\n\ntypeof(1.0)\n\nHere, Float64 denotes decimal-valued numbers. Integer-valued numbers instead have the Int64 type.\n\ntypeof(1)\n\nThere exists a large number of Julia types (with even more being defined by various packages). Additional examples include Strings (defined by enclosing text within \" \"):\n\n\"Hello world!\"\n\nand Symbols (defined by pre-appending an expression with :):\n\n:Julia\n\nFinally, we note that the first time some code is run in Julia, it has to be compiled. However, this is only required once per Julia session. Hence, the second time the same code is run, it runs much faster. E.g. try running this line of code first one time, and then one additional time. You will note that the second run is much faster.\n\nrand(100, 100)^3.5\nnothing # hide\n\n(This code creates a random 100x100 matrix, and takes it to the power of 3.5)\n\nThis is useful to know when you e.g. declare, simulate, or plot, a Catalyst model. The first time you run a command there might be a slight delay. However, subsequent runs will be much quicker. This holds even if you make minor adjustments before the second run (such as changing simulation initial conditions).","category":"section"},{"location":"introduction_to_catalyst/catalyst_for_new_julia_users/#catalyst_for_new_julia_users_packages_intro","page":"Introduction to Catalyst and Julia for New Julia users","title":"Installing and activating packages","text":"Due to its native package manager (Pkg), and a registry of almost all packages of relevancy, package management in Julia is unusually easy. Here, we will briefly describe how to install and activate Catalyst (and two additional packages relevant to this tutorial).\n\nTo import a Julia package into a session, you can use the using PackageName command (where PackageName is the name of the package you wish to import). However, before you can do so, it must first be installed on your computer. This is done through the Pkg.add(\"PackageName\") command:\n\nusing Pkg\nPkg.add(\"Catalyst\")\n\nHere, the Julia package manager package (Pkg) is by default installed on your computer when Julia is installed, and can be activated directly. Next, we install an ODE solver from a sub-library of the larger OrdinaryDiffEq package, and install the Plots package for making graphs. We will import the recommended default solver from the OrdinaryDiffEqDefault sub-library. A full list of OrdinaryDiffEq solver sublibraries can be found on the sidebar of this page.\n\nPkg.add(\"OrdinaryDiffEqDefault\")\nPkg.add(\"Plots\")\n\nOnce a package has been installed through the Pkg.add command, this command does not have to be repeated if we restart our Julia session. We can now import all three packages into our current session with:\n\nusing Catalyst\nusing OrdinaryDiffEqDefault\nusing Plots\n\nHere, if we restart Julia, these using commands must be rerun.\n\nA more comprehensive (but still short) introduction to package management in Julia is provided at the end of this documentation page. It contains some useful information and is hence highly recommended reading. For a more detailed introduction to Julia package management, please read the Pkg documentation.","category":"section"},{"location":"introduction_to_catalyst/catalyst_for_new_julia_users/#Simulating-a-basic-Catalyst-model","page":"Introduction to Catalyst and Julia for New Julia users","title":"Simulating a basic Catalyst model","text":"Now that we have some basic familiarity with Julia, and have installed and imported the required packages, we will create and simulate a basic chemical reaction network model using Catalyst.\n\nCatalyst models are created through the @reaction_network macro. For more information on macros, please read the Julia documentation on macros. This documentation is, however, rather advanced (and not required to use Catalyst). We instead recommend that you simply familiarise yourself with the Catalyst syntax, without studying in detail how macros work and what they are.\n\nThe @reaction_network command is followed by the begin keyword, which is followed by one line for each reaction of the model. Each reaction consists of a reaction rate, followed by the reaction itself. The reaction contains a set of substrates and a set of products (what is consumed and produced by the reaction, respectively). These are separated by a --> arrow. Finally, the model ends with the end keyword.\n\nHere, we create a simple birth-death model, where a single species (X) is created at rate b, and degraded at rate d. The model is stored in the variable rn.\n\nrn = @reaction_network begin\n    b, 0 --> X\n    d, X --> 0\nend\n\nFor more information on how to use the Catalyst model creator (also known as the Catalyst DSL), please read the corresponding documentation.\n\nNext, we wish to simulate our model. To do this, we need to provide some additional information to the simulator. This is\n\nThe initial condition. That is, the concentration (or copy numbers) of each species at the start of the simulation.\nThe time span. That is, the time frame over which we wish to run the simulation.\nThe parameter values. That is, the values of the model's parameters for this simulation.\n\nThe initial condition is given as a Vector. This is a type which collects several different values. To declare a vector, the values are specific within brackets, [], and separated by ,. Since we only have one species, the vector holds a single element. In this element, we set the value of X using the :X => 1.0 syntax. Here, we first denote the name of the species (with a : pre-appended, which creates a Symbol), next follows a => and then the value of X. Since we wish to simulate the concentration of X over time, we will let the initial condition be decimal valued.\n\nu0 = [:X => 1.0]\n\nThe timespan sets the time point at which we start the simulation (typically 0.0 is used) and the final time point of the simulation. These are combined into a two-valued tuple. Tuples are similar to vectors, but are enclosed by () and not []. Again, we will let both time points be decimal valued.\n\ntspan = (0.0, 10.0)\n\nFinally, the parameter values are, like the initial conditions, given in a vector. Since we have two parameters (b and d), the parameter vector has two values. We use a similar notation for setting the parameter values as the initial condition (first the parameter, then an arrow, then the value).\n\nparams = [:b => 1.0, :d => 0.2]\n\nPlease read here for more information on vectors and tuples.\n\nNext, before we can simulate our model, we bundle all the required information together in a so-called ODEProblem. Note that the order in which the input (the model, the initial condition, the timespan, and the parameter values) is provided to the ODEProblem matters. E.g. the parameter values cannot be provided as the first argument, but have to be the fourth argument. Here, we save our ODEProblem in the oprob variable.\n\noprob = ODEProblem(rn, u0, tspan, params)\n\nWe can now simulate our model. We do this by providing the ODEProblem to the solve function. We save the output to the sol variable.\n\nsol = solve(oprob)\n\nFinally, we can plot the solution through the plot function.\n\nplot(sol)\n\nHere, the plot shows the time evolution of the concentration of the species X from its initial condition.\n\nFor more information about the numerical simulation package, please see the DifferentialEquations documentation. For more information about the plotting package, please see the Plots documentation.","category":"section"},{"location":"introduction_to_catalyst/catalyst_for_new_julia_users/#Additional-modelling-example","page":"Introduction to Catalyst and Julia for New Julia users","title":"Additional modelling example","text":"To make this introduction more comprehensive, we here provide another example, using a more complicated model. Instead of simulating our model as concentrations evolve over time, we will now simulate the individual reaction events through the Gillespie algorithm (a common approach for adding noise to models).\n\nRemember (unless we have restarted Julia) we do not need to activate our packages (through the using command) again. However, we do need to install, and then import, the JumpProcesses package (just to perform Gillespie, and other jump, simulations)\n\nPkg.add(\"JumpProcesses\")\nusing JumpProcesses\n\nThis time, we will declare a so-called SIR model for an infectious disease. Note that even if this model does not describe a set of chemical reactions, it can be modelled using the same framework. The model consists of 3 species:\n\nS, the amount of susceptible individuals.\nI, the amount of infected individuals.\nR, the amount of recovered (or removed) individuals.\n\nIt also has 2 reaction events:\n\nInfection, where a susceptible individual meets an infected individual and also becomes infected.\nRecovery, where an infected individual recovers from the infection.\n\nEach reaction is also associated with a specific rate (corresponding to a parameter).\n\nb, the infection rate.\nk, the recovery rate.\n\nWe declare the model using the @reaction_network macro, and store it in the sir_model variable.\n\nsir_model = @reaction_network begin\n    b, S + I --> 2I\n    k, I --> R\nend\n\nNote that the first reaction contains two different substrates (separated by a + sign). While there is only a single product (I), two copies of I are produced. The 2 in front of the product I denotes this.\n\nNext, we declare our initial condition, time span, and parameter values. Since we want to simulate the individual reaction events that discretely change the state of our model, we want our initial conditions to be integer-valued. We will start with a mostly susceptible population, but to which a single infected individual has been introduced.\n\nu0 = [:S => 50, :I => 1, :R => 0]\ntspan = (0.0, 10.0)\nparams = [:b => 0.2, :k => 1.0]\nnothing # hide\n\nPreviously we have bundled this information into an ODEProblem (denoting a deterministic ordinary differential equation). Now we wish to simulate our model as a jump process (where each reaction event corresponds to a discrete change in the state of the system). We do this by first processing the inputs to work in a jump model – an extra step needed for jump models that can be avoided for ODE/SDE models – and then creating a JumpProblem from the inputs:\n\nusing JumpProcesses # hide\njprob = JumpProblem(sir_model, u0, tspan, params)\nnothing # hide\n\nFinally, we can now simulate our model using the solve function, and plot the solution using the plot function.\n\nsol = solve(jprob)\nsol = solve(jprob; seed=1234) # hide\nplot(sol)\n\nExercise: Try simulating the model several times. Note that the epidemic doesn't always take off, but sometimes dies out without spreading through the population. Try changing the infection rate (b), determining how this value affects the probability that the epidemic goes through the population.","category":"section"},{"location":"introduction_to_catalyst/catalyst_for_new_julia_users/#catalyst_for_new_julia_users_packages","page":"Introduction to Catalyst and Julia for New Julia users","title":"Package management in Julia","text":"We have previously introduced how to install and activate Julia packages. While this is enough to get started with Catalyst, for long-term users, there are some additional considerations for a smooth experience. These are described here.","category":"section"},{"location":"introduction_to_catalyst/catalyst_for_new_julia_users/#catalyst_for_new_julia_users_packages_environments","page":"Introduction to Catalyst and Julia for New Julia users","title":"Setting up a new Julia environment","text":"Whenever you run Julia, it will run in a specific environment. You can specify any folder on your computer as a Julia environment. Some modes of running Julia will automatically use the environment corresponding to the folder you start Julia in. Others (or if you start Julia in a folder without an environment), will use your default environment. In these cases you can, during your session, switch to another environment. While it is possible to not worry about environments (and always use the default one), this can lead to long-term problems as more packages are installed.\n\nTo activate your current folder as an environment, run the following commands:\n\nusing Pkg\nPkg.activate(\".\")\n\nThis will:\n\nIf your current folder (which can be displayed using the pwd() command) is not designated as a possible Julia environment, designate it as such.\nSwitch your current Julia session to use the current folder's environment.\n\nnote: Note\nIf you check any folder which has been designated as a Julia environment, it contains a Project.toml and a Manifest.toml file. These store all information regarding the corresponding environment. For non-advanced users, it is recommended to never touch these files directly (and instead do so using various functions from the Pkg package, the important ones which are described in the next two subsections).","category":"section"},{"location":"introduction_to_catalyst/catalyst_for_new_julia_users/#catalyst_for_new_julia_users_packages_installing","page":"Introduction to Catalyst and Julia for New Julia users","title":"Installing and importing packages in Julia","text":"Package installation and import have been described previously. However, for the sake of this extended tutorial, let us repeat the description by demonstrating how to install the Latexify.jl package (which enables e.g. displaying Catalyst models in Latex format). First, we import the Julia Package manager (Pkg) (which is required to install Julia packages):\n\nusing Pkg\n\nLatexify is a registered package, so it can be installed directly using:\n\nPkg.add(\"Latexify\")\n\nFinally, to import Latexify into our current Julia session we use:\n\nusing Latexify\n\nHere, using Latexify must be rerun whenever you restart a Julia session. However, you only need to run Pkg.add(\"Latexify\") once to install it on your computer (but possibly additional times to add it to new environments, see the next section).","category":"section"},{"location":"introduction_to_catalyst/catalyst_for_new_julia_users/#catalyst_for_new_julia_users_packages_environment_importance","page":"Introduction to Catalyst and Julia for New Julia users","title":"Why environments are important","text":"We have previously described how to set up new Julia environments, how to install Julia packages, and how to import them into a current session. Let us say that you were to restart Julia in a new folder and activate this as a separate environment. If you then try to import Latexify through using Latexify you will receive an error claiming that Latexify was not found. The reason is that the Pkg.add(\"Latexify\") command actually carries out two separate tasks:\n\nIf Latexify is not already installed on your computer, install it.\nAdd Latexify as an available package to your current environment.\n\nHere, while Catalyst has previously been installed on your computer, it has not been added to the new environment you created. To do so, simply run\n\nusing Pkg\nPkg.add(\"Latexify\")\n\nafter which Catalyst can be imported through using Latexify. You can get a list of all packages available in your current environment using:\n\nPkg.status()\n\nSo, why is this required, and why cannot we simply import any package installed on our computer? The reason is that most packages depend on other packages, and these dependencies may be restricted to only specific versions of these packages. This creates complicated dependency graphs that restrict what versions of what packages are compatible with each other. When you use Pkg.add(\"PackageName\"), only a specific version of that package is actually added (the latest possible version as permitted by the dependency graph). Here, Julia environments both define what packages are available and their respective versions (these versions are also displayed by the Pkg.status() command). By doing this, Julia can guarantee that the packages (and their versions) specified in an environment are compatible with each other.\n\nThe reason why all this is important is that it is highly recommended to, for each project, define a separate environment. To these, only add the required packages. General-purpose environments with a large number of packages often, in the long term, produce package incompatibility issues. While these might not prevent you from installing all desired package, they often mean that you are unable to use the latest version of some packages.\n\nnote: Note\nA not-infrequent cause for reported errors with Catalyst (typically the inability to replicate code in tutorials) is package incompatibilities in large environments preventing the latest version of Catalyst from being installed. Hence, whenever an issue is encountered, it is useful to run Pkg.status() to check whenever the latest version of Catalyst is being used.\n\nSome additional useful Pkg commands are:\n\nPk.rm(\"PackageName\") removes a package from the current environment.\nPkg.update(\"PackageName\"): updates the designated package.\nPkg.update(): updates all packages.\n\nnote: Note\nA useful feature of Julia's environment system is that enables the exact definition of what packages and versions were used to execute a script. This supports e.g. reproducibility in academic research. Here, by providing the corresponding Project.toml and Manifest.toml files, you can enable someone to reproduce the exact program used to perform some set of analyses.\n\n","category":"section"},{"location":"introduction_to_catalyst/catalyst_for_new_julia_users/#Feedback","page":"Introduction to Catalyst and Julia for New Julia users","title":"Feedback","text":"If you are a new Julia user who has used this tutorial, and there was something you struggled with or would have liked to have explained better, please raise an issue. That way, we can continue improving this tutorial. The same goes for any part of the Catalyst documentation: It is written to help new users understand how to use the package, and if it is not doing so successfully we would like to know!\n\n","category":"section"},{"location":"introduction_to_catalyst/catalyst_for_new_julia_users/#References","page":"Introduction to Catalyst and Julia for New Julia users","title":"References","text":"[1]: Torkel E. Loman, Yingbo Ma, Vasily Ilin, Shashi Gowda, Niklas Korsbo, Nikhil Yewale, Chris Rackauckas, Samuel A. Isaacson, Catalyst: Fast and flexible modeling of reaction networks, PLOS Computational Biology (2023).\n\n[2]: Jeff Bezanson, Alan Edelman, Stefan Karpinski, Viral B. Shah, Julia: A Fresh Approach to Numerical Computing, SIAM Review (2017).","category":"section"},{"location":"spatial_modelling/spatial_jump_simulations/#spatial_dspace_jump_simulations","page":"Spatial Jump Simulations","title":"Spatial Jump Simulations","text":"Our introduction to spatial space simulations has already described how to simulate DiscreteSpaceReactionSystems using ODEs. Jump simulations of DiscreteSpaceReactionSystem are carried out using an almost identical approach. Spatial jump simulations in Catalyst are built on top of JumpProcesses.jl's spatial jump simulators, more details on which can be found here.\n\nBelow we perform a spatial jump simulation of a simple birth-death process.\n\nusing Catalyst, JumpProcesses\nbd_model = @reaction_network begin\n    (p,d), 0 <--> X\nend\ndiffusion_rx = @transport_reaction D X\nspace = CartesianGrid((10,10))\ndsrs = DiscreteSpaceReactionSystem(bd_model, [diffusion_rx], space)\n\nu0 = [:X => 0]\ntspan = (0.0, 200.0)\nps = [:p => 10.0, :d => 1.0, :D => 0.1]\ndprob = DiscreteProblem(dsrs, u0, tspan, ps)\njprob = JumpProblem(dsrs, u0, tspan, ps)\nsol = solve(jprob, SSAStepper())\nnothing # hide\n\nWe can now access the values of sol using the interfaces described here, and plot it using the functions described here.\n\nAggregators can be designated through the aggregator key word argument, By default, NSM() is used. However, DirectCRDirect() is also available and expected to perform better for large networks.\n\nnote: Note\nCurrently, spatial jump simulations are only supported when all reaction of the non-spatial ReactionSystem are MassActionJumps, i.e. have constant rates. This means that reactions with e.g. Michaelis-Menten rates are currently not supported.","category":"section"},{"location":"model_creation/dsl_advanced/#dsl_advanced_options","page":"The Catalyst DSL - Advanced Features and Options","title":"The Catalyst DSL - Advanced Features and Options","text":"<details><summary><strong>Environment setup and package installation</strong></summary>\n\nThe following code sets up an environment for running the code on this page.\n\nusing Pkg\nPkg.activate(; temp = true) # Creates a temporary environment, which is deleted when the Julia session ends.\nPkg.add(\"Catalyst\")\nPkg.add(\"Latexify\")\nPkg.add(\"OrdinaryDiffEqDefault\")\nPkg.add(\"OrdinaryDiffEqTsit5\")\nPkg.add(\"Plots\")\n\n</details>\n\n\n   Within the Catalyst DSL, each line can represent either a reaction or an option. The previous DSL tutorial described how to create reactions. This one will focus on options. These are typically used to supply a model with additional information. Examples include the declaration of initial condition/parameter default values, or the creation of observables. \n\nAll option designations begin with a declaration starting with @, followed by its input. E.g. the @observables option allows for the generation of observables. Each option can only be used once within each use of @reaction_network. This tutorial will also describe some additional advanced DSL features that do not involve using an option. \n\nAs a first step, we import Catalyst (which is required to run the tutorial):\n\nusing Catalyst","category":"section"},{"location":"model_creation/dsl_advanced/#dsl_advanced_options_declaring_species_and_parameters","page":"The Catalyst DSL - Advanced Features and Options","title":"Explicit specification of network species and parameters","text":"Previously, we mentioned that the DSL automatically determines which symbols correspond to species and which to parameters. This is done by designating everything that appears as either a substrate or a product as a species, and all remaining quantities as parameters (i.e. those only appearing within rates or stoichiometric constants). Sometimes, one might want to manually override this default behaviour for a given symbol. I.e. consider the following model, where the conversion of a protein P from its inactive form (Pᵢ) to its active form (Pₐ) is catalysed by an enzyme (E). Using the most natural description:\n\ncatalysis_sys = @reaction_network begin\n    k*E, Pᵢ --> Pₐ\nend\n\nE (as well as k) will be considered a parameter, something we can confirm directly:\n\nparameters(catalysis_sys)\n\nIf we want E to be considered a species, we can designate this using the @species option:\n\ncatalysis_sys = @reaction_network begin\n    @species E(t)\n    k*E, Pᵢ --> Pₐ\nend\nparameters(catalysis_sys)\n\nnote: Note\nWhen declaring species using the @species option, the species symbol must be followed by (t). The reason is that species are time-dependent variables, and this time-dependency must be explicitly specified (designation of non-t dependant species is also possible).\n\nSimilarly, the @parameters option can be used to explicitly designate something as a parameter:\n\ncatalysis_sys = @reaction_network begin\n    @parameters k\n    k*E, Pᵢ --> Pₐ\nend\n\nHere, while k is explicitly defined as a parameter, no information is provided about E. Hence, the default case will be used (setting E to a parameter). The @species and @parameter options can be used simultaneously (although a quantity cannot be declared both as a species and a parameter). They may be followed by a full list of all species/parameters, or just a subset.\n\nWhile designating something which would default to a parameter as a species is straightforward, the reverse (creating a parameter which occurs as a substrate or product) is more involved. This is, however, possible, and described here.\n\nRather than listing all species/parameters on a single line after the options, a begin ... end block can be used (listing one species/parameter on each line). E.g. in the following example we use this notation to explicitly designate all species and parameters of the system:\n\ncatalysis_sys = @reaction_network begin\n    @species begin \n        E(t)\n        Pᵢ(t)\n        Pₐ(t)\n    end\n    @parameters begin\n        k\n    end\n    k*E, Pᵢ --> Pₐ\nend\n\nA side-effect of using the @species and @parameter options is that they specify the order in which the species and parameters are stored. I.e. lets check the order of the parameters in the parameters in the following dimerisation model:\n\ndimerisation = @reaction_network begin\n    (p,d), 0 <--> X\n    (kB,kD), 2X <--> X2\nend\nparameters(dimerisation)\n\nThe default order is typically equal to the order with which the parameters (or species) are encountered in the DSL (this is, however, not guaranteed). If we specify the parameters using @parameters, the order used within the option is used instead:\n\ndimerisation = @reaction_network begin\n    @parameters kB kD p d\n    (p,d), 0 <--> X\n    (kB,kD), 2X <--> X2\nend\nparameters(dimerisation)\n\ndanger: Danger\nGenerally, Catalyst and the SciML ecosystem do not guarantee that parameter and species order are preserved throughout various operations on a model. Writing programs that depend on these orders is strongly discouraged. There are, however, some legacy packages which still depend on order (one example can be found here). In these situations, this might be useful. However, in these cases, it is recommended that the user is extra wary, and also checks the order manually. \n\nnote: Note\nThe syntax of the @species and @parameters options is identical to that used by the @species and @parameters macros used in programmatic modelling in Catalyst (for e.g. designating metadata or initial conditions). Hence, if one has learnt how to specify species/parameters using either approach, that knowledge can be transferred to the other one.\n\nGenerally, there are four main reasons for specifying species/parameters using the @species and @parameters options:\n\nTo designate a quantity, that would otherwise have defaulted to a parameter, as a species.\nTo designate default values for parameters/species initial conditions (described here).\nTo designate metadata for species/parameters (described here).\nTo designate a species or parameters that do not occur in reactions, but are still part of the model (e.g a parametric initial condition)\n\nwarning: Warning\nCatalyst's DSL automatically infer species and parameters from the input. However, it only does so for quantities that appear in reactions. Until now this has not been relevant. However, this tutorial will demonstrate cases where species/parameters that are not part of reactions are used. These must be designated using either the @species or @parameters options (or the @variables option, which is described later).","category":"section"},{"location":"model_creation/dsl_advanced/#dsl_advanced_options_default_vals","page":"The Catalyst DSL - Advanced Features and Options","title":"Setting default values for species and parameters","text":"When declaring species/parameters using the @species and @parameters options, one can also assign them default values (by appending them with = followed by the desired default value). E.g here we set X's default initial condition value to 10, and p and d's default values to 10 and 02, respectively:\n\nusing Catalyst # hide\nrn = @reaction_network begin\n    @species X(t)=1.0\n    @parameters p=1.0 d=0.1\n    (p,d), 0 <--> X\nend\n\nNext, if we simulate the model, we do not need to provide values for species or parameters that have default values. In this case all have default values, so both u0 and ps can be empty vectors:\n\nusing OrdinaryDiffEqDefault, Plots\nu0 = []\ntspan = (0.0, 10.0)\np = []\noprob = ODEProblem(rn, u0, tspan, p)\nsol = solve(oprob)\nplot(sol)\n\nIt is still possible to provide values for some (or all) initial conditions/parameters in u0/ps (in which case these overrides the default values):\n\nu0 = [:X => 4.0]\np = [:d => 0.5]\noprob = ODEProblem(rn, u0, tspan, p)\nsol = solve(oprob)\nplot(sol)\n\nIt is also possible to declare a model with default values for only some initial conditions/parameters:\n\nusing Catalyst # hide\nrn = @reaction_network begin\n    @species X(t)=1.0\n    (p,d), 0 <--> X\nend\n\ntspan = (0.0, 10.0)\np = [:p => 1.0, :d => 0.2]\noprob = ODEProblem(rn, u0, tspan, p)\nsol = solve(oprob)\nplot(sol)","category":"section"},{"location":"model_creation/dsl_advanced/#dsl_advanced_options_parametric_initial_conditions","page":"The Catalyst DSL - Advanced Features and Options","title":"Setting parametric initial conditions","text":"In the previous section, we designated default values for initial conditions and parameters. However, the right-hand side of the designation accepts any valid expression (not only numeric values). While this can be used to set up some advanced default values, the most common use case is to designate a species's initial condition as a parameter. E.g. in the following example we represent the initial condition of X using the parameter X₀. \n\nrn = @reaction_network begin\n    @species X(t)=X₀\n    @parameters X₀\n    (p,d), 0 <--> X\nend\n\nPlease note that as the parameter X₀ does not occur as part of any reactions, Catalyst's DSL cannot infer whether it is a species or a parameter. This must hence be explicitly declared. We can now simulate our model while providing X's value through the X₀ parameter:\n\nusing OrdinaryDiffEqTsit5\nu0 = []\np = [:X₀ => 1.0, :p => 1.0, :d => 0.5]\noprob = ODEProblem(rn, u0, tspan, p)\nsol = solve(oprob, Tsit5())\nplot(sol)\n\nWhen setting parametric default values these are called bindings. Bindings act similar to defaults, however, they cannot be override. E.g. in the example above, X's initial condition must be specified through the X₀ paraemter, and can no longer be provided through the u0 vector.","category":"section"},{"location":"model_creation/dsl_advanced/#dsl_advanced_options_species_and_parameters_metadata","page":"The Catalyst DSL - Advanced Features and Options","title":"Designating metadata for species and parameters","text":"Catalyst permits the user to define metadata for species and parameters. This permits the user to assign additional information to these, which can be used for a variety of purposes. Some Catalyst features depend on using metadata (with each such case describing specifically how this is done). Here we will introduce how to set metadata, and describe some common metadata types. \n\nWhenever a species/parameter is declared using the @species/@parameters options, it can be followed by a [] within which the metadata is given. Each metadata entry consists of the metadata's name, followed by a =, followed by its value. E.g. the description metadata allows you to attach a String to a species/parameter. Here we create a simple model where we add descriptions to all species and parameters.\n\nusing Catalyst # hide\ntwo_state_system = @reaction_network begin\n    @species Xᵢ(t) [description=\"X's inactive form\"] Xₐ(t) [description=\" X's active form\"]\n    @parameters kA [description=\"Activation rate\"] kD [description=\"Deactivation rate\"]\n    (ka,kD), Xᵢ <--> Xₐ\nend\n\nA metadata can be given to only a subset of a system's species/parameters, and a quantity can be given several metadata entries. To give several metadata, separate each by a ,. Here we only provide a description for kA, for which we also provide a bounds metadata,\n\ntwo_state_system = @reaction_network begin\n    @parameters kA [description=\"Activation rate\", bounds=(0.01,10.0)]\n    (ka,kD), Xᵢ <--> Xₐ\nend\n\nIt is possible to add both default values and metadata to a parameter/species. In this case, first provide the default value, and next the metadata. I.e. to in the above example set kA's default value to 10 we use\n\ntwo_state_system = @reaction_network begin\n    @parameters kA=1.0 [description=\"Activation rate\", bounds=(0.01,10.0)]\n    (ka,kD), Xᵢ <--> Xₐ\nend\n\nWhen designating metadata for species/parameters in begin ... end blocks the syntax changes slightly. Here, a , must be inserted before the metadata (but after any potential default value). I.e. a version of the previous example can be written as\n\ntwo_state_system = @reaction_network begin\n    @parameters begin\n        kA, [description=\"Activation rate\", bounds=(0.01,10.0)]\n        kD = 1.0, [description=\"Deactivation rate\"]\n    end\n    (kA,kD), Xᵢ <--> Xₐ\nend\n\nEach metadata has its own getter functions. E.g. we can get the description of the parameter kA using ModelingToolkitBase.getdescription:\n\nModelingToolkitBase.getdescription(two_state_system.kA)","category":"section"},{"location":"model_creation/dsl_advanced/#dsl_advanced_options_constant_species","page":"The Catalyst DSL - Advanced Features and Options","title":"Designating constant-valued/fixed species parameters","text":"Catalyst enables the designation of parameters as constantspecies. These parameters can be used as species in reactions, however, their values are not changed by the reaction and remain constant throughout the simulation (unless changed by e.g. the occurrence of an event). Practically, this is done by setting the parameter's isconstantspecies metadata to true. Here, we create a simple reaction where the species X is converted to Xᴾ at rate k. By designating X as a constant species parameter, we ensure that its quantity is unchanged by the occurrence of the reaction.\n\nusing Catalyst # hide\nrn = @reaction_network begin\n    @parameters X [isconstantspecies=true]\n    k, X --> Xᴾ\nend\n\nWe can confirm that Xᴾ is the only species of the system:\n\nspecies(rn)\n\nHere, the produced model is actually identical to if X had simply been a parameter in the reaction's rate:\n\nrn = @reaction_network begin\n    k*X, 0 --> Xᴾ\nend\n\nA common use-case for constant species is when modelling systems where some species are present in such surplus that their amounts the reactions' effect on it is negligible. A system which is commonly modelled this way is the Brusselator.","category":"section"},{"location":"model_creation/dsl_advanced/#dsl_advanced_options_parameter_types","page":"The Catalyst DSL - Advanced Features and Options","title":"Designating parameter types","text":"Sometimes it is desired to designate that a parameter should have a specific type. When supplying this parameter's value to e.g. an ODEProblem, that parameter will then be restricted to that specific type. Designating a type is done by appending the parameter with :: followed by its type. E.g. in the following example we specify that the parameter n (the number of X molecules in the Xn polymer) must be an integer (Int64)\n\nusing Catalyst # hide\npolymerisation_network = @reaction_network begin\n    @parameters n::Int64\n    (kB,kD), n*X <--> Xn\nend\nnothing # hide\n\nGenerally, when simulating models with mixed parameter types, it is recommended to declare parameter values as tuples, rather than vectors, e.g.:\n\nps = (:kB => 0.2, :kD => 1.0, :n => 2)\nnothing # hide\n\nIf a parameter has a type, metadata, and a default value, they are designated in the following order:\n\npolymerisation_network = @reaction_network begin\n    @parameters n::Int64 = 2 [description=\"Parameter n, an integer with defaults value 2.\"]\n    (kB,kD), n*X <--> Xn\nend\nnothing # hide","category":"section"},{"location":"model_creation/dsl_advanced/#dsl_advanced_options_vector_variables","page":"The Catalyst DSL - Advanced Features and Options","title":"Vector-valued species or parameters","text":"Sometimes, one wishes to declare a large number of similar parameters or species. This can be done by creating them as vectors. E.g. below we create a two-state system. However, instead of declaring X1 and X2 (and k1 and k2) as separate entities, we declare them as vectors:\n\nusing Catalyst # hide\ntwo_state_model = @reaction_network begin\n    @parameters k[1:2]\n    @species (X(t))[1:2]\n    (k[1],k[2]), X[1] <--> X[2]\nend\n\nNow, we can also declare our initial conditions and parameter values as vectors as well:\n\nusing OrdinaryDiffEqDefault, Plots # hide\nu0 = [:X => [0.0, 2.0]]\ntspan = (0.0, 1.0)\nps = [:k => [1.0, 2.0]]\noprob = ODEProblem(two_state_model, u0, tspan, ps)\nsol = solve(oprob)\nplot(sol)","category":"section"},{"location":"model_creation/dsl_advanced/#dsl_advanced_options_require_dec","page":"The Catalyst DSL - Advanced Features and Options","title":"Turning off species, parameter, and variable inference","text":"In some cases it may be desirable for Catalyst to not infer species and parameters from the DSL, as in the case of reaction networks with very many variables, or as a sanity check that variable names are written correctly. To turn off inference, simply use the @require_declaration option when using the @reaction_network DSL. This will require every single variable, species, or parameter used within the DSL to be explicitly declared using the @variable, @species, or @parameter options. In the case that the DSL parser encounters an undeclared symbolic, it will error with an UndeclaredSymbolicError and print the reaction or equation that the undeclared symbolic was found in. \n\nusing Catalyst\nrn = @reaction_network begin\n    @require_declaration\n    (k1, k2), A <--> B\nend\n\nRunning the code above will yield the following error: \n\nLoadError: UndeclaredSymbolicError: Unrecognized variables A detected in reaction expression: \"((k1, k2), A <--> B)\". Since the flag @require_declaration is declared, all species must be explicitly declared with the @species macro.\n\nIn order to avoid the error in this case all the relevant species and parameters will have to be declared.\n\n# The following case will not error. \nusing Catalyst\nt = default_t()\nrn = @reaction_network begin\n    @require_declaration\n    @species A(t) B(t)\n    @parameters k1 k2\n    (k1, k2), A <--> B\nend\n\nThe following cases in which the DSL would normally infer variables will all throw errors if @require_declaration is set and the variables are not explicitly declared.\n\nOccurrence of an undeclared species in a reaction, as in the example above.\nOccurrence of an undeclared parameter in a reaction rate expression, as in the reaction line k*n, A --> B.\nOccurrence of an undeclared parameter in the stoichiometry of a species, as in the reaction line k, n*A --> B.\nOccurrence of an undeclared variable in coupled differential or algebraic equation, as in A in @equations D(A) ~ 1 - A^2.\nOccurrence of an undeclared differential variables in coupled differential equation, as in A in @equations D(A) ~ 1 - A^2.\nOccurrence of an undeclared observable in an @observables expression, such as @observables Xtot ~ X1 + X2.","category":"section"},{"location":"model_creation/dsl_advanced/#dsl_advanced_options_naming","page":"The Catalyst DSL - Advanced Features and Options","title":"Naming reaction networks","text":"Each reaction network model has a name. It can be accessed using the nameof function. By default, some generic name is used:\n\nusing Catalyst # hide\nrn = @reaction_network begin\n    (p,d), 0 <--> X\nend\nnameof(rn)\n\nA specific name can be given as an argument between the @reaction_network and the begin. E.g. to name a network my_network we can use:\n\nrn = @reaction_network my_network begin\n    (p,d), 0 <--> X\nend\nnameof(rn)\n\nSetting model names is primarily useful for hierarchical modelling, where network names are appended to the display names of subnetworks' species and parameters.","category":"section"},{"location":"model_creation/dsl_advanced/#dsl_advanced_options_observables","page":"The Catalyst DSL - Advanced Features and Options","title":"Creating observables","text":"Sometimes one might want to use observable variables. These are variables with values that can be computed directly from a system's state (rather than having their values implicitly given by reactions or equations). Observables can be designated using the @observables option. Here, the @observables option is followed by a begin ... end block with one line for each observable. Each line first gives the observable, followed by a ~ (not a =!), followed by an expression describing how to compute it.\n\nLet us consider a model where two species (X and Y) can bind to form a complex (XY, which also can dissociate back into X and Y). If we wish to create a representation for the total amount of X and Y in the system, we can do this by creating observables Xtot and Ytot:\n\nusing Catalyst # hide\nrn = @reaction_network begin\n    @observables begin\n        Xtot ~ X + XY\n        Ytot ~ Y + XY\n    end\n    (kB,kD), X + Y <--> XY\nend\n\nWe can now simulate our model using normal syntax (initial condition values for observables should not, and can not, be provided):\n\nusing OrdinaryDiffEqTsit5\nu0 = [:X => 1.0, :Y => 2.0, :XY => 0.0]\ntspan = (0.0, 10.0)\nps = [:kB => 1.0, :kD => 1.5]\noprob = ODEProblem(rn, u0, tspan, ps)\nsol = solve(oprob, Tsit5())\nnothing # hide\n\nNext, we can use symbolic indexing of our solution object, but with the observable as input. E.g. we can use \n\nsol[:Xtot]\nnothing # hide\n\nto get a vector with Xtot's value throughout the simulation. We can also use\n\nusing Plots\nplot(sol; idxs = :Xtot)\nplot!(ylimit = (minimum(sol[:Xtot])*0.95, maximum(sol[:Xtot])*1.05)) # hide\n\nto plot the observables (rather than the species).\n\nObservables can be defined using complicated expressions containing species, parameters, and variables (but not other observables). In the following example (which uses a parametric stoichiometry) X polymerises to form a complex Xn containing n copies of X. Here, we create an observable describing the total number of X molecules in the system:\n\nrn = @reaction_network begin\n    @observables Xtot ~ X + n*Xn\n    (kB,kD), n*X <--> Xn\nend\nnothing # hide\n\nnote: Note\nIf only a single observable is declared, the begin ... end block is not required and the observable can be declared directly after the @observables option.\n\nMetadata can be supplied to an observable directly after its declaration (but before its formula). If so, the metadata must be separated from the observable with a ,, and the observable plus the metadata encapsulated by (). E.g. to add a description metadata to our observable we can use\n\nrn = @reaction_network begin\n    @observables (Xtot, [description=\"The total amount of X in the system.\"]) ~ X + n*Xn\n    (kB,kD), n*X <--> Xn\nend\nnothing # hide\n\nObservables are by default considered variables (not species). To designate them as a species, they can be pre-declared using the @species option. I.e. Here Xtot becomes a species:\n\nrn = @reaction_network begin\n    @species Xtot(t)\n    @observables Xtot ~ X + n*Xn  \n    (kB,kD), n*X <--> Xn\nend\nnothing # hide\n\nSome final notes regarding observables:\n\nThe left-hand side of the observable declaration must contain a single symbol only (with the exception of metadata, which can also be supplied).\nAll quantities appearing on the right-hand side must be declared elsewhere within the @reaction_network call (either by being part of a reaction, or through the @species, @parameters, or @variables options).\nObservables may not depend on other observables.\nObservables have their dependent variable(s) automatically assigned as the union of the dependent variables of the species and variables on which it depends.","category":"section"},{"location":"model_creation/dsl_advanced/#dsl_advanced_options_ivs","page":"The Catalyst DSL - Advanced Features and Options","title":"Specifying non-time independent variables","text":"Catalyst's ReactionSystem models depend on a time independent variable, and potentially one or more spatial independent variables. By default, the independent variable t is used. We can declare another independent variable (which is automatically used as the default one) using the @ivs option. E.g. to use τ instead of t we can use\n\nusing Catalyst # hide\nrn = @reaction_network begin\n    @ivs τ\n    (ka,kD), Xᵢ <--> Xₐ\nend\nnothing # hide\n\nWe can confirm that Xᵢ and Xₐ depend on τ (and not t):\n\nspecies(rn)\n\nIt is possible to designate several independent variables using @ivs. If so, the first one is considered the default (time) independent variable, while the following one(s) are considered spatial independent variable(s). If we want some species to depend on a non-default independent variable, this has to be explicitly declared. In the following example, both Xᵢ and Xₐ depends both on the t and x independent variables.\n\nrn = @reaction_network begin\n    @ivs t x\n    @species Xᵢ(t,x) Xₐ(t,x)\n    (ka,kD), Xᵢ <--> Xₐ\nend\nspecies(rn)\n\nnote: Note\nSetting spatial independent variables is primarily intended for the modelling of spatial systems on continuous domains. Catalyst's support for this is currently under development. Hence, the utility of specifying spatial independent variables is limited.","category":"section"},{"location":"model_creation/dsl_advanced/#dsl_advanced_options_reaction_metadata","page":"The Catalyst DSL - Advanced Features and Options","title":"Setting reaction metadata","text":"It is possible to supply reactions with metadata, containing some additional information of the reaction. A reaction's metadata follows after its declaration (first using the metadata's name, then a =, then its value) and is encapsulated by [] (where individual entries are separated by ,). Here, we add a description metadata to the reactions of a birth-death process:\n\nusing Catalyst # hide\nbd_model = @reaction_network begin\n    p, 0 --> X, [description=\"Production reaction\"]\n    d, X --> 0, [description=\"Degradation reaction\"]\nend\nnothing # hide\n\nWhen bundling reactions, reaction metadata can be bundled using the same rules as rates. Bellow we re-declare our birth-death process, but on a single line:\n\nbd_model = @reaction_network begin\n    (p,d), 0 <--> X, ([description=\"Production reaction\"], [description=\"Degradation reaction\"])\nend\nnothing # hide\n\nHere we declare a model where we also provide a misc metadata (which can hold any quantity we require) to our birth reaction:\n\nbd_model = @reaction_network begin\n    p, 0 --> X, [description=\"Production reaction\", misc=:value]\n    d, X --> 0, [description=\"Degradation reaction\"]\nend\nnothing # hide\n\nA reaction's metadata can be accessed using specific functions, e.g. Catalyst.hasdescription and Catalyst.getdescription can be used to check if a reaction have a description metadata, and to retrieve it, respectively:\n\nrx = @reaction p, 0 --> X, [description=\"A production reaction\"]\nCatalyst.getdescription(rx)\n\nA list of all available reaction metadata can be found in the api.","category":"section"},{"location":"model_creation/dsl_advanced/#dsl_advanced_options_symbolics_and_DSL","page":"The Catalyst DSL - Advanced Features and Options","title":"Working with symbolic variables and the DSL","text":"We have previously described how Catalyst represents its models symbolically (enabling e.g. symbolic differentiation of expressions stored in models). While Catalyst utilises this for many internal operation, these symbolic representations can also be accessed and harnessed by the user. Primarily, doing so is much easier during programmatic (as opposed to DSL-based) modelling. Indeed, the section on programmatic modelling goes into more details about symbolic representation in models, and how these can be used. It is, however, also ways to utilise these methods during DSL-based modelling. Below we briefly describe two methods for doing so.","category":"section"},{"location":"model_creation/dsl_advanced/#dsl_advanced_options_symbolics_and_DSL_unpack","page":"The Catalyst DSL - Advanced Features and Options","title":"Using @unpack to extract symbolic variables from ReactionSystems","text":"Let us consider a simple birth-death process created using the DSL:\n\nusing Catalyst # hide\nbd_model = @reaction_network begin\n    (p,d), 0 <--> X\nend\nnothing # hide\n\nSince we have not explicitly declared p, d, and X using @parameters and @species, we cannot represent these symbolically (only using Symbols). If we wish to do so, however, we can fetch these into our current scope using the @unpack macro:\n\n@unpack p, d, X = bd_model\nnothing # hide\n\nThis lists first the quantities we wish to fetch (does not need to be the model's full set of parameters and species), then =, followed by the model variable. p, d and X are now symbolic variables in the current scope, just as if they had been declared using @parameters or @species. We can confirm this:\n\nX\n\nNext, we can now use these to e.g. designate initial conditions and parameter values for model simulations:\n\nusing OrdinaryDiffEqDefault, Plots # hide\nu0 = [X => 0.1]\ntspan = (0.0, 10.0)\nps = [p => 1.0, d => 0.2]\noprob = ODEProblem(bd_model, u0, tspan, ps)\nsol = solve(oprob)\nplot(sol)\n\nwarning: Warning\nJust like when using @parameters and @species, @unpack will overwrite any variables in the current scope which share name with the imported quantities.","category":"section"},{"location":"model_creation/dsl_advanced/#dsl_advanced_options_symbolics_and_DSL_interpolation","page":"The Catalyst DSL - Advanced Features and Options","title":"Interpolating variables into the DSL","text":"Catalyst's DSL allows Julia variables to be interpolated for the network name, within rate constant expressions, or for species/stoichiometries within reactions. Using the lower-level symbolic interface we can then define symbolic variables and parameters outside of @reaction_network, which can then be used within expressions in the DSL. \n\nInterpolation is carried out by pre-appending the interpolating variable with a $. For example, here we declare the parameters and species of a birth-death model, and interpolate these into the model:\n\nusing Catalyst # hide\nt = default_t()\n@species X(t)\n@parameters p d\nbd_model = @reaction_network begin\n    ($p, $d), 0 <--> $X\nend\n\nAdditional information (such as default values or metadata) supplied to p, d, and X is carried through to the DSL. However, interpolation for this purpose is of limited value, as such information can be declared within the DSL. However, it is possible to interpolate larger algebraic expressions into the DSL, e.g. here\n\n@species X1(t) X2(t) X3(t) E(t)\n@parameters d\nd_rate = d/(1 + E)\ndegradation_model = @reaction_network begin\n    $d_rate, X1 --> 0\n    $d_rate, X2 --> 0\n    $d_rate, X3 --> 0\nend\n\nwe declare an expression d_rate, which then can be inserted into the DSL via interpolation.\n\nIt is also possible to use interpolation in combination with the @reaction macro. E.g. the reactions of the above network can be declared individually using\n\nrxs = [\n    @reaction $d_rate, $X1 --> 0\n    @reaction $d_rate, $X2 --> 0\n    @reaction $d_rate, $X3 --> 0\n]\nnothing # hide \n\nnote: Note\nWhen using interpolation, expressions like 2$spec won't work; the multiplication symbol must be explicitly included like 2*$spec.","category":"section"},{"location":"model_creation/dsl_advanced/#dsl_advanced_options_reaction_macro","page":"The Catalyst DSL - Advanced Features and Options","title":"Creating individual reactions using the @reaction macro","text":"Catalyst exports a macro @reaction, which can be used to generate a singular Reaction object of the same type which is stored within the ReactionSystem structure (which in turn can be generated by @reaction_network). Generally, @reaction follows identical rules to those of @reaction_network for writing and interpreting reactions (however, bi-directional reactions are not permitted). E.g. here we create a simple dimerisation reaction:\n\nusing Catalyst # hide\nrx_dimerisation = @reaction kD, 2X --> X2\n\nHere, @reaction is followed by a single line consisting of three parts:\n\nA rate (at which the reaction occurs).\nAny number of substrates (which are consumed by the reaction).\nAny number of products (which are produced by the reaction).\n\nIn the next example, we first create a simple SIR model. Then, we specify the same model by instead creating its individual reaction components using the @reaction macro. Finally, we confirm that these are identical to those stored in the initial model (using the reactions function).\n\nsir_model = @reaction_network begin\n α, S + I --> 2I\n β, I --> R\nend\ninfection_rx = @reaction α, S + I --> 2I\nrecovery_rx = @reaction β, I --> R\nsir_rxs = [infection_rx, recovery_rx]\nissetequal(reactions(sir_model), sir_rxs)\n\nOne of the primary uses of the @reaction macro is to provide some of the convenience of the DSL to *programmatic modelling. E.g. here we can combine our reactions to create a ReactionSystem directly, and also confirm that this is identical to the model created through the DSL:\n\nsir_programmatic = complete(ReactionSystem(sir_rxs, default_t(); name = :sir))\nsir_programmatic == sir_model\n\nDuring programmatic modelling, it can be good to keep in mind that already declared symbolic variables can be interpolated. E.g. here we create two production reactions both depending on the same Michaelis-Menten function:\n\nt = default_t()\n@species X(t)\n@parameters v K\nmm_term = Catalyst.mm(X, v, K)\nrx1 = @reaction $mm_term, 0 --> P1\nrx2 = @reaction $mm_term, 0 --> P2\nnothing # hide","category":"section"},{"location":"model_creation/dsl_advanced/#dsl_advanced_options_disable_ma","page":"The Catalyst DSL - Advanced Features and Options","title":"Disabling mass action for reactions","text":"As described previously, Catalyst uses mass action kinetics to generate ODEs from reactions. Here, each reaction generates a term for each of its reactants, which consists of the reaction's rate, substrates, and the reactant's stoichiometry. E.g. the following reaction:\n\nusing Catalyst # hide\nrn = @reaction_network begin\n  k, X --> ∅\nend\n\ngenerates a single term -k*X:\n\nusing Latexify\nlatexify(rn; form = :ode)\n\nIt is possible to remove the substrate contribution by using any of the following non-filled arrows when declaring the reaction: <=, ⇐, ⟽, =>, ⇒, ⟾, ⇔, ⟺. This means that the reaction\n\nrn = @reaction_network begin\n  k, X => ∅\nend\nlatexify(rn; form = :ode)\n\nwill occur at rate dXdt = -k (which might become a problem since X will be degraded at a constant rate even when very small or equal to 0). This functionality allows the user to fully customise the ODEs generated by their models. \n\nNote, stoichiometric coefficients are still included, i.e. the reaction\n\nrn = @reaction_network begin\n  k, 2*X ⇒ ∅\nend\nlatexify(rn; form = :ode)\n\nhas rate dXdt = -2 k.","category":"section"},{"location":"api/core_api/#api","page":"Catalyst.jl API","title":"Catalyst.jl API","text":"","category":"section"},{"location":"api/core_api/#Reaction-network-generation-and-representation","page":"Catalyst.jl API","title":"Reaction network generation and representation","text":"Catalyst provides the @reaction_network macro for generating a complete network, stored as a ReactionSystem, which in turn is composed of Reactions. ReactionSystems can be converted to other ModelingToolkitBase.AbstractSystems, specifically a ModelingToolkitBase.System representing an ODE, SDE, or jump model.\n\nWhen using the @reaction_network macro, Catalyst will automatically attempt to detect what is a species and what is a parameter. Everything that appear as a substrate or product in some reaction will be treated as a species, while all remaining symbols will be considered parameters (corresponding to those symbols that only appear within rate expressions and/or as stoichiometric coefficients). I.e. in\n\nrn = @reaction_network begin\n    k*X, Y --> W\nend\n\nY and W will all be classified as chemical species, while k and X will be classified as parameters.\n\nThe ReactionSystem generated by the @reaction_network macro is a ModelingToolkitBase.AbstractSystem that symbolically represents a system of chemical reactions. In some cases it can be convenient to bypass the macro and directly generate a collection of symbolic Reactions and a corresponding ReactionSystem encapsulating them. Below we illustrate with a simple SIR example how a system can be directly constructed, and demonstrate how to then generate from the ReactionSystem and solve corresponding chemical reaction ODE models, chemical Langevin equation SDE models, and stochastic chemical kinetics jump process models.\n\nusing Catalyst, OrdinaryDiffEqTsit5, StochasticDiffEq, JumpProcesses, Plots\nt = default_t()\n@parameters β γ\n@species S(t) I(t) R(t)\n\nrxs = [Reaction(β, [S,I], [I], [1,1], [2])\n       Reaction(γ, [I], [R])]\n@named rs = ReactionSystem(rxs, t)\nrs = complete(rs)\n\nu₀map    = [S => 999.0, I => 1.0, R => 0.0]\nparammap = [β => 1/10000, γ => 0.01]\ntspan    = (0.0, 250.0)\n\n# solve as ODEs\noprob = ODEProblem(rs, u₀map, tspan, parammap)\nsol = solve(oprob, Tsit5())\np1 = plot(sol, title = \"ODE\")\n\n# solve as SDEs\nsprob = SDEProblem(rs, u₀map, tspan, parammap)\nsol = solve(sprob, EM(), dt=.01, saveat = 2.0)\np2 = plot(sol, title = \"SDE\")\n\n# solve as jump process\nu₀map    = [S => 999, I => 1, R => 0]\njprob = JumpProblem(rs, u₀map, tspan, parammap)\nsol = solve(jprob)\nsol = solve(jprob; seed = 1234) # hide\np3 = plot(sol, title = \"jump\")\nplot(p1, p2, p3; layout = (3,1))\nCatalyst.PNG(plot(p1, p2, p3; layout = (3,1), fmt = :png, dpi = 200)) # hide","category":"section"},{"location":"api/core_api/#api_dsl_options","page":"Catalyst.jl API","title":"Options for the @reaction_network DSL","text":"We have previously described how options permit the user to supply non-reaction information to ReactionSystem created through the DSL. Here follows a list of all options currently available.\n\nparameters: Allows the designation of a set of symbols as system parameters.\nspecies: Allows the designation of a set of symbols as system species.\nvariables: Allows the designation of a set of symbols as system non-species variables.\nivs: Allows the designation of a set of symbols as system independent variables.\ncompounds: Allows the designation of compound species.\nobservables: Allows the designation of compound observables.\ndefault_noise_scaling: Enables the setting of a default noise scaling expression.\ndifferentials: Allows the designation of differentials.\nequations: Allows the creation of algebraic and/or differential equations.\ncontinuous_events: Allows the creation of continuous events.\ndiscrete_events: Allows the creation of discrete events.\nbrownians: Allows the creation of brownian processes that can be used to add noise to non-species variables.\npoissonians: Allows the creation of poissonian processes that can be used to add jump events to non-species variables.\ndiscretes: Creates discrete parameters, i.e. time-dependent parameters.\ncombinatoric_ratelaws: Takes a single option (true or false), which sets whether to use combinatorial rate laws.\nunit_checks: Takes a single option (true or false) controlling whether unit validation runs during DSL construction (false by default).","category":"section"},{"location":"api/core_api/#api_accessor_functions","page":"Catalyst.jl API","title":"ModelingToolkitBase and Catalyst accessor functions","text":"A ReactionSystem is an instance of a ModelingToolkitBase.AbstractSystem, and has a number of fields that can be accessed using the Catalyst API and the ModelingToolkitBase.jl Abstract System Interface. Below we overview these components.\n\nThere are three basic sets of convenience accessors that will return information either from a top-level system, the top-level system and all sub-systems that are also ReactionSystems (i.e. the full reaction-network), or the top-level system, all subs-systems, and all constraint systems (i.e. the full model). To retrieve info from just a base ReactionSystem rn, ignoring sub-systems of rn, one can use the ModelingToolkit accessors (these provide direct access to the corresponding internal fields of the ReactionSystem)\n\nModelingToolkitBase.get_unknowns(rn) is a vector that collects all the species defined within rn, ordered by species and then non-species variables.\nCatalyst.get_species(rn) is a vector of all the species variables in the system. The entries in get_species(rn) correspond to the first length(get_species(rn)) components in get_unknowns(rn).\nModelingToolkitBase.get_ps(rn) is a vector that collects all the parameters defined within reactions in rn.\nModelingToolkitBase.get_eqs(rn) is a vector that collects all the Reactions and Symbolics.Equation defined within rn, ordering all Reactions before Equations.\nCatalyst.get_rxs(rn) is a vector of all the Reactions in rn, and corresponds to the first length(get_rxs(rn)) entries in get_eqs(rn).\nModelingToolkitBase.get_iv(rn) is the independent variable used in the system (usually t to represent time).\nModelingToolkitBase.get_systems(rn) is a vector of all sub-systems of rn.\nModelingToolkitBase.get_initial_conditions(rn) is a dictionary of all the initial values for parameters and species in rn.\n\nThe preceding accessors do not allocate, directly accessing internal fields of the ReactionSystem.\n\nTo retrieve information from the full reaction network represented by a system rn, which corresponds to information within both rn and all sub-systems, one can call:\n\nModelingToolkitBase.unknowns(rn) returns all species and variables across the system, all sub-systems. Species are ordered before non-species variables in unknowns(rn), with the first numspecies(rn) entries in unknowns(rn) being the same as species(rn).\nspecies(rn) is a vector collecting all the chemical species within the system and any sub-systems that are also ReactionSystems.\nModelingToolkitBase.parameters(rn) returns all parameters across the system, all sub-systems.\nModelingToolkitBase.equations(rn) returns all Reactions and all Symbolics.Equations defined across the system, all sub-systems. Reactions  are ordered ahead of Equations with the first numreactions(rn) entries in equations(rn) being the same as reactions(rn).\nreactions(rn) is a vector of all the Reactions within the system and any sub-systems that are also ReactionSystems.\n\nThese accessors will generally allocate new arrays to store their output unless there are no subsystems. In the latter case the usually return the same vector as the corresponding get_* function.\n\nBelow we list the remainder of the Catalyst API accessor functions mentioned above.","category":"section"},{"location":"api/core_api/#Basic-system-properties","page":"Catalyst.jl API","title":"Basic system properties","text":"See Programmatic Construction of Symbolic Reaction Systems for examples and ModelingToolkitBase and Catalyst Accessor Functions for more details on the basic accessor functions.","category":"section"},{"location":"api/core_api/#Coupled-reaction/equation-system-properties","page":"Catalyst.jl API","title":"Coupled reaction/equation system properties","text":"The following system property accessor functions are primarily relevant to reaction system coupled to differential and/or algebraic equations.","category":"section"},{"location":"api/core_api/#Basic-species-properties","page":"Catalyst.jl API","title":"Basic species properties","text":"The following functions permits the querying of species properties.","category":"section"},{"location":"api/core_api/#Symbolic-variable-properties","page":"Catalyst.jl API","title":"Symbolic variable properties","text":"The following function from SymbolicIndexingInterface.jl is useful for getting the name of individual symbolic variables (species, parameters, or non-species variables).","category":"section"},{"location":"api/core_api/#Basic-reaction-properties","page":"Catalyst.jl API","title":"Basic reaction properties","text":"","category":"section"},{"location":"api/core_api/#api_rx_metadata","page":"Catalyst.jl API","title":"Reaction metadata","text":"The following functions permits the retrieval of reaction metadata.","category":"section"},{"location":"api/core_api/#api_system_metadata","page":"Catalyst.jl API","title":"System-level metadata","text":"The following types and functions allow storing and retrieving system-level metadata on a ReactionSystem. These are primarily intended for use by file parsers that need to attach extra mapping data (e.g., initial condition or parameter value maps) to a system.","category":"section"},{"location":"api/core_api/#api_network_extension_and_modification","page":"Catalyst.jl API","title":"Functions to extend or modify a network","text":"ReactionSystems can be programmatically extended using ModelingToolkitBase.extend and ModelingToolkitBase.compose.","category":"section"},{"location":"api/core_api/#network_visualization","page":"Catalyst.jl API","title":"Network visualization","text":"Latexify can be used to convert networks to LaTeX equations by\n\nusing Latexify\nlatexify(rn)\n\nAn optional argument, form allows using latexify to display a reaction network's ODE (as generated by the reaction rate equation) or SDE (as generated by the chemical Langevin equation) form:\n\nlatexify(rn; form=:ode)\n\nlatexify(rn; form=:sde)\n\n(As of writing this, an upstream bug causes the SDE form to be erroneously displayed as the ODE form)\n\nFinally, another optional argument (expand_functions=true) automatically expands functions defined by Catalyst (such as mm). To disable this, set expand_functions=false.\n\nReaction networks can be plotted using the GraphMakie extension, which is loaded whenever all of Catalyst, GraphMakie, and NetworkLayout are loaded (note that a Makie backend, like CairoMakie, must be loaded as well). The two functions for plotting networks are plot_network and plot_complexes, which are two distinct representations. ","category":"section"},{"location":"api/core_api/#api_rate_laws","page":"Catalyst.jl API","title":"Rate laws","text":"As the underlying ReactionSystem is comprised of ModelingToolkitBase expressions, one can directly access the generated rate laws, and using ModelingToolkitBase tooling generate functions or Julia Exprs from them.","category":"section"},{"location":"api/core_api/#Transformations","page":"Catalyst.jl API","title":"Transformations","text":"","category":"section"},{"location":"api/core_api/#Chemistry-related-functionalities","page":"Catalyst.jl API","title":"Chemistry-related functionalities","text":"Various functionalities primarily relevant to modelling of chemical systems (but potentially also in biology).","category":"section"},{"location":"api/core_api/#Unit-validation","page":"Catalyst.jl API","title":"Unit validation","text":"","category":"section"},{"location":"api/core_api/#api_dspace_simulations","page":"Catalyst.jl API","title":"Spatial modelling","text":"The first step of spatial modelling is to create a so-called DiscreteSpaceReactionSystem:\n\nThe following functions can be used to querying the properties of DiscreteSpaceReactionSystems:\n\nIn addition, most accessor functions for normal ReactionSystems (such as species and parameters) works when applied to DiscreteSpaceReactionSystems as well.\n\nThe following two helper functions can be used to create non-uniform parameter values.\n\nThe following functions can be used to access, or change, species or parameter values stored in problems, integrators, and solutions that are based on DiscreteSpaceReactionSystems.\n\nFinally, we provide the following helper functions to plot and animate spatial discrete space simulations.","category":"section"},{"location":"api/core_api/#Catalyst.@reaction_network","page":"Catalyst.jl API","title":"Catalyst.@reaction_network","text":"@reaction_network\n\nMacro for generating chemical reaction network models (Catalyst ReactionSystems). See the (DSL introduction and advantage usage) sections of the Catalyst documentation for more details on the domain-specific language (DSL) that the macro implements. The macro's output (a ReactionSystem structure) is central to Catalyst and its functionality. How to e.g. simulate these is described in the Catalyst documentation.\n\nReturns:\n\nA Catalyst ReactionSystem, i.e. a symbolic model for the reaction network. The returned\n\nsystem is marked complete. To obtain a ReactionSystem that is not marked complete, for example to then use in compositional modelling, see the otherwise equivalent @network_component macro.\n\nExamples: Here we create a basic SIR model. It contains two reactions (infection and recovery):\n\nsir_model = @reaction_network begin\n    c1, S + I --> 2I\n    c2, I --> R\nend\n\nNext, we create a self-activation loop. Here, a single component (X) activates its own production with a Michaelis-Menten function:\n\nsa_loop = @reaction_network begin\n    mm(X,v,K), 0 --> X\n    d, X --> 0\nend\n\nThis model also contains production and degradation reactions, where 0 denotes that there are either no substrates or no products in a reaction.\n\nOptions: In addition to reactions, the macro also supports \"option\" inputs (permitting e.g. the addition of observables). Each option is designated by a tag starting with a @ followed by its input. A list of options can be found here.\n\n\n\n\n\n","category":"macro"},{"location":"api/core_api/#Catalyst.@network_component","page":"Catalyst.jl API","title":"Catalyst.@network_component","text":"@network_component\n\nEquivalent to @reaction_network except the generated ReactionSystem is not marked as complete.\n\n\n\n\n\n","category":"macro"},{"location":"api/core_api/#Catalyst.make_empty_network","page":"Catalyst.jl API","title":"Catalyst.make_empty_network","text":"make_empty_network(; iv=DEFAULT_IV, name=gensym(:ReactionSystem))\n\nConstruct an empty ReactionSystem. iv is the independent variable, usually time, and name is the name to give the ReactionSystem.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.@reaction","page":"Catalyst.jl API","title":"Catalyst.@reaction","text":"@reaction\n\nMacro for generating a single Reaction object using a similar syntax as the @reaction_network macro (but permitting only a single reaction). A more detailed introduction to the syntax can be found in the description of @reaction_network.\n\nThe @reaction macro is followed by a single line consisting of three parts:\n\nA rate (at which the reaction occurs).\nAny number of substrates (which are consumed by the reaction).\nAny number of products (which are produced by the reaction).\n\nThe output is a reaction (just like created using the Reaction constructor).\n\nExamples: Here we create a simple binding reaction and store it in the variable rx:\n\nrx = @reaction k, X + Y --> XY\n\nThe macro will automatically deduce X, Y, and XY to be species (as these occur as reactants) and k as a parameter (as it does not occur as a reactant).\n\nThe @reaction macro provides a more concise notation to the Reaction constructor. I.e. here we create the same reaction using both approaches, and also confirm that they are identical.\n\n# Creates a reaction using the `@reaction` macro.\nrx = @reaction k*v, A + B --> C + D\n\n# Creates a reaction using the `Reaction` constructor.\nt = default_t()\n@parameters k v\n@species A(t) B(t) C(t) D(t)\nrx2 = Reaction(k*v, [A, B], [C, D])\n\n# Confirms that the two approaches yield identical results:\nrx1 == rx2\n\nInterpolation of already declared symbolic variables into @reaction is possible:\n\nt = default_t()\n@parameters k b\n@species A(t)\nex = k*A^2 + t\nrx = @reaction b*$ex*$A, $A --> C\n\nNotes:\n\n@reaction does not support bi-directional type reactions (using <-->) or reaction bundling\n\n(e.g. d, (X,Y) --> 0).\n\nInterpolation of Julia variables into the macro works similarly to the @reaction_network\n\nmacro. See The Reaction DSL tutorial for more details.\n\n\n\n\n\n","category":"macro"},{"location":"api/core_api/#Catalyst.Reaction","page":"Catalyst.jl API","title":"Catalyst.Reaction","text":"struct Reaction{S, T}\n\nOne chemical reaction.\n\nFields\n\nrate: The rate function (excluding mass action terms).\nsubstrates: Reaction substrates.\nproducts: Reaction products.\nsubstoich: The stoichiometric coefficients of the reactants.\nprodstoich: The stoichiometric coefficients of the products.\nnetstoich: The net stoichiometric coefficients of all species changed by the reaction.\nonly_use_rate: false (default) if rate should be multiplied by mass action terms to give the rate law. true if rate represents the full reaction rate law.\n\nmetadata: Contain additional data, such whenever the reaction have a specific noise-scaling expression for the chemical Langevin equation.\n\nExamples\n\nusing Catalyst\nt = default_t()\n@parameters k[1:20]\n@species A(t) B(t) C(t) D(t)\nrxs = [Reaction(k[1], nothing, [A]),            # 0 -> A\n       Reaction(k[2], [B], nothing),            # B -> 0\n       Reaction(k[3],[A],[C]),                  # A -> C\n       Reaction(k[4], [C], [A,B]),              # C -> A + B\n       Reaction(k[5], [C], [A], [1], [2]),      # C -> A + A\n       Reaction(k[6], [A,B], [C]),              # A + B -> C\n       Reaction(k[7], [B], [A], [2], [1]),      # 2B -> A\n       Reaction(k[8], [A,B], [A,C]),            # A + B -> A + C\n       Reaction(k[9], [A,B], [C,D]),            # A + B -> C + D\n       Reaction(k[10], [A], [C,D], [2], [1,1]), # 2A -> C + D\n       Reaction(k[11], [A], [A,B], [2], [1,1]), # 2A -> A + B\n       Reaction(k[12], [A,B,C], [C,D], [1,3,4], [2, 3]),          # A+3B+4C -> 2C + 3D\n       Reaction(k[13], [A,B], nothing, [3,1], nothing),           # 3A+B -> 0\n       Reaction(k[14], nothing, [A], nothing, [2]),               # 0 -> 2A\n       Reaction(k[15]*A/(2+A), [A], nothing; only_use_rate=true), # A -> 0 with custom rate\n       Reaction(k[16], [A], [B]; only_use_rate=true),             # A -> B with custom rate.\n       Reaction(k[17]*A*exp(B), [C], [D], [2], [1]),              # 2C -> D with non constant rate.\n       Reaction(k[18]*B, nothing, [B], nothing, [2]),             # 0 -> 2B with non constant rate.\n       Reaction(k[19]*t, [A], [B]),                                # A -> B with non constant rate.\n       Reaction(k[20]*t*A, [B,C], [D],[2,1],[2])                  # 2A +B -> 2C with non constant rate.\n  ]\n\nNotes:\n\nnothing can be used to indicate a reaction that has no reactants or no products. In this case the corresponding stoichiometry vector should also be set to nothing.\nThe three-argument form assumes all reactant and product stoichiometric coefficients are one.\nPass unit_checks = true to validate unit consistency at construction time. When enabled, checks that all substrates and products share the same units, and that the rate expression has internally consistent additive terms. Default is false.\n\n\n\n\n\n","category":"type"},{"location":"api/core_api/#Catalyst.ReactionSystem","page":"Catalyst.jl API","title":"Catalyst.ReactionSystem","text":"struct ReactionSystem{V<:Catalyst.NetworkProperties} <: ModelingToolkitBase.AbstractSystem\n\nA system of chemical reactions.\n\nFields\n\neqs: The equations (reactions and algebraic/differential) defining the system.\nrxs: The Reactions defining the system.\niv: Independent variable (usually time).\nsivs: Spatial independent variables\nunknowns: All dependent (unknown) variables, species and non-species. Must not contain the independent variable.\nspecies: Dependent unknown variables representing species\nps: Parameter variables. Must not contain the independent variable.\nvar_to_name: Maps Symbol to corresponding variable.\nobserved: Equations for observed variables.\nname: The name of the system\nsystems: Internal sub-systems\ninitial_conditions: The initial values to use when initial conditions and/or parameters are not supplied to problem constructors.\n\nnetworkproperties: NetworkProperties object that can be filled in by API functions. INTERNAL – not considered part of the public API.\ncombinatoric_ratelaws: Sets whether to use combinatoric scalings in rate laws. true by default.\ncontinuous_events: continuous_events: A Vector{SymbolicContinuousCallback} that model events. The integrator will use root finding to guarantee that it steps at each zero crossing.\n\ndiscrete_events: discrete_events: A Vector{SymbolicDiscreteCallback} that models events. Symbolic analog to SciMLBase.DiscreteCallback that executes an affect when a given condition is true at the end of an integration step.\n\ntstops: tstops: A Vector{Any} of extra time points for the integrator to stop at. These can be numeric values or symbolic expressions of parameters and time.\n\nbrownians: Brownian variables for non-reaction noise, created via @brownians.\npoissonians: Poissonian variables for Poisson jump noise, created via @poissonians.\njumps: Non-reaction jumps (VariableRateJump, ConstantRateJump, MassActionJump).\nmetadata: Metadata for the system, to be used by downstream packages.\n\ncomplete: complete: if a model sys is complete, then sys.x no longer performs namespacing.\n\nparent: The hierarchical parent system before simplification that MTK now seems to require for hierarchical namespacing to work in indexing.\n\nExample\n\nContinuing from the example in the Reaction definition:\n\n# simple constructor that infers species and parameters\n@named rs = ReactionSystem(rxs, t)\n\n# allows specification of species and parameters\n@named rs = ReactionSystem(rxs, t, [A,B,C,D], k)\n\nKeyword Arguments:\n\nobserved::Vector{Equation}, equations specifying observed variables.\nsystems::Vector{ReactionSystem}, vector of sub-ReactionSystems.\nname::Symbol, the name of the system (must be provided, or @named must be used).\ninitial_conditions::SymmapT, a dictionary mapping parameters and species to their initial values.\nchecks = true, boolean for whether to run structural checks at construction time.\nunit_checks = false, boolean for whether to perform unit validation at construction time. Uses Catalyst.validate_units / Catalyst.assert_valid_units. Units should be specified with symbolic units (us\"...\") from DynamicQuantities.\nnetworkproperties = NetworkProperties(), cache for network properties calculated via API functions.\ncombinatoric_ratelaws = true, sets the default value of combinatoric_ratelaws used in conversion functions (ode_model, sde_model, jump_model, ss_ode_model, hybrid_model) or when calling problem constructors with the ReactionSystem.\nbalanced_bc_check = true, sets whether to check that BC species appearing in reactions are balanced (i.e appear as both a substrate and a product with the same stoichiometry).\ntstops = [], a vector of extra time points for the integrator to stop at. These can be numeric values or symbolic expressions of parameters and time.\nbrownians, vector of Brownian variables for non-reaction SDE noise (created via @brownians). Auto-discovered from equations in the two-argument constructor.\npoissonians, vector of Poissonian variables for Poisson jump noise (created via @poissonians). Auto-discovered from equations in the two-argument constructor.\njumps, vector of non-reaction jump processes.\n\n\n\n\n\n","category":"type"},{"location":"api/core_api/#Catalyst.species","page":"Catalyst.jl API","title":"Catalyst.species","text":"species(network)\n\nGiven a ReactionSystem, return a vector of all species defined in the system and any subsystems that are of type ReactionSystem. To get the species and non-species variables in the system and all subsystems, including non-ReactionSystem subsystems, uses unknowns(network).\n\nNotes:\n\nIf ModelingToolkitBase.get_systems(network) is non-empty will allocate.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.get_species","page":"Catalyst.jl API","title":"Catalyst.get_species","text":"get_species(sys::ReactionSystem)\n\nReturn the current dependent variables that represent species in sys (toplevel system only).\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.nonspecies","page":"Catalyst.jl API","title":"Catalyst.nonspecies","text":"nonspecies(network)\n\nReturn the non-species variables within the network, i.e. those unknowns for which isspecies == false.\n\nNotes:\n\nAllocates a new array to store the non-species variables.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.reactions","page":"Catalyst.jl API","title":"Catalyst.reactions","text":"reactions(network)\n\nGiven a ReactionSystem, return a vector of all Reactions in the system.\n\nNotes:\n\nIf ModelingToolkitBase.get_systems(network) is not empty, will allocate.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.get_rxs","page":"Catalyst.jl API","title":"Catalyst.get_rxs","text":"get_rxs(sys::ReactionSystem)\n\nReturn the system's Reaction vector (toplevel system only).\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.nonreactions","page":"Catalyst.jl API","title":"Catalyst.nonreactions","text":"nonreactions(network)\n\nReturn the non-reaction equations within the network (i.e. algebraic and differential equations).\n\nNotes:\n\nAllocates a new array to store the non-species variables.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.numspecies","page":"Catalyst.jl API","title":"Catalyst.numspecies","text":"numspecies(network)\n\nReturn the total number of species within the given ReactionSystem and subsystems that are ReactionSystems.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.numparams","page":"Catalyst.jl API","title":"Catalyst.numparams","text":"numparams(network)\n\nReturn the total number of parameters within the given system and all subsystems.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.numreactions","page":"Catalyst.jl API","title":"Catalyst.numreactions","text":"numreactions(network)\n\nReturn the total number of reactions within the given ReactionSystem and subsystems that are ReactionSystems.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.speciesmap","page":"Catalyst.jl API","title":"Catalyst.speciesmap","text":"speciesmap(network)\n\nGiven a ReactionSystem, return a Dictionary mapping species that participate in Reactions to their index within species(network).\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.paramsmap","page":"Catalyst.jl API","title":"Catalyst.paramsmap","text":"paramsmap(network)\n\nGiven a ReactionSystem, return a Dictionary mapping from all parameters that appear within the system to their index within parameters(network).\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.isautonomous","page":"Catalyst.jl API","title":"Catalyst.isautonomous","text":"isautonomous(rs::ReactionSystem)\n\nChecks if a system is autonomous (i.e. no rate or equation depend on the independent variable(s)). Example:\n\nrs1 = @reaction_system\n    (p,d), 0 <--> X\nend\nisautonomous(rs1) # Returns `true`.\n\nrs2 = @reaction_system\n    (p/t,d), 0 <--> X\nend\nisautonomous(rs2) # Returns `false`.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#ModelingToolkitBase.has_alg_equations","page":"Catalyst.jl API","title":"ModelingToolkitBase.has_alg_equations","text":"has_alg_equations(sys::AbstractSystem)\n\nFor a system, returns true if it contain at least one algebraic equation (i.e. that does not contain any differentials).\n\nExample:\n\nusing ModelingToolkitBase\nusing ModelingToolkitBase: t_nounits as t, D_nounits as D\n@parameters p d\n@variables X(t)\neq1 = D(X) ~ p - d*X\neq2 = 0 ~ p - d*X\n@named osys1 = System([eq1], t)\n@named osys2 = System([eq2], t)\n\nhas_alg_equations(osys1) # returns `false`.\nhas_alg_equations(osys2) # returns `true`.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#ModelingToolkitBase.alg_equations","page":"Catalyst.jl API","title":"ModelingToolkitBase.alg_equations","text":"alg_equations(sys::AbstractSystem)\n\nFor a system, returns a vector of all its algebraic equations (i.e. that does not contain any differentials).\n\nExample:\n\nusing ModelingToolkitBase\nusing ModelingToolkitBase: t_nounits as t, D_nounits as D\n@parameters p d\n@variables X(t)\neq1 = D(X) ~ p - d*X\neq2 = 0 ~ p - d*X\n@named osys = System([eq1, eq2], t)\n\nalg_equations(osys) # returns `[0 ~ p - d*X(t)]`.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#ModelingToolkitBase.has_diff_equations","page":"Catalyst.jl API","title":"ModelingToolkitBase.has_diff_equations","text":"has_diff_equations(sys::AbstractSystem)\n\nFor a system, returns true if it contain at least one differential equation (i.e. that contain a differential).\n\nExample:\n\nusing ModelingToolkitBase\nusing ModelingToolkitBase: t_nounits as t, D_nounits as D\n@parameters p d\n@variables X(t)\neq1 = D(X) ~ p - d*X\neq2 = 0 ~ p - d*X\n@named osys1 = System([eq1], t)\n@named osys2 = System([eq2], t)\n\nhas_diff_equations(osys1) # returns `true`.\nhas_diff_equations(osys2) # returns `false`.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#ModelingToolkitBase.diff_equations","page":"Catalyst.jl API","title":"ModelingToolkitBase.diff_equations","text":"diff_equations(sys::AbstractSystem)\n\nFor a system, returns a vector of all its differential equations (i.e. that does contain a differential).\n\nExample:\n\nusing ModelingToolkitBase\nusing ModelingToolkitBase: t_nounits as t, D_nounits as D\n@parameters p d\n@variables X(t)\neq1 = D(X) ~ p - d*X\neq2 = 0 ~ p - d*X\n@named osys = System([eq1, eq2], t)\n\ndiff_equations(osys) # returns `[Differential(t)(X(t)) ~ p - d*X(t)]`.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.isspecies","page":"Catalyst.jl API","title":"Catalyst.isspecies","text":"isspecies(s)\n\nTests if the given symbolic variable corresponds to a chemical species.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.isconstant","page":"Catalyst.jl API","title":"Catalyst.isconstant","text":"Catalyst.isconstant(s)\n\nTests if the given symbolic variable corresponds to a constant species.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.isbc","page":"Catalyst.jl API","title":"Catalyst.isbc","text":"Catalyst.isbc(s)\n\nTests if the given symbolic variable corresponds to a boundary condition species.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.isvalidreactant","page":"Catalyst.jl API","title":"Catalyst.isvalidreactant","text":"isvalidreactant(s)\n\nTest if a species is valid as a reactant (i.e. a species variable or a constant parameter).\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#SymbolicIndexingInterface.getname","page":"Catalyst.jl API","title":"SymbolicIndexingInterface.getname","text":"getname(x)::Symbol\n\nGet the name of a symbolic variable as a Symbol. Acts as the identity function for x::Symbol.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.ismassaction","page":"Catalyst.jl API","title":"Catalyst.ismassaction","text":"ismassaction(rx, rs; rxvars = get_variables(rx.rate),\n                              haveivdep = nothing,\n                              unknownset = Set(unknowns(rs)),\n                              ivset = nothing)\n\nTrue if a given reaction is of mass action form, i.e. rx.rate does not depend on any chemical species that correspond to unknowns of the system, and does not depend explicitly on the independent variable (usually time).\n\nArguments\n\nrx, the Reaction.\nrs, a ReactionSystem containing the reaction.\nOptional: rxvars, Variables which are not in rxvars are ignored as possible dependencies.\nOptional: haveivdep, true if the Reaction rate field explicitly depends on any independent variable (i.e. t or for spatial systems x,y,etc). If not set, will be automatically calculated.\nOptional: unknownset, set of unknowns which if the rxvars are within mean rx is non-mass action.\nOptional: ivset, a Set of the independent variables of the system. If not provided and the system is spatial, i.e. isspatial(rs) == true, it will be created with all the spatial variables and the time variable. If the rate expression contains any element of ivset, then ismassaction(rx,rs) == false. Pass a custom set to control this behavior.\n\nNotes:\n\nNon-integer stoichiometry is treated as non-mass action. This includes symbolic variables/terms or floating point numbers for stoichiometric coefficients.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.dependents","page":"Catalyst.jl API","title":"Catalyst.dependents","text":"dependents(rx, network)\n\nGiven a Reaction and a ReactionSystem, return a vector of the non-constant species and variables the reaction rate law depends on. e.g., for\n\nk*W, 2X + 3Y --> 5Z + W\n\nthe returned vector would be [W(t),X(t),Y(t)].\n\nNotes:\n\nAllocates\nDoes not check for dependents within any subsystems.\nConstant species are not considered dependents since they are internally treated as parameters.\nIf the rate expression depends on a non-species unknown variable that will be included in the dependents, i.e. in\nt = default_t()\n@parameters k\n@variables V(t)\n@species A(t) B(t) C(t)\nrx = Reaction(k*V, [A, B], [C])\n@named rs = ReactionSystem([rx], t)\nissetequal(dependents(rx, rs), [A,B,V]) == true\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.dependants","page":"Catalyst.jl API","title":"Catalyst.dependants","text":"dependents(rx, network)\n\nSee documentation for dependents.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.substoichmat","page":"Catalyst.jl API","title":"Catalyst.substoichmat","text":"substoichmat(rn; sparse=false)\n\nReturns the substrate stoichiometry matrix, S, with S_i j the stoichiometric coefficient of the ith substrate within the jth reaction.\n\nNote:\n\nSet sparse=true for a sparse matrix representation\nNote that constant species are not considered substrates, but just components that modify the associated rate law.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.prodstoichmat","page":"Catalyst.jl API","title":"Catalyst.prodstoichmat","text":"prodstoichmat(rn; sparse=false)\n\nReturns the product stoichiometry matrix, P, with P_i j the stoichiometric coefficient of the ith product within the jth reaction.\n\nNote:\n\nSet sparse=true for a sparse matrix representation\nNote that constant species are not treated as products, but just components that modify the associated rate law.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.netstoichmat","page":"Catalyst.jl API","title":"Catalyst.netstoichmat","text":"netstoichmat(rn, sparse=false)\n\nReturns the net stoichiometry matrix, N, with N_i j the net stoichiometric coefficient of the ith species within the jth reaction.\n\nNotes:\n\nSet sparse=true for a sparse matrix representation\nCaches the matrix internally within rn so subsequent calls are fast.\nNote that constant species are not treated as reactants, but just components that modify the associated rate law. As such they do not contribute to the net stoichiometry matrix.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.reactionrates","page":"Catalyst.jl API","title":"Catalyst.reactionrates","text":"reactionrates(network)\n\nGiven a ReactionSystem, returns a vector of the symbolic reaction rates for each reaction.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.hasnoisescaling","page":"Catalyst.jl API","title":"Catalyst.hasnoisescaling","text":"hasnoisescaling(reaction::Reaction)\n\nReturns true if the input reaction has the noise_scaing metadata field assigned, else false.\n\nArguments:\n\nreaction: The reaction we wish to check for the noise_scaing metadata field.\n\nExample:\n\nreaction = @reaction k, 0 --> X, [noise_scaling=0.0]\nhasnoisescaling(reaction)\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.getnoisescaling","page":"Catalyst.jl API","title":"Catalyst.getnoisescaling","text":"getnoisescaling(reaction::Reaction)\n\nReturns noise_scaing metadata field for the input reaction.\n\nArguments:\n\nreaction: The reaction we wish to retrieve the noise_scaing metadata field.\n\nExample:\n\nreaction = @reaction k, 0 --> X, [noise_scaling=0.0]\ngetnoisescaling(reaction)\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.hasdescription","page":"Catalyst.jl API","title":"Catalyst.hasdescription","text":"hasdescription(reaction::Reaction)\n\nReturns true if the input reaction has the description metadata field assigned, else false.\n\nArguments:\n\nreaction: The reaction we wish to check for the description metadata field.\n\nExample:\n\nreaction = @reaction k, 0 --> X, [description=\"A reaction\"]\nhasdescription(reaction)\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.getdescription","page":"Catalyst.jl API","title":"Catalyst.getdescription","text":"getdescription(reaction::Reaction)\n\nReturns description metadata field for the input reaction.\n\nArguments:\n\nreaction: The reaction we wish to retrieve the description metadata field.\n\nExample:\n\nreaction = @reaction k, 0 --> X, [description=\"A reaction\"]\ngetdescription(reaction)\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.hasmisc","page":"Catalyst.jl API","title":"Catalyst.hasmisc","text":"hasmisc(reaction::Reaction)\n\nReturns true if the input reaction has the misc metadata field assigned, else false.\n\nArguments:\n\nreaction: The reaction we wish to check for the misc metadata field.\n\nExample:\n\nreaction = @reaction k, 0 --> X, [misc=\"A reaction\"]\nhasmisc(reaction)\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.getmisc","page":"Catalyst.jl API","title":"Catalyst.getmisc","text":"getmisc(reaction::Reaction)\n\nReturns misc metadata field for the input reaction.\n\nArguments:\n\nreaction: The reaction we wish to retrieve the misc metadata field.\n\nExample:\n\nreaction = @reaction k, 0 --> X, [misc=\"A reaction\"]\ngetmisc(reaction)\n\nNotes:\n\nThe misc field can contain any valid Julia structure. This mean that Catalyst cannot check it\n\nfor symbolic variables that are added here. This means that symbolic variables (e.g. parameters of species) that are stored here are not accessible to Catalyst. This can cause troubles when e.g. creating a ReactionSystem programmatically (in which case any symbolic variables stored in the misc metadata field should also be explicitly provided to the ReactionSystem constructor).\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.U0Map","page":"Catalyst.jl API","title":"Catalyst.U0Map","text":"U0Map\n\nMetadata key for storing initial condition / species value mappings on a ReactionSystem. Intended for use by file parsers that need to preserve mappings in a format different from initial_conditions.\n\nSee also: has_u0_map, get_u0_map, set_u0_map\n\n\n\n\n\n","category":"type"},{"location":"api/core_api/#Catalyst.ParameterMap","page":"Catalyst.jl API","title":"Catalyst.ParameterMap","text":"ParameterMap\n\nMetadata key for storing parameter value mappings on a ReactionSystem. Intended for use by file parsers that need to preserve parameter mappings in a format different from initial_conditions.\n\nSee also: has_parameter_map, get_parameter_map, set_parameter_map\n\n\n\n\n\n","category":"type"},{"location":"api/core_api/#Catalyst.has_u0_map","page":"Catalyst.jl API","title":"Catalyst.has_u0_map","text":"has_u0_map(rs::ReactionSystem)\n\nReturns true if the ReactionSystem has a U0Map metadata entry.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.get_u0_map","page":"Catalyst.jl API","title":"Catalyst.get_u0_map","text":"get_u0_map(rs::ReactionSystem)\n\nReturns the U0Map metadata from the ReactionSystem, or nothing if no U0Map has been set.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.set_u0_map","page":"Catalyst.jl API","title":"Catalyst.set_u0_map","text":"set_u0_map(rs::ReactionSystem, u0map)\n\nReturns a new ReactionSystem with the U0Map metadata set to u0map. The original system is not modified.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.has_parameter_map","page":"Catalyst.jl API","title":"Catalyst.has_parameter_map","text":"has_parameter_map(rs::ReactionSystem)\n\nReturns true if the ReactionSystem has a ParameterMap metadata entry.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.get_parameter_map","page":"Catalyst.jl API","title":"Catalyst.get_parameter_map","text":"get_parameter_map(rs::ReactionSystem)\n\nReturns the ParameterMap metadata from the ReactionSystem, or nothing if no ParameterMap has been set.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.set_parameter_map","page":"Catalyst.jl API","title":"Catalyst.set_parameter_map","text":"set_parameter_map(rs::ReactionSystem, pmap)\n\nReturns a new ReactionSystem with the ParameterMap metadata set to pmap. The original system is not modified.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#ModelingToolkitBase.extend","page":"Catalyst.jl API","title":"ModelingToolkitBase.extend","text":"extend(\n    sys::ModelingToolkitBase.AbstractSystem,\n    basesys::ModelingToolkitBase.AbstractSystem;\n    name,\n    description,\n    gui_metadata\n) -> ReactionSystem{Catalyst.NetworkProperties{Int64, SymbolicUtils.BasicSymbolicImpl.var\"typeof(BasicSymbolicImpl)\"{SymReal}}}\n\n\nExtend basesys with sys. This can be thought of as the merge operation on systems. Values in sys take priority over duplicates in basesys (for example, initial conditions).\n\nBy default, the resulting system inherits sys's name and description.\n\nThe & operator can also be used for this purpose. sys & basesys is equivalent to extend(sys, basesys).\n\nSee also compose.\n\n\n\n\n\nextend(\n    sys,\n    basesys::Array{T<:ModelingToolkitBase.AbstractSystem, 1}\n) -> Any\n\n\nExtend sys with all systems in basesys in order.\n\n\n\n\n\nModelingToolkitBase.extend(sys::ReactionSystem, rs::ReactionSystem; name::Symbol=nameof(sys))\n\nExtends the indicated ReactionSystem with another ReactionSystem.\n\nNotes:\n\nOnly ReactionSystems can be used to extend a ReactionSystem.\nReturns a new ReactionSystem and does not modify rs.\nBy default, the new ReactionSystem will have the same name as sys.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#ModelingToolkitBase.compose","page":"Catalyst.jl API","title":"ModelingToolkitBase.compose","text":"compose(sys, systems; name)\n\n\nCompose multiple systems together. This adds all of systems as subsystems of sys. The resulting system inherits the name of sys by default.\n\nThe ∘ operator can also be used for this purpose. sys ∘ basesys is equivalent to compose(sys, basesys).\n\nSee also extend.\n\n\n\n\n\ncompose(syss...; name) -> Any\n\n\nSyntactic sugar for adding all systems in syss as the subsystems of first(syss).\n\n\n\n\n\nModelingToolkitBase.compose(sys::ReactionSystem, systems::AbstractArray; name = nameof(sys))\n\nCompose the indicated ReactionSystem with one or more ReactionSystems.\n\nNotes:\n\nOnly ReactionSystems can be composed with a ReactionSystem.\nReturns a new ReactionSystem and does not modify sys.\nBy default, the new ReactionSystem will have the same name as sys.\nBrownians and jumps from subsystems are collected at flatten time via recursive accessors.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#ModelingToolkitBase.flatten","page":"Catalyst.jl API","title":"ModelingToolkitBase.flatten","text":"flatten(sys::System) -> System\nflatten(sys::System, noeqs) -> System\n\n\nFlatten the hierarchical structure of a system, collecting all equations, unknowns, etc. into one top-level system after namespacing appropriately.\n\n\n\n\n\nModelingToolkitBase.flatten(rs::ReactionSystem)\n\nMerges all subsystems of the given ReactionSystem up into rs.\n\nNotes:\n\nReturns a new ReactionSystem that represents the flattened system.\nAll Reactions within subsystems are namespaced and merged into the list of Reactions of rs. The merged list is then available as reactions(rs).\nAll algebraic and differential equations are merged in the equations of rs.\nOnly ReactionSystems are supported as subsystems when flattening.\nrs.networkproperties is reset upon flattening.\nThe default value of combinatoric_ratelaws will be the logical or of all ReactionSystems.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.plot_network-Tuple{ReactionSystem}","page":"Catalyst.jl API","title":"Catalyst.plot_network","text":"plot_network(rn::ReactionSystem; kwargs...)\n\nConverts a ReactionSystem into a GraphMakie plot of the species reaction graph (or Petri net representation). Reactions correspond to small green circles, and  species to blue circles.\n\nNotes:\n\nBlack arrows from species to reactions indicate reactants, and are labelled with their input stoichiometry.\nBlack arrows from reactions to species indicate products, and are labelled with their output stoichiometry.\nRed arrows from species to reactions indicate that species is used within the rate expression. For example, in the reaction k*A, B --> C, there would be a red arrow from A to the reaction node. In k*A, A+B --> C, there would be red and black arrows from A to the reaction node.\n\nFor a list of accepted keyword arguments to the graph plot, please see the GraphMakie documentation.\n\n\n\n\n\n","category":"method"},{"location":"api/core_api/#Catalyst.plot_complexes-Tuple{ReactionSystem}","page":"Catalyst.jl API","title":"Catalyst.plot_complexes","text":"plot_complexes(rn::ReactionSystem; show_rate_labels = false, kwargs...)\n\nCreates a GraphMakie plot of the Catalyst.ReactionComplexs in rn. Reactions correspond to arrows and reaction complexes to blue circles.\n\nNotes:\n\nBlack arrows from complexes to complexes indicate reactions whose rate is a parameter or a Number. i.e. k, A --> B.\nRed arrows from complexes to complexes indicate reactions whose rate constants\n\ndepends on species. i.e. k*C, A --> B for C a species.\n\nThe show_rate_labels keyword, if set to true, will annotate each edge\n\nwith the rate constant for the reaction.\n\nFor a list of accepted keyword arguments to the graph plot, please see the GraphMakie documentation.\n\n\n\n\n\n","category":"method"},{"location":"api/core_api/#Catalyst.oderatelaw","page":"Catalyst.jl API","title":"Catalyst.oderatelaw","text":"oderatelaw(rx; combinatoric_ratelaw=true)\n\nGiven a Reaction, return the symbolic reaction rate law used in generated ODEs for the reaction. Note, for a reaction defined by\n\nk*X*Y, X+Z --> 2X + Y\n\nthe expression that is returned will be k*X(t)^2*Y(t)*Z(t). For a reaction of the form\n\nk, 2X+3Y --> Z\n\nthe expression that is returned will be k * (X(t)^2/2) * (Y(t)^3/6).\n\nNotes:\n\nAllocates\ncombinatoric_ratelaw=true uses factorial scaling factors in calculating the   rate law, i.e. for 2S -> 0 at rate k the ratelaw would be k*S^2/2!. If   combinatoric_ratelaw=false then the ratelaw is k*S^2, i.e. the scaling   factor is ignored.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.jumpratelaw","page":"Catalyst.jl API","title":"Catalyst.jumpratelaw","text":"jumpratelaw(rx; combinatoric_ratelaw=true)\n\nGiven a Reaction, return the symbolic reaction rate law used in generated stochastic chemical kinetics model SSAs for the reaction. Note, for a reaction defined by\n\nk*X*Y, X+Z --> 2X + Y\n\nthe expression that is returned will be k*X^2*Y*Z. For a reaction of the form\n\nk, 2X+3Y --> Z\n\nthe expression that is returned will be k * binomial(X,2) * binomial(Y,3).\n\nNotes:\n\nAllocates\ncombinatoric_ratelaw=true uses binomials in calculating the rate law, i.e. for 2S -> 0 at rate k the ratelaw would be k*S*(S-1)/2. If combinatoric_ratelaw=false then the ratelaw is k*S*(S-1), i.e. the rate law is not normalized by the scaling factor.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.mm","page":"Catalyst.jl API","title":"Catalyst.mm","text":"mm(X,v,K) = v*X / (X + K)\n\nA Michaelis-Menten rate function.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.mmr","page":"Catalyst.jl API","title":"Catalyst.mmr","text":"mmr(X,v,K) = v*K / (X + K)\n\nA repressive Michaelis-Menten rate function.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.hill","page":"Catalyst.jl API","title":"Catalyst.hill","text":"hill(X,v,K,n) = v*(X^n) / (X^n + K^n)\n\nA Hill rate function.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.hillr","page":"Catalyst.jl API","title":"Catalyst.hillr","text":"hillr(X,v,K,n) = v*(K^n) / (X^n + K^n)\n\nA repressive Hill rate function.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.hillar","page":"Catalyst.jl API","title":"Catalyst.hillar","text":"hillar(X,Y,v,K,n) = v*(X^n) / (X^n + Y^n + K^n)\n\nAn activation/repressing Hill rate function.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#ModelingToolkitBase.mtkcompile","page":"Catalyst.jl API","title":"ModelingToolkitBase.mtkcompile","text":"mtkcompile(\n    sys;\n    additional_passes,\n    inputs,\n    outputs,\n    disturbance_inputs,\n    split,\n    kwargs...\n)\n\n\nCompile the given system into a form that ModelingToolkitBase can generate code for. Also performs order reduction for ODEs and handles simple discrete/implicit-discrete systems.\n\nKeyword Arguments\n\nfully_determined=true controls whether or not an error will be thrown if the number of equations don't match the number of inputs, outputs, and equations.\ninputs, outputs and disturbance_inputs are passed as keyword arguments.All inputs get converted to parameters and are allowed to be unconnected, allowing models where n_unknowns = n_equations - n_inputs.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.set_default_noise_scaling","page":"Catalyst.jl API","title":"Catalyst.set_default_noise_scaling","text":"setdefaultnoisescaling(rs::ReactionSystem, noisescaling)\n\nCreates an updated ReactionSystem. This is the old ReactionSystem, but each Reaction that does not have a noise_scaling metadata have its noisescaling metadata updated. The input ReactionSystem is not mutated. Any subsystems of rs have their `noisescaling` metadata updated as well.\n\nArguments:\n\nrs::ReactionSystem: The ReactionSystem which you wish to remake.\nnoise_scaling: The updated noise scaling terms\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.@compound","page":"Catalyst.jl API","title":"Catalyst.@compound","text":"@compound\n\nMacro that creates a compound species, which is composed of smaller component species.\n\nExample:\n\nt = default_t()\n@species C(t) O(t)\n@compound CO2(t) ~ C + 2O\n\nNotes:\n\nThe component species must be defined before using the @compound macro.\n\n\n\n\n\n","category":"macro"},{"location":"api/core_api/#Catalyst.@compounds","page":"Catalyst.jl API","title":"Catalyst.@compounds","text":"@compounds\n\nMacro that creates several compound species, which each is composed of smaller component species. Uses the same syntax as @compound, but with one compound species per line.\n\nExample:\n\nt = default_t()\n@species C(t) H(t) O(t)\n@compounds\n    CH4(t) = C + 4H\n    O2(t) = 2O\n    CO2(t) = C + 2O\n    H2O(t) = 2H + O\nend\n\nNotes:\n\nThe component species must be defined before using the @compound macro.\n\n\n\n\n\n","category":"macro"},{"location":"api/core_api/#Catalyst.iscompound","page":"Catalyst.jl API","title":"Catalyst.iscompound","text":"iscompound(s)\n\nReturns true if the input is a compound species (else false).\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.components","page":"Catalyst.jl API","title":"Catalyst.components","text":"components(s)\n\nReturns a vector with a list of all the components of a compound species (created using e.g. the @compound macro).\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.coefficients","page":"Catalyst.jl API","title":"Catalyst.coefficients","text":"coefficients(s)\n\nReturns a vector with a list of all the stoichiometric coefficients of the components of a compound species (created using e.g. the @compound macro).\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.component_coefficients","page":"Catalyst.jl API","title":"Catalyst.component_coefficients","text":"component_coefficients(s)\n\nReturns a Vector{Pair{Symbol,Int64}} listing, for a compound species (created using e.g. the @compound macro), all the components and their stoichiometric coefficients.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.validate_units-Tuple{Reaction}","page":"Catalyst.jl API","title":"Catalyst.validate_units","text":"validate_units(rx::Reaction; info::String = \"\", warn::Bool = true)\n\nCheck that all substrates and products within the given Reaction have the same units, and that the units of the reaction's rate expression are internally consistent (i.e. if the rate involves sums, each term in the sum has the same units).\n\n\n\n\n\n","category":"method"},{"location":"api/core_api/#Catalyst.validate_units-Tuple{ReactionSystem}","page":"Catalyst.jl API","title":"Catalyst.validate_units","text":"validate_units(rs::ReactionSystem; info::String=\"\", warn::Bool = true)\n\nCheck that all species in the ReactionSystem have the same units, and that the rate laws of all reactions reduce to units of (species units) / (time units). Also validates unit consistency of non-reaction equations. \n\nUses catalyst_get_unit for SymbolicDimensions-preserving unit inference, avoiding the floating-point precision loss that occurs with MTKBase's get_unit when using non-SI units like M or μM.\n\nNotes:\n\nCorrectly handles only_use_rate=true reactions (does not multiply substrate units into the rate).\nAssumes reaction-local rate-expression checks (e.g. additive-term consistency) were already performed on each Reaction (for example via unit_checks = true at Reaction construction time, or by calling validate_units(rx) separately).\nIf all species/time/parameters are unitless, reaction-rate dimensional checks are skipped. This mode assumes rate/equation expressions do not include literal dimensional quantities (for example us\"...\" constants), which are currently unsupported model inputs.\nDoes not check subsystems, use flatten(rs) and then call validate_units on the flattened system if you want to check the full composed system.\nDoes not require that non-species variables have consistent units (outside of the equations in which they appear).\nDoes not handle events or user-provided jumps.\n\n\n\n\n\n","category":"method"},{"location":"api/core_api/#Catalyst.assert_valid_units-Tuple{Reaction}","page":"Catalyst.jl API","title":"Catalyst.assert_valid_units","text":"assert_valid_units(rx::Reaction; info::String = \"\")\n\nRun strict unit validation on a Reaction. Throws UnitValidationError if any unit inconsistency is detected.\n\n\n\n\n\n","category":"method"},{"location":"api/core_api/#Catalyst.assert_valid_units-Tuple{ReactionSystem}","page":"Catalyst.jl API","title":"Catalyst.assert_valid_units","text":"assert_valid_units(rs::ReactionSystem; info::String = \"\")\n\nRun strict unit validation on a ReactionSystem. Throws UnitValidationError if any unit inconsistency is detected.\n\n\n\n\n\n","category":"method"},{"location":"api/core_api/#Catalyst.unit_validation_report-Tuple{Reaction}","page":"Catalyst.jl API","title":"Catalyst.unit_validation_report","text":"unit_validation_report(rx::Reaction; info::String = \"\")\n\nRun unit validation on a Reaction and return a UnitValidationReport containing both overall validity and structured issue diagnostics.\n\n\n\n\n\n","category":"method"},{"location":"api/core_api/#Catalyst.unit_validation_report-Tuple{ReactionSystem}","page":"Catalyst.jl API","title":"Catalyst.unit_validation_report","text":"unit_validation_report(rs::ReactionSystem; info::String = \"\")\n\nRun unit validation on a ReactionSystem and return a UnitValidationReport containing both overall validity and structured issue diagnostics.\n\n\n\n\n\n","category":"method"},{"location":"api/core_api/#Catalyst.UnitValidationIssue","page":"Catalyst.jl API","title":"Catalyst.UnitValidationIssue","text":"UnitValidationIssue\n\nA single unit-validation diagnostic entry.\n\n\n\n\n\n","category":"type"},{"location":"api/core_api/#Catalyst.UnitValidationReport","page":"Catalyst.jl API","title":"Catalyst.UnitValidationReport","text":"UnitValidationReport\n\nStructured output from unit validation.\n\n\n\n\n\n","category":"type"},{"location":"api/core_api/#Catalyst.UnitValidationError","page":"Catalyst.jl API","title":"Catalyst.UnitValidationError","text":"UnitValidationError\n\nException thrown by strict unit-validation entrypoints (e.g. assert_valid_units). Wraps the full validation report.\n\n\n\n\n\n","category":"type"},{"location":"api/core_api/#Catalyst.DiscreteSpaceReactionSystem","page":"Catalyst.jl API","title":"Catalyst.DiscreteSpaceReactionSystem","text":"struct DiscreteSpaceReactionSystem{Q, R, S, T} <: ModelingToolkitBase.AbstractSystem\n\nA representation of a spatial system of chemical reactions on a discrete space.\n\nFields\n\nreactionsystem: The (non-spatial) reaction system within each vertex.\nspatial_reactions: The spatial reactions defined between individual vertices.\ndspace: The discrete space on which the (discrete) spatial system is defined.\nnum_verts: The number of vertices (compartments).\nnum_edges: The number of edges.\nnum_species: The number of species.\nspatial_species: List of species that may move spatially.\nparameters: All parameters related to the discrete space reaction system (both those whose values are tied to vertices and edges).\n\nvertex_parameters: Parameters which values are tied to vertices, e.g. that possibly could have unique values at each vertex of the system.\n\nedge_parameters: Parameters whose values are tied to edges (adjacencies), e.g. that possibly could have unique values at each edge of the system.\n\nedge_iterator: An iterator over all the discrete space's edges. Currently, the format is always a Vector{Pair{Int64,Int64}}. However, in the future, different types could potentially be used for different types of discrete space (E.g. for a Cartesian grid, we do not technically need to enumerate each edge)\n\nname: The name of the discrete space reaction system. Typically taken directly from the base ReactionSystem.\n\nArguments:\n\nrs: The non-spatial ReactionSystem model that is expanded to a spatial model.\nsrs: A vector of spatial reactions. These provide the rules for how species may move spatially.\ndspace: Either a Cartesian grid, a masked grid, or a graph. This describes the discrete space\n\nto which the non-spatial model is expanded.\n\nKeyword Arguments:\n\ndiagonal_connections = false: Only relevant for Cartesian and masked spaces. If true,\n\ndiagonally adjacent compartments are considered adjacent, and spatial reactions in between these are possible.\n\nExample:\n\n# Fetch packages.\nusing Catalyst, OrdinaryDiffEqDefault\nimport CairoMakie\n\n# Creates the `DiscreteSpaceReactionSystem` model.\nrs = @reaction_network begin\n    (p,d), 0 <--> X\nend\ndiffusion_rx = @transport_reaction D X\nspace = CartesianGrid((5,5))\ndsrs = DiscreteSpaceReactionSystem(rs, [diffusion_rx], space)\n\n# Simulates the model (using ODE and jumps).\nu0 = [:X => rand(5,5)]\ntspan = (0.0, 1.0)\nps = [:p => 1.0, :d => 0.5, :D => 0.1]\noprob = ODEProblem(dsrs, u0, tspan, ps)\nosol = solve(oprob)\n\n# Saves an animation of the solution to the file \"dspace_animation.mp4\".\ndspace_animation(osol, :X, dsrs, \"dspace_animation.mp4\")\n\nNotes:\n\nSpatial modelling in Catalyst is still a work in progress, any feedback (or contributions) to this\n\nis highly welcome.\n\nDiscreteSpaceReactionSystems are primarily intended to model systems in discrete space. Modelling\n\ncontinuous space systems with them is possible, but requires the user to determine the discretisation (the space). Better support for continuous space models is a work in progress.\n\nCatalyst contains extensive documentation on spatial modelling, which can be found here.\n\n\n\n\n\n","category":"type"},{"location":"api/core_api/#Catalyst.reactionsystem","page":"Catalyst.jl API","title":"Catalyst.reactionsystem","text":"reactionsystem(dsrs::DiscreteSpaceReactionSystem)\n\nReturns the non-spatial ReactionSystem stored in a DiscreteSpaceReactionSystem.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.spatial_reactions","page":"Catalyst.jl API","title":"Catalyst.spatial_reactions","text":"spatial_reactions(dsrs::DiscreteSpaceReactionSystem)\n\nReturns a vector with all the spatial reactions stored in a DiscreteSpaceReactionSystem.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.dspace","page":"Catalyst.jl API","title":"Catalyst.dspace","text":"dspace(dsrs::DiscreteSpaceReactionSystem)\n\nReturns the dspace (i.e. discrete space) stored in a DiscreteSpaceReactionSystem.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.num_verts","page":"Catalyst.jl API","title":"Catalyst.num_verts","text":"num_verts(dsrs::DiscreteSpaceReactionSystem)\n\nReturns the number of vertices (i.e. compartments) in the discrete space stored in a DiscreteSpaceReactionSystem.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.num_edges","page":"Catalyst.jl API","title":"Catalyst.num_edges","text":"num_edges(dsrs::DiscreteSpaceReactionSystem)\n\nReturns the number of edges (i.e. connections between vertices) in the discrete space stored in a DiscreteSpaceReactionSystem.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.num_species","page":"Catalyst.jl API","title":"Catalyst.num_species","text":"num_species(dsrs::DiscreteSpaceReactionSystem)\n\nReturns the number of species that a DiscreteSpaceReactionSystem contains.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.spatial_species","page":"Catalyst.jl API","title":"Catalyst.spatial_species","text":"spatial_species(dsrs::DiscreteSpaceReactionSystem)\n\nReturns the species that can move spatially in a DiscreteSpaceReactionSystem.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.vertex_parameters","page":"Catalyst.jl API","title":"Catalyst.vertex_parameters","text":"vertex_parameters(dsrs::DiscreteSpaceReactionSystem)\n\nReturns all the parameters of a DiscreteSpaceReactionSystem whose values are tied to vertices.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.edge_parameters","page":"Catalyst.jl API","title":"Catalyst.edge_parameters","text":"edge_parameters(dsrs::DiscreteSpaceReactionSystem)\n\nReturns all the parameters of a DiscreteSpaceReactionSystem whose values are tied to edges.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.edge_iterator","page":"Catalyst.jl API","title":"Catalyst.edge_iterator","text":"edge_iterator(dsrs::DiscreteSpaceReactionSystem)\n\nReturns an iterator over all of the edges in the discrete space stored in a DiscreteSpaceReactionSystem. Each edge is a Pair{Int64, Int64}, taking the source vertex to the destination vertex.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.is_transport_system","page":"Catalyst.jl API","title":"Catalyst.is_transport_system","text":"is_transport_system(dsrs::DiscreteSpaceReactionSystem)\n\nReturns true if all spatial reactions in dsrs are TransportReactions.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.has_cartesian_dspace","page":"Catalyst.jl API","title":"Catalyst.has_cartesian_dspace","text":"has_cartesian_dspace(dsrs::DiscreteSpaceReactionSystem)\n\nReturns true if dsrs was created using a cartesian grid discrete space (e.g. created via CartesianGrid(5,5)). Otherwise, returns false.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.has_masked_dspace","page":"Catalyst.jl API","title":"Catalyst.has_masked_dspace","text":"has_masked_dspace(dsrs::DiscreteSpaceReactionSystem)\n\nReturns true if dsrs was created using a masked grid discrete space (e.g. created via [true true; true false]). Otherwise, returns false.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.has_grid_dspace","page":"Catalyst.jl API","title":"Catalyst.has_grid_dspace","text":"has_grid_dspace(dsrs::DiscreteSpaceReactionSystem)\n\nReturns true if dsrs was created using a cartesian or masked grid discrete space. Otherwise, returns false.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.has_graph_dspace","page":"Catalyst.jl API","title":"Catalyst.has_graph_dspace","text":"has_graph_dspace(dsrs::DiscreteSpaceReactionSystem)\n\nReturns true if dsrs was created using a graph grid discrete space (e.g. created via path_graph(5)). Otherwise, returns false.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.grid_size","page":"Catalyst.jl API","title":"Catalyst.grid_size","text":"grid_size(dsrs::DiscreteSpaceReactionSystem)\n\nReturns the size of dsrs's discrete space (only if it is a cartesian or masked grid discrete space). E.g. for a discrete space CartesianGrid(4,6), (4,6) is returned.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.grid_dims","page":"Catalyst.jl API","title":"Catalyst.grid_dims","text":"grid_dims(dsrs::DiscreteSpaceReactionSystem)\n\nReturns the number of dimensions of dsrs's discrete space (only if it is a cartesian or masked grid discrete space). The output is either 1, 2, or 3.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.make_edge_p_values","page":"Catalyst.jl API","title":"Catalyst.make_edge_p_values","text":"make_edge_p_values(dsrs::DiscreteSpaceReactionSystem, make_edge_p_value::Function)\n\nGenerates edge parameter values for a discrete space reaction system. Only works for (Cartesian or masked) grid discrete spaces (without diagonal adjacencies).\n\nInput:\n\ndsrs: The discrete space reaction system for which values should be generated.\nmake_edge_p_value: a function describing a rule for generating the edge parameter values.\n\nOutput:     - ep_vals: A sparse matrix of size (numverts,numverts) (where numverts is the number of     vertices in dsrs). Here, eps[i,j] is filled only if there is an edge going from vertex i to     vertex j. The value of eps[i,j] is determined by `makeedgepvalue`.\n\nHere, make_edge_p_value should take two arguments, src_vert and dst_vert, which correspond to the grid indices of an edge's source and destination vertices, respectively. It outputs a single value, which is the value assigned to that edge.\n\nExample:     In the following example, we assign the value 0.1 to all edges, except for the one leading from     vertex (1,1) to vertex (1,2), to which we assign the value 1.0.\n\nusing Catalyst\nrn = @reaction_network begin\n    (p,d), 0 <--> X\nend\ntr = @transport_reaction D X\nspace = CartesianGrid((5,5))\ndsrs = DiscreteSpaceReactionSystem(rn, [tr], space)\n\nfunction make_edge_p_value(src_vert, dst_vert)\n    if src_vert == (1,1) && dst_vert == (1,2)\n        return 1.0\n    else\n        return 0.1\n    end\nend\n\nD_vals = make_edge_p_values(dsrs, make_edge_p_value)\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.make_directed_edge_values","page":"Catalyst.jl API","title":"Catalyst.make_directed_edge_values","text":"make_directed_edge_values(dsrs::DiscreteSpaceReactionSystem, x_vals::Tuple{T,T}, y_vals::Tuple{T,T} = (undef,undef),\n                 z_vals::Tuple{T,T} = (undef,undef)) where {T}\n\nGenerates edge parameter values for a discrete space reaction system. Only works for (Cartesian or masked) grid discrete spaces (without diagonal adjacencies). Each dimension (x, and possibly y and z), and direction has assigned its own constant edge parameter value.\n\nInput:     - dsrs: The discrete space reaction system for which values should be generated.     - x_vals::Tuple{T,T}: The values in the increasing (from a lower x index to a higher x index)     and decreasing (from a higher x index to a lower x index) direction along the x dimension.     - y_vals::Tuple{T,T}: The values in the increasing and decreasing direction along the y dimension.     Should only be used for 2 and 3-dimensional grids.     - z_vals::Tuple{T,T}: The values in the increasing and decreasing direction along the z dimension.     Should only be used for 3-dimensional grids.\n\nOutput:     - ep_vals: A sparse matrix of size (numverts,numverts) (where numverts is the number of     vertices in dsrs). Here, eps[i,j] is filled only if there is an edge going from vertex i to     vertex j. The value of eps[i,j] is determined by the `xvals,yvals, andzvals` Tuples,     and vertices i and j's relative position in the grid.\n\nIt should be noted that two adjacent vertices will always be different in exactly a single dimension (x, y, or z). The corresponding tuple determines which value is assigned.\n\nExample:     In the following example, we wish to have diffusion in the x dimension, but a constant flow from     low y values to high y values (so not transportation from high to low y). We achieve it in the     following manner:\n\nusing Catalyst\nrn = @reaction_network begin\n    (p,d), 0 <--> X\nend\ntr = @transport_reaction D X\nspace = CartesianGrid((5,5))\ndsrs = DiscreteSpaceReactionSystem(rn, [tr], space)\n\nD_vals = make_directed_edge_values(dsrs, (0.1, 0.1), (0.1, 0.0))\n\nHere, since we have a 2d grid, we only provide the first two Tuples to make_directed_edge_values.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.spat_getu","page":"Catalyst.jl API","title":"Catalyst.spat_getu","text":"spat_getu(sim_struct, sp, dsrs::DiscreteSpaceReactionSystem)\n\nFor a problem or integrators, retrieves its u values. For non-discrete space models, this is can be done through direct interfacing (e.g. prob[X]). However, for DiscreteSpaceReactionSystem-based problems and integrators, this function must be used instead. The output format depends on the discrete space (a dense array for cartesian grid discrete spaces, a sparse array for masked grid discrete spaces, and a vector for graph discrete spaces). This format is similar to which is used to designate species initial conditions.\n\nArguments:\n\nsim_struct: The simulation structure which u value we wish to retrieve. Can be either a ODEProblem, JumpProblem, or an integrator derived from either of these.\nsp: The species which value we wish to update. Can be provided either in its symbolic form (e.g. X) or as a symbol (e.g. :X).\ndsrs: The DiscreteSpaceReactionSystem which was used to generate the structure we wish to modify.\n\nNotes:\n\nEven if the species is spatially uniform, a full array with its values across all vertices will be retrieved.\n\nExample:\n\n# Prepare `DiscreteSpaceReactionSystem`s.\nusing Catalyst\nrs = @reaction_network begin\n    (k1,k2), X1 <--> X2\nend\ntr = @transport_reaction D X1\ndsrs = DiscreteSpaceReactionSystem(rs, [tr], CartesianGrid((2,3)))\n\n# Prepares a corresponding ODEProblem.\nu0 = [:X1 => [1.0 2.0 3.0; 4.0 5.0 6.0], :X2 => 2.0]\ntspan = (0.0, 50.0)\nps = [:k1 => 2.0, :k2 => 1.0, :D => 0.01]\noprob = ODEProblem(dsrs, u0, tspan, ps)\n\n# Updates the `ODEProblem`.\nspat_getu(oprob, :X1, dsrs) # Retrieves the value of `X1`.\n\n\n\n\n\nspat_getu(sol, sp, dsrs::DiscreteSpaceReactionSystem; t = nothing)\n\nA function for retrieving the solution of a DiscreteSpaceReactionSystem-based simulation on various desired forms. Generally, for DiscreteSpaceReactionSystems, the values in sol is ordered in a way which is not directly interpretable by the user. Furthermore, the normal Catalyst interface for solutions (e.g. sol[:X]) does not work for these solutions. Hence this function is used instead.\n\nThe output is a vector, which in each position contains sp's value (either at a time step of time, depending on the input t). Its shape depends on the discrete space (using a similar form as heterogeneous initial conditions). I.e. for a NxM cartesian grid, the values are NxM matrices. For a masked grid, the values are sparse matrices. For a graph discrete space, the values are vectors (where the value in the n'th position corresponds to sp's value in the n'th vertex).\n\nArguments:\n\nsol: The solution from which we wish to retrieve some values.\nsp: The species which value we wish to update. Can be provided either in its symbolic form (e.g. X) or as a symbol (e.g. :X).\ndsrs: The DiscreteSpaceReactionSystem which was simulated to generate the solution.\nt = nothing: If nothing, we simply return the solution across all saved time steps (default). If t instead is a vector (or range of values), returns the solution interpolated at these time points.\n\nNotes:\n\nThe spat_getu is not optimised for performance. However, it should still be quite performant, but there might be some limitations if called a very large number of times.\nLong-term it is likely that this function gets replaced with a sleeker interface.\n\nExample:\n\nusing Catalyst, OrdinaryDiffEqDefault\n\n# Prepare `DiscreteSpaceReactionSystem`s.\nrs = @reaction_network begin\n    (k1,k2), X1 <--> X2\nend\ntr = @transport_reaction D X1\ndsrs = DiscreteSpaceReactionSystem(rs, [tr], CartesianGrid((2,2)))\n\n# Create problems.\nu0 = [:X1 => 1, :X2 => 2]\ntspan = (0.0, 10.0)\nps = [:k1 => 1, :k2 => 2.0, :D => 0.1]\n\noprob = ODEProblem(dsrs1, u0, tspan, ps)\nosol = solve(oprob)\nspat_getu(osol, :X1, dsrs) # Returns the value of X1 at each time step.\nspat_getu(osol, :X1, dsrs; t = 0.0:10.0) # Returns the value of X1 at times 0.0, 1.0, ..., 10.0\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.spat_setu!","page":"Catalyst.jl API","title":"Catalyst.spat_setu!","text":"spat_setu!(sim_struct, sp, dsrs::DiscreteSpaceReactionSystem, u)\n\nFor a problem or integrators, update its u vector with the input u. For non-discrete space models, this is can be done through direct interfacing (e.g. prob[X] = 1.0). However, for DiscreteSpaceReactionSystem-based problems and integrators, this function must be used instead.\n\nArguments:\n\nsim_struct: The simulation structure which u value we wish to update. Can be either a ODEProblem, JumpProblem, or an integrator derived from either of these.\nsp: The species which value we wish to update. Can be provided either in its symbolic form (e.g. X) or as a symbol (e.g. :X).\ndsrs: The DiscreteSpaceReactionSystem which was used to generate the structure we wish to modify.\nu: The species's new values. Must be given in a form which is also a valid initial input to the ODEProblem/JumpProblem.\n\nExample:\n\n# Prepare `DiscreteSpaceReactionSystem`s.\nusing Catalyst\nrs = @reaction_network begin\n    (k1,k2), X1 <--> X2\nend\ntr = @transport_reaction D X1\ndsrs = DiscreteSpaceReactionSystem(rs, [tr], CartesianGrid((2,3)))\n\n# Prepares a corresponding ODEProblem.\nu0 = [:X1 => [1.0 2.0 3.0; 4.0 5.0 6.0], :X2 => 2.0]\ntspan = (0.0, 50.0)\nps = [:k1 => 2.0, :k2 => 1.0, :D => 0.01]\noprob = ODEProblem(dsrs, u0, tspan, ps)\n\n# Updates the `ODEProblem`.\nspat_setu!(oprob, :X1, dsrs, 0.0) # Sets `X1` to uniformly 0 across the discrete space.\nspat_setu!(oprob, :X2, dsrs, [1.0 0.0 0.0; 0.0 0.0 0.0]) # Sets `X2` to `1.0` in one vertex, and 0 elsewhere.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.spat_getp","page":"Catalyst.jl API","title":"Catalyst.spat_getp","text":"spat_getp(sim_struct, p, dsrs::DiscreteSpaceReactionSystem)\n\nFor a problem or integrators, retrieves its p values. For non-discrete space models, this is can be done through direct interfacing (e.g. prob[p]). However, for DiscreteSpaceReactionSystem-based problems and integrators, this function must be used instead. The output format depends on the discrete space (a dense array for cartesian grid discrete spaces, a sparse array for masked grid discrete spaces, and a vector for graph discrete spaces). This format is similar to what is used to designate parameter initial values.\n\nArguments:\n\nsim_struct: The simulation structure which p value we wish to retrieve. Can be either a ODEProblem,\n\nJumpProblem, or an integrator derived from either of these.\n\np: The species which value we wish to update. Can be provided either in its symbolic form (e.g. k) or as a symbol (e.g. :k).\ndsrs: The DiscreteSpaceReactionSystem which was used to generate the structure we wish to modify.\n\nNotes:\n\nEven if the parameter is spatially uniform, a full array with its values across all vertices will be retrieved.\n\nExample:\n\n# Prepare `DiscreteSpaceReactionSystem`s.\nusing Catalyst\nrs = @reaction_network begin\n    (k1,k2), X1 <--> X2\nend\ntr = @transport_reaction D X1\ndsrs = DiscreteSpaceReactionSystem(rs, [tr], CartesianGrid((2,3)))\n\n# Prepares a corresponding ODEProblem.\nu0 = [:X1 => 1.0, :X2 => 2.0]\ntspan = (0.0, 50.0)\nps = [:k1 => [1.0 2.0 3.0; 4.0 5.0 6.0], :k2 => 1.0, :D => 0.01]\noprob = ODEProblem(dsrs, u0, tspan, ps)\n\n# Updates the `ODEProblem`.\nspat_getp(oprob, :k1, dsrs) # Retrieves the value of `k1`.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.spat_setp!","page":"Catalyst.jl API","title":"Catalyst.spat_setp!","text":"spat_setp!(sim_struct, p, dsrs::DiscreteSpaceReactionSystem, p_val)\n\nFor a problem or integrators, update its p vector with the input p_val. For non-discrete space models, this is can be done through direct interfacing (e.g. prob[p] = 1.0). However, for DiscreteSpaceReactionSystem-based problems and integrators, this function must be used instead.\n\nArguments:\n\nsim_struct: The simulation structure which u value we wish to update. Can be either a ODEProblem, JumpProblem, or an integrator derived from either of these.\np: The species which value we wish to update. Can be provided either in its symbolic form (e.g. k) or as a symbol (e.g. :k).\ndsrs: The DiscreteSpaceReactionSystem which was used to generate the structure we wish to modify.\np_val: The parameter's new values. Must be given in a form which is also a valid initial input to the ODEProblem/JumpProblem.\n\nExample:\n\n# Prepare `DiscreteSpaceReactionSystem`s.\nusing Catalyst\nrs = @reaction_network begin\n    (k1,k2), X1 <--> X2\nend\ntr = @transport_reaction D X1\ndsrs = DiscreteSpaceReactionSystem(rs, [tr], CartesianGrid((2,3)))\n\n# Prepares a corresponding ODEProblem.\nu0 = [:X1 => 1.0, :X2 => 2.0]\ntspan = (0.0, 50.0)\nps = [:k1 => [1.0 2.0 3.0; 4.0 5.0 6.0], :k2 => 1.0, :D => 0.01]\noprob = ODEProblem(dsrs, u0, tspan, ps)\n\n# Updates the `ODEProblem`.\nspat_setp!(oprob, :k1, dsrs, 0.0) # Sets `k1` to uniformly 0 across the discrete space.\nspat_setp!(oprob, :k2, dsrs, [1.0 0.0 0.0; 0.0 0.0 0.0]) # Sets `k2` to `1.0` in one vertex, and 0 elsewhere.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.rebuild_spat_internals!","page":"Catalyst.jl API","title":"Catalyst.rebuild_spat_internals!","text":"rebuild_spat_internals!(sciml_struct)\n\nRebuilds the internal functions for simulating a DiscreteSpaceReactionSystem. Whenever a problem or integrator has had its parameter values updated, this function should be called for the update to be taken into account. For ODE simulations, rebuild_spat_internals! needs only to be called when\n\nAn edge parameter has been updated.\nWhen a parameter with spatially homogeneous values has been given spatially heterogeneous values (or vice versa).\n\nArguments:\n\nsciml_struct: The problem (e.g. an ODEProblem) or an integrator which we wish to rebuild.\n\nNotes:\n\nCurrently does not work for DiscreteProblems, JumpProblems, or their integrators.\nThe function is not built with performance in mind, so avoid calling it multiple times in performance-critical applications.\n\nExample:\n\n# Creates an initial `ODEProblem`\nrs = @reaction_network begin\n    (k1,k2), X1 <--> X2\nend\ntr = @transport_reaction D X1\ngrid = CartesianGrid((2,2))\ndsrs = DiscreteSpaceReactionSystem(rs, [tr], grid)\n\nu0 = [:X1 => 2, :X2 => [5 6; 7 8]]\ntspan = (0.0, 10.0)\nps = [:k1 => 1.5, :k2 => [1.0 1.5; 2.0 3.5], :D => 0.1]\n\noprob = ODEProblem(dsrs, u0, tspan, ps)\n\n# Updates parameter values.\noprob.ps[:ks] = [2.0 2.5; 3.0 4.5]\noprob.ps[:D] = 0.05\n\n# Rebuilds `ODEProblem` to make changes have an effect.\nrebuild_spat_internals!(oprob)\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.dspace_plot","page":"Catalyst.jl API","title":"Catalyst.dspace_plot","text":"dspace_plot(sol, sp, dsrs::DiscreteSpaceReactionSystem, filename::String; t = sol.tspan[2], kwargs...)\n\nCreates a plot of a DiscreteSpaceReactionSystem simulation. The plot is created at the time point specified by t (defaults to the simulation's final time point).\n\nArguments (all spaces):\n\nsol: The simulation we wish to plot.\nsp: The species whose values we wish to plot. Can be provided either in its symbolic form or as a symbol.\ndsrs: The DiscreteSpaceReactionSystem which was simulated.\nt = sol.t[end]: The time point at which we wish to plot the solution\n\nIn addition, depending on the type of space used, the following optional arguments might be relevant.\n\nArguments (1d spaces):\n\nmarkersize = 20: The size of the markers marking each compartment's value.\n\nArguments (Graph & 2d spaces):\n\ncolormap = :BuGn_7: The colour map with which we display the species amounts in the animation.\nplot_min = nothing: The minimum value for the colour scale (values less than this will be set at this value when the colour scale is computed). If nothing, use the simulation's minimum value (across the entire simulation, not just at the plotted time value).\nplot_max = nothing: The maximum value for the colour scale (values more than this will be set at this value when the colour scale is computed). If nothing, use the simulation's minimum value (across the entire simulation, not just at the plotted time value).\n\nArguments (Graph spaces):\n\nnode_size = 50: The size of the compartments in the plot.\nlayout = Spring(): The layout for the graph nodes in the plot. Can be provided as a vector, where the i'th element is a 2-valued tuple (determining the i'th compartment's y and x positions, respectively).\n\nNotes: \n\nFor masked spaces, there are no value displayed for grid points which do not correspond to a compartments.\nThe current plotting interface is a work in progress, and modifications are expected. if you have any feedback, please contact the package authors.\nAdditional arguments can be passed to dspace_plot, which then will be passed to Makie's lines plotting command.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.dspace_animation","page":"Catalyst.jl API","title":"Catalyst.dspace_animation","text":"dspace_animation(sol, sp, dsrs::DiscreteSpaceReactionSystem, filename::String; kwargs...)\n\nCreates an animation of a DiscreteSpaceReactionSystem simulation. The animation is saved to a file,  whose name is provided in the filename argument.\n\nArguments (all spaces):\n\nsol: The simulation we wish to animate.\nsp: The species which values we wish to animate. Can be provided either in its symbolic form or as a symbol.\ndsrs: The DiscreteSpaceReactionSystem which was simulated.\nfilename: The name of the file to which we wish to save the animation.\nnframes = 200: The number of frames in the animation (these are evenly samples across the simulation).\nframerate = 20: The frame rate of the animation.\nttitle = true: Whether to add a title showing the simulation's time throughout the animation.\n\nIn addition, depending on the type of space used, the following optional arguments might be relevant.\n\nArguments (1d spaces):\n\nmarkersize = 20: The size of the markers marking each compartment's value.\nplot_min = nothing: The y-scale's minimum. If nothing, use the simulation's minimum value.\nplot_max = nothing: The y-scale's maximum. If nothing, use the simulation's maximum value.\n\nArguments (Graph & 2d spaces):\n\ncolormap = :BuGn_7: The colour map with which we display the species amounts in the animation.\nplot_min = nothing: The minimum value for the colour scale (values less than this will be set at this value when the colour scale is computed). If nothing, use the simulation's minimum value.\nplot_max = nothing: The maximum value for the colour scale (values more than this will be set at this value when the colour scale is computed). If nothing, use the simulation's minimum value.\n\nArguments (Graph spaces):\n\nnode_size = 50: The size of the compartments in the plot.\nlayout = Spring(): The layout for the graph nodes in the plot. Can be provided as a vector, where the i'th element is a 2-valued tuple (determining the i'th compartment's y and x positions, respectively).\n\nNotes: \n\nFor masked spaces, there are no value displayed for grid points which do not correspond to a compartments.\nThe current animation interface if a work in progress, and modifications are expected. if you have any feedback, please contact the package authors.\nAdditional arguments can be passed to dspace_animation, which then will be passed to Makie's heatmap plotting command.\n\n\n\n\n\n","category":"function"},{"location":"api/core_api/#Catalyst.dspace_kymograph","page":"Catalyst.jl API","title":"Catalyst.dspace_kymograph","text":"dspace_kymograph(sol, sp, dsrs::DiscreteSpaceReactionSystem, kwargs...)\n\nCreates a kymograph of a DiscreteSpaceReactionSystem simulation based on a Cartesian or masked space.  The plot shows the compartments on the y-axis, and the time development of the system's state along the x-axis. Species amounts are shown as a heatmap.\n\nArguments (all spaces):\n\nsol: The simulation we wish to plot.\nsp: The species whose values we wish to plot. Can be provided either in its symbolic form or as a symbol.\ndsrs: The DiscreteSpaceReactionSystem which was simulated.\ncolormap = :BuGn_7: The colour map with which we display the species amounts in the kymograph.\nnframes = 200: The number of time samples which the time series is sampled with.\nplot_min = nothing: The minimum value for the colour scale (values less than this will be set at this value when the colour scale is computed). If nothing, use the simulation's minimum value.\nplot_max = nothing: The maximum value for the colour scale (values more than this will be set at this value when the colour scale is computed). If nothing, use the simulation's minimum value.\n\nNotes: \n\nFor masked spaces, there are no value displayed for grid points which do not correspond to a compartments.\nThe current plotting interface is a work in progress, and modifications are expected. if you have any feedback, please contact the package authors.\nAdditional arguments can be passed to dspace_plot, which then will be passed to Makie's heatmap plotting command.\n\n\n\n\n\n","category":"function"},{"location":"model_creation/coupled_non_crn_models/#coupled_models","page":"Coupled non-reaction network equations to models","title":"Coupled non-reaction network equations to models","text":"<details><summary><strong>Environment setup and package installation</strong></summary>\n\nThe following code sets up an environment for running the code on this page.\n\nusing Pkg\nPkg.activate(; temp = true) # Creates a temporary environment, which is deleted when the Julia session ends.\nPkg.add(\"Catalyst\")\nPkg.add(\"OrdinaryDiffEqTsit5\")\nPkg.add(\"Plots\")\n\n</details>\n\n\n Non-reaction model components can be inserted directly in a Catalyst model. Here we will briefly describe the simplest case: adding an ODE to a model declared through the @reaction_network DSL. The equation is added using the @equations option, after which the equation is written (with D(V) denoting differential with respect to time).\n\nusing Catalyst, OrdinaryDiffEqDefault, Plots\n\n# Create model with a ODE for the variable `V` (e.g. denoting Volume).\nrn = @reaction_network begin\n    @parameters Vmax # Parameters appearing in equations must be explicitly declared as such.\n    @equations D(V) ~ X * (Vmax - V) # `~` (not `=`!) separates equation left and right-hand sides.\n    (p,d), 0 <--> X\nend\n\n# The model can be simulated using normal syntax.\nu0 = [:X => 2.0, :V => 0.5] # Must give an initial condition for `V` as well as `X`.\nps = [:p => 2.0, :d => 0.25, :Vmax => 2.0]\noprob = ODEProblem(rn, u0, 10.0, ps)\nsol = solve(oprob)\nplot(sol)\n\nOther types of model components (algebraic equations and brownian and poissonian processes) are also possible, as are non-DSL approaches for adding these to a model.\n\n</details>\n\n\n In many applications one has additional algebraic or differential equations for non-chemical species that can be coupled to a chemical reaction network model. Catalyst supports coupled differential and algebraic equations, and currently allows conversion of such coupled systems to ModelingToolkitBase ODE, SDE, and nonlinear Systems.\n\nIn this tutorial we'll illustrate how to make use of coupled (i.e. ODE/algebraic) equations. Let's consider a model of a cell with volume V(t) that grows at a rate lambda. For now we'll assume the cell can grow indefinitely. We'll also keep track of one protein P(t), which is produced at a rate proportional V and can be degraded.","category":"section"},{"location":"model_creation/coupled_non_crn_models/#coupled_models_dsl","page":"Coupled non-reaction network equations to models","title":"Coupling ODE via the DSL","text":"The easiest way to include ODEs and algebraic equations is to just include them when using the DSL to specify a model. Here we include an ODE for V(t) along with degradation and production reactions for P(t):\n\nusing Catalyst, OrdinaryDiffEqTsit5, Plots\n\nrn = @reaction_network growing_cell begin\n    # the growth rate\n    @parameters λ = 1.0\n\n    # assume there is no protein initially\n    @species P(t) = 0.0\n\n    # set the initial volume to 1.0\n    @variables V(t) = 1.0\n\n    # the reactions\n    V, 0 --> P\n    1.0, P --> 0\n\n    # the coupled ODE for V(t)\n    @equations begin\n        D(V) ~ λ * V\n    end\nend\n\nWe can now create an ODEProblem from our model and solve it to see how V(t) and P(t) evolve in time:\n\noprob = ODEProblem(rn, [], (0.0, 1.0))\nsol = solve(oprob, Tsit5())\nplot(sol)","category":"section"},{"location":"model_creation/coupled_non_crn_models/#Coupling-ODEs-via-directly-building-a-ReactionSystem","page":"Coupled non-reaction network equations to models","title":"Coupling ODEs via directly building a ReactionSystem","text":"As an alternative to the previous approach, we could have also constructed our ReactionSystem all at once using the symbolic interface:\n\nusing Catalyst, OrdinaryDiffEqTsit5, Plots\n\nt = default_t()\nD = default_time_deriv()\n\n@parameters λ = 1.0\n@variables V(t) = 1.0\n@species P(t) = 0.0\neq = D(V) ~ λ * V\nrx1 = @reaction $V, 0 --> $P\nrx2 = @reaction 1.0, $P --> 0\n@named growing_cell = ReactionSystem([rx1, rx2, eq], t)\ngrowing_cell = complete(growing_cell)\n\noprob = ODEProblem(growing_cell, [], (0.0, 1.0))\nsol = solve(oprob, Tsit5())\nplot(sol)","category":"section"},{"location":"model_creation/coupled_non_crn_models/#coupled_models_extending_systems","page":"Coupled non-reaction network equations to models","title":"Coupling ODEs via extending a system","text":"Finally, we could also construct our model by using compositional modeling. Here we create separate ReactionSystems, one with the reaction part of the model, and one with the differential equation part. Next, we extend the first ReactionSystem with the second one. Let's begin by creating these two systems.\n\nHere, to create differentials with respect to time (for our differential equations), we must import the time differential operator from Catalyst. We do this through D = default_time_deriv(). Here, D(V) denotes the differential of the variable V with respect to time.\n\nusing Catalyst, OrdinaryDiffEqTsit5, Plots\n\nt = default_t()\nD = default_time_deriv()\n\n# set the growth rate to 1.0\n@parameters λ = 1.0\n\n# set the initial volume to 1.0\n@variables V(t) = 1.0\n\n# build the ReactionSystem for dV/dt\neq = [D(V) ~ λ * V]\n@named osys = ReactionSystem(eq, t)\n\n# build the ReactionSystem with no protein initially\nrn = @network_component begin\n    @species P(t) = 0.0\n    $V,   0 --> P\n    1.0, P --> 0\nend\n\nNotice, here we interpolated the variable V with $V to ensure we use the same symbolic unknown variable in the rn as we used in building osys. See the doc section on interpolation of variables for more information. We also use @network_component instead of @reaction_network as when merging systems together Catalyst requires that the systems have not been marked as complete (which indicates to Catalyst that a system is finalized).\n\nWe can now merge the two systems into one complete ReactionSystem model using ModelingToolkitBase.extend:\n\n@named growing_cell = extend(osys, rn)\ngrowing_cell = complete(growing_cell)\n\nWe see that the combined model now has both the reactions and ODEs as its equations. To solve and plot the model we proceed like normal\n\noprob = ODEProblem(growing_cell, [], (0.0, 1.0))\nsol = solve(oprob, Tsit5())\nplot(sol)","category":"section"},{"location":"model_simulation/examples/activation_time_distribution_measurement/#activation_time_distribution_measurement","page":"Measuring the Distribution of System Activation Times","title":"Measuring the Distribution of System Activation Times","text":"<details><summary><strong>Environment setup and package installation</strong></summary>\n\nThe following code sets up an environment for running the code on this page.\n\nusing Pkg\nPkg.activate(; temp = true) # Creates a temporary environment, which is deleted when the Julia session ends.\nPkg.add(\"Catalyst\")\nPkg.add(\"Plots\")\nPkg.add(\"StochasticDiffEq\")\n\n</details>\n\n\n   In this example we will consider a model which, while initially inactive, activates in response to an input. The model is stochastic, causing the activation times to be random. By combining events, callbacks, and stochastic ensemble simulations, we will measure the probability distribution of the activation times (so-called first passage times).\n\nOur model will be a version of the simple self-activation loop (the ensemble simulations of which we have considered previously). Here, we will consider the activation threshold parameter (K) to be activated by an input (at an input time t = 0). Before the input, K is very large (essentially keeping the system inactive). After the input, it is reduced to a lower value (which permits the system to activate). We will model this using two an additional parameter, Kₐ, which represents the active value of K. At the input time (t = 0), an event will change K's value to Kₐ.\n\nusing Catalyst\nsa_model = @reaction_network begin\n    @parameters Kₐ\n    @discrete_events [0.0] => [K => Kₐ]\n    v0 + hill(X,v,K,n), 0 --> X\n    deg, X --> 0\nend\n\nNext, to perform stochastic simulations of the system we will create an SDEProblem. Note that we start the simulation at a time t = -200  0. This ensures that by the input time (t = 0), the system has (more or less) reached its (inactive) steady state distribution. It also means that the activation time can be measured exactly as the simulation time at the time of activation (as this will be the time from the input at t = 0).\n\nu0 = [:X => 10.0]\ntspan = (-200.0, 2000.0)\nps = [:v0 => 0.1, :v => 2.5, :K => 1000.0, :Kₐ => 40.0, :n => 3.0, :deg => 0.01]\nsprob = SDEProblem(sa_model, u0, tspan, ps)\nnothing # hide \n\nWe can now create a simple EnsembleProblem and perform an ensemble simulation (as described here). Please note that the system has an event which modifies its parameters, hence we must add the safetycopy = true argument to EnsembleProblem (else, subsequent simulations would start with K = Kₐ).\n\nusing Plots, StochasticDiffEq\neprob = EnsembleProblem(sprob; safetycopy = true)\nesol = solve(eprob, ImplicitEM(); trajectories = 10)\nplot(esol)\n\nHere we see how, after the input time, the system (randomly) switches from the inactive state to the active one (several examples of this, bistability-based, activation have been studied in the literature, both in models and experiments[1][2]).\n\nNext, we wish to measure the distribution of these activation times. First we will create a callback which terminates the simulation once it has reached a threshold. This both ensures that we do not have to expend unnecessary computer time on the simulation after its activation, and also enables us to measure the activation time as the final time point of the simulation. Here we will use a discrete callback. By looking at the previous simulations, we determine X = 100 as a suitable activation threshold. We use the terminate! function to terminate the simulation once this value has been reached.\n\ncondition(u, t, integrator) = integrator[:X] > 100.0\naffect!(integrator) = terminate!(integrator)\ncallback = DiscreteCallback(condition, affect!)\nnothing # hide\n\nNext, we will perform our ensemble simulation. By default, for each simulation, these save the full trajectory. Here, however, we are only interested in the activation time. This enables us to utilise an output function. This will be called at the end of every single simulation and determine what to save (it can also be used to potentially rerun individual simulations, however, we will not use this feature here). Here we create an output function which saves only the simulation's final time point. We also make it throw a warning if the simulation reaches the end of the simulation time frame (t = 2000) (this indicates a simulation that never activated, the occurrences of such simulation would cause us to underestimate the activation times). Finally, just like previously, we must set safetycopy = true.\n\nfunction output_func(sol, i)\n    (sol.t[end] == tspan[2]) && @warn \"A simulation did not activate during the given time span.\"\n    return (sol.t[end], false)\nend\neprob = EnsembleProblem(sprob; output_func, safetycopy = true)\nesol = solve(eprob, ImplicitEM(); trajectories = 250, callback)\nnothing # hide\n\nFinally, we can plot the distribution of activation times. For this, we will use the histogram function (with the normalize = true argument to create a probability density function). An alternative we also recommend is StatsPlots.jl's density function (which creates a smoothed histogram that is also easier to combine with other plots). The input to density is the activation times (which our output function has saved to esol.u).\n\nhistogram(esol.u; normalize = true, label = \"Activation time distribution\", xlabel = \"Activation time\")\n\nHere we that the activation times take some form of long tail distribution (for non-trivial models like this one, it is generally not possible to identify the activation times as any known statistical distribution).\n\n","category":"section"},{"location":"model_simulation/examples/activation_time_distribution_measurement/#References","page":"Measuring the Distribution of System Activation Times","title":"References","text":"[1]: David Frigola, Laura Casanellas, José M. Sancho, Marta Ibañes, Asymmetric Stochastic Switching Driven by Intrinsic Molecular Noise, PLoS One (2012).\n\n[2]: Christian P Schwall et al., Tunable phenotypic variability through an autoregulatory alternative sigma factor circuit, Molecular Systems Biology (2021).","category":"section"},{"location":"spatial_modelling/discrete_space_simulation_structure_interaction/#dspace_simulation_structure_interaction","page":"Interfacing with Lattice Problems, Integrators, and Solutions","title":"Interfacing with Lattice Problems, Integrators, and Solutions","text":"We have previously described how to retrieve species and parameter values stored in non-spatial problems, integrators, and solutions. This section describes similar workflows for simulations based on DiscreteSpaceReactionSystems.\n\nGenerally, while for non-spatial systems these operations can typically be done by indexing a structure directly, e.g. through\n\nsol[:X]\n\nthere are no equally straightforward interfaces for spatial simulations. Typically, helper functions have to be used, e.g\n\nspat_getu(sol, :X, dsrs)\n\nFurthermore, there are some cases of interfacing which are currently not supported (e.g. updating parameter values in JumpProblems). It is likely that these interfaces will be improved in the future (i.e. by introducing a similar syntax to the current non-spatial one). Finally, we note that many of the functions presented below have not been as extensively optimised for performance as other parts of the Catalyst package. Hence, you should take care when designing workflows which requires using them a large number of times.\n\nnote: Note\nBelow we will describe various features using ODE simulations as examples. However, all interfaces (unless where else is stated) work identically for jump simulations.","category":"section"},{"location":"spatial_modelling/discrete_space_simulation_structure_interaction/#dspace_simulation_structure_interaction_simulation_species","page":"Interfacing with Lattice Problems, Integrators, and Solutions","title":"Retrieving values from space simulations","text":"Let us consider a simulation of a DiscreteSpaceReactionSystem:\n\nusing Catalyst, OrdinaryDiffEqDefault\ntwo_state_model = @reaction_network begin\n    (k1,k2), X1 <--> X2\nend\ndiffusion_rx = @transport_reaction D X1\nspace = CartesianGrid((2,3))\ndsrs = DiscreteSpaceReactionSystem(two_state_model, [diffusion_rx], space)\n\nu0 = [:X1 => [0.0 0.0 0.0; 2.0 2.0 2.0], :X2 => 0.0]\ntspan = (0.0, 1.0)\nps = [:k1 => 2.0, :k2 => 1.0, :D => 0.1]\noprob = ODEProblem(dsrs, u0, tspan, ps)\nsol = solve(oprob)\nnothing # hide\n\nTo retrieve the values of X1 across the simulation we use the spat_getu function. It takes three arguments:\n\nThe solution objects from which we wish to retrieve values.\nThe species which values we wish to retrieve.\nThe DiscreteSpaceReactionSystem which was simulated.\n\nspat_getu(sol, :X1, dsrs)\n\nHere, the output is a vector with X1's value at each simulation time step. How the species's value is represented at each time step depends on the space which was originally used to create the DiscreteSpaceReactionSystem:\n\nFor Cartesian spaces, an array of the same size as the Cartesian space is used. Each array element corresponds to the species's value in the corresponding compartment.\nFor masked spaces, a sparse array of the same size as the masked space is used. Each filled array element corresponds to the species's value in the corresponding compartment. Unfilled array elements correspond to positions without compartments.\nFor unstructured (graph) spaces, vectors are used. The i'th element in the vectors corresponds to the species's value in the i'th compartment.\n\nUnlike for non-spatial simulations, spat_getu does not take vector (e.g. spat_getu(sol, [:X1, :X2], dsrs)) or symbolic expression (e.g. spat_getu(sol, [X1 + X2], dsrs)) inputs. However, it is possible to use symbolic variables as input (e.g. spat_getu(sol, two_state_model.X1, dsrs)).","category":"section"},{"location":"spatial_modelling/discrete_space_simulation_structure_interaction/#dspace_simulation_structure_interaction_simulation_species_ts","page":"Interfacing with Lattice Problems, Integrators, and Solutions","title":"Retrieving space simulations values at specific time points","text":"Just like for non-spatial solutions, it is possible to access the simulation's values at designated time points. This is possible even if the simulation did not stop at those specific time points (in which case an interpolated value is returned). To do this, the desired time points to sample are provided as a vector to spat_getu using the optional argument t. E.g. here we retrieve the simulation's (interpolated) values at time points 0.5 and 0.75:\n\nspat_getu(sol, :X1, dsrs; t = [0.5, 0.75])","category":"section"},{"location":"spatial_modelling/discrete_space_simulation_structure_interaction/#dspace_simulation_structure_interaction_prob_int_species","page":"Interfacing with Lattice Problems, Integrators, and Solutions","title":"Retrieving and updating species values in problems and integrators","text":"Let us consider a spatial ODEProblem\n\nusing Catalyst, OrdinaryDiffEqDefault\ntwo_state_model = @reaction_network begin\n    (k1,k2), X1 <--> X2\nend\ndiffusion_rx = @transport_reaction D X1\nspace = CartesianGrid((2,3))\ndsrs = DiscreteSpaceReactionSystem(two_state_model, [diffusion_rx], space)\n\nu0 = [:X1 => [0.0 0.0 0.0; 2.0 2.0 2.0], :X2 => 0.0]\ntspan = (0.0, 1.0)\nps = [:k1 => 2.0, :k2 => 1.0, :D => 0.1]\noprob = ODEProblem(dsrs, u0, tspan, ps)\nnothing # hide\n\nWe can retrieve the species values stored in oprob using the spat_getu function. It uses identical syntax as for simulations (except that you cannot specify a time point). However, it returns a single set of species values (while for simulations it returns a vector across different time steps):\n\nspat_getu(oprob, :X1, dsrs)\n\nAgain, the format used corresponds to the space used to create the original DiscreteSpaceReactionSystem. Here, even if a species has homogeneous values, the full format is used.\n\nspat_getu(oprob, :X2, dsrs)\n\nFor both problems and integrators, species values can be updated using the spat_setu! function. It uses a similar syntax as spat_getu, but takes a fourth argument which is the new values to use for the designated species:\n\nspat_setu!(oprob, :X1, dsrs, [1.0 2.0 3.0; 4.0 5.0 6.0])\n\nHere, the same format (which depends on the used space) is used for the species's new values, as which is used when initially designating their initial conditions. I.e. to make X1's initial condition values uniform we can call\n\nspat_setu!(oprob, :X1, dsrs, 1.0)\n\nnote: Note\nIt is currently not possible to change a species value at a single compartment only. To do so, you must first retrieve its values across all compartments using spat_getu, then modify this at the desired compartment, and then use the modified version as input to spat_setu!.\n\nSpecies values in integrators can be interfaced with using identical syntax as for problems.","category":"section"},{"location":"spatial_modelling/discrete_space_simulation_structure_interaction/#dspace_simulation_structure_interaction_prob_int_parameters","page":"Interfacing with Lattice Problems, Integrators, and Solutions","title":"Retrieving and updating parameter values in problems and integrators","text":"Retrieval and updating of parameter values for problems and integrators works similarly as for species, but with the following differences:\n\nThe spat_getp and spat_setp! functions are used.\nIt is currently not possible to interface with parameter values of JumpProblems and their integrators.\nAfter parameter values are modified, the rebuild_spat_internals! function must be applied before the problem/integrator can be used for further analysis.\nUpdating of edge parameters is limited and uses a different interface. \n\nLet us consider the spatial ODEProblem we previously declared. We can check the value of k1 by using spat_getp\n\nspat_getp(oprob, :k1, dsrs)\n\nNext, we can update it using spat_setp! (here we also confirm that it now has the updated values):\n\nspat_setp!(oprob, :k1, dsrs, [1.0 2.0 3.0; 4.0 5.0 6.0])\nspat_getp(oprob, :k1, dsrs)\n\nIf we now were to simulate oprob, the simulation would not take the updated value of k1 into account. For our changes to take effect we might first need to call rebuild_spat_internals! with oprob as an input\n\nrebuild_spat_internals!(oprob)\n\nThere are two different circumstances when rebuild_spat_internals! must be called:\n\nWhen modifying the value of an edge parameter.\nWhen changing a parameter from having spatially uniform values to spatially non-uniform values, or the other way around.\n\nParameter values of integrators can be interfaced with just like for problems (this is primarily relevant when using callbacks). Again, after doing so, the rebuild_spat_internals! function might need to be applied to the integrator.","category":"section"},{"location":"spatial_modelling/discrete_space_simulation_structure_interaction/#dspace_simulation_structure_interaction_prob_int_parameters_edge_ps","page":"Interfacing with Lattice Problems, Integrators, and Solutions","title":"Retrieving and updatingedge  parameter values in problems and integrators","text":"The spat_getp and spat_setp! functions cannot currently be applied to edge parameters. Instead, to access the value of an edge parameter, use \n\noprob.ps[:D]\n\nTo update an edge parameter's value, use\n\noprob.ps[:D] = [0.2]\nnothing # hide\n\nThis interface is somewhat limited, and the following aspects should be noted:\n\nEdge parameter values can only be interfaced with if the edge parameter's value is spatially uniform.\nWhen accessing an (spatially uniform) edge parameter's value, its single value will be encapsulated in a vector.\nWhen setting an (spatially uniform) edge parameter's value, you must encapsulate the new value in a vector.","category":"section"},{"location":"model_creation/examples/programmatic_generative_linear_pathway/#programmatic_generative_linear_pathway","page":"Programmatic, Generative, Modelling of a Linear Pathway","title":"Programmatic, Generative, Modelling of a Linear Pathway","text":"<details><summary><strong>Environment setup and package installation</strong></summary>\n\nThe following code sets up an environment for running the code on this page.\n\nusing Pkg\nPkg.activate(; temp = true) # Creates a temporary environment, which is deleted when the Julia session ends.\nPkg.add(\"Catalyst\")\nPkg.add(\"OrdinaryDiffEqDefault\")\nPkg.add(\"Plots\")\n\n</details>\n\n\n This example will show how to use programmatic, generative, modelling to model a system implicitly. I.e. rather than listing all system reactions explicitly, the reactions are implicitly generated from a simple set of rules. This example is specifically designed to show how programmatic modelling enables generative workflows (demonstrating one of its advantages as compared to DSL-based modelling). In our example, we will model linear pathways, so we will first introduce these. Next, we will model them first using the DSL, and then using a generative programmatic workflow.","category":"section"},{"location":"model_creation/examples/programmatic_generative_linear_pathway/#programmatic_generative_linear_pathway_intro","page":"Programmatic, Generative, Modelling of a Linear Pathway","title":"Linear pathways","text":"Linear pathways consists of a series of species (X_0, X_1, X_2, ..., X_n) where each activates the subsequent one[1]. These are often modelled through the following reaction system:\n\nX_i-1tauhspace033cm  to X_i\n1tauhspace033cm X_i to \n\nfor i = 1  n, where the activation of X_1 depends on some input species X_0. \n\nA common use of these linear pathways is the implementation of time delays. Consider a species X(t) which is activated by species X_0(t). This can be modelled by making the production rate of X(t) a function of the time-delayed value of X_0(t):\n\nf(X_0(t-tau))hspace033cm  to X\n\nThis is a so-called discrete-delay (which will generate a delay differential equation). However, in reality, X(t) probably does not depend on only f(X_0(t-tau)), but rather a distribution of previous X_0(t) values. This can be modelled through a distributed delays\n\nf(int_0^inf X_0(t-tau)g(tau) dtau)hspace033cm  to X\n\nfor some kernel g(tau). Here, a common kernel is a gamma distribution, which generates a gamma-distributed delay:\n\ng(tau alpha beta) = fracbeta^alphatau^alpha-1Gamma(alpha)e^-betatau\n\nWhen this is converted to an ODE, this generates an integro-differential equation. These (as well as the simpler delay differential equations) can be difficult to solve and analyse (especially when SDE or jump simulations are desired). Here, the linear chain trick can be used to instead model the delay as a linear pathway of the form described above[2]. A result by Fargue shows that this is equivalent to a gamma-distributed delay, where alpha is equivalent to n (the number of species in our linear pathway) and beta to %\\tau$ (the delay length term)[3]. While modelling time delays using the linear chain trick introduces additional system species, it is often advantageous as it enables simulations using standard ODE, SDE, and Jump methods.","category":"section"},{"location":"model_creation/examples/programmatic_generative_linear_pathway/#programmatic_generative_linear_pathway_dsl","page":"Programmatic, Generative, Modelling of a Linear Pathway","title":"Modelling linear pathways using the DSL","text":"It is known that two linear pathways have similar delays if the following equality holds:\n\nfrac1tau_1 n_1 = frac1tau_2 n_2\n\nHowever, the shape of the delay depends on the number of intermediaries (n). Here we wish to investigate this shape for two choices of n (n = 3 and n = 10). We do so by implementing two models using the DSL, one for each n. \n\nusing Catalyst\n\nlp_n3 = @reaction_network begin\n    1/τ, X0 --> 0\n    (X0/τ, 1/τ), 0 <--> X1\n    (X1/τ, 1/τ), 0 <--> X2\n    (X2/τ, 1/τ), 0 <--> X3\nend\n\nlp_n10 = @reaction_network begin\n    1/τ, X0 --> 0\n    (X0/τ, 1/τ), 0 <--> X1\n    (X1/τ, 1/τ), 0 <--> X2\n    (X2/τ, 1/τ), 0 <--> X3\n    (X3/τ, 1/τ), 0 <--> X4\n    (X4/τ, 1/τ), 0 <--> X5\n    (X5/τ, 1/τ), 0 <--> X6\n    (X6/τ, 1/τ), 0 <--> X7\n    (X7/τ, 1/τ), 0 <--> X8\n    (X8/τ, 1/τ), 0 <--> X9\n    (X9/τ, 1/τ), 0 <--> X10\nend\nnothing # hide\n\nNext, we prepare an ODE for each model (scaling the initial concentration of X_0 and the value of tau appropriately for each model).\n\nusing OrdinaryDiffEqDefault, Plots\nu0_n3 = [:X0 => 3*1.0, :X1 => 0.0, :X2 => 0.0, :X3 => 0.0]\nps_n3 = [:τ => 1.0/3]\noprob_n3 = ODEProblem(lp_n3, u0_n3, (0.0, 5.0), ps_n3)\n\nu0_n10 = [:X0 => 10*1.0, :X1 => 0.0, :X2 => 0.0, :X3 => 0.0, :X4 => 0.0, :X5 => 0.0, :X6 => 0.0, :X7 => 0.0, :X8 => 0.0, :X9 => 0.0, :X10 => 0.0]\nps_n10 = [:τ => 1.0/10.0]\noprob_n10 = ODEProblem(lp_n10, u0_n10, (0.0, 5.0), ps_n10)\nnothing # hide\n\nFinally, we plot the concentration of the final species in each linear pathway, noting that while the two pulses both peak at t = 10, their shapes depend on n.\n\nsol_n3 = solve(oprob_n3)\nsol_n10 = solve(oprob_n10)\nplot(sol_n3; idxs = :X3, label = \"n = 3\")\nplot!(sol_n10; idxs = :X10, label = \"n = 10\")","category":"section"},{"location":"model_creation/examples/programmatic_generative_linear_pathway/#programmatic_generative_linear_pathway_generative","page":"Programmatic, Generative, Modelling of a Linear Pathway","title":"Modelling linear pathways using programmatic, generative, modelling","text":"Above, we investigated the impact of linear pathways' lengths on their behaviours. Since the models were implemented using the DSL, we had to implement a new model for each pathway (in each case writing out all reactions). Here, we will instead show how programmatic modelling can be used to generate pathways of arbitrary lengths.\n\nFirst, we create a function, generate_lp, which creates a linear pathway model of length n. It utilises vector variables to create an arbitrary number of species, and also creates an observable for the final species of the chain.\n\nusing Catalyst # hide\nt = default_t()\n@parameters τ\nfunction generate_lp(n)\n    # Creates a vector `X` with n+1 species.\n    @species (X(t))[1:n+1]\n    @species Xend(t)\n\n    # Generate\n    #     (1) A degradation reaction for the input species.\n    #     (2) Production reactions for all intermediary species.\n    #     (2) Degradation reactions for all intermediary species.\n    rxs = [\n        Reaction(1/τ, [X[1]], []);\n        [Reaction(X[i]/τ, [], [X[i+1]]) for i in 1:n];\n        [Reaction(1/τ, [X[i+1]], []) for i in 1:n]     \n    ]\n\n    # Assembly and return a complete `ReactionSystem` (including an observable for the final species).\n    @named lp_n = ReactionSystem(rxs, t; observed = [Xend ~ X[end]])\n    return complete(lp_n)\nend\nnothing # hide\n\nNext, we create a function that generates an ODEProblem (with appropriate initial conditions and parameter values) for arbitrarily length linear pathway models.\n\nfunction generate_oprob(n)\n    lp = generate_lp(n)\n    X_init = fill(0.0, n + 1)\n    X_init[1] = n * 1.0\n    u0 = [lp.X => X_init]\n    ps = [τ => 1.0 / n]\n    return ODEProblem(lp, u0, (0.0, 5.0), ps)\nend\nnothing # hide\n\nWe can now simulate linear pathways of arbitrary lengths using a simple syntax. We use this to recreate our previous result from the DSL:\n\nusing OrdinaryDiffEqDefault, Plots # hide\nsol_n3 = solve(generate_oprob(3))\nsol_n10 = solve(generate_oprob(10))\nplot(sol_n3; idxs = :Xend, label = \"n = 3\")\nplot!(sol_n10; idxs = :Xend, label = \"n = 10\")\n\nIf we wish to investigate the behaviour of a pathway with a different length, we can easily add this to the plot\n\nsol_n20 = solve(generate_oprob(20))\nplot!(sol_n20; idxs = :Xend, label = \"n = 20\")\n\n","category":"section"},{"location":"model_creation/examples/programmatic_generative_linear_pathway/#References","page":"Programmatic, Generative, Modelling of a Linear Pathway","title":"References","text":"[1]: N. Korsbo, H. Jönsson It’s about time: Analysing simplifying assumptions for modelling multi-step pathways in systems biology, PLoS Computational Biology (2020).\n\n[2]: J. Metz, O. Diekmann The Abstract Foundations of Linear Chain Trickery (1991).\n\n[3]: D. Fargue Reductibilite des systemes hereditaires a des systemes dynamiques (regis par des equations differentielles aux derivees partielles), Comptes rendus de l'Académie des Sciences (1973).","category":"section"},{"location":"model_simulation/ensemble_simulations/#ensemble_simulations","page":"Ensemble/Monte Carlo Simulations","title":"Ensemble/Monte Carlo Simulations","text":"<details><summary><strong>Environment setup and package installation</strong></summary>\n\nThe following code sets up an environment for running the code on this page.\n\nusing Pkg\nPkg.activate(; temp = true) # Creates a temporary environment, which is deleted when the Julia session ends.\nPkg.add(\"Catalyst\")\nPkg.add(\"OrdinaryDiffEqTsit5\")\nPkg.add(\"Plots\")\nPkg.add(\"StochasticDiffEq\")\n\n</details>\n\n<details><summary><strong>Quick-start example</strong></summary>\n\nThe following code provides a brief example of how to run an ensemble simulation (here we perform multiple SDE simulations).\n\nusing Catalyst, StochasticDiffEq\n\n# First we create the normal problem we wish to run an ensemble simulation of.\nrn = @reaction_network begin\n    (p,d), 0 <--> X\nend\nu0 = [:X => 100.0]\nps = [:p => 50.0, :d => 0.25]\nsprob = SDEProblem(rn, u0, 10.0, ps)\n\n# Next we create an `EnsembleProblem`. When simulate it the `trajectories` argument designate the number of simulations.\neprob = EnsembleProblem(sprob)\nesol = solve(eprob, trajectories = 20)\nplot(esol) # By default, `plot` shows all trajectories.\n\n</details>\n\n\n   In many contexts, a single model is re-simulated under similar conditions. Examples include:\n\nPerforming Monte Carlo simulations of a stochastic model to gain insight in its behaviour.\nScanning a model's behaviour for different parameter values and/or initial conditions.\n\nWhile this can be handled using for loops, it is typically better to first create an EnsembleProblem, and then perform an ensemble simulation. Advantages include a more concise interface and the option for automatic simulation parallelisation. Here we provide a short tutorial on how to perform parallel ensemble simulations, with a more extensive documentation being available here.","category":"section"},{"location":"model_simulation/ensemble_simulations/#ensemble_simulations_monte_carlo","page":"Ensemble/Monte Carlo Simulations","title":"Monte Carlo simulations using unmodified conditions","text":"We will first consider Monte Carlo simulations where the simulation conditions are identical in-between simulations. First, we declare a simple self-activation loop model\n\nusing Catalyst\nsa_model = @reaction_network begin\n    v0 + hill(X,v,K,n), ∅ --> X\n    deg, X --> ∅\nend\nu0 = [:X => 10.0]\ntspan = (0.0, 1000.0)\nps = [:v0 => 0.1, :v => 2.5, :K => 40.0, :n => 4.0, :deg => 0.01]\nnothing # hide\n\nWe wish to simulate it as an SDE. Rather than performing a single simulation, however, we want to perform multiple ones. Here, we first create a normal SDEProblem, and use it as the single input to a EnsembleProblem (EnsembleProblem are created similarly for ODE and jump simulations, but the ODEProblem or JumpProblem is used instead).\n\nusing StochasticDiffEq\nsprob = SDEProblem(sa_model, u0, tspan, ps)\neprob = EnsembleProblem(sprob)\nnothing # hide\n\nNext, the EnsembleProblem can be used as input to the solve command. Here, we use exactly the same inputs that we use for single simulations, however, we add a trajectories argument to denote how many simulations we wish to carry out. Here we perform 10 simulations:\n\nsols = solve(eprob, STrapezoid(); trajectories = 10)\nnothing # hide\n\nFinally, we can use our ensemble simulation solution as input to plot (just like normal simulations):\n\nusing Plots\nplot(sols)\n\nHere, each simulation is displayed as an individual trajectory.\n\nnote: Note\nWhile not used here, the la plotting option (which modifies line transparency) can help improve the plot visual when a large number of (overlapping) lines are plotted.\n\nVarious convenience functions are available for analysing and plotting ensemble simulations (a full list can be found here). Here, we use these to first create an EnsembleSummary (retrieving each simulation's value at time points 0.0, 1.0, 2.0, ... 1000.0). Next, we use this as an input to the plot command, which automatically plots the mean X activity across the ensemble, while also displaying the 5% and 95% quantiles as the shaded area:\n\ne_summary = EnsembleAnalysis.EnsembleSummary(sols, 0.0:1.0:1000.0)\nplot(e_summary)","category":"section"},{"location":"model_simulation/ensemble_simulations/#ensemble_simulations_varying_conditions","page":"Ensemble/Monte Carlo Simulations","title":"Ensemble simulations using varying simulation conditions","text":"Previously, we assumed that each simulation used the same initial conditions and parameter values. If this is not the case (when e.g. performing a parameter scan), a prob_func optional argument must be supplied to the EnsembleProblem, this describes how the problem should be modified for each individual simulation of the ensemble.\n\nHere, we first create an ODEProblem of our previous self-activation loop:\n\nusing OrdinaryDiffEqTsit5\noprob = ODEProblem(sa_model, u0, tspan, ps)\nnothing # hide\n\nNext, we wish to simulate the model for a range of initial conditions of X`. To do this we create a problem function, which takes the following arguments:\n\nprob: The problem given to our EnsembleProblem (which is the problem that prob_func modifies in each iteration).\ni: The number of this specific Monte Carlo iteration in the interval 1:trajectories.\nrepeat: The iteration of the repeat of the simulation. Typically 1, but potentially higher if the simulation re-running option is used.\n\nHere we will use the following problem function (utilising remake), which will provide a uniform range of initial concentrations of X:\n\nfunction prob_func(prob, i, repeat)\n    remake(prob; u0 = [:X => i * 5.0])\nend\nnothing # hide\n\nNext, we create our EnsembleProblem, and simulate it 10 times:\n\neprob = EnsembleProblem(oprob; prob_func)\nsols = solve(eprob, Tsit5(); trajectories = 10)\nplot(sols)\n\nHere we see that the deterministic model (unlike the stochastic one), only activates for some initial conditions (while other tends to an inactive state).\n\nThe EnsembleProblem constructor accepts a few additional optional options (output_func, reduction, u_init, and safetycopy), which are described in more detail here. These can be used to e.g. rerun an individual simulation which does not fulfil some condition, or extract and save only some selected information from a simulation (rather than the full simulation).","category":"section"},{"location":"model_creation/programmatic_CRN_construction/#programmatic_CRN_construction","page":"Programmatic Construction of Symbolic Reaction Systems","title":"Programmatic Construction of Symbolic Reaction Systems","text":"<details><summary><strong>Environment setup and package installation</strong></summary>\n\nThe following code sets up an environment for running the code on this page.\n\nusing Pkg\nPkg.activate(; temp = true) # Creates a temporary environment, which is deleted when the Julia session ends.\nPkg.add(\"Catalyst\")\n\n</details>\n\n\n   While the DSL provides a simple interface for creating ReactionSystems, it can often be convenient to build or augment a ReactionSystem programmatically. In this tutorial we show how to build the repressilator model of the Introduction to Catalyst tutorial directly using symbolic variables, and then summarize the basic API functionality for accessing information stored within ReactionSystems.","category":"section"},{"location":"model_creation/programmatic_CRN_construction/#Directly-building-the-repressilator-with-ReactionSystems","page":"Programmatic Construction of Symbolic Reaction Systems","title":"Directly building the repressilator with ReactionSystems","text":"We first load Catalyst\n\nusing Catalyst\n\nand then define symbolic variables for each parameter and species in the system (the latter corresponding to a variable or unknown in ModelingToolkitBase terminology)\n\nt = default_t()\n@parameters α K n δ γ β μ\n@species m₁(t) m₂(t) m₃(t) P₁(t) P₂(t) P₃(t)\nnothing    # hide\n\nNote: each species is declared as a function of time. Here, we first import the time independent variable, and stores it in t, using t = default_t(), and then use it to declare out species.\n\nnote: Note\nFor users familiar with ModelingToolkitBase, chemical species must be declared    via the @species macro, and not the ModelingToolkitBase.@variables macro.    @species wraps @variables, adding additional metadata to the symbolic    variables that represent species which is used internally in Catalyst.\n\nNext, we specify the chemical reactions that comprise the system using Catalyst Reactions\n\nrxs = [Reaction(hillr(P₃,α,K,n), nothing, [m₁]),\n       Reaction(hillr(P₁,α,K,n), nothing, [m₂]),\n       Reaction(hillr(P₂,α,K,n), nothing, [m₃]),\n       Reaction(δ, [m₁], nothing),\n       Reaction(γ, nothing, [m₁]),\n       Reaction(δ, [m₂], nothing),\n       Reaction(γ, nothing, [m₂]),\n       Reaction(δ, [m₃], nothing),\n       Reaction(γ, nothing, [m₃]),\n       Reaction(β, [m₁], [m₁,P₁]),\n       Reaction(β, [m₂], [m₂,P₂]),\n       Reaction(β, [m₃], [m₃,P₃]),\n       Reaction(μ, [P₁], nothing),\n       Reaction(μ, [P₂], nothing),\n       Reaction(μ, [P₃], nothing)]\nnothing    # hide\n\nHere we use nothing where the DSL used varnothing. Finally, we are ready to construct our ReactionSystem as\n\n@named repressilator = ReactionSystem(rxs, t)\nrepressilator = complete(repressilator)\nnothing     # hide\n\nNotice, the model is named repressilator. A name must always be specified when directly constructing a ReactionSystem (the DSL will auto-generate one if left out). Using @named when constructing a ReactionSystem causes the name of the system to be the same as the name of the variable storing the system. Alternatively, one can use the name = :repressilator keyword argument to the ReactionSystem constructor.\n\nwarning: Warning\nAll ReactionSystems created via the symbolic interface (i.e. by calling ReactionSystem with some input, rather than using @reaction_network) are not marked as complete. To simulate them, they must first be marked as complete, indicating to Catalyst and ModelingToolkitBase that they represent finalized models. This can be done using the complete function, as above. An expanded description on completeness can be found here.\n\nFor more options in building ReactionSystems, see the ReactionSystem API docs. For a more extensive example of how to programmatically create a ReactionSystem, see the Smoluchowski Coagulation Equation example.","category":"section"},{"location":"model_creation/programmatic_CRN_construction/#More-general-Reactions","page":"Programmatic Construction of Symbolic Reaction Systems","title":"More general Reactions","text":"In the example above all the specified Reactions were first or zero order. The three-argument form of Reaction implicitly assumes all species have a stoichiometric coefficient of one, i.e. for substrates [S₁,...,Sₘ] and products [P₁,...,Pₙ] it has the possible forms\n\n# rate, S₁ + ... + Sₘ --> P₁ + ... + Pₙ\nReaction(rate, [S₁,...,Sₘ], [P₁,...,Pₙ])\n\n# rate, S₁ + ... + Sₘ --> ∅\nReaction(rate, [S₁,...,Sₘ], nothing)\n\n# rate, ∅ --> P₁ + ... + Pₙ\nReaction(rate, nothing, [P₁,...,Pₙ])\n\nTo allow for other stoichiometric coefficients we also provide a five argument form\n\n# rate, α₁*S₁ + ... + αₘ*Sₘ --> β₁*P₁ + ... + βₙ*Pₙ\nReaction(rate, [S₁,...,Sₘ], [P₁,...,Pₙ], [α₁,...,αₘ], [β₁,...,βₙ])\n\n# rate, α₁*S₁ + ... + αₘ*Sₘ --> ∅\nReaction(rate, [S₁,...,Sₘ], nothing, [α₁,...,αₘ], nothing)\n\n# rate, ∅ --> β₁*P₁ + ... + βₙ*Pₙ\nReaction(rate, nothing, [P₁,...,Pₙ], nothing, [β₁,...,βₙ])\n\nFinally, we note that the rate constant, rate above, does not need to be a constant or fixed function, but can be a general symbolic expression:\n\nt = default_t()\n@parameters α, β\n@species A(t), B(t)\nrx = Reaction(α + β*t*A, [A], [B])\n\nSee the FAQs for info on using general user-specified functions for the rate constant.","category":"section"},{"location":"model_creation/programmatic_CRN_construction/#The-@reaction-macro-for-constructing-Reactions","page":"Programmatic Construction of Symbolic Reaction Systems","title":"The @reaction macro for constructing Reactions","text":"In some cases one wants to build reactions incrementally, as in the repressilator example, but it would be nice to still have a short hand as in the @reaction_network DSL. In this case one can construct individual reactions using the @reaction macro.\n\nFor example, the repressilator reactions could also have been constructed like\n\nt = default_t()\n@species P₁(t) P₂(t) P₃(t)\nrxs = [(@reaction hillr($P₃,α,K,n), ∅ --> m₁),\n       (@reaction hillr($P₁,α,K,n), ∅ --> m₂),\n       (@reaction hillr($P₂,α,K,n), ∅ --> m₃),\n       (@reaction δ, m₁ --> ∅),\n       (@reaction γ, ∅ --> m₁),\n       (@reaction δ, m₂ --> ∅),\n       (@reaction γ, ∅ --> m₂),\n       (@reaction δ, m₃ --> ∅),\n       (@reaction γ, ∅ --> m₃),\n       (@reaction β, m₁ --> m₁ + P₁),\n       (@reaction β, m₂ --> m₂ + P₂),\n       (@reaction β, m₃ --> m₃ + P₃),\n       (@reaction μ, P₁ --> ∅),\n       (@reaction μ, P₂ --> ∅),\n       (@reaction μ, P₃ --> ∅)]\n@named repressilator = ReactionSystem(rxs, t)\n\nNote, there are a few differences when using the @reaction macro to specify one reaction versus using the full @reaction_network macro to create a ReactionSystem. First, only one reaction (i.e. a single forward arrow type) can be used, i.e. reversible arrows like <--> will not work (since they define more than one reaction). Second, the @reaction macro does not have an option for designating what should be considered a species or parameter, and instead assumes that any symbol that appears as either a substrate or a product is a species, and everything else (including stoichiometric coefficients) are parameters. As such, the following are equivalent\n\nrx = @reaction hillr(P,α,K,n), A --> B\n\nis equivalent to\n\nt = default_t()\n@parameters P α K n\n@variables A(t) B(t)\nrx = Reaction(hillr(P,α,K,n), [A], [B])\n\nHere (P,α,K,n) are parameters and (A,B) are species.\n\nThis behavior is the reason that in the repressilator example above we pre-declared (P₁(t),P₂(t),P₃(t)) as variables, and then used them via interpolating their values into the rate law expressions using $ in the macro. This ensured they were properly treated as species and not parameters. See the @reaction macro docstring for more information.","category":"section"},{"location":"model_creation/programmatic_CRN_construction/#Basic-querying-of-ReactionSystems","page":"Programmatic Construction of Symbolic Reaction Systems","title":"Basic querying of ReactionSystems","text":"The Catalyst.jl API provides a large variety of functionality for querying properties of a reaction network. Here we go over a few of the most useful basic functions. Given the repressillator defined above we have that\n\nspecies(repressilator)\n\nparameters(repressilator)\n\nreactions(repressilator)\n\nWe can test if a Reaction is mass action, i.e. the rate does not depend on t or other species, as\n\n# Catalyst.hillr(P₃(t), α, K, n), ∅ --> m₁\nrx1 = reactions(repressilator)[1]\nismassaction(rx1,repressilator)\n\nwhile\n\n# δ, m₁ --> ∅\nrx2 = reactions(repressilator)[4]\nismassaction(rx2,repressilator)\n\nSimilarly, we can determine which species a reaction's rate law will depend on like\n\nrn = @reaction_network begin\n       k*W, 2X + 3Y --> 5Z + W\n     end\ndependents(reactions(rn)[1], rn)\n\nBasic stoichiometry matrices can be obtained from a ReactionSystem as\n\nsubstoichmat(repressilator)\n\nprodstoichmat(repressilator)\n\nnetstoichmat(repressilator)\n\nHere the (ij) entry gives the corresponding stoichiometric coefficient of species i for reaction j.\n\nFinally, we can directly access fields of individual reactions like\n\nrx1.rate\n\nrx1.substrates\n\nrx1.products\n\nrx1.substoich\n\nrx1.prodstoich\n\nrx1.netstoich\n\nSee the Catalyst.jl API for much more detail on the various querying and analysis functions provided by Catalyst.","category":"section"},{"location":"model_creation/model_visualisation/#visualisation","page":"Model Visualisation","title":"Model Visualisation","text":"<details><summary><strong>Environment setup and package installation</strong></summary>\n\nThe following code sets up an environment for running the code on this page.\n\nusing Pkg\nPkg.activate(; temp = true) # Creates a temporary environment, which is deleted when the Julia session ends.\nPkg.add(\"CairoMakie\")\nPkg.add(\"Catalyst\")\nPkg.add(\"GLMakie\")\nPkg.add(\"GraphMakie\")\nPkg.add(\"Latexify\")\nPkg.add(\"NetworkLayout\")\n\n</details>\n\n\n   Catalyst-created ReactionSystem models can be visualised either as LaTeX code (of either the model reactions or its equations) or as a network graph. This section describes both functionalities.","category":"section"},{"location":"model_creation/model_visualisation/#visualisation_latex","page":"Model Visualisation","title":"Displaying models using LaTeX","text":"Once a model has been created, the Latexify.jl package can be used to generate LaTeX code of the model. This can either be used for easy model inspection (e.g. to check which equations are being simulated), or to generate code which can be directly pasted into a LaTeX document.\n\nLet us consider a simple Brusselator model:\n\nusing Catalyst\nbrusselator = @reaction_network begin\n    A, ∅ --> X\n    1, 2X + Y --> 3X\n    B, X --> Y\n    1, X --> ∅\nend\n\nTo display its reaction (using LaTeX formatting) we run latexify with our model as input:\n\nusing Latexify\nlatexify(brusselator)\nbrusselator # hide\n\nHere, we note that the output of latexify(brusselator) is identical to how a model is displayed by default. Indeed, the reason is that Catalyst internally uses Latexify's latexify function to display its models. It is also possible to display the ODE equations a model would generate by adding the form = :ode argument:\n\nlatexify(brusselator; form = :ode)\nlatexify(brusselator; form = :ode, math_delimiters = true) # hide\n\nnote: Note\nInternally, latexify(brusselator; form = :ode) generates the corresponding ODE equations from the model and renders them through Catalyst's LaTeX display pipeline. The same can be done for SDEs using form = :sde.\n\nIf you wish to copy the output to your clipboard (e.g. so that you can paste it into a LaTeX document), run copy_to_clipboard(true) before you run latexify. A more throughout description of Latexify's features can be found in its documentation.\n\nnote: Note\nFor a model to be nicely displayed you have to use an IDE that actually supports this (such as a notebook). Other environments (such as the Julia REPL) will simply return the full LaTeX code which would generate the desired expression. ","category":"section"},{"location":"model_creation/model_visualisation/#visualisation_graphs","page":"Model Visualisation","title":"Displaying model networks","text":"Catalyst uses GraphMakie.jl to display representations of chemical reaction networks, including the complex graph and the species-reaction graph (which is similar to the Petri net representation). To get started, import Catalyst, GraphMakie, and NetworkLayout to load the CatalystGraphMakieExtension extension, and then load a Makie backend (CairoMakie is a good lightweight choice). Here, while Catalyst primarily uses Plots.jl for plotting, Makie is used for displaying network graphs. Makie can also be used for plotting more generally (and is also a preferred option for some).\n\nusing Catalyst, GraphMakie, NetworkLayout\nusing CairoMakie\nnothing # hide\n\nLet's declare a Brusselator model to see this plotting functionality. The functions plot_network and plot_complexes are used to create the species-reaction and complex graphs, respectively. For a more thorough description of these two representations, please see the network visualization section of the API, but the gist is that the species-reaction graph has species and reactions as nodes, and the complex graph has reaction complexes as nodes. Below we will plot the species-reaction graph using plot_network. \n\nbrusselator = @reaction_network begin\n    A, ∅ --> X\n    1, 2X + Y --> 3X\n    B, X --> Y\n    1, X --> ∅\nend\nplot_network(brusselator)\n\nThe species-reaction graph (or network graph) represents species as blue nodes and reactions as green dots. Black arrows from species to reactions indicate substrates, and are labelled with their respective stoichiometries. Similarly, black arrows from reactions to species indicate products (also labelled with their respective stoichiometries). If there are any reactions where a species affect the rate, but does not participate as a reactant, this is displayed with a dashed red arrow. This can be seen in the following Repressilator model:\n\nrepressilator = @reaction_network begin\n    hillr(Z,v,K,n), ∅ --> X\n    hillr(X,v,K,n), ∅ --> Y\n    hillr(Y,v,K,n), ∅ --> Z\n    d, (X, Y, Z) --> ∅\nend\nplot_network(repressilator)\n\nA generated graph can be saved using Makie's save function. \n\nrepressilator_graph = plot_network(repressilator)\nsave(\"repressilator_graph.png\", repressilator_graph)\n\nFinally, a network's reaction complexes (and the reactions in between these) can be displayed using the plot_complexes(brusselator) function:\n\nplot_complexes(brusselator)\n\nHere, reaction complexes are displayed as blue nodes, and reactions between complexes are displayed as black arrows. Red arrows indicate that the rate constantof a reaction has a species-dependence. Edges can be optionally labeled with their rate expressions by calling with the option show_rate_labels.\n\nplot_complexes(brusselator, show_rate_labels = true)","category":"section"},{"location":"model_creation/model_visualisation/#Customizing-Plots","page":"Model Visualisation","title":"Customizing Plots","text":"In this section we demonstrate some of the ways that plot objects can be manipulated to give nicer images. Let's start with our brusselator plot once again. Note that the plot function returns three objects: the Figure, the Axis, and the Plot, which can each be customized independently. See the general Makie documentation for more information.\n\nf, ax, p = plot_complexes(brusselator, show_rate_labels = true)\n\nIt seems like a bit of the top node is cut off. Let's increase the top and bottom margins by increasing yautolimitmargin.\n\nax.yautolimitmargin = (0.3, 0.3) # defaults to (0.15, 0.15)\nax.aspect = DataAspect()\nf\n\nThere are many keyword arguments that can be passed to plot_network or plot_complexes to change the look of the graph (which get passed to the graphplot Makie recipe). Let's change the color of the nodes and make the inner labels a bit smaller. Let's also give the plot a title. \n\nf, ax, p = plot_complexes(brusselator, show_rate_labels = true, node_color = :yellow, ilabels_fontsize = 10)\nax.title = \"Brusselator\"\nf\n\nMost of the kwargs that modify the nodes or edges will also accept a vector with the same length as the number of nodes or edges, respectively. See here for a full list of keyword arguments to graph_plot. Note that plot_complexes and plot_network default to layout = Stress() rather than layout = Spring(), since Stress() is better at generating plots with fewer edge crossings. More layout options and customizations (such as pinning nodes to certain positions) can be found in the NetworkLayout documentation.\n\nOnce a graph is already created we can also change the keyword arguments by modifying the fields of the Plot object p.\n\np.node_color = :orange\nf\n\nCustom node positions can also be given, if the automatic layout is unsatisfactory.\n\nfixedlayout = [(0,0), (1,0), (0,1), (1,1), (2,0)]\np.layout = fixedlayout\nautolimits!(ax)\nf\n\nMakie graph plots can be made to be interactive, allowing one to drag nodes and edges. To do this, we retrieve the axis from the GraphMakie plot, and then register the interactions. Note that this can only be done if GLMakie is the installed Makie backend. See the GraphMakie docs for more information about the types of interactions one can register. Below is a non-interactive code example that shows how to do this:\n\nusing GLMakie\nf, ax, p = plot_network(brusselator)\nderegister_interaction!(ax, :rectanglezoom)\nregister_interaction!(ax, :ndrag, NodeDrag(p))\nregister_interaction!(ax, :edrag, EdgeDrag(p))\n\nThe equivalent of show for Makie plots is the display function. \n\nf = plot_network(brusselator)\ndisplay(f)\n\nOnce you are happy with the graph plot, you can save it using the save function. \n\nsave(\"fig.png\", f)","category":"section"},{"location":"network_analysis/network_properties/#network_analysis_caching_properties","page":"Caching of Network Properties in ReactionSystems","title":"Caching of Network Properties in ReactionSystems","text":"<details><summary><strong>Environment setup and package installation</strong></summary>\n\nThe following code sets up an environment for running the code on this page.\n\nusing Pkg\nPkg.activate(; temp = true) # Creates a temporary environment, which is deleted when the Julia session ends.\nPkg.add(\"Catalyst\")\n\n</details>\n\n\n When calling many of the network API functions, Catalyst calculates and caches in rn a variety of information. For example the first call to\n\nrcs,B = reactioncomplexes(rn)\n\ncalculates, caches, and returns the reaction complexes, rcs, and the incidence matrix, B, of rn. Subsequent calls simply return rcs and B from the cache.\n\nSimilarly, the first call to\n\nN = netstoichmat(rn)\n\ncalculates, caches and returns the net stoichiometry matrix. Subsequent calls then simply return the cached value of N. Caching such information means users do not need to manually know which subsets of network properties are needed for a given calculation (like the deficiency). Generally only\n\nrcs,B = reactioncomplexes(rn)    # must be called once to cache rcs and B\nany_other_network_property(rn)\n\nshould work to calculate a desired network property, with the API doc strings indicating when reactioncomplexes(rn) must be called at least once before a given function is used.\n\nBecause of the caching of network properties, subsequent calls to most API functions will be fast, simply returning the previously calculated and cached values. In some cases it may be desirable to reset the cache and recalculate these properties. This can be done by calling\n\nCatalyst.reset_networkproperties!(rn)\n\nNetwork property functions will then recalculate their associated properties and cache the new values the next time they are called.","category":"section"},{"location":"network_analysis/odes/#network_analysis_odes","page":"Decomposing the Reaction Network ODEs","title":"Decomposing the Reaction Network ODEs","text":"<details><summary><strong>Environment setup and package installation</strong></summary>\n\nThe following code sets up an environment for running the code on this page.\n\nusing Pkg\nPkg.activate(; temp = true) # Creates a temporary environment, which is deleted when the Julia session ends.\nPkg.add(\"CairoMakie\")\nPkg.add(\"Catalyst\")\nPkg.add(\"GraphMakie\")\nPkg.add(\"NetworkLayout\")\n\n</details>\n\n\n In this tutorial we will discuss the specific mathematical  structure of the ODEs that arise from the mass-action dynamics of chemical reaction networks, and decompose them as a product of matrices that describe the network.  A complete summary of  the exported functions is given in the API section Network Analysis and Representations. Please consult Feinberg's Foundations of Chemical Reaction Network Theory[1] for more discussion about the concepts on this page.\n\nNote, currently API functions for network analysis and conservation law analysis do not work with constant species (which are generated by SBML, and can be declared in Catalyst as well.","category":"section"},{"location":"network_analysis/odes/#network_analysis_repressilator_representation","page":"Decomposing the Reaction Network ODEs","title":"Network representation of the Repressilator ReactionSystem","text":"We first load Catalyst and construct our model of the repressilator\n\nusing Catalyst, CairoMakie, GraphMakie, NetworkLayout\nrepressilator = @reaction_network Repressilator begin\n       hillr(P₃,α,K,n), ∅ --> m₁\n       hillr(P₁,α,K,n), ∅ --> m₂\n       hillr(P₂,α,K,n), ∅ --> m₃\n       (δ,γ), m₁ <--> ∅\n       (δ,γ), m₂ <--> ∅\n       (δ,γ), m₃ <--> ∅\n       β, m₁ --> m₁ + P₁\n       β, m₂ --> m₂ + P₂\n       β, m₃ --> m₃ + P₃\n       μ, P₁ --> ∅\n       μ, P₂ --> ∅\n       μ, P₃ --> ∅\nend\n\nIn the Model Visualization tutorial we showed how the above network could be visualized as a species-reaction graph. There, species are represented by the nodes of the graph and edges show the reactions in which a given species is a substrate or product.\n\ng = plot_network(repressilator)\n\nWe also showed in the Introduction to Catalyst tutorial that the reaction rate equation (RRE) ODE model for the repressilator is\n\nbeginaligned\nfracdm_1(t)dt = fracalpha K^nK^n + left( P_3left( t right) right)^n - delta m_1left( t right) + gamma \nfracdm_2(t)dt = fracalpha K^nK^n + left( P_1left( t right) right)^n - delta m_2left( t right) + gamma \nfracdm_3(t)dt = fracalpha K^nK^n + left( P_2left( t right) right)^n - delta m_3left( t right) + gamma \nfracdP_1(t)dt = beta m_1left( t right) - mu P_1left( t right) \nfracdP_2(t)dt = beta m_2left( t right) - mu P_2left( t right) \nfracdP_3(t)dt = beta m_3left( t right) - mu P_3left( t right)\nendaligned","category":"section"},{"location":"network_analysis/odes/#network_analysis_matrix_vector_representation","page":"Decomposing the Reaction Network ODEs","title":"Matrix-vector reaction rate equation representation","text":"In general, reaction rate equation (RRE) ODE models for chemical reaction networks can be represented as a first-order system of ODEs in a compact matrix-vector notation. Suppose we have a reaction network with K reactions and M species, labelled by the state vector\n\nmathbfx(t) = beginpmatrix x_1(t)  vdots  x_M(t)) endpmatrix\n\nFor the repressilator, mathbfx(t) is just\n\nx = species(repressilator)\n\nThe RRE ODEs satisfied by mathbfx(t) are then\n\nfracdmathbfxdt = N mathbfv(mathbfx(t)t)\n\nwhere N is a constant M by K matrix with N_m k the net stoichiometric coefficient of species m in reaction k. mathbfv(mathbfx(t)t) is the rate law vector, with v_k(mathbfx(t)t) the rate law for the kth reaction. For example, for the first reaction of the repressilator above, the rate law is\n\nv_1(mathbfx(t)t) = fracalpha K^nK^n + left( P_3(t) right)^n\n\nWe can calculate each of these in Catalyst via\n\nN = netstoichmat(repressilator)\n\nand by using the oderatelaw function\n\nrxs = reactions(repressilator)\nν = oderatelaw.(rxs)\n\nNote, as oderatelaw takes just one reaction as input we use broadcasting to apply it to each element of rxs.\n\nLet's check that this really gives the same ODEs as Catalyst. Here is what Catalyst generates by converting to an ODE System\n\nosys = ode_model(repressilator)\n\n# for display purposes we just pull out the right side of the equations\nodes = [eq.rhs for eq in equations(osys)]\n\nwhereas our matrix-vector representation gives\n\nodes2 = N * ν\n\nLet's check these are equal symbolically\n\nisequal(odes, odes2)","category":"section"},{"location":"network_analysis/odes/#network_analysis_reaction_complexes","page":"Decomposing the Reaction Network ODEs","title":"Reaction complex representation","text":"We now introduce a further decomposition of the RRE ODEs, which has been used to facilitate analysis of a variety of reaction network properties. Consider a simple reaction system like\n\nrn = @reaction_network begin\n k, 2A + 3B --> A + 2C + D\n b, C + D --> 2A + 3B\nend\n\nWe can think of the first reaction as converting the reaction complex, 2A+3B to the complex A+2C+D with rate k. Suppose we order our species the same way as Catalyst does, i.e.\n\nbeginpmatrix\nx_1(t)\nx_2(t)\nx_3(t)\nx_4(t)\nendpmatrix =\nbeginpmatrix\nA(t)\nB(t)\nC(t)\nD(t)\nendpmatrix\n\nwhich should be the same as\n\nspecies(rn)\n\nWe can describe a given reaction complex by the stoichiometric coefficients of each species within the complex. For the reactions in rn these vectors would be\n\nbeginalign*\n2A+3B = beginpmatrix\n2\n3\n0\n0\nendpmatrix \nA+2C+D = beginpmatrix\n1\n0\n2\n1\nendpmatrix\n \nC+D = beginpmatrix\n0\n0\n1\n1\nendpmatrix\nendalign*\n\nCatalyst can calculate these representations as the columns of the complex stoichiometry matrix,\n\nZ = complexstoichmat(rn)\n\nIf we have C complexes, Z is a M by C matrix with Z_m c giving the stoichiometric coefficient of species m within complex c.\n\nWe can use this representation to provide another representation of the RRE ODEs. The net stoichiometry matrix can be factored as N = Z B, where B is called the incidence matrix of the reaction network,\n\nB = incidencemat(rn)\n\nHere B is a C by K matrix with B_c k = 1 if complex c appears as a product of reaction k, and B_c k = -1 if complex c is a substrate of reaction k.\n\nUsing our decomposition of N, the RRE ODEs become\n\nfracdxdt = Z B mathbfv(mathbfx(t)t)\n\nLet's verify that N = Z B,\n\nN = netstoichmat(rn)\nN == Z*B\n\nReaction complexes give an alternative way to visualize a reaction network graph. Catalyst's plot_complexes command will calculate the complexes of a network and then show how they are related. For example, we can run\n\nplot_complexes(rn)\n\nwhile for the repressilator we find\n\nplot_complexes(repressilator)\n\nHere ∅ represents the empty complex, black arrows show reactions converting substrate complexes into product complexes where the rate is just a number or parameter, and red arrows indicate the conversion of substrate complexes into product complexes where the rate is an expression involving chemical species.","category":"section"},{"location":"network_analysis/odes/#Full-decomposition-of-the-reaction-network-ODEs-(flux-matrix-and-mass-action-vector)","page":"Decomposing the Reaction Network ODEs","title":"Full decomposition of the reaction network ODEs (flux matrix and mass-action vector)","text":"So far we have covered two equivalent descriptions of the chemical reaction network ODEs: \n\nbeginalign\nfracdmathbfxdt = N mathbfv(mathbfx(t)t) \n= Z B mathbfv(mathbfx(t)t)\nendalign\n\nIn this section we discuss a further decomposition of the ODEs. Recall the reaction rate vector mathbfv, which is a vector of length R whose elements are the rate expressions for each reaction. Its elements can be written as \n\nmathbfv_y rightarrow y = k_y rightarrow y mathbfx^y\n\nwhere mathbfx^y = prod_s x_s^y_s denotes the mass-action product of the substrate complex y from the y rightarrow y reaction. We can define a new vector called the mass action vector Phi(mathbfx(t)), a vector of length C whose elements are the mass action products of each complex: \n\nΦ = massactionvector(rn)\n\nAn important thing to note is this function assumes combinatoric ratelaws, meaning that mass-action products will get rescaled by factorial factors. For instance, note that the mass-action product for the complex 2A + 3B has a factor of 1/12, corresponding to 1/(2! 3!). This option can be turned off with combinatoric_ratelaws = false.\n\nΦ_2 = massactionvector(rn; combinatoric_ratelaws = false)\n\nThen the reaction rate vector mathbfv can be written as \n\nmathbfv(mathbfx(t)) = K Phi(mathbfx(t))\n\nwhere K is an R-by-C matrix called the flux matrix, where K_rc is the rate constant of reaction r if c is the index of the substrate complex of reaction r, and 0 otherwise. In Catalyst, the API function for K is fluxmat: \n\nK = fluxmat(rn)\n\nSince we have that mathbfv = KPhi, we can rewrite the above decompositions as follows: \n\nbeginalign\nfracdmathbfxdt = N mathbfv(mathbfx(t)t) \n= N K Phi(mathbfx(t)t) \n= Z B K Phi(mathbfx(t)t) \nendalign\n\nThe final matrix to discuss is the product of A_k = BK, which is a C-by-C matrix that turns out to be exactly the negative of the graph Laplacian of the weighted graph whose nodes are reaction complexes and whose edges represent reactions, weighted by the rate constants. The API function for A_k is the laplacianmat: \n\nA_k = laplacianmat(rn)\n\nWe can check that \n\nisequal(A_k, B * K)\n\nNote that we have used isequal instead of == here because laplacianmat returns a Matrix{Num}, since some of its entries are symbolic rate constants (symbolic variables and Nums cannot be compared using ==, since a == b is interpreted as a symbolic expression).\n\nIn summary, we have that\n\nbeginalign\nfracdmathbfxdt = N mathbfv(mathbfx(t)t) \n= N K Phi(mathbfx(t)t) \n= Z B K Phi(mathbfx(t)t)  \n= Z A_k Phi(mathbfx(t)t) \nendalign\n\nAll three of the objects introduced in this section (the flux matrix, mass-action vector, and Laplacian matrix) will return symbolic outputs by default, but can be made to return numerical outputs if values are specified.  For example, massactionvector will return a numerical output if a set of species concentrations is supplied using a dictionary, tuple, or vector of Symbol-value pairs.\n\nconcmap = Dict([:A => 3., :B => 5., :C => 2.4, :D => 1.5])\nmassactionvector(rn, concmap)\n\nfluxmat and laplacianmat will return numeric matrices if a set of rate constants and other parameters are supplied the same way.\n\nparammap = Dict([:k => 12., :b => 8.])\nfluxmat(rn, parammap)\n\nlaplacianmat(rn, parammap)","category":"section"},{"location":"network_analysis/odes/#Symbolic-ODE-functions","page":"Decomposing the Reaction Network ODEs","title":"Symbolic ODE functions","text":"In some cases it might be useful to generate the function defining the system of ODEs as a symbolic Julia function that can be used for further analysis. This can be done using Symbolics' build_function, which takes a symbolic expression and a set of desired arguments, and converts it into a Julia function taking those arguments.\n\nLet's build the full symbolic function corresponding to our ODE system. build_function will return two expressions, one for a function that outputs a new vector for the result, and one for a function that modifies the input in-place. Either expression can then be evaluated to return a Julia function.\n\nparammap = Dict([:k => 12., :b => 8.])\nK = fluxmat(rn, parammap)\nodes = N * K * Φ\nf_oop_expr, f_iip_expr = Symbolics.build_function(odes, species(rn))\nf = eval(f_oop_expr)\n\nc = [3., 5., 2., 6.]\nf(c)\n\nThe generated f now corresponds to the f(mathbfx(t)) on the right-hand side of fracdmathbfx(t)dt = f(mathbfx(t)). Given a vector of species concentrations c, f will return the rate of change of each species. Steady state concentration vectors c_ss will satisfy f(c_ss) = zeros(length(species(rn))).\n\nAbove we have generated a numeric rate matrix to substitute the rate constants into the symbolic expressions. We could have used a symbolic rate matrix, but then we would need to define the parameters k, b, so that the function f knows what k and b in its output refer to.\n\n@parameters k b \nK = fluxmat(rn)\nodes = N * K * Φ\nf_oop_expr, f_iip_expr = Symbolics.build_function(odes, species(rn))\nf = eval(f_oop_expr)\n\nc = [3., 5., 2., 6.]\nf(c)\n\nAlternatively, if we use a symbolic rate matrix, we could define our function to take in both species concentrations and parameter values as arguments:\n\nK = fluxmat(rn)\nodes = N * K * Φ\nf_oop_expr, f_iip_expr = Symbolics.build_function(odes, species(rn), parameters(rn))\nf = eval(f_oop_expr)\n\nc = [3., 5., 2., 6]; ks = [12., 4.]\nf(c, ks)\n\nNote also that f can take any vector with the right dimension (i.e. the number of species), not just a vector of Number, so it can be used to build, e.g. a vector of polynomials in Nemo for commutative algebraic methods.","category":"section"},{"location":"network_analysis/odes/#Properties-of-matrix-null-spaces","page":"Decomposing the Reaction Network ODEs","title":"Properties of matrix null spaces","text":"The null spaces of the matrices discussed in this section often have special meaning. Below we will discuss some of these properties.\n\nRecall that we may write the net stoichiometry matrix N = ZB, where Z is the complex stoichiometry matrix and B is the incidence matrix of the graph.\n\nConservation laws arise as left null eigenvectors of the net stoichiometry matrix N, and cycles arise as right null eigenvectors of the stoichiometry matrix. A cycle may be understood as a sequence of reactions that leaves the overall species composition unchanged. These do not necessarily have to correspond to actual cycles in the graph.\n\nComplex balance can be compactly formulated as the following: a set of steady state reaction fluxes is complex-balanced if it is in the nullspace of the incidence matrix B.","category":"section"},{"location":"network_analysis/odes/#API-Section-for-matrices-and-vectors","page":"Decomposing the Reaction Network ODEs","title":"API Section for matrices and vectors","text":"We have that: \n\nN is the netstoichmat\nZ is the complexstoichmat\nB is the incidencemat\nK is the fluxmat\nA_k is the laplacianmat\nPhi is the massactionvector\n\n","category":"section"},{"location":"network_analysis/odes/#References","page":"Decomposing the Reaction Network ODEs","title":"References","text":"[1]: Feinberg, M. Foundations of Chemical Reaction Network Theory, Applied Mathematical Sciences 202, Springer (2019).","category":"section"},{"location":"network_analysis/odes/#Catalyst.netstoichmat-network_analysis-odes","page":"Decomposing the Reaction Network ODEs","title":"Catalyst.netstoichmat","text":"netstoichmat(rn, sparse=false)\n\nReturns the net stoichiometry matrix, N, with N_i j the net stoichiometric coefficient of the ith species within the jth reaction.\n\nNotes:\n\nSet sparse=true for a sparse matrix representation\nCaches the matrix internally within rn so subsequent calls are fast.\nNote that constant species are not treated as reactants, but just components that modify the associated rate law. As such they do not contribute to the net stoichiometry matrix.\n\n\n\n\n\n","category":"function"},{"location":"network_analysis/odes/#Catalyst.complexstoichmat-network_analysis-odes","page":"Decomposing the Reaction Network ODEs","title":"Catalyst.complexstoichmat","text":"complexstoichmat(network::ReactionSystem; sparse=false)\n\nGiven a ReactionSystem and vector of reaction complexes, return a matrix with positive entries of size number of species by number of complexes, where the non-zero positive entries in the kth column denote stoichiometric coefficients of the species participating in the kth reaction complex.\n\nNotes:\n\nSet sparse=true for a sparse matrix representation\n\n\n\n\n\n","category":"function"},{"location":"network_analysis/odes/#Catalyst.incidencemat-network_analysis-odes","page":"Decomposing the Reaction Network ODEs","title":"Catalyst.incidencemat","text":"incidencemat(rn::ReactionSystem; sparse=false)\n\nCalculate the incidence matrix of rn, see reactioncomplexes.\n\nNotes:\n\nIs cached in rn so that future calls, assuming the same sparsity, will also be fast.\n\n\n\n\n\n","category":"function"},{"location":"network_analysis/odes/#Catalyst.fluxmat-network_analysis-odes","page":"Decomposing the Reaction Network ODEs","title":"Catalyst.fluxmat","text":"fluxmat(rn::ReactionSystem, pmap = Dict(); sparse=false)\n\nReturn an r×c matrix K such that, if complex j is the substrate complex of reaction i, then K_ij = k, the rate constant for this reaction. Mostly a helper function for the network Laplacian, laplacianmat. Has the useful property that fracdxdt = S*K*Φ(x), where S is the netstoichmat or net stoichiometry matrix and Φ(x) is the massactionvector.     Returns a symbolic matrix by default, but will return a numerical matrix if rate constants are specified as a Tuple, Vector, or Dict of symbol-value pairs via pmap.\n\nWarning: Unlike other Catalyst functions, the fluxmat function will return a Matrix{Num} in the symbolic case. This is to allow easier computation of the matrix decomposition of the ODEs, and to ensure that multiplying the sparse form of the matrix will work.\n\n\n\n\n\n","category":"function"},{"location":"network_analysis/odes/#Catalyst.laplacianmat-network_analysis-odes","page":"Decomposing the Reaction Network ODEs","title":"Catalyst.laplacianmat","text":"laplacianmat(rn::ReactionSystem, pmap=Dict(); sparse=false)\n\nReturn the negative of the graph Laplacian of the reaction network. The ODE system of a chemical reaction network can be factorized as fracdxdt = Y A_k Φ(x), where Y is the complexstoichmat and A_k is the negative of the graph Laplacian, and Φ is the massactionvector. A_k is an n-by-n matrix, where n is the number of complexes, where A_ij = k_ij if a reaction exists between the two complexes and 0 otherwise.\n\nReturns a symbolic matrix by default, but will return a numerical matrix if parameter values are specified via pmap.\n\nWarning: Unlike other Catalyst functions, the laplacianmat function will return a Matrix{Num} in the symbolic case. This is to allow easier computation of the matrix decomposition of the ODEs, and to ensure that multiplying the sparse form of the matrix will work.\n\n\n\n\n\n","category":"function"},{"location":"network_analysis/odes/#Catalyst.massactionvector-network_analysis-odes","page":"Decomposing the Reaction Network ODEs","title":"Catalyst.massactionvector","text":"massactionvector(rn::ReactionSystem, scmap = Dict(); combinatoric_ratelaws = true)\n\nReturn the vector whose entries correspond to the \"mass action products\" of each complex. For example, given the complex A + B, the corresponding entry of the vector would be A*B, and for the complex 2X + Y, the corresponding entry would be X^2*Y. The ODE system of a chemical reaction network can be factorized as racdxdt = Y A_k Φ(x), where Y is the complexstoichmat and A_k is the negative of the laplacianmat. This utility returns Φ(x).\n\nReturns a symbolic vector by default, but will return a numerical vector if species concentrations are specified as a tuple, vector, or dictionary via scmap.\n\nIf the combinatoric_ratelaws option is set, will include prefactors for that (see introduction to Catalyst's rate laws. Will default to the default for the system.\n\nWarning: Unlike other Catalyst functions, the massactionvector function will return a Vector{Num} in the symbolic case. This is to allow easier computation of the matrix decomposition of the ODEs.\n\n\n\n\n\n","category":"function"},{"location":"steady_state_functionality/dynamical_systems/#dynamical_systems","page":"Analysing Model Steady State Properties with DynamicalSystems.jl","title":"Analysing Model Steady State Properties with DynamicalSystems.jl","text":"<details><summary><strong>Environment setup and package installation</strong></summary>\n\nThe following code sets up an environment for running the code on this page.\n\nusing Pkg\nPkg.activate(; temp = true) # Creates a temporary environment, which is deleted when the Julia session ends.\nPkg.add(\"CairoMakie\")\nPkg.add(\"Catalyst\")\nPkg.add(\"DynamicalSystems\")\nPkg.add(\"OrdinaryDiffEqRosenbrock\")\nPkg.add(\"OrdinaryDiffEqTsit5\")\nPkg.add(\"Plots\")\n\n</details>\n\n\n   The DynamicalSystems.jl package implements a wide range of methods for analysing dynamical systems[1][2]. This includes both continuous-time systems (i.e. ODEs) and discrete-times ones (difference equations, however, these are not relevant to chemical reaction network modelling). Here we give two examples of how DynamicalSystems.jl can be used, with the package's documentation describing many more features. Finally, it should also be noted that DynamicalSystems.jl contain several tools for analysing data measured from dynamical systems.","category":"section"},{"location":"steady_state_functionality/dynamical_systems/#dynamical_systems_basins_of_attraction","page":"Analysing Model Steady State Properties with DynamicalSystems.jl","title":"Finding basins of attraction","text":"Given enough time, an ODE will eventually reach a so-called attractor. For chemical reaction networks (CRNs), this will typically be either a steady state or a limit cycle. Since ODEs are deterministic, which attractor a simulation will reach is uniquely determined by the initial condition (assuming parameter values are fixed). Conversely, each attractor is associated with a set of initial conditions such that model simulations originating in these will tend to that attractor. These sets are called basins of attraction. Here, phase space (the space of all possible states of the system) can be divided into a number of basins of attraction equal to the number of attractors. \n\nDynamicalSystems.jl provides a simple interface for finding an ODE's basins of attraction across any given subspace of phase space. In this example we will use the bistable Wilhelm model (which steady states we have previous computed using homotopy continuation). As a first step, we create an ODEProblem corresponding to the model which basins of attraction we wish to compute. For this application, u0 and tspan is unused, and their values are of little importance (the only exception is than tspan, for implementation reason, must provide a not too small interval, we recommend minimum (0.0, 1.0)). \n\nusing Catalyst\nwilhelm_model = @reaction_network begin\n    k1, Y --> 2X\n    k2, 2X --> X + Y\n    k3, X + Y --> Y\n    k4, X --> 0\nend\n\nu0 = [:X => 0.0, :Y => .0]\ntspan = (0.0, 10.0)\nps = [:k1 => 8.0, :k2 => 2.0, :k3 => 1.0, :k4 => 1.5]\noprob = ODEProblem(wilhelm_model, u0, tspan, ps)\nnothing # hide\n\nNext, for any application of DynamicalSystems.jl, our ODEProblem must be converted into a so-called CoupledODEs structure. This is done by combining the ODE with the solver (and potential solver options) with which we wish to simulate it (just like when it is simulated using solve). Here, we will simply designate the Tsit5 numeric solver (but provide no other options).\n\nusing DynamicalSystems, OrdinaryDiffEqTsit5\nds = CoupledODEs(oprob, (alg = Tsit5(),))\n\nWe can now compute the basins of attraction. This is done by first creating a grid that designates which subspace of phase-space we wish to investigate (here, the corresponding basin of attraction is found for every point on the grid). Next, we create a AttractorsViaRecurrences struct, that maps initial conditions to attractors, and then use that as input to the basins_of_attraction function.\n\n# We provide one grid of values for each species. These are then bundled into a tuple.\nx_grid = 0.0:0.03:6.0\ny_grid = 0.0:0.03:9.0\ngrid = (x_grid, y_grid)\navr = AttractorsViaRecurrences(ds, grid)\nbasins, attractors = basins_of_attraction(avr, grid; show_progress = false)\nattractors\n\nHere, attractors is a dictionary that maps attractor labels (integers) to attractors. In this case we have two fixed points, one at (0000) and one at (4560). Next, basins is a matrix of equal size to grid, where each value is an integer describing to which attractor's basin that state belongs.\n\nDynamicalSystems.jl also provides a simple interface for plotting the resulting basins. This uses Makie.jl, an alternative plotting package to Plots.jl (which is typically the preferred plotting package within the context of Catalyst). Generally, Makie is good at creating animations or interactive graphics (however, it is also a popular competitor to Plots.jl for general-purpose plotting). \n\nusing CairoMakie\nheatmap_basins_attractors(grid, basins, attractors)\n\nHere, in addition to the basins of attraction, the system's three steady states are marked (the one at the intersection of the two basins is unstable).\n\nwarning: Warning\nBoth Makie and Plots.jl exports a function called plot. Hence, if both these packages are imported into the same session, calls to plot must be prepended with the package one wishes to use (e.g. Plots.plot(sol)).\n\nMore information on how to compute basins of attractions for ODEs using DynamicalSystems.jl can be found here.","category":"section"},{"location":"steady_state_functionality/dynamical_systems/#dynamical_systems_lyapunov_exponents","page":"Analysing Model Steady State Properties with DynamicalSystems.jl","title":"Computing Lyapunov exponents","text":"Lyapunov exponents are scalar values that can be computed for any attractor of an ODE. For an ODE with n variables, for each state, a total of n Lyapunov exponents can be computed (and these are collectively called the Lyapunov spectrum). Positive Lyapunov exponents indicate that trajectories initially infinitesimally close diverge from each other. Conversely, negative Lyapunov exponents suggests that trajectories converge to each others. \n\nWhile Lyapunov exponents can be used for other purposes, they are primarily used to characterise chaotic behaviours (where small changes in initial conditions has large effect on the resulting trajectories). Generally, an ODE exhibit chaotic behaviour if its attractor(s) have at least one positive Lyapunov exponent. Practically, Lyapunov exponents can be computed using DynamicalSystems.jl's lyapunovspectrum function. Here we will use it to investigate two models, one which exhibits chaos and one which do not.\n\nFirst, let us consider the Willamowski–Rössler model, which is known to exhibit chaotic behaviour.\n\nusing Catalyst\nwr_model = @reaction_network begin\n    k1, 2X --> 3X\n    k2, X --> 2X\n    k3, Z + 2X --> 2Z\n    k4, Y + X --> 2Y\n    k5, Y --> ∅\n    k6, 2Z --> ∅\n    k7, Z --> ∅\nend\n\nWe can simulate the model, noting that its behaviour seems chaotic.\n\nusing OrdinaryDiffEqRosenbrock, Plots\n\nu0 = [:X => 1.5, :Y => 1.5, :Z => 1.5]\ntspan = (0.0, 100.0)\np = [:k1 => 2.1, :k2 => 0.7, :k3 => 2.9, :k4 => 1.1, :k5 => 1.0, :k6 => 0.5, :k7 => 2.7]\n\noprob = ODEProblem(wr_model, u0, tspan, p)\nsol = solve(oprob, Rodas5P())\nplot(sol; idxs=(:X, :Y, :Z))\n\nNext, like when we computed basins of attraction, we create a CoupledODEs corresponding to the model and state for which we wish to compute our Lyapunov spectrum. Like previously, tspan must provide some small interval (at least (0.0, 1.0) is recommended), but else have no impact on the computed Lyapunov spectrum.\n\nusing DynamicalSystems\nds = CoupledODEs(oprob, (alg = Rodas5P(autodiff = false),))\nnothing # hide\n\nHere, the autodiff = false argument is required when Lyapunov spectrums are computed. We can now provide our CoupledODEs (ds) to lyapunovspectrum to compute the lyapunov spectrum. This function requires a second argument (here set to 100). Generally setting this to a higher value will increase accuracy, but also increase runtime (since lyapunovspectrum is fast for most systems, setting this to a large value is recommended).\n\nlyapunovspectrum(ds, 100)\n\nHere, the largest exponent is positive, suggesting that the model is chaotic (or, more accurately, it has at least one chaotic attractor, to which is approached from the initial condition (151515)).\n\nNext, we consider the [Brusselator] model. First we simulate the model for two similar initial conditions, confirming that they converge to the same limit cycle:\n\nusing OrdinaryDiffEqTsit5\nbrusselator = @reaction_network begin\n    A, ∅ --> X\n    1, 2X + Y --> 3X\n    B, X --> Y\n    1, X --> ∅\nend\n\nu0_1 = [:X => 1.0, :Y => 1.0]\nu0_2 = [:X => 1.2, :Y => 1.0]\ntspan = (0., 25.)\nps = [:A => 1.0, :B => 4.0]\n\noprob1 = ODEProblem(brusselator, u0_1, tspan, ps)\noprob2 = ODEProblem(brusselator, u0_2, tspan, ps)\nosol1  = solve(oprob1, Tsit5())\nosol2  = solve(oprob2, Tsit5())\nplot(osol1; idxs = (:X, :Y))\nplot!(osol2; idxs = (:X, :Y))\n\nNext, we compute the Lyapunov spectrum at one of the initial conditions:\n\nds = CoupledODEs(oprob1, (alg = Rodas5P(autodiff = false),))\nlyapunovspectrum(ds, 100)\n\nHere, all Lyapunov exponents are negative, confirming that the brusselator is non-chaotic.\n\nMore details on how to compute Lyapunov exponents using DynamicalSystems.jl can be found here. A full overview of tools for analysing chaotic behaviours (using the \"ChaosTools.jl subpackage) can be found here.\n\n","category":"section"},{"location":"steady_state_functionality/dynamical_systems/#dynamical_systems_citations","page":"Analysing Model Steady State Properties with DynamicalSystems.jl","title":"Citations","text":"If you use this functionality in your research, in addition to Catalyst, please cite the following paper to support the author of the DynamicalSystems.jl package:\n\n@article{DynamicalSystems.jl-2018,\n  doi = {10.21105/joss.00598},\n  url = {https://doi.org/10.21105/joss.00598},\n  year  = {2018},\n  month = {mar},\n  volume = {3},\n  number = {23},\n  pages = {598},\n  author = {George Datseris},\n  title = {DynamicalSystems.jl: A Julia software library for chaos and nonlinear dynamics},\n  journal = {Journal of Open Source Software}\n}\n\n","category":"section"},{"location":"steady_state_functionality/dynamical_systems/#Learning-more","page":"Analysing Model Steady State Properties with DynamicalSystems.jl","title":"Learning more","text":"If you want to learn more about analysing dynamical systems, including chaotic behaviour, see the textbook Nonlinear Dynamics. It utilizes DynamicalSystems.jl and provides a concise, hands-on approach to learning nonlinear dynamics and analysing dynamical systems [3].\n\n","category":"section"},{"location":"steady_state_functionality/dynamical_systems/#References","page":"Analysing Model Steady State Properties with DynamicalSystems.jl","title":"References","text":"[1]: S. H. Strogatz, Nonlinear Dynamics and Chaos, Westview Press (1994).\n\n[2]: A. M. Lyapunov, The general problem of the stability of motion, International Journal of Control (1992).\n\n[3]: G. Datseris, U. Parlitz, Nonlinear dynamics: A concise introduction interlaced with code, Springer (2022).","category":"section"}]
}
