<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Approaches for modelling system noise · Catalyst.jl</title><meta name="title" content="Approaches for modelling system noise · Catalyst.jl"/><meta property="og:title" content="Approaches for modelling system noise · Catalyst.jl"/><meta property="twitter:title" content="Approaches for modelling system noise · Catalyst.jl"/><meta name="description" content="Documentation for Catalyst.jl."/><meta property="og:description" content="Documentation for Catalyst.jl."/><meta property="twitter:description" content="Documentation for Catalyst.jl."/><meta property="og:url" content="https://docs.sciml.ai/Catalyst/stable/model_creation/examples/noise_modelling_approaches/"/><meta property="twitter:url" content="https://docs.sciml.ai/Catalyst/stable/model_creation/examples/noise_modelling_approaches/"/><link rel="canonical" href="https://docs.sciml.ai/Catalyst/stable/model_creation/examples/noise_modelling_approaches/"/><script async src="https://www.googletagmanager.com/gtag/js?id=UA-90474609-3"></script><script>  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());
  gtag('config', 'UA-90474609-3', {'page_path': location.pathname + location.search + location.hash});
</script><script data-outdated-warner src="../../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../../assets/documenter.js"></script><script src="../../../search_index.js"></script><script src="../../../siteinfo.js"></script><script src="../../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../../assets/themeswap.js"></script><link href="../../../assets/favicon.ico" rel="icon" type="image/x-icon"/></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../../../"><img src="../../../assets/logo.png" alt="Catalyst.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href="../../../">Catalyst.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../../../">Home</a></li><li><input class="collapse-toggle" id="menuitem-2" type="checkbox"/><label class="tocitem" for="menuitem-2"><span class="docs-label">Introduction to Catalyst</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../../../introduction_to_catalyst/catalyst_for_new_julia_users/">Introduction to Catalyst and Julia for New Julia users</a></li><li><a class="tocitem" href="../../../introduction_to_catalyst/introduction_to_catalyst/">Introduction to Catalyst</a></li><li><a class="tocitem" href="../../../introduction_to_catalyst/math_models_intro/">Mathematical Models Catalyst can Generate</a></li></ul></li><li><input class="collapse-toggle" id="menuitem-3" type="checkbox" checked/><label class="tocitem" for="menuitem-3"><span class="docs-label">Model creation and properties</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../../dsl_basics/">The Catalyst DSL - Introduction</a></li><li><a class="tocitem" href="../../dsl_advanced/">The Catalyst DSL - Advanced Features and Options</a></li><li><a class="tocitem" href="../../programmatic_CRN_construction/">Programmatic Construction of Symbolic Reaction Systems</a></li><li><a class="tocitem" href="../../compositional_modeling/">Compositional Modeling of Reaction Systems</a></li><li><a class="tocitem" href="../../coupled_non_crn_models/">Coupled non-reaction network equations to models</a></li><li><a class="tocitem" href="../../events/">Modelling Events</a></li><li><a class="tocitem" href="../../conservation_laws/">Working with Conservation Laws</a></li><li><a class="tocitem" href="../../parametric_stoichiometry/">Symbolic Stoichiometries</a></li><li><a class="tocitem" href="../../functional_parameters/">Inputs and time-dependent (or functional) parameters</a></li><li><a class="tocitem" href="../../model_file_loading_and_export/">Loading Chemical Reaction Network Models from Files</a></li><li><a class="tocitem" href="../../model_visualisation/">Model Visualisation</a></li><li><a class="tocitem" href="../../reactionsystem_content_accessing/">Accessing Model Properties</a></li><li><a class="tocitem" href="../../chemistry_related_functionality/">Chemistry-related Functionality</a></li><li><input class="collapse-toggle" id="menuitem-3-14" type="checkbox" checked/><label class="tocitem" for="menuitem-3-14"><span class="docs-label">Examples</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../basic_CRN_library/">Library of Basic Chemical Reaction Network Models</a></li><li><a class="tocitem" href="../programmatic_generative_linear_pathway/">Programmatic, Generative, Modelling of a Linear Pathway</a></li><li><a class="tocitem" href="../hodgkin_huxley_equation/">Hodgkin-Huxley Equation</a></li><li><a class="tocitem" href="../smoluchowski_coagulation_equation/">Smoluchowski Coagulation Equation</a></li><li class="is-active"><a class="tocitem" href>Approaches for modelling system noise</a><ul class="internal"><li><a class="tocitem" href="#noise_modelling_approaches_model_intro"><span>The repressilator model</span></a></li><li><a class="tocitem" href="#noise_modelling_approaches_model_intrinsic"><span>Using intrinsic noise</span></a></li><li><a class="tocitem" href="#noise_modelling_approaches_model_extrinsic"><span>Using extrinsic noise</span></a></li><li><a class="tocitem" href="#noise_modelling_approaches_model_input_noise"><span>Using a noisy input process</span></a></li><li><a class="tocitem" href="#noise_modelling_approaches_model_noisy_oscillation_mean"><span>Investigating the mean of noisy oscillations</span></a></li><li><a class="tocitem" href="#noise_modelling_approaches_references"><span>References</span></a></li></ul></li></ul></li></ul></li><li><input class="collapse-toggle" id="menuitem-4" type="checkbox"/><label class="tocitem" for="menuitem-4"><span class="docs-label">Model simulation and visualization</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../../../model_simulation/simulation_introduction/">Model Simulation Introduction</a></li><li><a class="tocitem" href="../../../model_simulation/simulation_plotting/">Simulation Plotting</a></li><li><a class="tocitem" href="../../../model_simulation/simulation_structure_interfacing/">Interfacing Problems, Integrators, and Solutions</a></li><li><a class="tocitem" href="../../../model_simulation/ensemble_simulations/">Ensemble/Monte Carlo Simulations</a></li><li><a class="tocitem" href="../../../model_simulation/ode_simulation_performance/">Advice for Performant ODE Simulations</a></li><li><a class="tocitem" href="../../../model_simulation/sde_simulation_performance/">Advice for Performant SDE Simulations</a></li><li><input class="collapse-toggle" id="menuitem-4-7" type="checkbox"/><label class="tocitem" for="menuitem-4-7"><span class="docs-label">Examples</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../../../model_simulation/examples/periodic_events_simulation/">Modelling a Periodic Event During ODE and Jump Simulations</a></li><li><a class="tocitem" href="../../../model_simulation/examples/activation_time_distribution_measurement/">Measuring the Distribution of System Activation Times</a></li><li><a class="tocitem" href="../../../model_simulation/examples/interactive_brusselator_simulation/">Interactive Simulation and Plotting</a></li></ul></li></ul></li><li><input class="collapse-toggle" id="menuitem-5" type="checkbox"/><label class="tocitem" for="menuitem-5"><span class="docs-label">Network Analysis</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../../../network_analysis/odes/">Decomposing the Reaction Network ODEs</a></li><li><a class="tocitem" href="../../../network_analysis/crn_theory/">Chemical Reaction Network Theory</a></li><li><a class="tocitem" href="../../../network_analysis/network_properties/">Caching of Network Properties in <code>ReactionSystems</code></a></li></ul></li><li><input class="collapse-toggle" id="menuitem-6" type="checkbox"/><label class="tocitem" for="menuitem-6"><span class="docs-label">Steady state analysis</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../../../steady_state_functionality/homotopy_continuation/">Finding Steady States through Homotopy Continuation</a></li><li><a class="tocitem" href="../../../steady_state_functionality/nonlinear_solve/">Finding Steady States using NonlinearSolve.jl and SteadyStateDiffEq.jl</a></li><li><a class="tocitem" href="../../../steady_state_functionality/steady_state_stability_computation/">Steady state stability computation</a></li><li><a class="tocitem" href="../../../steady_state_functionality/bifurcation_diagrams/">Bifurcation Diagrams</a></li><li><a class="tocitem" href="../../../steady_state_functionality/dynamical_systems/">Analysing Model Steady State Properties with DynamicalSystems.jl</a></li><li><input class="collapse-toggle" id="menuitem-6-6" type="checkbox"/><label class="tocitem" for="menuitem-6-6"><span class="docs-label">Examples</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../../../steady_state_functionality/examples/nullcline_plotting/">Plotting Nullclines and Steady States in Phase Space</a></li><li><a class="tocitem" href="../../../steady_state_functionality/examples/bifurcationkit_periodic_orbits/">Computing Periodic Orbits (Oscillations) Using BifurcationKit.jl</a></li><li><a class="tocitem" href="../../../steady_state_functionality/examples/bifurcationkit_codim2/">Tracking Bifurcation Point w.r.t. Secondary Parameters using BifurcationKit.jl</a></li></ul></li></ul></li><li><input class="collapse-toggle" id="menuitem-7" type="checkbox"/><label class="tocitem" for="menuitem-7"><span class="docs-label">Inverse problems</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../../../inverse_problems/optimization_ode_param_fitting/">Parameter Fitting for ODEs using Optimization.jl</a></li><li><a class="tocitem" href="../../../inverse_problems/behaviour_optimisation/">Optimization for Non-data Fitting Purposes</a></li><li><a class="tocitem" href="../../../inverse_problems/structural_identifiability/">Structural Identifiability Analysis</a></li><li><a class="tocitem" href="../../../inverse_problems/global_sensitivity_analysis/">Global Sensitivity Analysis</a></li></ul></li><li><input class="collapse-toggle" id="menuitem-8" type="checkbox"/><label class="tocitem" for="menuitem-8"><span class="docs-label">Spatial modelling</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../../../spatial_modelling/discrete_space_reaction_systems/">Introduction to Spatial Modelling with Catalyst</a></li><li><a class="tocitem" href="../../../spatial_modelling/discrete_space_simulation_structure_interaction/">Interfacing with Lattice Problems, Integrators, and Solutions</a></li><li><a class="tocitem" href="../../../spatial_modelling/discrete_space_simulation_plotting/">Plotting Spatial Simulations</a></li><li><a class="tocitem" href="../../../spatial_modelling/spatial_ode_simulations/">Spatial ODE simulations</a></li><li><a class="tocitem" href="../../../spatial_modelling/spatial_jump_simulations/">Spatial Jump Simulations</a></li></ul></li><li><a class="tocitem" href="../../../faqs/">FAQs</a></li><li><input class="collapse-toggle" id="menuitem-10" type="checkbox"/><label class="tocitem" for="menuitem-10"><span class="docs-label">API</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../../../api/core_api/">Catalyst.jl API</a></li><li><a class="tocitem" href="../../../api/network_analysis_api/">Network analysis and representations</a></li></ul></li><li><a class="tocitem" href="../../../devdocs/dev_guide/">Developer Documentation</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Model creation and properties</a></li><li><a class="is-disabled">Examples</a></li><li class="is-active"><a href>Approaches for modelling system noise</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Approaches for modelling system noise</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/SciML/Catalyst.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/SciML/Catalyst.jl/blob/master/docs/src/model_creation/examples/noise_modelling_approaches.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="noise_modelling_approaches"><a class="docs-heading-anchor" href="#noise_modelling_approaches">Approaches for modelling system noise</a><a id="noise_modelling_approaches-1"></a><a class="docs-heading-anchor-permalink" href="#noise_modelling_approaches" title="Permalink"></a></h1><details><summary><strong>Environment setup and package installation</strong></summary><p>The following code sets up an environment for running the code on this page.</p><pre><code class="language-julia hljs">using Pkg
Pkg.activate(; temp = true) # Creates a temporary environment, which is deleted when the Julia session ends.
Pkg.add(&quot;Catalyst&quot;)
Pkg.add(&quot;DataInterpolations&quot;)
Pkg.add(&quot;Distributions&quot;)
Pkg.add(&quot;OrdinaryDiffEqDefault&quot;)
Pkg.add(&quot;Plots&quot;)
Pkg.add(&quot;StochasticDiffEq&quot;)</code></pre></details><p><br/>   Catalyst&#39;s primary tools for modelling stochasticity include the creation of <code>SDEProblem</code>s or <code>JumpProblem</code>s from reaction network models. However, other approaches for incorporating model noise exist, some of which will be discussed here. We will first consider <em>intrinsic</em> and <em>extrinsic</em> noise. These are well-established terms, both of which we will describe below (however, to our knowledge, no generally agreed-upon definition of these terms exists)<sup class="footnote-reference"><a id="citeref-1" href="#footnote-1" class="footnote-ref">[1]</a><span class="footnote-preview" id="fn-1"></span></sup>. Finally, we will demonstrate a third approach, the utilisation of a noisy input process to an otherwise deterministic system. This approach is infrequently used, however, as it is encountered in the literature, we will demonstrate it here as well. </p><p>We note that these approaches can all be combined. E.g. an intrinsic noise model (using an SDE) can be combined with extrinsic noise (using randomised parameter values), while also feeding a noisy input process into the system.</p><div class="admonition is-info" id="Note-5826fd05891ba7de"><header class="admonition-header">Note<a class="admonition-anchor" href="#Note-5826fd05891ba7de" title="Permalink"></a></header><div class="admonition-body"><p>Here we use intrinsic and extrinsic noise as descriptions of two of our modelling approaches. It should be noted that while these are established terminologies for noisy biological systems<sup class="footnote-reference"><a id="citeref-1" href="#footnote-1" class="footnote-ref">[1]</a><span class="footnote-preview" id="fn-1"></span></sup>, our use of these terms to describe different approaches for modelling noise is only inspired by this terminology, and nothing that is established in the field.  Please consider the <a href="#noise_modelling_approaches_references">references</a> for more information on intrinsic and extrinsic noise.</p></div></div><h2 id="noise_modelling_approaches_model_intro"><a class="docs-heading-anchor" href="#noise_modelling_approaches_model_intro">The repressilator model</a><a id="noise_modelling_approaches_model_intro-1"></a><a class="docs-heading-anchor-permalink" href="#noise_modelling_approaches_model_intro" title="Permalink"></a></h2><p>For this tutorial we will use the oscillating <a href="../basic_CRN_library/#basic_CRN_library_repressilator">repressilator</a> model.</p><pre><code class="language-julia hljs">using Catalyst
repressilator = @reaction_network begin
    hillr(Z,v,K,n), ∅ --&gt; X
    hillr(X,v,K,n), ∅ --&gt; Y
    hillr(Y,v,K,n), ∅ --&gt; Z
    d, (X, Y, Z) --&gt; ∅
end</code></pre><p class="math-container">\[ \begin{align*}
\varnothing &amp;\xrightarrow{\frac{K^{n} v}{Z^{n} + K^{n}}} \mathrm{X} \\
\varnothing &amp;\xrightarrow{\frac{K^{n} v}{K^{n} + X^{n}}} \mathrm{Y} \\
\varnothing &amp;\xrightarrow{\frac{K^{n} v}{Y^{n} + K^{n}}} \mathrm{Z} \\
\mathrm{X} &amp;\xrightarrow{d} \varnothing \\
\mathrm{Y} &amp;\xrightarrow{d} \varnothing \\
\mathrm{Z} &amp;\xrightarrow{d} \varnothing  
 \end{align*}
 \]</p><h2 id="noise_modelling_approaches_model_intrinsic"><a class="docs-heading-anchor" href="#noise_modelling_approaches_model_intrinsic">Using intrinsic noise</a><a id="noise_modelling_approaches_model_intrinsic-1"></a><a class="docs-heading-anchor-permalink" href="#noise_modelling_approaches_model_intrinsic" title="Permalink"></a></h2><p>Generally, intrinsic noise is randomness inherent to a system itself. This means that it cannot be controlled for, or filtered out by, experimental settings. Low-copy number cellular systems, were reaction occurs due to the encounters of molecules due to random diffusion, is an example of intrinsic noise. In practise, this can be modelled exactly through <a href="../../../model_simulation/simulation_introduction/#simulation_intro_SDEs">SDE</a> (chemical Langevin equations) or <a href="../../../model_simulation/simulation_introduction/#simulation_intro_jumps">jump</a> (stochastic chemical kinetics) simulations. </p><p>In Catalyst, intrinsic noise is accounted for whenever an <code>SDEProblem</code> or <code>JumpProblem</code> is created and simulated. Here we will model intrinsic noise through SDEs, which means creating an <code>SDEProblem</code> using the standard approach.</p><pre><code class="language-julia hljs">u0 = [:X =&gt; 45.0, :Y =&gt; 20.0, :Z =&gt; 20.0]
tend = 200.0
ps = [:v =&gt; 10.0, :K =&gt; 20.0, :n =&gt; 3, :d =&gt; 0.1]
sprob = SDEProblem(repressilator, u0, tend, ps)</code></pre><p>Next, to illustrate the system&#39;s noisiness, we will perform multiple simulations. We do this by <a href="../../../model_simulation/ensemble_simulations/#ensemble_simulations_monte_carlo">creating an <code>EnsembleProblem</code></a>. From it, we perform, and plot, 4 simulations.</p><pre><code class="language-julia hljs">using StochasticDiffEq, Plots
eprob_intrinsic = EnsembleProblem(sprob)
sol_intrinsic = solve(eprob_intrinsic, ImplicitEM(); trajectories = 4)
plot(sol_intrinsic; idxs = :X)</code></pre><img src="9942210b.svg" alt="Example block output"/><p>Here, each simulation is performed from the same system using the same settings. Despite this, due to the noise, the individual trajectories are different.</p><h2 id="noise_modelling_approaches_model_extrinsic"><a class="docs-heading-anchor" href="#noise_modelling_approaches_model_extrinsic">Using extrinsic noise</a><a id="noise_modelling_approaches_model_extrinsic-1"></a><a class="docs-heading-anchor-permalink" href="#noise_modelling_approaches_model_extrinsic" title="Permalink"></a></h2><p>Next, we consider extrinsic noise. This is randomness caused by stochasticity external to, yet affecting, a system. Examples could be different bacteria experiencing different microenvironments or cells being in different parts of the cell cycle. This is noise which (in theory) can be controlled for experimentally (e.g. by ensuring a uniform environment). Whenever a specific source of noise is intrinsic and extrinsic to a system may depend on how one defines the system itself (this is a reason why giving an exact definition of these terms is difficult).</p><p>In Catalyst we can model extrinsic noise by letting the model parameters be probability distributions. Here, at the beginning of each simulation, random parameter values are drawn from their distributions. Let us imagine that our repressilator circuit was inserted into a bacterial population. Here, while each bacteria would have the same circuit, their individual number of e.g. ribosomes (which will be random) might affect the production rates (which while constant within each bacteria, might differ between the individuals).</p><p>Again we will perform ensemble simulation. Instead of creating an <code>SDEProblem</code>, we will create an <code>ODEProblem</code>, as well as a <a href="../../../model_simulation/ensemble_simulations/#ensemble_simulations_varying_conditions">problem function</a> which draws random parameter values for each simulation. Here we have implemented the parameter&#39;s probability distributions as <a href="https://en.wikipedia.org/wiki/Normal_distribution">normal distributions</a> using the <a href="https://github.com/JuliaStats/Distributions.jl">Distributions.jl</a> package.</p><pre><code class="language-julia hljs">using Distributions
p_dists = Dict([:v =&gt; Normal(10.0, 2.0), :K =&gt; Normal(20.0, 5.0), :n =&gt; Normal(3, 0.2), :d =&gt; Normal(0.1, 0.02)])
function prob_func(prob, i, repeat)
    p = [par =&gt; rand(p_dists[par]) for par in keys(p_dists)]
    return remake(prob; p)
end</code></pre><p>Next, we again perform 4 simulations. While the individual trajectories are performed using deterministic simulations, the randomised parameter values create heterogeneity across the ensemble.</p><pre><code class="language-julia hljs">using OrdinaryDiffEqDefault
oprob = ODEProblem(repressilator, u0, tend, ps)
eprob_extrinsic = EnsembleProblem(oprob; prob_func)
sol_extrinsic = solve(eprob_extrinsic; trajectories = 4)
plot(sol_extrinsic; idxs = :X)</code></pre><img src="43abcb3f.svg" alt="Example block output"/><p>We note that a similar approach can be used to also randomise the initial conditions. In a very detailed model, the parameter values could fluctuate during a single simulation, something which could be implemented using the approach from the next section.</p><h2 id="noise_modelling_approaches_model_input_noise"><a class="docs-heading-anchor" href="#noise_modelling_approaches_model_input_noise">Using a noisy input process</a><a id="noise_modelling_approaches_model_input_noise-1"></a><a class="docs-heading-anchor-permalink" href="#noise_modelling_approaches_model_input_noise" title="Permalink"></a></h2><p>Finally, we will consider the case where we have a deterministic system, but which is exposed to a noisy input process. One example could be a <a href="../../functional_parameters/#functional_parameters_circ_rhythm">light sensitive system, where the amount of experienced sunlight is stochastic due to e.g. variable cloud cover</a>. Practically, this can be considered as extrinsic noise, however, we will generate the noise using a different approach from in the previous section. Here, we pre-simulate a random process in time, which we then feed into the system as a functional, time-dependent, parameter. A more detailed description of functional parameters can be found <a href="../../functional_parameters/#time_dependent_parameters">here</a>.</p><p>We assume that our repressilator has an input, which corresponds to the <span>$K$</span> value that controls <span>$X$</span>&#39;s production. First we create a function, <code>make_K_series</code>, which creates a randomised time series representing <span>$K$</span>&#39;s value over time. </p><pre><code class="language-julia hljs">using DataInterpolations
function make_K_series(; K_mean = 20.0, n = 500, θ = 0.01)
    t_samples = range(0.0, stop = tend, length = n)
    K_series = fill(K_mean, n)
    for i = 2:n
        K_series[i] = K_series[i-1] + (rand() - 0.5) - θ*(K_series[i-1] - K_mean)
    end
    return LinearInterpolation(K_series, t_samples)
end
plot(make_K_series())</code></pre><img src="b85c8a1e.svg" alt="Example block output"/><p>Next, we create an updated repressilator model, where the input <span>$K$</span> value is modelled as a time-dependent parameter.</p><pre><code class="language-julia hljs">@parameters (K_in::typeof(make_K_series()))(..)
K_in = K_in(default_t())
repressilator_Kin = @reaction_network begin
    hillr(Z,v,$K_in,n), ∅ --&gt; X
    hillr(X,v,K,n), ∅ --&gt; Y
    hillr(Y,v,K,n), ∅ --&gt; Z
    d, (X, Y, Z) --&gt; ∅
end</code></pre><p>Finally, we will again perform ensemble simulations of our model. This time, at the beginning of each simulation, we will use <code>make_K_series</code> to generate a new <span>$K$</span>, and set this as the <code>K_in</code> parameter&#39;s value.</p><pre><code class="language-julia hljs">function prob_func_Kin(prob, i, repeat)
    p = [ps; :K_in =&gt; make_K_series()]
    return remake(prob; p)
end
oprob = ODEProblem(repressilator_Kin, u0, tend, [ps; :K_in =&gt; make_K_series()])
eprob_inputnoise = EnsembleProblem(oprob; prob_func = prob_func_Kin)
sol_inputnoise = solve(eprob_inputnoise; trajectories = 4)
plot(sol_inputnoise; idxs = :X)</code></pre><img src="bf76b6fd.svg" alt="Example block output"/><p>Like in the previous two cases, this generates heterogeneous trajectories across our ensemble.</p><h2 id="noise_modelling_approaches_model_noisy_oscillation_mean"><a class="docs-heading-anchor" href="#noise_modelling_approaches_model_noisy_oscillation_mean">Investigating the mean of noisy oscillations</a><a id="noise_modelling_approaches_model_noisy_oscillation_mean-1"></a><a class="docs-heading-anchor-permalink" href="#noise_modelling_approaches_model_noisy_oscillation_mean" title="Permalink"></a></h2><p>Finally, we will observe an interesting phenomenon for ensembles of stochastic oscillators. First, we create ensemble simulations with a larger number of trajectories.</p><pre><code class="language-julia hljs">sol_intrinsic = solve(eprob_intrinsic, ImplicitEM(); trajectories = 200)
sol_extrinsic = solve(eprob_extrinsic; trajectories = 200)</code></pre><p>Next, we can use the <code>EnsembleSummary</code> interface to plot each ensemble&#39;s mean activity (as well as 5% and 95% quantiles) over time:</p><pre><code class="language-julia hljs">e_summary_intrinsic = EnsembleAnalysis.EnsembleSummary(sol_intrinsic, 0.0:1.0:tend)
e_summary_extrinsic = EnsembleAnalysis.EnsembleSummary(sol_extrinsic, 0.0:1.0:tend)
plot(e_summary_intrinsic; label = &quot;Intrinsic noise&quot;, idxs = 1)
plot!(e_summary_extrinsic; label = &quot;Extrinsic noise&quot;, idxs = 1)</code></pre><img src="a1b9ca59.svg" alt="Example block output"/><p>Here we can see that, over time, the systems&#39; mean <span>$X$</span> activity reaches a constant level around <span>$30$</span>.</p><p>This is a well-known phenomenon (especially in circadian biology<sup class="footnote-reference"><a id="citeref-2" href="#footnote-2" class="footnote-ref">[2]</a><span class="footnote-preview" id="fn-2"></span></sup>). Here, as stochastic oscillators evolve from a common initial condition the mean behaves as a damped oscillator. This can be caused by two different phenomena:</p><ul><li>The individual trajectories are themselves damped.</li><li>The individual trajectories&#39;s phases get de-synchronised.</li></ul><p>However, if we only observe the mean behaviour (and not the single trajectories), we cannot know which of these cases we are encountering. Here, by checking the single-trajectory plots from the previous sections, we note that this is due to trajectory de-synchronisation. Stochastic oscillators have often been cited as a reason for the importance to study cellular systems at the <em>single-cell</em> level, and not just in bulk.</p><hr/><h2 id="noise_modelling_approaches_references"><a class="docs-heading-anchor" href="#noise_modelling_approaches_references">References</a><a id="noise_modelling_approaches_references-1"></a><a class="docs-heading-anchor-permalink" href="#noise_modelling_approaches_references" title="Permalink"></a></h2><section class="footnotes is-size-7"><ul><li class="footnote" id="footnote-1"><a class="tag is-link" href="#citeref-1">1</a><a href="https://www.science.org/doi/10.1126/science.1070919">Michael B. Elowitz, Arnold J. Levine, Eric D. Siggia, Peter S. Swain, <em>Stochastic Gene Expression in a Single Cell</em>, Science (2002).</a></li><li class="footnote" id="footnote-2"><a class="tag is-link" href="#citeref-2">2</a><a href="https://www.science.org/doi/10.1126/science.1181759">Qiong Yang, Bernardo F. Pando, Guogang Dong, Susan S. Golden, Alexander van Oudenaarden, <em>Circadian Gating of the Cell Cycle Revealed in Single Cyanobacterial Cells</em>, Science (2010).</a></li></ul></section></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../smoluchowski_coagulation_equation/">« Smoluchowski Coagulation Equation</a><a class="docs-footer-nextpage" href="../../../model_simulation/simulation_introduction/">Model Simulation Introduction »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.17.0 on <span class="colophon-date" title="Friday 27 February 2026 13:54">Friday 27 February 2026</span>. Using Julia version 1.12.5.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
